// @qwen-code/qwen-code/chunks/ca-IUFWESZL.js
// Force strict mode and setup for ESM
"use strict";
import {
  init_esbuild_shims
} from "./chunk-A4BMJM77.js";
import "./chunk-J2S4EL5Y.js";

// packages/cli/src/i18n/locales/ca.js
init_esbuild_shims();
var ca_default = {
  // ============================================================================
  // Ajuda / Components de la interfície
  // ============================================================================
  "\u2191 to manage attachments": "\u2191 per gestionar els adjunts",
  "\u2190 \u2192 select, Delete to remove, \u2193 to exit": "\u2190 \u2192 seleccionar, Delete per eliminar, \u2193 per sortir",
  "Attachments: ": "Adjunts: ",
  "Basics:": "B\xE0sic:",
  "Add context": "Afegir context",
  "Use {{symbol}} to specify files for context (e.g., {{example}}) to target specific files or folders.": "Useu {{symbol}} per especificar fitxers de context (p. ex., {{example}}) per seleccionar fitxers o carpetes espec\xEDfics.",
  "@": "@",
  "@src/myFile.ts": "@src/myFile.ts",
  "Shell mode": "Mode shell",
  "YOLO mode": "Mode YOLO",
  "Auto mode": "Mode auto",
  "plan mode": "mode de planificaci\xF3",
  "auto-accept edits": "acceptaci\xF3 autom\xE0tica de canvis",
  "Accepting edits": "Acceptant canvis",
  "(shift + tab to cycle)": "(Shift + Tab per canviar)",
  "(tab to cycle)": "(Tab per canviar)",
  "Execute shell commands via {{symbol}} (e.g., {{example1}}) or use natural language (e.g., {{example2}}).": "Executeu ordres shell amb {{symbol}} (p. ex., {{example1}}) o useu el llenguatge natural (p. ex., {{example2}}).",
  "!": "!",
  "!npm run start": "!npm run start",
  "start server": "iniciar el servidor",
  "Commands:": "Ordres:",
  "shell command": "ordre shell",
  "Model Context Protocol command (from external servers)": "Ordre del protocol de context del model (des de servidors externs)",
  "Keyboard Shortcuts:": "Dreceres de teclat:",
  "Toggle this help display": "Mostrar/amagar aquesta ajuda",
  "Toggle shell mode": "Canviar el mode shell",
  "Open command menu": "Obrir el men\xFA d'ordres",
  "Add file context": "Afegir context de fitxer",
  "Accept suggestion / Autocomplete": "Acceptar suggeriment / Autocompleci\xF3",
  "Reverse search history": "Cerca inversa a l'historial",
  "Press ? again to close": "Premeu ? de nou per tancar",
  "for shell mode": "per al mode shell",
  "for commands": "per a les ordres",
  "for file paths": "per als camins de fitxers",
  "to clear input": "per esborrar l'entrada",
  "to cycle approvals": "per canviar les aprovacions",
  "to quit": "per sortir",
  "for newline": "per a nova l\xEDnia",
  "to clear screen": "per netejar la pantalla",
  "to search history": "per cercar a l'historial",
  "to paste images": "per enganxar imatges",
  "for external editor": "per a editor extern",
  "to toggle compact mode": "per canviar el mode compacte",
  "Jump through words in the input": "Saltar entre paraules a l'entrada",
  "Close dialogs, cancel requests, or quit application": "Tancar di\xE0legs, cancel\xB7lar peticions o sortir de l'aplicaci\xF3",
  "New line": "Nova l\xEDnia",
  "New line (Alt+Enter works for certain linux distros)": "Nova l\xEDnia (Alt+Enter funciona en certes distribucions de Linux)",
  "Clear the screen": "Netejar la pantalla",
  "Open input in external editor": "Obrir l'entrada en un editor extern",
  "Send message": "Enviar missatge",
  "Initializing...": "Inicialitzant...",
  "Connecting to MCP servers... ({{connected}}/{{total}})": "Connectant a MCP servers... ({{connected}}/{{total}})",
  "Type your message or @path/to/file": "Escriviu el vostre missatge o @cam\xED/al/fitxer",
  "? for shortcuts": "? per a dreceres",
  "Press 'i' for INSERT mode and 'Esc' for NORMAL mode.": "Premeu 'i' per al mode INSERCI\xD3 i 'Esc' per al mode NORMAL.",
  "Cancel operation / Clear input (double press)": "Cancel\xB7lar operaci\xF3 / Esborrar entrada (doble premuda)",
  "Cycle approval modes": "Canviar els modes d'aprovaci\xF3",
  "Cycle through your prompt history": "Navegar per l'historial de missatges",
  "For a full list of shortcuts, see {{docPath}}": "Per a una llista completa de dreceres, vegeu {{docPath}}",
  "docs/keyboard-shortcuts.md": "docs/keyboard-shortcuts.md",
  "for help on Qwen Code": "per a ajuda sobre Qwen Code",
  "show version info": "mostrar informaci\xF3 de la versi\xF3",
  "submit a bug report": "enviar un informe d'error",
  Status: "Estat",
  // ============================================================================
  // Informació del sistema
  // ============================================================================
  "Qwen Code": "Qwen Code",
  Runtime: "Entorn d'execuci\xF3",
  OS: "SO",
  Auth: "Autenticaci\xF3",
  Model: "Model",
  "Fast Model": "Model r\xE0pid",
  Sandbox: "Entorn a\xEFllat",
  "Session ID": "ID de sessi\xF3",
  "Base URL": "Base URL",
  Proxy: "Proxy",
  "Memory Usage": "\xDAs de mem\xF2ria",
  "IDE Client": "Client IDE",
  // ============================================================================
  // Ordres - General
  // ============================================================================
  "Analyzes the project and creates a tailored QWEN.md file.": "Analitza el projecte i crea un fitxer QWEN.md personalitzat.",
  "List available Qwen Code tools. Usage: /tools [desc]": "Llistar les eines disponibles de Qwen Code. \xDAs: /tools [desc]",
  "Open the skills panel (browse, search, toggle, pick).": "Obrir el panell d'habilitats (explorar, cercar, activar, triar).",
  "Manage Skills": "Gestionar habilitats",
  "Skills configuration saved.": "Configuraci\xF3 d'habilitats desada.",
  "Skills configuration saved, but refresh failed: {{error}}. Restart to ensure the new state is applied.": "Configuraci\xF3 d'habilitats desada, per\xF2 l'actualitzaci\xF3 ha fallat: {{error}}. Reinicia per assegurar-te que el nou estat s'apliqui.",
  "Workspace is untrusted; workspace settings are ignored by the merged config. Run /trust first to persist skills changes here, or edit ~/.qwen/settings.json directly to manage skills at user scope.": "L'espai de treball no \xE9s de confian\xE7a; els par\xE0metres de l'espai de treball s'ignoren a la configuraci\xF3 fusionada. Executa /trust primer, o edita ~/.qwen/settings.json directament per gestionar habilitats a l'\xE0mbit d'usuari.",
  "SkillManager not available.": "SkillManager no disponible.",
  "Loading skills\u2026": "Carregant habilitats\u2026",
  "Failed to load skills: {{error}}": "No s\u2019han pogut carregar les habilitats: {{error}}",
  "Failed to save skills configuration: {{error}}": "No s'ha pogut desar la configuraci\xF3 d'habilitats: {{error}}",
  "All available skills are disabled. Edit ~/.qwen/settings.json or .qwen/settings.json (skills.disabled) to re-enable.": "Totes les habilitats disponibles estan desactivades. Edita ~/.qwen/settings.json o .qwen/settings.json (skills.disabled) per tornar-les a activar.",
  "Press esc to close.": "Prem Esc per tancar.",
  "{{count}} skills \xB7 ": "{{count}} habilitats \xB7 ",
  "{{matched}} / {{total}} skills \xB7 ": "{{matched}} / {{total}} habilitats \xB7 ",
  "Space toggle \xB7 Enter pick (fill input) \xB7 Esc save & exit \xB7 workspace scope": "Espai alternar \xB7 Enter triar (omple l'entrada) \xB7 Esc desar i sortir \xB7 \xE0mbit d'espai de treball",
  "Search:": "Cerca:",
  "type to filter\u2026": "escriu per filtrar\u2026",
  "No skills are currently available.": "No hi ha habilitats disponibles actualment.",
  "All available skills are locked at a higher scope (see below).": "Totes les habilitats disponibles estan bloquejades en un \xE0mbit superior (veure a sota).",
  "No skills match the search.": "Cap habilitat coincideix amb la cerca.",
  "Locked by higher-scope settings (cannot toggle here):": "Bloquejades per par\xE0metres d'\xE0mbit superior (aqu\xED no es poden commutar):",
  "higher scope": "\xE0mbit superior",
  "  {{name}} {{description}}  [locked: {{scope}}]": "  {{name}} {{description}}  [bloquejada: {{scope}}]",
  "\u2191/\u2193 navigate \xB7 backspace edits search": "\u2191/\u2193 navega \xB7 Retroc\xE9s edita la cerca",
  Bundled: "Integrada",
  "Available Qwen Code CLI tools:": "Eines del CLI de Qwen Code disponibles:",
  "No tools available": "No hi ha eines disponibles",
  "View or change the approval mode for tool usage": "Veure o canviar el mode d'aprovaci\xF3 per a l'\xFAs d'eines",
  'Invalid approval mode "{{arg}}". Valid modes: {{modes}}': `Mode d'aprovaci\xF3 no v\xE0lid "{{arg}}". Modes v\xE0lids: {{modes}}`,
  'Approval mode set to "{{mode}}"': `Mode d'aprovaci\xF3 establert a "{{mode}}"`,
  "View or change the language setting": "Veure o canviar la configuraci\xF3 d'idioma",
  "List background tasks (text dump \u2014 interactive dialog opens via the footer pill)": "Llistar les tasques en segon pla (sortida de text; el di\xE0leg interactiu es pot obrir des de l'indicador del peu de p\xE0gina)",
  "Delete a previous session": "Suprimir una sessi\xF3 anterior",
  "Run installation and environment diagnostics": "Executar diagn\xF2stics d'instal\xB7laci\xF3 i d'entorn",
  "Browse dynamic model catalogs and choose which models stay enabled locally": "Explorar els cat\xE0legs din\xE0mics de models i triar quins models continuen activats localment",
  "Generate a one-line session recap now": "Generar ara un resum de la sessi\xF3 en una sola l\xEDnia",
  "Rename the current conversation. --auto lets the fast model pick a title.": "Canviar el nom de la conversa actual. --auto permet que el model r\xE0pid tri\xEF un t\xEDtol.",
  "Rewind conversation to a previous turn": "Rebobinar la conversa fins a un torn anterior",
  "Rewind Conversation": "Rebobinar la conversa",
  "No user turns to rewind to.": "No hi ha torns d'usuari per rebobinar.",
  "Rewind to: ": "Rebobinar a: ",
  "Restore code and conversation": "Restaura el codi i la conversa",
  "Restore conversation only": "Restaura nom\xE9s la conversa",
  "Restore code only": "Restaura nom\xE9s el codi",
  "Never mind": "Tant \xE9s",
  "Computing file changes...": "S'estan calculant els canvis als fitxers...",
  "Restoring...": "S'est\xE0 restaurant...",
  "Restored {{count}} file(s).": "S'han restaurat {{count}} fitxer(s).",
  "Failed to restore files: {{error}}": "Error en restaurar els fitxers: {{error}}",
  "Rewind failed: {{error}}": "Error en retrocedir: {{error}}",
  "Cannot rewind conversation: no active model client.": "No es pot retrocedir la conversa: cap client de model actiu.",
  "Code restored, but conversation could not be rewound (no active client).": "Codi restaurat, per\xF2 la conversa no s\u2019ha pogut retrocedir (cap client actiu).",
  "Conversation rewound. Edit your prompt and press Enter to continue.": "Conversa retrocedida. Edita la teva indicaci\xF3 i prem Retorn per continuar.",
  "Rewinding does not affect files edited manually or via shell commands.": "El retroc\xE9s no afecta els fitxers editats manualment o mitjan\xE7ant comandes de shell.",
  "Cannot rewind to a turn that was compressed. Try a more recent turn.": "No es pot retrocedir a un torn que ha estat comprimit. Prova amb un torn m\xE9s recent.",
  "File restore is unavailable for this turn (no captured file changes, or this turn predates the current session).": "La restauraci\xF3 de fitxers no est\xE0 disponible per a aquest torn (no s\u2019han capturat canvis, o aquest torn \xE9s anterior a la sessi\xF3 actual).",
  "(+{{insertions}} -{{deletions}} in {{count}} file)": "(+{{insertions}} -{{deletions}} en {{count}} fitxer)",
  "(+{{insertions}} -{{deletions}} in {{count}} files)": "(+{{insertions}} -{{deletions}} en {{count}} fitxers)",
  "Failed to restore {{count}} file(s): {{files}}": "Error en restaurar {{count}} fitxer(s): {{files}}",
  "Cannot restore files: this turn was created before file checkpointing was enabled.": "No es poden restaurar els fitxers: aquest torn es va crear abans que el punt de control de fitxers estigu\xE9s habilitat.",
  "No files needed to be restored.": "Cap fitxer necessitava restauraci\xF3.",
  "\u2191\u2193 to navigate \xB7 Enter to select \xB7 Esc to go back": "\u2191\u2193 per navegar \xB7 Enter per seleccionar \xB7 Esc per tornar",
  "\u2191\u2193 to navigate \xB7 Enter to select \xB7 Esc to cancel": "\u2191\u2193 per navegar \xB7 Enter per seleccionar \xB7 Esc per cancel\xB7lar",
  "Enter/Y to confirm \xB7 Esc/N to go back": "Enter/Y per confirmar \xB7 Esc/N per tornar",
  "change the theme": "canviar el tema",
  "Select Theme": "Seleccionar tema",
  Preview: "Previsualitzaci\xF3",
  "(Use Enter to select, Tab to configure scope)": "(Useu Enter per seleccionar, Tab per configurar l'\xE0mbit)",
  "(Use Enter to apply scope, Tab to go back)": "(Useu Enter per aplicar l'\xE0mbit, Tab per tornar enrere)",
  "Theme configuration unavailable due to NO_COLOR env variable.": "La configuraci\xF3 del tema no est\xE0 disponible degut a la variable d'entorn NO_COLOR.",
  'Theme "{{themeName}}" not found.': 'Tema "{{themeName}}" no trobat.',
  'Theme "{{themeName}}" not found in selected scope.': `Tema "{{themeName}}" no trobat en l'\xE0mbit seleccionat.`,
  "Clear conversation history and free up context": "Esborrar l'historial de la conversa i alliberar context",
  "Compresses the context by replacing it with a summary.": "Comprimeix el context substituint-lo per un resum.",
  "open full Qwen Code documentation in your browser": "obrir la documentaci\xF3 completa de Qwen Code al navegador",
  "Configuration not available.": "Configuraci\xF3 no disponible.",
  "Connect an LLM provider": "Connectar un prove\xEFdor LLM",
  "Copy the last AI response to clipboard (/copy N for Nth-latest)": "Copia l'\xFAltima resposta de la IA al porta-retalls (/copy N per a l'N-\xE8sima)",
  // ============================================================================
  // Ordres - Agents
  // ============================================================================
  "Manage subagents for specialized task delegation.": "Gestionar subagents per a la delegaci\xF3 de tasques especialitzades.",
  "Manage existing subagents (view, edit, delete).": "Gestionar subagents existents (veure, editar, eliminar).",
  "Create a new subagent with guided setup.": "Crear un nou subagent amb configuraci\xF3 guiada.",
  // ============================================================================
  // Agents - Diàleg de gestió
  // ============================================================================
  Agents: "Agents",
  "Choose Action": "Triar acci\xF3",
  "Edit {{name}}": "Editar {{name}}",
  "Edit Tools: {{name}}": "Editar eines: {{name}}",
  "Edit Color: {{name}}": "Editar color: {{name}}",
  "Delete {{name}}": "Eliminar {{name}}",
  "Unknown Step": "Pas desconegut",
  "Esc to close": "Esc per tancar",
  "Enter to select, \u2191\u2193 to navigate, Esc to close": "Enter per seleccionar, \u2191\u2193 per navegar, Esc per tancar",
  "Esc to go back": "Esc per tornar enrere",
  "Enter to confirm, Esc to cancel": "Enter per confirmar, Esc per cancel\xB7lar",
  "Enter to select, \u2191\u2193 to navigate, Esc to go back": "Enter per seleccionar, \u2191\u2193 per navegar, Esc per tornar enrere",
  "Enter to submit, Esc to go back": "Enter per enviar, Esc per tornar enrere",
  "Invalid step: {{step}}": "Pas no v\xE0lid: {{step}}",
  "No subagents found.": "No s'han trobat subagents.",
  "Use '/agents create' to create your first subagent.": "Useu '/agents create' per crear el vostre primer subagent.",
  "(built-in)": "(integrat)",
  "(overridden by project level agent)": "(sobreescrit per un agent de nivell de projecte)",
  "Project Level ({{path}})": "Nivell de projecte ({{path}})",
  "User Level ({{path}})": "Nivell d'usuari ({{path}})",
  "Built-in Agents": "Agents integrats",
  "Extension Agents": "Agents d'extensi\xF3",
  "Using: {{count}} agents": "En \xFAs: {{count}} agents",
  "View Agent": "Veure agent",
  "Edit Agent": "Editar agent",
  "Delete Agent": "Eliminar agent",
  Back: "Enrere",
  "No agent selected": "Cap agent seleccionat",
  "File Path: ": "Cam\xED del fitxer: ",
  "Tools: ": "Eines: ",
  "Color: ": "Color: ",
  "Description:": "Descripci\xF3:",
  "System Prompt:": "Missatge del sistema:",
  "Open in editor": "Obrir a l'editor",
  "Edit tools": "Editar eines",
  "Edit color": "Editar color",
  "\u274C Error:": "\u274C Error:",
  'Are you sure you want to delete agent "{{name}}"?': `Esteu segur que voleu eliminar l'agent "{{name}}"?`,
  // ============================================================================
  // Agents - Assistent de creació
  // ============================================================================
  "Project Level (.qwen/agents/)": "Nivell de projecte (.qwen/agents/)",
  "User Level (~/.qwen/agents/)": "Nivell d'usuari (~/.qwen/agents/)",
  "\u2705 Subagent Created Successfully!": "\u2705 Subagent creat correctament!",
  'Subagent "{{name}}" has been saved to {{level}} level.': `El subagent "{{name}}" s'ha desat al nivell {{level}}.`,
  "Name: ": "Nom: ",
  "Location: ": "Ubicaci\xF3: ",
  "\u274C Error saving subagent:": "\u274C Error en desar el subagent:",
  "Warnings:": "Advert\xE8ncies:",
  'Name "{{name}}" already exists at {{level}} level - will overwrite existing subagent': 'El nom "{{name}}" ja existeix al nivell {{level}} - sobreescriur\xE0 el subagent existent',
  'Name "{{name}}" exists at user level - project level will take precedence': `El nom "{{name}}" existeix al nivell d'usuari - el nivell de projecte tindr\xE0 prioritat`,
  'Name "{{name}}" exists at project level - existing subagent will take precedence': 'El nom "{{name}}" existeix al nivell de projecte - el subagent existent tindr\xE0 prioritat',
  "Description is over {{length}} characters": "La descripci\xF3 supera els {{length}} car\xE0cters",
  "System prompt is over {{length}} characters": "El missatge del sistema supera els {{length}} car\xE0cters",
  "Step {{n}}: Choose Location": "Pas {{n}}: Triar ubicaci\xF3",
  "Step {{n}}: Choose Generation Method": "Pas {{n}}: Triar m\xE8tode de generaci\xF3",
  "Generate with Qwen Code (Recommended)": "Generar amb Qwen Code (Recomanat)",
  "Manual Creation": "Creaci\xF3 manual",
  "Describe what this subagent should do and when it should be used. (Be comprehensive for best results)": "Descriviu qu\xE8 ha de fer aquest subagent i quan s'ha d'usar. (Sigueu exhaustiu per obtenir els millors resultats)",
  "e.g., Expert code reviewer that reviews code based on best practices...": "p. ex., Revisor de codi expert que revisa el codi seguint les millors pr\xE0ctiques...",
  "Generating subagent configuration...": "Generant la configuraci\xF3 del subagent...",
  "Failed to generate subagent: {{error}}": "Error en generar el subagent: {{error}}",
  "Step {{n}}: Describe Your Subagent": "Pas {{n}}: Descriure el vostre subagent",
  "Step {{n}}: Enter Subagent Name": "Pas {{n}}: Introduir el nom del subagent",
  "Step {{n}}: Enter System Prompt": "Pas {{n}}: Introduir el missatge del sistema",
  "Step {{n}}: Enter Description": "Pas {{n}}: Introduir la descripci\xF3",
  "Step {{n}}: Select Tools": "Pas {{n}}: Seleccionar eines",
  "All Tools (Default)": "Totes les eines (per defecte)",
  "All Tools": "Totes les eines",
  "Read-only Tools": "Eines de nom\xE9s lectura",
  "Read & Edit Tools": "Eines de lectura i edici\xF3",
  "Read & Edit & Execution Tools": "Eines de lectura, edici\xF3 i execuci\xF3",
  "All tools selected, including MCP tools": "Totes les eines seleccionades, inclosos MCP tools",
  "Selected tools:": "Eines seleccionades:",
  "Read-only tools:": "Eines de nom\xE9s lectura:",
  "Edit tools:": "Eines d'edici\xF3:",
  "Execution tools:": "Eines d'execuci\xF3:",
  "Step {{n}}: Choose Background Color": "Pas {{n}}: Triar el color de fons",
  "Step {{n}}: Confirm and Save": "Pas {{n}}: Confirmar i desar",
  "Esc to cancel": "Esc per cancel\xB7lar",
  "Press Enter to save, e to save and edit, Esc to go back": "Premeu Enter per desar, e per desar i editar, Esc per tornar enrere",
  "Press Enter to continue, {{navigation}}Esc to {{action}}": "Premeu Enter per continuar, {{navigation}}Esc per {{action}}",
  cancel: "cancel\xB7lar",
  "go back": "tornar enrere",
  "\u2191\u2193 to navigate, ": "\u2191\u2193 per navegar, ",
  "Enter a clear, unique name for this subagent.": "Introdu\xEFu un nom clar i \xFAnic per a aquest subagent.",
  "e.g., Code Reviewer": "p. ex., Revisor de codi",
  "Name cannot be empty.": "El nom no pot estar buit.",
  "Write the system prompt that defines this subagent's behavior. Be comprehensive for best results.": "Escriviu el missatge del sistema que defineix el comportament d'aquest subagent. Sigueu exhaustiu per obtenir els millors resultats.",
  "e.g., You are an expert code reviewer...": "p. ex., Sou un revisor de codi expert...",
  "System prompt cannot be empty.": "El missatge del sistema no pot estar buit.",
  "Describe when and how this subagent should be used.": "Descriviu quan i com s'ha d'usar aquest subagent.",
  "e.g., Reviews code for best practices and potential bugs.": "p. ex., Revisa el codi seguint les millors pr\xE0ctiques i detectant errors potencials.",
  "Description cannot be empty.": "La descripci\xF3 no pot estar buida.",
  "Failed to launch editor: {{error}}": "Error en iniciar l'editor: {{error}}",
  "Failed to save and edit subagent: {{error}}": "Error en desar i editar el subagent: {{error}}",
  // ============================================================================
  // Extensions - Diàleg de gestió
  // ============================================================================
  "Manage Extensions": "Gestionar extensions",
  "Extension Details": "Detalls de l'extensi\xF3",
  "View Extension": "Veure l'extensi\xF3",
  "Update Extension": "Actualitzar l'extensi\xF3",
  "Disable Extension": "Desactivar l'extensi\xF3",
  "Enable Extension": "Activar l'extensi\xF3",
  "Uninstall Extension": "Desinstal\xB7lar l'extensi\xF3",
  "Select Scope": "Seleccionar l'\xE0mbit",
  "User Scope": "\xC0mbit d'usuari",
  "Workspace Scope": "\xC0mbit de l'espai de treball",
  "No extensions found.": "No s'han trobat extensions.",
  "Updating...": "Actualitzant...",
  Unknown: "Desconegut",
  Error: "Error",
  "Stopped because": "Aturat perqu\xE8",
  "Version:": "Versi\xF3:",
  "Status:": "Estat:",
  'Are you sure you want to uninstall extension "{{name}}"?': `Esteu segur que voleu desinstal\xB7lar l'extensi\xF3 "{{name}}"?`,
  "This action cannot be undone.": "Aquesta acci\xF3 no es pot desfer.",
  'Extension "{{name}}" updated successfully.': `L'extensi\xF3 "{{name}}" s'ha actualitzat correctament.`,
  "Name:": "Nom:",
  "MCP Servers:": "MCP Servers:",
  "Settings:": "Configuraci\xF3:",
  active: "activa",
  disabled: "desactivada",
  enabled: "activada",
  "View Details": "Veure detalls",
  "Update failed:": "Error en l'actualitzaci\xF3:",
  "Updating {{name}}...": "Actualitzant {{name}}...",
  "Update complete!": "Actualitzaci\xF3 completada!",
  "User (global)": "Usuari (global)",
  "Workspace (project-specific)": "Espai de treball (espec\xEDfic del projecte)",
  'Disable "{{name}}" - Select Scope': 'Desactivar "{{name}}" - Seleccionar \xE0mbit',
  'Enable "{{name}}" - Select Scope': 'Activar "{{name}}" - Seleccionar \xE0mbit',
  "No extension selected": "Cap extensi\xF3 seleccionada",
  "{{count}} extensions installed": "{{count}} extensions instal\xB7lades",
  "Use '/extensions install' to install your first extension.": "Useu '/extensions install' per instal\xB7lar la vostra primera extensi\xF3.",
  "up to date": "al dia",
  "update available": "actualitzaci\xF3 disponible",
  "checking...": "comprovant...",
  "not updatable": "no actualitzable",
  error: "error",
  // ============================================================================
  // Ordres - General (continuació)
  // ============================================================================
  "View and edit Qwen Code settings": "Veure i editar la configuraci\xF3 de Qwen Code",
  Settings: "Configuraci\xF3",
  "To see changes, Qwen Code must be restarted. Press r to exit and apply changes now.": "Per veure els canvis, cal reiniciar Qwen Code. Premeu r per sortir i aplicar els canvis ara.",
  // ============================================================================
  // Etiquetes de configuració
  // ============================================================================
  "Vim Mode": "Mode Vim",
  "Attribution: commit": "Atribuci\xF3: commit",
  "Terminal Bell Notification": "Notificaci\xF3 de campana del terminal",
  "Enable Usage Statistics": "Activar estad\xEDstiques d'\xFAs",
  Theme: "Tema",
  "Preferred Editor": "Editor preferit",
  "Auto-connect to IDE": "Connexi\xF3 autom\xE0tica a IDE",
  "Debug Keystroke Logging": "Registre de tecles per a depuraci\xF3",
  "Language: UI": "Idioma: Interf\xEDcie",
  "Language: Model": "Idioma: Model",
  "Output Format": "Format de sortida",
  "Hide Window Title": "Amagar el t\xEDtol de la finestra",
  "Show Status in Title": "Mostrar l'estat al t\xEDtol",
  "Hide Tips": "Amagar consells",
  "Show Line Numbers in Code": "Mostrar n\xFAmeros de l\xEDnia al codi",
  "Show Citations": "Mostrar cites",
  "Custom Witty Phrases": "Frases enginyoses personalitzades",
  "Show Welcome Back Dialog": "Mostrar el di\xE0leg de benvinguda",
  "Enable User Feedback": "Activar les valoracions dels usuaris",
  "How is Qwen doing this session? (optional)": "Com va Qwen en aquesta sessi\xF3? (opcional)",
  Bad: "Malament",
  Fine: "B\xE9",
  Good: "Molt b\xE9",
  Dismiss: "Descartar",
  "Screen Reader Mode": "Mode de lector de pantalla",
  "Max Session Turns": "Torns m\xE0xims de sessi\xF3",
  "Skip Next Speaker Check": "Ometre la comprovaci\xF3 del proper parlant",
  "Skip Loop Detection": "Ometre la detecci\xF3 de bucles",
  "Skip Startup Context": "Ometre el context d'inici",
  "Enable OpenAI Logging": "Activar el registre d'OpenAI",
  "OpenAI Logging Directory": "Directori de registres d'OpenAI",
  Timeout: "Temps d'espera",
  "Max Retries": "Reintents m\xE0xims",
  "Load Memory From Include Directories": "Carregar mem\xF2ria des dels directoris inclosos",
  "Respect .gitignore": "Respectar .gitignore",
  "Respect .qwenignore": "Respectar .qwenignore",
  "Enable Recursive File Search": "Activar la cerca recursiva de fitxers",
  "Interactive Shell (PTY)": "Shell interactiva (PTY)",
  "Show Color": "Mostrar color",
  "Auto Accept": "Acceptaci\xF3 autom\xE0tica",
  "Use Ripgrep": "Usar Ripgrep",
  "Use Builtin Ripgrep": "Usar Ripgrep integrat",
  "Tool Output Truncation Threshold": "Llindar de truncament de la sortida d'eines",
  "Tool Output Truncation Lines": "L\xEDnies de truncament de la sortida d'eines",
  "Folder Trust": "Confian\xE7a de carpeta",
  "Tool Schema Compliance": "Compliment de Tool Schema",
  "Auto (detect from system)": "Autom\xE0tic (detectar del sistema)",
  "Auto (detect terminal theme)": "Autom\xE0tic (detectar el tema del terminal)",
  Auto: "Autom\xE0tic",
  Text: "Text",
  JSON: "JSON",
  Plan: "Planificaci\xF3",
  "Ask permissions": "Demanar permisos",
  "Auto Edit": "Edici\xF3 autom\xE0tica",
  YOLO: "YOLO",
  "toggle vim mode on/off": "activar/desactivar el mode Vim",
  "Show model-specific usage statistics.": "Mostrar les estad\xEDstiques d'\xFAs espec\xEDfiques del model.",
  "Show tool-specific usage statistics.": "Mostrar les estad\xEDstiques d'\xFAs espec\xEDfiques de les eines.",
  "Show daily token usage statistics.": "Mostrar les estad\xEDstiques di\xE0ries d'\xFAs de tokens.",
  "Show monthly token usage statistics.": "Mostrar les estad\xEDstiques mensuals d'\xFAs de tokens.",
  "Export token usage statistics to CSV or JSON.": "Exportar les estad\xEDstiques d'\xFAs de tokens a CSV o JSON.",
  "No usage data.": "No hi ha dades d'\xFAs.",
  "{{label}}: {{tokens}} tokens ({{requests}} requests)": "{{label}}: {{tokens}} tokens ({{requests}} sol\xB7licituds)",
  "Daily token usage for {{value}}": "\xDAs diari de tokens per a {{value}}",
  "Monthly token usage for {{value}}": "\xDAs mensual de tokens per a {{value}}",
  "Total: {{tokens}} tokens": "Total: {{tokens}} tokens",
  "Requests: {{requests}}": "Sol\xB7licituds: {{requests}}",
  "Breakdown:": "Desglossament:",
  "Input: {{tokens}}": "Entrada: {{tokens}}",
  "Output: {{tokens}}": "Sortida: {{tokens}}",
  "Cached (included in Input): {{tokens}}": "Mem\xF2ria cau (inclosa a l\u2019entrada): {{tokens}}",
  "Thoughts: {{tokens}}": "Raonament: {{tokens}}",
  "By model:": "Per model:",
  "By auth type:": "Per tipus d'autenticaci\xF3:",
  "By model/auth type:": "Per model/tipus d'autenticaci\xF3:",
  "By source:": "Per origen:",
  "Failed to load token usage stats: {{error}}": "No s'han pogut carregar les estad\xEDstiques d'\xFAs de tokens: {{error}}",
  "Expected --format csv or --format json.": "S'esperava --format csv o --format json.",
  "Expected a file path after --output.": "S'esperava una ruta de fitxer despr\xE9s de --output.",
  "Unexpected argument: {{argument}}": "Argument inesperat: {{argument}}",
  "Usage: /stats export <daily|monthly> [YYYY-MM-DD|YYYY-MM] [--format csv|json] [--output path]": "\xDAs: /stats export <daily|monthly> [YYYY-MM-DD|YYYY-MM] [--format csv|json] [--output path]",
  "Token usage export path must be within the project working directory.": "La ruta d'exportaci\xF3 de l'\xFAs de tokens ha d'estar dins del directori de treball del projecte.",
  "Export target does not exist: {{path}}": "La destinaci\xF3 d'exportaci\xF3 no existeix: {{path}}",
  "Cannot resolve export path within the working directory.": "No s'ha pogut resoldre la ruta d'exportaci\xF3 dins del directori de treball.",
  "Could not create a temporary export file.": "No s'ha pogut crear un fitxer temporal d'exportaci\xF3.",
  "Token usage exported to {{format}}: {{path}}": "\xDAs de tokens exportat a {{format}}: {{path}}",
  "Failed to export token usage stats: {{error}}": "No s'han pogut exportar les estad\xEDstiques d'\xFAs de tokens: {{error}}",
  "Unclosed quote in arguments.": "Cometes sense tancar als arguments.",
  "Note: generation timing (TTFT/TPS) belongs to generation metrics.": "Nota: el temps de generaci\xF3 (TTFT/TPS) pertany a les m\xE8triques de generaci\xF3.",
  "exit the cli": "sortir del CLI",
  "Manage workspace directories": "Gestionar els directoris de l'espai de treball",
  "Add directories to the workspace. Use comma to separate multiple paths": "Afegir directoris a l'espai de treball. Useu comes per separar m\xFAltiples camins",
  "Show all directories in the workspace": "Mostrar tots els directoris de l'espai de treball",
  "set external editor preference": "establir la prefer\xE8ncia d'editor extern",
  "Select Editor": "Seleccionar editor",
  "Editor Preference": "Prefer\xE8ncia d'editor",
  "These editors are currently supported. Please note that some editors cannot be used in sandbox mode.": "Aquests editors estan suportats. Cal tenir en compte que alguns editors no es poden usar en mode a\xEFllat.",
  "Your preferred editor is:": "El vostre editor preferit \xE9s:",
  "Manage extensions": "Gestionar extensions",
  "Manage installed extensions": "Gestionar les extensions instal\xB7lades",
  "Disable an extension": "Desactivar una extensi\xF3",
  "Enable an extension": "Activar una extensi\xF3",
  "Install an extension from a git repo or local path": "Instal\xB7lar una extensi\xF3 des d'un repositori git o cam\xED local",
  "Uninstall an extension": "Desinstal\xB7lar una extensi\xF3",
  "No extensions installed.": "No hi ha extensions instal\xB7lades.",
  'Extension "{{name}}" not found.': 'Extensi\xF3 "{{name}}" no trobada.',
  "No extensions to update.": "No hi ha extensions per actualitzar.",
  "Usage: /extensions install <source>": "\xDAs: /extensions install <font>",
  'Installing extension from "{{source}}"...': 'Instal\xB7lant extensi\xF3 des de "{{source}}"...',
  'Extension "{{name}}" installed successfully.': `L'extensi\xF3 "{{name}}" s'ha instal\xB7lat correctament.`,
  'Failed to install extension from "{{source}}": {{error}}': `Error en instal\xB7lar l'extensi\xF3 des de "{{source}}": {{error}}`,
  "Do you want to continue? [Y/n]: ": "Voleu continuar? [S/n]: ",
  "Do you want to continue?": "Voleu continuar?",
  'Installing extension "{{name}}".': `Instal\xB7lant l'extensi\xF3 "{{name}}".`,
  "**Extensions may introduce unexpected behavior. Ensure you have investigated the extension source and trust the author.**": "**Les extensions poden introduir comportaments inesperats. Assegureu-vos d'haver investigat la font de l'extensi\xF3 i de confiar en l'autor.**",
  "This extension will run the following MCP servers:": "Aquesta extensi\xF3 executar\xE0 els seg\xFCents MCP servers:",
  local: "local",
  remote: "remot",
  "This extension will add the following commands: {{commands}}.": "Aquesta extensi\xF3 afegir\xE0 les ordres seg\xFCents: {{commands}}.",
  "This extension will append info to your QWEN.md context using {{fileName}}": "Aquesta extensi\xF3 afegir\xE0 informaci\xF3 al vostre context QWEN.md usant {{fileName}}",
  "This extension will install the following skills:": "Aquesta extensi\xF3 instal\xB7lar\xE0 les habilitats seg\xFCents:",
  "This extension will install the following subagents:": "Aquesta extensi\xF3 instal\xB7lar\xE0 els subagents seg\xFCents:",
  'Installation cancelled for "{{name}}".': 'Instal\xB7laci\xF3 cancel\xB7lada per a "{{name}}".',
  "You are installing an extension from {{originSource}}. Some features may not work perfectly with Qwen Code.": "Esteu instal\xB7lant una extensi\xF3 des de {{originSource}}. Algunes funcions poden no funcionar perfectament amb Qwen Code.",
  "--ref and --auto-update are not applicable for marketplace extensions.": "--ref i --auto-update no s'apliquen a les extensions del mercat.",
  'Extension "{{name}}" installed successfully and enabled.': `L'extensi\xF3 "{{name}}" s'ha instal\xB7lat i activat correctament.`,
  "The github URL, local path, or marketplace source (marketplace-url:plugin-name) of the extension to install.": "La URL de GitHub, el cam\xED local o la font del mercat (marketplace-url:nom-del-connector) de l'extensi\xF3 a instal\xB7lar.",
  "The git ref to install from.": "La refer\xE8ncia git des de la qual instal\xB7lar.",
  "Enable auto-update for this extension.": "Activar l'actualitzaci\xF3 autom\xE0tica per a aquesta extensi\xF3.",
  "Enable pre-release versions for this extension.": "Activar les versions preliminars per a aquesta extensi\xF3.",
  "Acknowledge the security risks of installing an extension and skip the confirmation prompt.": "Acceptar els riscos de seguretat d'instal\xB7lar una extensi\xF3 i ometre el missatge de confirmaci\xF3.",
  "The source argument must be provided.": "Cal proporcionar l'argument font.",
  'Extension "{{name}}" successfully uninstalled.': `L'extensi\xF3 "{{name}}" s'ha desinstal\xB7lat correctament.`,
  "Uninstalls an extension.": "Desinstal\xB7la una extensi\xF3.",
  "The name or source path of the extension to uninstall.": "El nom o cam\xED font de l'extensi\xF3 a desinstal\xB7lar.",
  "Please include the name of the extension to uninstall as a positional argument.": "Incloeu el nom de l'extensi\xF3 a desinstal\xB7lar com a argument posicional.",
  "Enables an extension.": "Activa una extensi\xF3.",
  "The name of the extension to enable.": "El nom de l'extensi\xF3 a activar.",
  "The scope to enable the extenison in. If not set, will be enabled in all scopes.": "L'\xE0mbit en el qual activar l'extensi\xF3. Si no s'estableix, s'activar\xE0 en tots els \xE0mbits.",
  'Extension "{{name}}" successfully enabled for scope "{{scope}}".': `L'extensi\xF3 "{{name}}" s'ha activat correctament per a l'\xE0mbit "{{scope}}".`,
  'Extension "{{name}}" successfully enabled in all scopes.': `L'extensi\xF3 "{{name}}" s'ha activat correctament en tots els \xE0mbits.`,
  "Invalid scope: {{scope}}. Please use one of {{scopes}}.": "\xC0mbit no v\xE0lid: {{scope}}. Useu un dels seg\xFCents: {{scopes}}.",
  "Disables an extension.": "Desactiva una extensi\xF3.",
  "The name of the extension to disable.": "El nom de l'extensi\xF3 a desactivar.",
  "The scope to disable the extenison in.": "L'\xE0mbit en el qual desactivar l'extensi\xF3.",
  'Extension "{{name}}" successfully disabled for scope "{{scope}}".': `L'extensi\xF3 "{{name}}" s'ha desactivat correctament per a l'\xE0mbit "{{scope}}".`,
  'Extension "{{name}}" successfully updated: {{oldVersion}} \u2192 {{newVersion}}.': `L'extensi\xF3 "{{name}}" s'ha actualitzat correctament: {{oldVersion}} \u2192 {{newVersion}}.`,
  'Unable to install extension "{{name}}" due to missing install metadata': `No es pot instal\xB7lar l'extensi\xF3 "{{name}}" per manca de metadades d'instal\xB7laci\xF3`,
  'Extension "{{name}}" is already up to date.': `L'extensi\xF3 "{{name}}" ja \xE9s al dia.`,
  "Updates all extensions or a named extension to the latest version.": "Actualitza totes les extensions o una extensi\xF3 espec\xEDfica a l'\xFAltima versi\xF3.",
  "Update all extensions.": "Actualitzar totes les extensions.",
  "The name of the extension to update.": "El nom de l'extensi\xF3 a actualitzar.",
  "Either an extension name or --all must be provided": "Cal proporcionar un nom d'extensi\xF3 o --all",
  "Lists installed extensions.": "Llista les extensions instal\xB7lades.",
  "Path:": "Cam\xED:",
  "Source:": "Font:",
  "Type:": "Tipus:",
  "Release tag:": "Etiqueta de versi\xF3:",
  "Enabled (User):": "Activada (Usuari):",
  "Enabled (Workspace):": "Activada (Espai de treball):",
  "Context files:": "Fitxers de context:",
  "Skills:": "Habilitats:",
  "Agents:": "Agents:",
  "MCP servers:": "MCP servers:",
  "Link extension failed to install.": "No s'ha pogut instal\xB7lar l'extensi\xF3 d'enlla\xE7.",
  'Extension "{{name}}" linked successfully and enabled.': `L'extensi\xF3 "{{name}}" s'ha enlla\xE7at i activat correctament.`,
  "Links an extension from a local path. Updates made to the local path will always be reflected.": "Enlla\xE7a una extensi\xF3 des d'un cam\xED local. Els canvis al cam\xED local sempre es reflectiran.",
  "The name of the extension to link.": "El nom de l'extensi\xF3 a enlla\xE7ar.",
  "Set a specific setting for an extension.": "Establir una configuraci\xF3 espec\xEDfica per a una extensi\xF3.",
  "Name of the extension to configure.": "Nom de l'extensi\xF3 a configurar.",
  "The setting to configure (name or env var).": "La configuraci\xF3 a establir (nom o variable d'entorn).",
  "The scope to set the setting in.": "L'\xE0mbit on establir la configuraci\xF3.",
  "List all settings for an extension.": "Llistar tota la configuraci\xF3 d'una extensi\xF3.",
  "Name of the extension.": "Nom de l'extensi\xF3.",
  'Extension "{{name}}" has no settings to configure.': `L'extensi\xF3 "{{name}}" no t\xE9 cap configuraci\xF3.`,
  'Settings for "{{name}}":': 'Configuraci\xF3 per a "{{name}}":',
  "(workspace)": "(espai de treball)",
  "(user)": "(usuari)",
  "[not set]": "[no establert]",
  "[value stored in keychain]": "[valor emmagatzemat al clauer]",
  "Value:": "Valor:",
  "Manage extension settings.": "Gestionar la configuraci\xF3 de les extensions.",
  "You need to specify a command (set or list).": "Cal especificar una ordre (set o list).",
  // ============================================================================
  // Selecció de connector / Mercat
  // ============================================================================
  "No plugins available in this marketplace.": "No hi ha connectors disponibles en aquest mercat.",
  'Select a plugin to install from marketplace "{{name}}":': 'Seleccioneu un connector per instal\xB7lar des del mercat "{{name}}":',
  "Plugin selection cancelled.": "Selecci\xF3 de connector cancel\xB7lada.",
  'Select a plugin from "{{name}}"': 'Seleccionar un connector de "{{name}}"',
  "Use \u2191\u2193 or j/k to navigate, Enter to select, Escape to cancel": "Useu \u2191\u2193 o j/k per navegar, Enter per seleccionar, Escape per cancel\xB7lar",
  "{{count}} more above": "{{count}} m\xE9s amunt",
  "{{count}} more below": "{{count}} m\xE9s avall",
  "manage IDE integration": "gestionar la integraci\xF3 de l'IDE",
  "check status of IDE integration": "comprovar l'estat de la integraci\xF3 de l'IDE",
  "install required IDE companion for {{ideName}}": "instal\xB7lar el complement IDE necessari per a {{ideName}}",
  "enable IDE integration": "activar la integraci\xF3 de l'IDE",
  "disable IDE integration": "desactivar la integraci\xF3 de l'IDE",
  "IDE integration is not supported in your current environment. To use this feature, run Qwen Code in one of these supported IDEs: VS Code or VS Code forks.": "La integraci\xF3 de l'IDE no \xE9s compatible en el vostre entorn actual. Per usar aquesta funci\xF3, executeu Qwen Code en un dels IDEs compatibles: VS Code o bifurcacions de VS Code.",
  "Set up GitHub Actions": "Configurar GitHub Actions",
  "Configure terminal keybindings for multiline input (VS Code, Cursor, Windsurf, Trae)": "Configurar les dreceres del terminal per a entrada multil\xEDnia (VS Code, Cursor, Windsurf, Trae)",
  "Please restart your terminal for the changes to take effect.": "Reinicieu el terminal perqu\xE8 els canvis tinguin efecte.",
  "Failed to configure terminal: {{error}}": "Error en configurar el terminal: {{error}}",
  "Could not determine {{terminalName}} config path on Windows: APPDATA environment variable is not set.": "No s'ha pogut determinar el cam\xED de configuraci\xF3 de {{terminalName}} a Windows: la variable d'entorn APPDATA no est\xE0 establerta.",
  "{{terminalName}} keybindings.json exists but is not a valid JSON array. Please fix the file manually or delete it to allow automatic configuration.": "{{terminalName}} keybindings.json existeix per\xF2 no \xE9s un array JSON v\xE0lid. Corregiu el fitxer manualment o elimineu-lo per permetre la configuraci\xF3 autom\xE0tica.",
  "File: {{file}}": "Fitxer: {{file}}",
  "Failed to parse {{terminalName}} keybindings.json. The file contains invalid JSON. Please fix the file manually or delete it to allow automatic configuration.": "Error en analitzar {{terminalName}} keybindings.json. El fitxer cont\xE9 JSON no v\xE0lid. Corregiu el fitxer manualment o elimineu-lo per permetre la configuraci\xF3 autom\xE0tica.",
  "Error: {{error}}": "Error: {{error}}",
  "Shift+Enter binding already exists": "La drecera Shift+Enter ja existeix",
  "Ctrl+Enter binding already exists": "La drecera Ctrl+Enter ja existeix",
  "Existing keybindings detected. Will not modify to avoid conflicts.": "S'han detectat dreceres existents. No es modificaran per evitar conflictes.",
  "Please check and modify manually if needed: {{file}}": "Comproveu i modifiqueu manualment si cal: {{file}}",
  "Added Shift+Enter and Ctrl+Enter keybindings to {{terminalName}}.": "S'han afegit les dreceres Shift+Enter i Ctrl+Enter a {{terminalName}}.",
  "Modified: {{file}}": "Modificat: {{file}}",
  "{{terminalName}} keybindings already configured.": "Les dreceres de {{terminalName}} ja estan configurades.",
  "Failed to configure {{terminalName}}.": "Error en configurar {{terminalName}}.",
  "Your terminal is already configured for an optimal experience with multiline input (Shift+Enter and Ctrl+Enter).": "El vostre terminal ja est\xE0 configurat per a una experi\xE8ncia \xF2ptima amb entrada multil\xEDnia (Shift+Enter i Ctrl+Enter).",
  // ============================================================================
  // Ordres - Hooks
  // ============================================================================
  "Manage Qwen Code hooks": "Gestionar els hooks de Qwen Code",
  "List all configured hooks": "Llistar tots els hooks configurats",
  Hooks: "Hooks",
  "Loading hooks...": "Carregant hooks...",
  "Error loading hooks:": "Error en carregar els hooks:",
  "Press Escape to close": "Premeu Escape per tancar",
  "Press Escape, Ctrl+C, or Ctrl+D to cancel": "Premeu Escape, Ctrl+C o Ctrl+D per cancel\xB7lar",
  "Press Space, Enter, or Escape to dismiss": "Premeu Space, Enter o Escape per descartar",
  "No hook selected": "Cap hook seleccionat",
  "No hook events found.": "No s'han trobat esdeveniments de hook.",
  "{{count}} hook configured": "{{count}} hook configurat",
  "{{count}} hooks configured": "{{count}} hooks configurats",
  "This menu is read-only. To add or modify hooks, edit settings.json directly or ask Qwen Code.": "Aquest men\xFA \xE9s de nom\xE9s lectura. Per afegir o modificar hooks, editeu settings.json directament o demaneu-ho a Qwen Code.",
  "Enter to select \xB7 Esc to cancel": "Enter per seleccionar \xB7 Esc per cancel\xB7lar",
  "Exit codes:": "Codis de sortida:",
  "Configured hooks:": "Hooks configurats:",
  "No hooks configured for this event.": "No hi ha hooks configurats per a aquest esdeveniment.",
  "To add hooks, edit settings.json directly or ask Qwen.": "Per afegir hooks, editeu settings.json directament o demaneu-ho a Qwen.",
  "Enter to select \xB7 Esc to go back": "Enter per seleccionar \xB7 Esc per tornar enrere",
  "Hook details": "Detalls del hook",
  "Event:": "Esdeveniment:",
  "Extension:": "Extensi\xF3:",
  "No hook config selected": "Cap configuraci\xF3 de hook seleccionada",
  "To modify or remove this hook, edit settings.json directly or ask Qwen to help.": "Per modificar o eliminar aquest hook, editeu settings.json directament o demaneu ajuda a Qwen.",
  "Hook Configuration - Disabled": "Configuraci\xF3 de hooks - Desactivats",
  "All hooks are currently disabled. You have {{count}} that are not running.": "Tots els hooks estan desactivats. En teniu {{count}} que no estan en execuci\xF3.",
  "{{count}} configured hook": "{{count}} hook configurat",
  "{{count}} configured hooks": "{{count}} hooks configurats",
  "When hooks are disabled:": "Quan els hooks estan desactivats:",
  "No hook commands will execute": "Cap ordre de hook s'executar\xE0",
  "StatusLine will not be displayed": "La barra d'estat no es mostrar\xE0",
  "Tool operations will proceed without hook validation": "Les operacions d'eines continuaran sense validaci\xF3 de hook",
  'To re-enable hooks, remove "disableAllHooks" from settings.json or ask Qwen Code.': 'Per tornar a activar els hooks, elimineu "disableAllHooks" de settings.json o demaneu-ho a Qwen Code.',
  Project: "Projecte",
  User: "Usuari",
  Skill: "Habilitat",
  System: "Sistema",
  Extension: "Extensi\xF3",
  "Local Settings": "Configuraci\xF3 local",
  "User Settings": "Configuraci\xF3 d'usuari",
  "System Settings": "Configuraci\xF3 del sistema",
  Extensions: "Extensions",
  "Session (temporary)": "Sessi\xF3 (temporal)",
  "Before tool execution": "Abans de l'execuci\xF3 de l'eina",
  "After tool execution": "Despr\xE9s de l'execuci\xF3 de l'eina",
  "After tool execution fails": "Quan falla l'execuci\xF3 de l'eina",
  "When notifications are sent": "Quan s'envien notificacions",
  "When the user submits a prompt": "Quan l'usuari envia un missatge",
  "When a slash command expands into a prompt": "Quan una ordre de barra s'expandeix en un missatge",
  "When a new session is started": "Quan s'inicia una nova sessi\xF3",
  "Right before Qwen Code concludes its response": "Immediatament abans que Qwen Code conclou la seva resposta",
  "When a subagent (Agent tool call) is started": "Quan s'inicia un subagent (crida a l'eina Agent)",
  "Right before a subagent concludes its response": "Immediatament abans que un subagent conclou la seva resposta",
  "Before conversation compaction": "Abans de la compactaci\xF3 de la conversa",
  "When a session is ending": "Quan una sessi\xF3 s'est\xE0 acabant",
  "When a permission dialog is displayed": "Quan es mostra un di\xE0leg de permisos",
  "Input to command is JSON of tool call arguments.": "L'entrada a l'ordre \xE9s JSON dels arguments de la crida a l'eina.",
  'Input to command is JSON with fields "inputs" (tool call arguments) and "response" (tool call response).': `L'entrada a l'ordre \xE9s JSON amb els camps "inputs" (arguments de la crida a l'eina) i "response" (resposta de la crida a l'eina).`,
  "Input to command is JSON with tool_name, tool_input, tool_use_id, error, error_type, is_interrupt, and is_timeout.": "L'entrada a l'ordre \xE9s JSON amb tool_name, tool_input, tool_use_id, error, error_type, is_interrupt i is_timeout.",
  "Input to command is JSON with notification message and type.": "L'entrada a l'ordre \xE9s JSON amb el missatge de notificaci\xF3 i el tipus.",
  "Input to command is JSON with original user prompt text.": "L'entrada a l'ordre \xE9s JSON amb el text original del missatge de l'usuari.",
  "Input to command is JSON with command_name, command_args, and expanded prompt text.": "L'entrada a l'ordre \xE9s JSON amb command_name, command_args i el text del missatge expandit.",
  "Input to command is JSON with session start source.": "L'entrada a l'ordre \xE9s JSON amb la font d'inici de sessi\xF3.",
  "Input to command is JSON with session end reason.": "L'entrada a l'ordre \xE9s JSON amb el motiu de fi de sessi\xF3.",
  "Input to command is JSON with agent_id and agent_type.": "L'entrada a l'ordre \xE9s JSON amb agent_id i agent_type.",
  "Input to command is JSON with agent_id, agent_type, and agent_transcript_path.": "L'entrada a l'ordre \xE9s JSON amb agent_id, agent_type i agent_transcript_path.",
  "Input to command is JSON with compaction details.": "L'entrada a l'ordre \xE9s JSON amb els detalls de compactaci\xF3.",
  "Input to command is JSON with tool_name, tool_input, and tool_use_id. Output JSON with hookSpecificOutput containing decision to allow or deny.": "L'entrada a l'ordre \xE9s JSON amb tool_name, tool_input i tool_use_id. La sortida JSON amb hookSpecificOutput cont\xE9 la decisi\xF3 de permetre o denegar.",
  "stdout/stderr not shown": "stdout/stderr no es mostra",
  "show stderr to model and continue conversation": "mostrar stderr al model i continuar la conversa",
  "show stderr to user only": "mostrar stderr nom\xE9s a l'usuari",
  "stdout shown in transcript mode (ctrl+o)": "stdout mostrat en mode transcripci\xF3 (ctrl+o)",
  "show stderr to model immediately": "mostrar stderr al model immediatament",
  "show stderr to user only but continue with tool call": "mostrar stderr nom\xE9s a l'usuari per\xF2 continuar amb la crida a l'eina",
  "block processing, erase original prompt, and show stderr to user only": "blocar el processament, esborrar el missatge original i mostrar stderr nom\xE9s a l'usuari",
  "block expanded prompt submission and show stderr to user only": "blocar l'enviament del missatge expandit i mostrar stderr nom\xE9s a l'usuari",
  "stdout shown to Qwen": "stdout mostrat a Qwen",
  "show stderr to user only (blocking errors ignored)": "mostrar stderr nom\xE9s a l'usuari (errors de bloqueig ignorats)",
  "command completes successfully": "l'ordre es completa correctament",
  "stdout shown to subagent": "stdout mostrat al subagent",
  "show stderr to subagent and continue having it run": "mostrar stderr al subagent i continuar la seva execuci\xF3",
  "stdout appended as custom compact instructions": "stdout afegit com a instruccions compactes personalitzades",
  "block compaction": "blocar la compactaci\xF3",
  "show stderr to user only but continue with compaction": "mostrar stderr nom\xE9s a l'usuari per\xF2 continuar amb la compactaci\xF3",
  "use hook decision if provided": "usar la decisi\xF3 del hook si es proporciona",
  "Config not loaded.": "Configuraci\xF3 no carregada.",
  "Hooks are not enabled. Enable hooks in settings to use this feature.": "Els hooks no estan activats. Activeu els hooks a la configuraci\xF3 per usar aquesta funci\xF3.",
  // ============================================================================
  // Ordres - Exportació de sessió
  // ============================================================================
  "Export current session message history to a file": "Exportar l'historial de missatges de la sessi\xF3 actual a un fitxer",
  "Export session to HTML format": "Exportar la sessi\xF3 en format HTML",
  "Export session to JSON format": "Exportar la sessi\xF3 en format JSON",
  "Export session to JSONL format (one message per line)": "Exportar la sessi\xF3 en format JSONL (un missatge per l\xEDnia)",
  "Export session to markdown format": "Exportar la sessi\xF3 en format markdown",
  // ============================================================================
  // Ordres - Idees
  // ============================================================================
  "generate personalized programming insights from your chat history": "generar idees de programaci\xF3 personalitzades a partir del vostre historial de xat",
  // ============================================================================
  // Ordres - Historial de sessió
  // ============================================================================
  "Resume a previous session": "Reprendre una sessi\xF3 anterior",
  "Fork the current conversation into a new session": "Bifurca la conversa actual en una sessi\xF3 nova",
  "Spawn a background agent that inherits the full conversation": "Inicia un agent en segon pla que hereta tota la conversa",
  "Please provide a directive. Usage: /fork <directive>": "Proporcioneu una directiva. \xDAs: /fork <directiva>",
  "Cannot fork while a response or tool call is in progress. Wait for it to finish or resolve the pending tool call.": "No es pot crear una bifurcaci\xF3 mentre hi ha una resposta o una crida a una eina en curs. Espereu que acabi o resolgueu la crida a l'eina pendent.",
  "Cannot fork before the first conversation turn.": "No es pot crear una bifurcaci\xF3 abans del primer torn de conversa.",
  "The agent tool is unavailable; cannot fork.": "L'eina d'agent no est\xE0 disponible; no es pot crear una bifurcaci\xF3.",
  "Failed to launch fork: {{error}}": "No s\u2019ha pogut iniciar la bifurcaci\xF3: {{error}}",
  "User launched a background fork via /fork: {{directive}}": "L'usuari ha iniciat una bifurcaci\xF3 en segon pla amb /fork: {{directive}}",
  "Forked into a background agent. It inherits this conversation and runs without blocking \u2014 track it in the background tasks panel; it reports back when done.": "S'ha bifurcat a un agent en segon pla. Hereta aquesta conversa i s'executa sense bloquejar \u2014 feu-ne el seguiment al tauler de tasques en segon pla; informar\xE0 quan acabi.",
  "Cannot branch while a response or tool call is in progress. Wait for it to finish or resolve the pending tool call.": "No es pot bifurcar mentre hi ha una resposta o una crida a una eina en curs. Espereu que acabi o resolgueu la crida a l'eina pendent.",
  "No conversation to branch.": "No hi ha cap conversa per bifurcar.",
  "Restore a tool call. This will reset the conversation and file history to the state it was in when the tool call was suggested": "Restaurar una crida a una eina. Aix\xF2 restablir\xE0 la conversa i l'historial de fitxers a l'estat en qu\xE8 es trobaven quan es va suggerir la crida a l'eina",
  "Could not detect terminal type. Supported terminals: VS Code, Cursor, Windsurf, and Trae.": "No s'ha pogut detectar el tipus de terminal. Terminals compatibles: VS Code, Cursor, Windsurf i Trae.",
  'Terminal "{{terminal}}" is not supported yet.': 'El terminal "{{terminal}}" no \xE9s compatible encara.',
  // ============================================================================
  // Ordres - Idioma
  // ============================================================================
  "Invalid language. Available: {{options}}": "Idioma no v\xE0lid. Disponibles: {{options}}",
  "Language subcommands do not accept additional arguments.": "Les subordres d'idioma no accepten arguments addicionals.",
  "Current UI language: {{lang}}": "Idioma actual de la interf\xEDcie: {{lang}}",
  "Current LLM output language: {{lang}}": "Idioma actual de la sortida del model: {{lang}}",
  "Set UI language": "Establir l'idioma de la interf\xEDcie",
  "Set LLM output language": "Establir l'idioma de sortida del model",
  "Usage: /language ui [{{options}}]": "\xDAs: /language ui [{{options}}]",
  "Usage: /language output <language>": "\xDAs: /language output <idioma>",
  "Example: /language output \u4E2D\u6587": "Exemple: /language output \u4E2D\u6587",
  "Example: /language output English": "Exemple: /language output English",
  "Example: /language output \u65E5\u672C\u8A9E": "Exemple: /language output \u65E5\u672C\u8A9E",
  "UI language changed to {{lang}}": "Idioma de la interf\xEDcie canviat a {{lang}}",
  "LLM output language set to {{lang}}": "Idioma de sortida del model establert a {{lang}}",
  "Please restart the application for the changes to take effect.": "Reinicieu l'aplicaci\xF3 perqu\xE8 els canvis tinguin efecte.",
  "Failed to generate LLM output language rule file: {{error}}": "Error en generar el fitxer de regles d'idioma de sortida del model: {{error}}",
  "Invalid command. Available subcommands:": "Ordre no v\xE0lida. Subordres disponibles:",
  "Available subcommands:": "Subordres disponibles:",
  "To request additional UI language packs, please open an issue on GitHub.": "Per sol\xB7licitar paquets d'idioma addicionals per a la interf\xEDcie, obriu una incid\xE8ncia a GitHub.",
  "Available options:": "Opcions disponibles:",
  "Set UI language to {{name}}": "Establir l'idioma de la interf\xEDcie a {{name}}",
  // ============================================================================
  // Ordres - Mode d'aprovació
  // ============================================================================
  "Tool Approval Mode": "Mode d'aprovaci\xF3 d'eines",
  "Analyze only, do not modify files or execute commands": "Analitzar nom\xE9s, sense modificar fitxers ni executar ordres",
  "Require approval for file edits or shell commands": "Requerir aprovaci\xF3 per a edicions de fitxers o ordres shell",
  "Automatically approve file edits": "Aprovar autom\xE0ticament les edicions de fitxers",
  "Use classifier to automatically approve safe tool calls": "Utilitzar el classificador per aprovar autom\xE0ticament les crides segures a eines",
  "Automatically approve all tools": "Aprovar autom\xE0ticament totes les eines",
  "Workspace approval mode exists and takes priority. User-level change will have no effect.": "Existeix un mode d'aprovaci\xF3 de l'espai de treball i t\xE9 prioritat. El canvi a nivell d'usuari no tindr\xE0 cap efecte.",
  "Apply To": "Aplicar a",
  "Workspace Settings": "Configuraci\xF3 de l'espai de treball",
  "Open auto-memory folder": "Obrir la carpeta de mem\xF2ria autom\xE0tica",
  "Auto-memory: {{status}}": "Mem\xF2ria autom\xE0tica: {{status}}",
  "Auto-dream: {{status}} \xB7 {{lastDream}} \xB7 /dream to run": "Auto-dream: {{status}} \xB7 {{lastDream}} \xB7 /dream per executar",
  "Auto-skill: {{status}}": "Habilitat autom\xE0tica: {{status}}",
  never: "mai",
  on: "activada",
  off: "desactivada",
  "Remove matching entries from managed auto-memory.": "Eliminar les entrades coincidents de la mem\xF2ria autom\xE0tica gestionada.",
  "Usage: /forget <memory text to remove>": "\xDAs: /forget <text de mem\xF2ria a eliminar>",
  "No managed auto-memory entries matched: {{query}}": "Cap entrada de mem\xF2ria autom\xE0tica gestionada coincideix: {{query}}",
  "Consolidate managed auto-memory topic files.": "Consolidar els fitxers de temes de mem\xF2ria autom\xE0tica gestionada.",
  "Open MCP management dialog": "Obrir el di\xE0leg de gesti\xF3 MCP",
  "Could not retrieve tool registry.": "No s'ha pogut recuperar el registre d'eines.",
  "Successfully authenticated and refreshed tools for '{{name}}'.": "S'ha autenticat correctament i s'han actualitzat les eines per a '{{name}}'.",
  "Re-discovering tools from '{{name}}'...": "Redescobrint les eines de '{{name}}'...",
  "Discovered {{count}} tool(s) from '{{name}}'.": "S'han descobert {{count}} eina(es) de '{{name}}'.",
  "Authentication complete. Returning to server details...": "Autenticaci\xF3 completada. Tornant als detalls del servidor...",
  "Authentication successful.": "Autenticaci\xF3 correcta.",
  // ============================================================================
  // Diàleg de gestió MCP
  // ============================================================================
  "Manage MCP servers": "Gestionar MCP servers",
  "Server Detail": "Detalls del servidor",
  Tools: "Eines",
  "Tool Detail": "Detalls de l'eina",
  "Loading...": "Carregant...",
  "Unknown step": "Pas desconegut",
  "Esc to back": "Esc per tornar",
  "\u2191\u2193 to navigate \xB7 Enter to select \xB7 Esc to close": "\u2191\u2193 per navegar \xB7 Enter per seleccionar \xB7 Esc per tancar",
  "\u2191\u2193 to navigate \xB7 Enter to select \xB7 Esc to back": "\u2191\u2193 per navegar \xB7 Enter per seleccionar \xB7 Esc per tornar",
  "\u2191\u2193 to navigate \xB7 Enter to confirm \xB7 Esc to back": "\u2191\u2193 per navegar \xB7 Enter per confirmar \xB7 Esc per tornar",
  "User Settings (global)": "Configuraci\xF3 d'usuari (global)",
  "Workspace Settings (project-specific)": "Configuraci\xF3 de l'espai de treball (espec\xEDfica del projecte)",
  "Disable server:": "Desactivar el servidor:",
  "Select where to add the server to the exclude list:": "Seleccioneu on afegir el servidor a la llista d'exclusi\xF3:",
  "Press Enter to confirm, Esc to cancel": "Premeu Enter per confirmar, Esc per cancel\xB7lar",
  "View tools": "Veure eines",
  Reconnect: "Reconnectar",
  Enable: "Activar",
  Disable: "Desactivar",
  Authenticate: "Autenticar",
  "Re-authenticate": "Tornar a autenticar",
  "Clear Authentication": "Esborrar l'autenticaci\xF3",
  "Server:": "Servidor:",
  "Command:": "Ordre:",
  "Working Directory:": "Directori de treball:",
  "No server selected": "Cap servidor seleccionat",
  prompts: "missatges",
  "Error:": "Error:",
  tool: "eina",
  tools: "eines",
  connected: "connectat",
  connecting: "connectant",
  disconnected: "desconnectat",
  "User MCPs": "MCPs de l'usuari",
  "Project MCPs": "MCPs del projecte",
  "Extension MCPs": "MCPs de les extensions",
  server: "servidor",
  servers: "servidors",
  "Add MCP servers to your settings to get started.": "Afegiu MCP servers a la configuraci\xF3 per comen\xE7ar.",
  "Run qwen --debug to see error logs": "Executeu qwen --debug per veure els registres d'errors",
  "OAuth Authentication": "Autenticaci\xF3 OAuth",
  "Authenticating... Please complete the login in your browser.": "Autenticant... Completeu l'inici de sessi\xF3 al vostre navegador.",
  "Press c to copy the authorization URL to your clipboard.": "Premeu c per copiar la URL d'autoritzaci\xF3 al porta-retalls.",
  "Copy request sent to your terminal. If paste is empty, copy the URL above manually.": "Sol\xB7licitud de c\xF2pia enviada al vostre terminal. Si el que enganxeu \xE9s buit, copieu la URL anterior manualment.",
  "Cannot write to terminal \u2014 copy the URL above manually.": "No es pot escriure al terminal \u2014 copieu la URL anterior manualment.",
  "No tools available for this server.": "No hi ha eines disponibles per a aquest servidor.",
  destructive: "destructiu",
  "read-only": "nom\xE9s lectura",
  "open-world": "m\xF3n obert",
  idempotent: "idempotent",
  "Tools for {{serverName}}": "Eines per a {{serverName}}",
  "{{current}}/{{total}}": "{{current}}/{{total}}",
  required: "obligatori",
  Parameters: "Par\xE0metres",
  "No tool selected": "Cap eina seleccionada",
  Server: "Servidor",
  "{{count}} invalid tools": "{{count}} eines no v\xE0lides",
  invalid: "no v\xE0lid",
  "invalid: {{reason}}": "no v\xE0lid: {{reason}}",
  "missing name": "nom absent",
  "missing description": "descripci\xF3 absent",
  "(unnamed)": "(sense nom)",
  "Warning: This tool cannot be called by the LLM": "Advert\xE8ncia: el model no pot cridar aquesta eina",
  Reason: "Motiu",
  "Tools must have both name and description to be used by the LLM.": "Les eines han de tenir nom i descripci\xF3 per poder ser usades pel model.",
  // ===========================================================
  // Ordres - Resum
  // ============================================================================
  "Generate a project summary and save it to .qwen/PROJECT_SUMMARY.md": "Generar un resum del projecte i desar-lo a .qwen/PROJECT_SUMMARY.md",
  "No chat client available to generate summary.": "No hi ha cap client de xat disponible per generar el resum.",
  "Already generating summary, wait for previous request to complete": "Ja s'est\xE0 generant el resum, espereu que acabi la sol\xB7licitud anterior",
  "No conversation found to summarize.": "No s'ha trobat cap conversa per resumir.",
  "Failed to generate project context summary: {{error}}": "Error en generar el resum del context del projecte: {{error}}",
  "Saved project summary to {{filePathForDisplay}}.": "Resum del projecte desat a {{filePathForDisplay}}.",
  "Saving project summary...": "Desant el resum del projecte...",
  "Generating project summary...": "Generant el resum del projecte...",
  "Processing summary...": "Processant el resum...",
  "Project summary generated and saved successfully!": "El resum del projecte s'ha generat i desat correctament!",
  "Saved to: {{filePath}}": "Desat a: {{filePath}}",
  "Failed to generate summary - no text content received from LLM response": "Error en generar el resum - no s'ha rebut contingut de text de la resposta del model",
  // ============================================================================
  // Ordres - Model
  // ============================================================================
  "Switch the model for this session (--fast for suggestion model, [model-id] to switch immediately).": "Canviar el model per a aquesta sessi\xF3 (--fast per al model de suggeriments)",
  "Set a lighter model for prompt suggestions and speculative execution": "Establir un model m\xE9s lleuger per a suggeriments de missatges i execuci\xF3 especulativa",
  "Content generator configuration not available.": "Configuraci\xF3 del generador de contingut no disponible.",
  "Authentication type not available.": "Tipus d'autenticaci\xF3 no disponible.",
  "No models available for the current authentication type ({{authType}}).": "No hi ha models disponibles per al tipus d'autenticaci\xF3 actual ({{authType}}).",
  // Needs translation
  " (not in model registry)": " (not in model registry)",
  // ============================================================================
  // Ordres - Netejar
  // ============================================================================
  "Starting a new session, resetting chat, and clearing terminal.": "Iniciant una nova sessi\xF3, restablint el xat i netejant el terminal.",
  "Starting a new session and clearing.": "Iniciant una nova sessi\xF3 i netejant.",
  // ============================================================================
  // Ordres - Comprimir
  // ============================================================================
  "Already compressing, wait for previous request to complete": "Ja s'est\xE0 comprimint, espereu que acabi la sol\xB7licitud anterior",
  "Failed to compress chat history.": "Error en comprimir l'historial del xat.",
  "Failed to compress chat history: {{error}}": "Error en comprimir l'historial del xat: {{error}}",
  "Compressing chat history": "Comprimint l'historial del xat",
  "Chat history compressed from {{originalTokens}} to {{newTokens}} tokens.": "L'historial del xat s'ha comprimit de {{originalTokens}} a {{newTokens}} tokens.",
  "Compression was not beneficial for this history size.": "La compressi\xF3 no ha estat beneficiosa per a aquesta mida d'historial.",
  "Chat history compression did not reduce size. This may indicate issues with the compression prompt.": "La compressi\xF3 de l'historial del xat no ha redu\xEFt la mida. Aix\xF2 pot indicar problemes amb el missatge de compressi\xF3.",
  "Could not compress chat history due to a token counting error.": "No s'ha pogut comprimir l'historial del xat per un error de recompte de tokens.",
  // ============================================================================
  // Ordres - Directori
  // ============================================================================
  "Configuration is not available.": "Configuraci\xF3 no disponible.",
  "Please provide at least one path to add.": "Proporcioneu almenys un cam\xED per afegir.",
  "The /directory add command is not supported in restrictive sandbox profiles. Please use --include-directories when starting the session instead.": "L'ordre /directory add no \xE9s compatible en perfils d'entorn a\xEFllat restrictius. En el seu lloc, useu --include-directories en iniciar la sessi\xF3.",
  "Error adding '{{path}}': {{error}}": "Error en afegir '{{path}}': {{error}}",
  "Successfully added QWEN.md files from the following directories if there are:\n- {{directories}}": "S'han afegit correctament els fitxers QWEN.md dels directoris seg\xFCents si n'hi ha:\n- {{directories}}",
  "Error refreshing memory: {{error}}": "Error en actualitzar la mem\xF2ria: {{error}}",
  "Successfully added directories:\n- {{directories}}": "S'han afegit correctament els directoris:\n- {{directories}}",
  "Current workspace directories:\n{{directories}}": "Directoris actuals de l'espai de treball:\n{{directories}}",
  // ============================================================================
  // Ordres - Documentació
  // ============================================================================
  "Please open the following URL in your browser to view the documentation:\n{{url}}": "Obriu la URL seg\xFCent al vostre navegador per veure la documentaci\xF3:\n{{url}}",
  "Opening documentation in your browser: {{url}}": "Obrint la documentaci\xF3 al vostre navegador: {{url}}",
  // ============================================================================
  // Diàlegs - Confirmació d'eines
  // ============================================================================
  "Do you want to proceed?": "Voleu continuar?",
  "Yes, allow once": "S\xED, permetre una vegada",
  "Allow always": "Permetre sempre",
  Yes: "S\xED",
  No: "No",
  "No (esc)": "No (esc)",
  "Modify in progress:": "Modificaci\xF3 en curs:",
  "Save and close external editor to continue": "Deseu i tanqueu l'editor extern per continuar",
  "Apply this change?": "Aplicar aquest canvi?",
  "Yes, allow always": "S\xED, permetre sempre",
  "Modify with external editor": "Modificar amb editor extern",
  "No, suggest changes (esc)": "No, suggerir canvis (esc)",
  "Allow execution of: '{{command}}'?": "Permetre l'execuci\xF3 de: '{{command}}'?",
  "Always allow in this project": "Permetre sempre en aquest projecte",
  "Always allow {{action}} in this project": "Permetre sempre {{action}} en aquest projecte",
  "Always allow for this user": "Permetre sempre per a aquest usuari",
  "Always allow {{action}} for this user": "Permetre sempre {{action}} per a aquest usuari",
  "Yes, restore previous mode ({{mode}})": "S\xED, restaurar el mode anterior ({{mode}})",
  "Yes, and auto-accept edits": "S\xED, i acceptar els canvis autom\xE0ticament",
  "Yes, and manually approve edits": "S\xED, i aprovar els canvis manualment",
  "No, keep planning (esc)": "No, seguir planificant (esc)",
  "URLs to fetch:": "URLs a recuperar:",
  "MCP Server: {{server}}": "MCP Server: {{server}}",
  "Tool: {{tool}}": "Eina: {{tool}}",
  'Allow execution of MCP tool "{{tool}}" from server "{{server}}"?': `Permetre l'execuci\xF3 de MCP tool "{{tool}}" des de MCP server "{{server}}"?`,
  // ============================================================================
  // Diàlegs - Confirmació de shell
  // ============================================================================
  "Shell Command Execution": "Execuci\xF3 d'ordres shell",
  "A custom command wants to run the following shell commands:": "Una ordre personalitzada vol executar les ordres shell seg\xFCents:",
  // ============================================================================
  // Diàlegs - Benvinguda
  // ============================================================================
  "Current Plan:": "Pla actual:",
  "Progress: {{done}}/{{total}} tasks completed": "Progr\xE9s: {{done}}/{{total}} tasques completades",
  ", {{inProgress}} in progress": ", {{inProgress}} en curs",
  "Pending Tasks:": "Tasques pendents:",
  "What would you like to do?": "Qu\xE8 voleu fer?",
  "Choose how to proceed with your session:": "Trieu com voleu continuar la vostra sessi\xF3:",
  "Start new chat session": "Iniciar una nova sessi\xF3 de xat",
  "Continue previous conversation": "Continuar la conversa anterior",
  "\u{1F44B} Welcome back! (Last updated: {{timeAgo}})": "\u{1F44B} Benvingut de nou! (Darrera actualitzaci\xF3: {{timeAgo}})",
  "\u{1F3AF} Overall Goal:": "\u{1F3AF} Objectiu general:",
  "Connect a Provider": "Connectar un prove\xEFdor",
  "You must connect a provider to proceed. Press Ctrl+C again to exit.": "Cal connectar un prove\xEFdor per continuar. Premeu Ctrl+C de nou per sortir.",
  "Terms of Services and Privacy Notice": "Termes de servei i av\xEDs de privacitat",
  "Qwen OAuth": "Qwen OAuth",
  "Discontinued \u2014 switch to Coding Plan or API Key": "Descontinuat \u2014 canvieu a Coding Plan o API Key",
  "Qwen OAuth free tier was discontinued on 2026-04-15. Please select Coding Plan or API Key instead.": "El nivell gratu\xEFt de Qwen OAuth es va descontinuar el 15-04-2026. Seleccioneu Coding Plan o API Key en el seu lloc.",
  "Qwen OAuth free tier was discontinued on 2026-04-15. Please select a model from another provider or run /auth to switch.": "El nivell gratu\xEFt de Qwen OAuth es va descontinuar el 15-04-2026. Seleccioneu un model d'un altre prove\xEFdor o executeu /auth per canviar.",
  "\n\u26A0 Qwen OAuth free tier was discontinued on 2026-04-15. Please select another option.\n": "\n\u26A0 El nivell gratu\xEFt de Qwen OAuth es va descontinuar el 15-04-2026. Seleccioneu una altra opci\xF3.\n",
  "Paid \xB7 Up to 6,000 requests/5 hrs \xB7 All Alibaba Cloud Coding Plan Models": "De pagament \xB7 Fins a 6.000 sol\xB7licituds/5 h \xB7 Tots els models de Coding Plan d'Alibaba Cloud",
  "Alibaba Cloud Coding Plan": "Coding Plan d'Alibaba Cloud",
  "Bring your own API key": "Porteu la vostra pr\xF2pia API Key",
  "Authentication is enforced to be {{enforcedType}}, but you are currently using {{currentType}}.": "L'autenticaci\xF3 ha de ser {{enforcedType}}, per\xF2 actualment esteu usant {{currentType}}.",
  "Qwen OAuth Authentication": "Autenticaci\xF3 Qwen OAuth",
  "Please visit this URL to authorize:": "Visiteu aquesta URL per autoritzar:",
  "Waiting for authorization": "Esperant l'autoritzaci\xF3",
  "Time remaining:": "Temps restant:",
  "Qwen OAuth Authentication Timeout": "Temps d'espera de l'autenticaci\xF3 Qwen OAuth esgotat",
  "OAuth token expired (over {{seconds}} seconds). Please select authentication method again.": "El token OAuth ha expirat (m\xE9s de {{seconds}} segons). Seleccioneu el m\xE8tode d'autenticaci\xF3 de nou.",
  "Press any key to return to authentication type selection.": "Premeu qualsevol tecla per tornar a la selecci\xF3 del tipus d'autenticaci\xF3.",
  "Waiting for Qwen OAuth authentication...": "Esperant l'autenticaci\xF3 Qwen OAuth...",
  "Authentication timed out. Please try again.": "L'autenticaci\xF3 ha expirat. Torneu-ho a intentar.",
  "Waiting for auth... (Press ESC or CTRL+C to cancel)": "Esperant l'autenticaci\xF3... (Premeu ESC o CTRL+C per cancel\xB7lar)",
  "Missing API key for OpenAI-compatible auth. Set settings.security.auth.apiKey, or set the {{envKeyHint}} environment variable.": "Manca l'API Key per a l'autenticaci\xF3 compatible amb OpenAI. Establiu settings.security.auth.apiKey o la variable d'entorn {{envKeyHint}}.",
  "{{envKeyHint}} environment variable not found. Please set it in your .env file or environment variables.": "La variable d'entorn {{envKeyHint}} no s'ha trobat. Establiu-la al fitxer .env o a les variables d'entorn.",
  "{{envKeyHint}} environment variable not found (or set settings.security.auth.apiKey). Please set it in your .env file or environment variables.": "La variable d'entorn {{envKeyHint}} no s'ha trobat (o establiu settings.security.auth.apiKey). Establiu-la al fitxer .env o a les variables d'entorn.",
  "Missing API key for OpenAI-compatible auth. Set the {{envKeyHint}} environment variable.": "Manca l'API Key per a l'autenticaci\xF3 compatible amb OpenAI. Establiu la variable d'entorn {{envKeyHint}}.",
  "Anthropic provider missing required baseUrl in modelProviders[].baseUrl.": "El prove\xEFdor Anthropic no t\xE9 la baseUrl obligat\xF2ria a modelProviders[].baseUrl.",
  "ANTHROPIC_BASE_URL environment variable not found.": "La variable d'entorn ANTHROPIC_BASE_URL no s'ha trobat.",
  "Invalid auth method selected.": "S'ha seleccionat un m\xE8tode d'autenticaci\xF3 no v\xE0lid.",
  "Failed to authenticate. Message: {{message}}": "Error en autenticar-se. Missatge: {{message}}",
  "Authenticated successfully with {{authType}} credentials.": "S'ha autenticat correctament amb les credencials {{authType}}.",
  'Invalid QWEN_DEFAULT_AUTH_TYPE value: "{{value}}". Valid values are: {{validValues}}': 'Valor de QWEN_DEFAULT_AUTH_TYPE no v\xE0lid: "{{value}}". Els valors v\xE0lids s\xF3n: {{validValues}}',
  // ============================================================================
  // Diàlegs - Model
  // ============================================================================
  "Select Model": "Seleccioneu el model",
  "API Key": "API Key",
  "(default)": "(per defecte)",
  "(not set)": "(no establert)",
  Modality: "Modalitat",
  "Context Window": "Fin. de context",
  text: "text",
  "text-only": "nom\xE9s text",
  image: "imatge",
  pdf: "pdf",
  audio: "\xE0udio",
  video: "v\xEDdeo",
  "not set": "no establert",
  none: "cap",
  unknown: "desconegut",
  // ============================================================================
  // Diàlegs - Permisos
  // ============================================================================
  "Manage folder trust settings": "Gestionar la configuraci\xF3 de confian\xE7a de carpetes",
  "Manage permission rules": "Gestionar permission rules",
  Allow: "Permetre",
  Ask: "Preguntar",
  Deny: "Denegar",
  Workspace: "Espai de treball",
  "Qwen Code won't ask before using allowed tools.": "Qwen Code no preguntar\xE0 abans d'usar les eines permeses.",
  "Qwen Code will ask before using these tools.": "Qwen Code preguntar\xE0 abans d'usar aquestes eines.",
  "Qwen Code is not allowed to use denied tools.": "Qwen Code no t\xE9 perm\xEDs per usar les eines denegades.",
  "Manage trusted directories for this workspace.": "Gestionar els directoris de confian\xE7a d'aquest espai de treball.",
  "Any use of the {{tool}} tool": "Qualsevol \xFAs de l'eina {{tool}}",
  "{{tool}} commands matching '{{pattern}}'": "Ordres de {{tool}} que coincideixen amb '{{pattern}}'",
  "From user settings": "Des de la configuraci\xF3 d'usuari",
  "From project settings": "Des de la configuraci\xF3 del projecte",
  "From session": "Des de la sessi\xF3",
  "Project settings": "Configuraci\xF3 del projecte",
  "Checked in at .qwen/settings.json": "Registrat a .qwen/settings.json",
  "User settings": "Configuraci\xF3 d'usuari",
  "Saved in at ~/.qwen/settings.json": "Desat a ~/.qwen/settings.json",
  "Add a new rule\u2026": "Afegir una nova regla\u2026",
  "Add {{type}} permission rule": "Afegir {{type}} permission rule",
  "Permission rules are a tool name, optionally followed by a specifier in parentheses.": "permission rules s\xF3n un nom d'eina, seguit opcionalment d'un especificador entre par\xE8ntesis.",
  "e.g.,": "p. ex.,",
  or: "o",
  "Enter permission rule\u2026": "Introdu\xEFu permission rule\u2026",
  "Enter to submit \xB7 Esc to cancel": "Enter per enviar \xB7 Esc per cancel\xB7lar",
  "Where should this rule be saved?": "On s'ha de desar aquesta regla?",
  "Enter to confirm \xB7 Esc to cancel": "Enter per confirmar \xB7 Esc per cancel\xB7lar",
  "Delete {{type}} rule?": "Eliminar la regla {{type}}?",
  "Are you sure you want to delete this permission rule?": "Esteu segur que voleu eliminar aquesta permission rule?",
  "Permissions:": "Permisos:",
  "(\u2190/\u2192 or tab to cycle)": "(\u2190/\u2192 o Tab per canviar)",
  "Press \u2191\u2193 to navigate \xB7 Enter to select \xB7 Type to search \xB7 Esc to cancel": "Premeu \u2191\u2193 per navegar \xB7 Enter per seleccionar \xB7 Escriviu per cercar \xB7 Esc per cancel\xB7lar",
  "Search\u2026": "Cercar\u2026",
  "Add directory\u2026": "Afegir directori\u2026",
  "Add directory to workspace": "Afegir directori a l'espai de treball",
  "Qwen Code can read files in the workspace, and make edits when auto-accept edits is on.": "Qwen Code pot llegir fitxers a l'espai de treball i fer canvis quan l'acceptaci\xF3 autom\xE0tica de canvis est\xE0 activada.",
  "Qwen Code will be able to read files in this directory and make edits when auto-accept edits is on.": "Qwen Code podr\xE0 llegir fitxers en aquest directori i fer canvis quan l'acceptaci\xF3 autom\xE0tica de canvis est\xE0 activada.",
  "Enter the path to the directory:": "Introdu\xEFu el cam\xED del directori:",
  "Enter directory path\u2026": "Introdu\xEFu el cam\xED del directori\u2026",
  "Tab to complete \xB7 Enter to add \xB7 Esc to cancel": "Tab per completar \xB7 Enter per afegir \xB7 Esc per cancel\xB7lar",
  "Remove directory?": "Eliminar el directori?",
  "Are you sure you want to remove this directory from the workspace?": "Esteu segur que voleu eliminar aquest directori de l'espai de treball?",
  "  (Original working directory)": "  (Directori de treball original)",
  "  (from settings)": "  (des de la configuraci\xF3)",
  "Directory does not exist.": "El directori no existeix.",
  "Path is not a directory.": "El cam\xED no \xE9s un directori.",
  "This directory is already in the workspace.": "Aquest directori ja \xE9s a l'espai de treball.",
  "Already covered by existing directory: {{dir}}": "Ja cobert per un directori existent: {{dir}}",
  // ============================================================================
  // Barra d'estat
  // ============================================================================
  "Using:": "En \xFAs:",
  "{{count}} open file": "{{count}} fitxer obert",
  "{{count}} open files": "{{count}} fitxers oberts",
  "(ctrl+g to view)": "(ctrl+g per veure)",
  "{{count}} {{name}} file": "{{count}} fitxer {{name}}",
  "{{count}} {{name}} files": "{{count}} fitxers {{name}}",
  "{{count}} MCP server": "{{count}} MCP server",
  "{{count}} MCP servers": "{{count}} MCP servers",
  "{{count}} Blocked": "{{count}} bloquejats",
  "(ctrl+t to view)": "(ctrl+t per veure)",
  "(ctrl+t to toggle)": "(ctrl+t per canviar)",
  "Press Ctrl+C again to exit.": "Premeu Ctrl+C de nou per sortir.",
  "Press Ctrl+D again to exit.": "Premeu Ctrl+D de nou per sortir.",
  "Press Esc again to clear.": "Premeu Esc de nou per esborrar.",
  "Press \u2191 to edit queued messages": "Premeu \u2191 per editar els missatges en cua",
  // ============================================================================
  // Estat MCP
  // ============================================================================
  "No MCP servers configured.": "No hi ha MCP servers configurats.",
  "\u23F3 MCP servers are starting up ({{count}} initializing)...": "\u23F3 MCP servers s'estan iniciant ({{count}} inicialitzant)...",
  "Note: First startup may take longer. Tool availability will update automatically.": "Nota: El primer inici pot tardar m\xE9s. La disponibilitat de les eines s'actualitzar\xE0 autom\xE0ticament.",
  "Configured MCP servers:": "MCP servers configurats:",
  Ready: "Preparat",
  "Starting... (first startup may take longer)": "Iniciant... (el primer inici pot tardar m\xE9s)",
  Disconnected: "Desconnectat",
  "{{count}} tool": "{{count}} eina",
  "{{count}} tools": "{{count}} eines",
  "{{count}} prompt": "{{count}} missatge",
  "{{count}} prompts": "{{count}} missatges",
  "(from {{extensionName}})": "(de {{extensionName}})",
  OAuth: "OAuth",
  "OAuth expired": "OAuth expirat",
  "OAuth not authenticated": "OAuth no autenticat",
  "tools and prompts will appear when ready": "les eines i els missatges apareixeran quan estiguin a punt",
  "{{count}} tools cached": "{{count}} eines en mem\xF2ria cau",
  "Tools:": "Eines:",
  "Parameters:": "Par\xE0metres:",
  "Prompts:": "Missatges:",
  "Resources:": "Recursos:",
  Blocked: "Bloquejat",
  "\u{1F4A1} Tips:": "\u{1F4A1} Consells:",
  Use: "Useu",
  "to show server and tool descriptions": "per mostrar les descripcions del servidor i de les eines",
  "to show tool parameter schemas": "per mostrar tool parameter schemas",
  "to hide descriptions": "per amagar les descripcions",
  "to authenticate with OAuth-enabled servers": "per autenticar-vos amb servidors OAuth",
  Press: "Premeu",
  "to toggle tool descriptions on/off": "per activar/desactivar les descripcions de les eines",
  "Starting OAuth authentication for MCP server '{{name}}'...": "Iniciant l'autenticaci\xF3 OAuth per a MCP server '{{name}}'...",
  // ============================================================================
  // Consells d'inici
  // ============================================================================
  "Tips:": "Consells:",
  "Use /compress when the conversation gets long to summarize history and free up context.": "Useu /compress quan la conversa sigui llarga per resumir l'historial i alliberar context.",
  "Start a fresh idea with /clear or /new; the previous session stays available in history.": "Comenceu una idea nova amb /clear o /new; la sessi\xF3 anterior segueix disponible a l'historial.",
  "Use /bug to submit issues to the maintainers when something goes off.": "Useu /bug per enviar incid\xE8ncies als mantenidors quan alguna cosa vagi malament.",
  "Switch auth type quickly with /auth.": "Canvieu r\xE0pidament el tipus d'autenticaci\xF3 amb /auth.",
  "You can run any shell commands from Qwen Code using ! (e.g. !ls).": "Podeu executar qualsevol ordre shell des de Qwen Code usant ! (p. ex. !ls).",
  "Type / to open the command popup; Tab autocompletes slash commands and saved prompts.": "Escriviu / per obrir el men\xFA emergent d'ordres; Tab completa autom\xE0ticament les ordres de barra i els missatges desats.",
  "You can resume a previous conversation by running qwen --continue or qwen --resume.": "Podeu reprendre una conversa anterior executant qwen --continue o qwen --resume.",
  "You can switch permission mode quickly with Shift+Tab or /approval-mode.": "Podeu canviar r\xE0pidament el mode de permisos amb Shift+Tab o /approval-mode.",
  "You can switch permission mode quickly with Tab or /approval-mode.": "Podeu canviar r\xE0pidament el mode de permisos amb Tab o /approval-mode.",
  "Try /insight to generate personalized insights from your chat history.": "Proveu /insight per generar idees personalitzades a partir del vostre historial de xat.",
  "Press Ctrl+O to toggle compact mode \u2014 hide tool output and thinking for a cleaner view.": "Premeu Ctrl+O per canviar el mode compacte \u2014 amagueu la sortida de les eines i el pensament per a una vista m\xE9s neta.",
  "Add a QWEN.md file to give Qwen Code persistent project context.": "Afegiu un fitxer QWEN.md per donar a Qwen Code un context persistent del projecte.",
  "Use /btw to ask a quick side question without disrupting the conversation.": "Useu /btw per fer una pregunta r\xE0pida sense interrompre la conversa.",
  "Context is almost full! Run /compress now or start /new to continue.": "El context gaireb\xE9 \xE9s ple! Executeu /compress ara o inicieu /new per continuar.",
  "Context is getting full. Use /compress to free up space.": "El context s'omple. Useu /compress per alliberar espai.",
  "Long conversation? /compress summarizes history to free context.": "Conversa llarga? /compress resumeix l'historial per alliberar context.",
  // ============================================================================
  // Pantalla de sortida / Estadístiques
  // ============================================================================
  "Agent powering down. Goodbye!": "L'agent s'apaga. Fins aviat!",
  "To continue this session, run": "Per continuar aquesta sessi\xF3, executeu",
  "Interaction Summary": "Resum de la interacci\xF3",
  "Session ID:": "ID de sessi\xF3:",
  "Tool Calls:": "Crides a eines:",
  "Success Rate:": "Taxa d'\xE8xit:",
  "User Agreement:": "Acord de l'usuari:",
  reviewed: "revisades",
  "Code Changes:": "Canvis de codi:",
  Performance: "Rendiment",
  "Wall Time:": "Temps real:",
  "Agent Active:": "Agent actiu:",
  "API Time:": "Temps de l'API:",
  "Tool Time:": "Temps d'eines:",
  "Session Stats": "Estad\xEDstiques de la sessi\xF3",
  "Model Usage": "\xDAs del model",
  Reqs: "Sol\xB7licituds",
  "Input Tokens": "Tokens d'entrada",
  "Output Tokens": "Tokens de sortida",
  "Savings Highlight:": "Estalvis destacats:",
  "of input tokens were served from the cache, reducing costs.": "dels tokens d'entrada s'han servit des de la mem\xF2ria cau, reduint els costos.",
  "Tip: For a full token breakdown, run `/stats model`.": "Consell: Per a un desglossament complet de tokens, executeu `/stats model`.",
  "Model Stats For Nerds": "Estad\xEDstiques del model per a nerds",
  "Tool Stats For Nerds": "Estad\xEDstiques d'eines per a nerds",
  Metric: "M\xE8trica",
  API: "API",
  Requests: "Sol\xB7licituds",
  Errors: "Errors",
  "Avg Latency": "Lat\xE8ncia mitjana",
  Tokens: "Tokens",
  Total: "Total",
  Prompt: "Missatge",
  Cached: "En mem\xF2ria cau",
  Thoughts: "Pensaments",
  Output: "Sortida",
  "No API calls have been made in this session.": "No s'ha realitzat cap crida a l'API en aquesta sessi\xF3.",
  "Tool Name": "Nom de l'eina",
  Calls: "Crides",
  "Success Rate": "Taxa d'\xE8xit",
  "Avg Duration": "Durada mitjana",
  "User Decision Summary": "Resum de decisions de l'usuari",
  "Total Reviewed Suggestions:": "Total de suggeriments revisats:",
  " \xBB Accepted:": " \xBB Acceptats:",
  " \xBB Rejected:": " \xBB Rebutjats:",
  " \xBB Modified:": " \xBB Modificats:",
  " Overall Agreement Rate:": " Taxa d'acord global:",
  "No tool calls have been made in this session.": "No s'ha realitzat cap crida a eines en aquesta sessi\xF3.",
  "Session start time is unavailable, cannot calculate stats.": "L'hora d'inici de la sessi\xF3 no est\xE0 disponible, no es poden calcular les estad\xEDstiques.",
  Activity: "Activitat",
  Efficiency: "Efici\xE8ncia",
  Today: "Avui",
  "Token Trend": "Tend\xE8ncia de Tokens",
  "Cache Hit Rate": "Taxa d'encert de cache",
  "Tool Success": "\xC8xit d'eines",
  "Tool Leaderboard": "Classificaci\xF3 d'eines",
  Time: "Temps",
  Success: "\xC8xit",
  Cache: "Cache",
  Latency: "Lat\xE8ncia",
  "Code Impact": "Impacte al codi",
  net: "net",
  streak: "ratxa",
  best: "r\xE8cord",
  // ============================================================================
  // Migració del format d'ordres
  // ============================================================================
  "Command Format Migration": "Migraci\xF3 del format d'ordres",
  "Found {{count}} TOML command file:": "S'ha trobat {{count}} fitxer d'ordres TOML:",
  "Found {{count}} TOML command files:": "S'han trobat {{count}} fitxers d'ordres TOML:",
  "Current tasks": "Tasques actuals",
  "Background tasks": "Tasques en segon pla",
  "No tasks currently running": "No hi ha cap tasca en execuci\xF3",
  "No entry to show.": "No hi ha cap entrada per mostrar.",
  "needs approval": "necessita aprovaci\xF3",
  "rejected \u2014 edit config to re-approve": "rebutjat \u2014 editeu la configuraci\xF3 per tornar a aprovar",
  "Background agent needs approval": "L'agent en segon pla necessita aprovaci\xF3",
  "Approve or deny the request above": "Aprova o denega la sol\xB7licitud de dalt",
  Running: "En execuci\xF3",
  Paused: "En pausa",
  Completed: "Completada",
  Failed: "Fallida",
  Stopped: "Aturada",
  Shell: "Shell",
  Monitor: "Monitor",
  Command: "Ordre",
  Dream: "Dream",
  "[dream] memory consolidation": "[dream] consolidaci\xF3 de mem\xF2ria",
  "[dream] memory consolidation (reviewing {{count}} session)": "[dream] consolidaci\xF3 de mem\xF2ria (revisant {{count}} sessi\xF3)",
  "[dream] memory consolidation (reviewing {{count}} sessions)": "[dream] consolidaci\xF3 de mem\xF2ria (revisant {{count}} sessions)",
  "... and {{count}} more": "... i {{count}} m\xE9s",
  "The TOML format is deprecated. Would you like to migrate them to Markdown format?": "El format TOML \xE9s obsolet. Voleu migrar-los al format Markdown?",
  "(Backups will be created and original files will be preserved)": "(Es crearan c\xF2pies de seguretat i els fitxers originals es conservaran)",
  // ============================================================================
  // Frases de càrrega
  // ============================================================================
  "Waiting for user confirmation...": "Esperant la confirmaci\xF3 de l'usuari...",
  // ============================================================================
  // Frases de càrrega enginyoses
  // ============================================================================
  WITTY_LOADING_PHRASES: [
    "Em sento afortunat",
    "Enviant el millor...",
    "Setze jutges d'un jutjat mengen fetge d'un penjat.",
    "Navegant pel fong mucilagin\xF3s...",
    "Consultant els esperits digitals...",
    "Desperta ferro...",
    "Escalfant els h\xE0msters de la IA...",
    "Preguntant a la petxina m\xE0gica...",
    "Generant una r\xE8plica enginyosa...",
    "Polint els algorismes...",
    "No correu la perfecci\xF3 (ni el meu codi)...",
    "Preparant bytes frescos...",
    "Comptant electrons...",
    "Activant els processadors cognitius...",
    "Buscant errors de sintaxi a l'univers...",
    "Un moment, optimitzant l'humor...",
    "Barrejant les gr\xE0cies...",
    "Desenredant les xarxes neuronals...",
    "Compilant la brillantor...",
    "Carregant gr\xE0cia.exe...",
    "Invocant el n\xFAvol de saviesa...",
    "Preparant una resposta enginyosa...",
    "Un segon, estic depurant la realitat...",
    "Donant els \xFAltims cops de...",
    "Afinant les freq\xFC\xE8ncies c\xF2smiques...",
    "Elaborant una resposta digna de la vostra paci\xE8ncia...",
    "Compilant els 1 i els 0...",
    "Resolent depend\xE8ncies... i crisis existencials...",
    "Desfragmentant records... tant de RAM com personals...",
    "Reiniciant el m\xF2dul de l'humor...",
    "Emmagatzemant en mem\xF2ria cau el necessari (principalment mems de gats)...",
    "Optimitzant per a velocitat rid\xEDcula",
    "Intercanviant bits... que no ho s\xE0piguen els bytes...",
    "Recollint brossa... torno de seguida...",
    "Assemblant les internets...",
    "Convertint caf\xE8 en codi...",
    "Actualitzant la sintaxi de la realitat...",
    "Reconnectant les sinapsis...",
    "Buscant un punt i coma mal posat...",
    "Engreixant els engranatges de la m\xE0quina...",
    "Precalfant els servidors...",
    "Calibrant el condensador de flux...",
    "Activant el motor de improbabilitat...",
    "Canalitzant la For\xE7a...",
    "Alineant les estrelles per a una resposta \xF2ptima...",
    "I tots ho diem...",
    "Carregant la propera gran idea...",
    "Un moment, estic en el meu element...",
    "Preparant-me per impressionar-vos amb brillantor...",
    "Un moment, polint el meu enginy...",
    "Aguanteu, estic creant una obra mestra...",
    "Un moment, depurant l'univers...",
    "Un moment, alineant els p\xEDxels...",
    "Un segon, optimitzant l'humor...",
    "Un moment, afinant els algorismes...",
    "Velocitat de curvatura activada...",
    "Preparant la seg\xFCent jugada mestre...",
    "No us espanteu...",
    "Seguint el conill blanc...",
    "La veritat \xE9s aqu\xED... en algun lloc...",
    "Bufant al cartutx...",
    "Carregant... Feu un gir de barril!",
    "Esperant la reaparici\xF3...",
    "Paci\xE8ncia, pensa que Rodalies encara va m\xE9s lent...",
    "El past\xEDs no \xE9s una mentida, simplement s'est\xE0 carregant...",
    "Tafanejant la pantalla de creaci\xF3 de personatge...",
    "Un moment, trobo el meme adequat...",
    "Prement 'A' per continuar...",
    "Pasturant gats digitals...",
    "Polint els p\xEDxels...",
    "Buscant un acudit per a la pantalla de c\xE0rrega...",
    "Distreu-vos amb aquesta frase enginyosa...",
    "Gaireb\xE9 a punt... probablement...",
    "Els nostres h\xE0msters treballen tan r\xE0pid com poden...",
    "Donant un copet al cap a Cloudy...",
    "Fent festes al gat...",
    "Endavant les atxes...",
    "Mai no us deixar\xE9 anar, mai no us decebr\xE9...",
    "Tocant el baix...",
    "Vaig a buscar ratafia...",
    "Vaig a tota velocitat, vaig a tota marxa...",
    "\xC9s la vida real? \xC9s sols fantasia?...",
    "Tinc bon pressentiment sobre aix\xF2...",
    "Tocant el tigre...",
    "Investigant els \xFAltims mems...",
    "Pensant com fer aix\xF2 m\xE9s enginy\xF3s...",
    "Hmm... deixeu-me pensar...",
    "Suant la cansalada...",
    "Trient el fetge per la boca...",
    "Posar fil a l'agulla...",
    "Un moment, ho tenim a tocar..",
    "Aix\xF2 \xE9s bufar i fer ampolles",
    "Qu\xE8 pots fer amb un llapis trencat? Res, no t\xE9 punta...",
    "Aplicant manteniment percussiu...",
    "Buscant l'orientaci\xF3 correcta de l'USB...",
    "Assegurant que el fum m\xE0gic quedi dins dels cables...",
    "Intentant sortir del Vim...",
    "Girant la roda del h\xE0mster...",
    "Aix\xF2 no \xE9s un error, \xE9s una caracter\xEDstica no documentada...",
    "Endavant.",
    "Tornar\xE9... amb una resposta.",
    "El meu altre proc\xE9s \xE9s una TARDIS...",
    "Posant oli als engranatges...",
    "Deixant que els pensaments macerin...",
    "Acabo de recordar on he deixat les claus...",
    "Ponderant l'orbe...",
    "He vist coses que no creur\xEDeu... com un usuari que llegeix els missatges de c\xE0rrega.",
    "Iniciant la mirada pensativa...",
    "Quin \xE9s el berenar preferit d'un computador? Xips micro.",
    "Per qu\xE8 els programadors de Java porten ulleres? Perqu\xE8 no veuen en C#.",
    "Carregant el l\xE0ser... piu piu!",
    "Dividint per zero... \xE9s broma!",
    "Buscant un supervisor adult... \xE9s a dir, processant.",
    "Fent que faci xup xup.",
    "Emmarcant... perqu\xE8 fins i tot les IA necessiten un moment.",
    "Entrella\xE7ant part\xEDcules qu\xE0ntiques per a una resposta m\xE9s r\xE0pida...",
    "Polint el crom... dels algorismes.",
    "No esteu entretinguts? (Hi estem treballant!)",
    "Invocant els follets del codi... per ajudar, \xE9s clar.",
    "Esperant que acabi el so del m\xF2dem de marcaci\xF3...",
    "Recalibrant el mesurament de l'humor.",
    "La meva altra pantalla de c\xE0rrega \xE9s fins i tot m\xE9s divertida.",
    "Estic bastant segur que hi ha un gat caminant per algun teclat...",
    "Millorant... millorant... encara carregant.",
    "No \xE9s un error, \xE9s una caracter\xEDstica... d'aquesta pantalla de c\xE0rrega.",
    "Heu provat apagar-ho i tornar-lo a encendre? (La pantalla de c\xE0rrega, no jo.)",
    "Construint pil\xF3 addicionals..."
  ],
  // ============================================================================
  // Entrada de configuració d'extensions
  // ============================================================================
  "Enter value...": "Introdu\xEFu el valor...",
  "Enter sensitive value...": "Introdu\xEFu el valor sensible...",
  "Press Enter to submit, Escape to cancel": "Premeu Enter per enviar, Escape per cancel\xB7lar",
  // ============================================================================
  // Eina de migració d'ordres
  // ============================================================================
  "Markdown file already exists: {{filename}}": "El fitxer Markdown ja existeix: {{filename}}",
  "TOML Command Format Deprecation Notice": "Av\xEDs d'obsolesc\xE8ncia del format d'ordres TOML",
  "Found {{count}} command file(s) in TOML format:": "S'ha(n) trobat {{count}} fitxer(s) d'ordres en format TOML:",
  "The TOML format for commands is being deprecated in favor of Markdown format.": "El format TOML per a ordres s'est\xE0 fent obsolet en favor del format Markdown.",
  "Markdown format is more readable and easier to edit.": "El format Markdown \xE9s m\xE9s llegible i f\xE0cil d'editar.",
  "You can migrate these files automatically using:": "Podeu migrar aquests fitxers autom\xE0ticament usant:",
  "Or manually convert each file:": "O convertiu cada fitxer manualment:",
  'TOML: prompt = "..." / description = "..."': 'TOML: prompt = "..." / description = "..."',
  "Markdown: YAML frontmatter + content": "Markdown: cap\xE7alera YAML + contingut",
  "The migration tool will:": "L'eina de migraci\xF3 far\xE0:",
  "Convert TOML files to Markdown": "Convertir fitxers TOML a Markdown",
  "Create backups of original files": "Crear c\xF2pies de seguretat dels fitxers originals",
  "Preserve all command functionality": "Preservar tota la funcionalitat de les ordres",
  "TOML format will continue to work for now, but migration is recommended.": "El format TOML seguir\xE0 funcionant de moment, per\xF2 es recomana la migraci\xF3.",
  // ============================================================================
  // Extensions - Ordre d'explorar
  // ============================================================================
  "Open extensions page in your browser": "Obrir la p\xE0gina d'extensions al vostre navegador",
  "Unknown extensions source: {{source}}.": "Font d'extensions desconeguda: {{source}}.",
  "Would open extensions page in your browser: {{url}} (skipped in test environment)": "Obriria la p\xE0gina d'extensions al vostre navegador: {{url}} (om\xE8s en entorn de proves)",
  "View available extensions at {{url}}": "Veure les extensions disponibles a {{url}}",
  "Opening extensions page in your browser: {{url}}": "Obrint la p\xE0gina d'extensions al vostre navegador: {{url}}",
  "Failed to open browser. Check out the extensions gallery at {{url}}": "Error en obrir el navegador. Visiteu la galeria d'extensions a {{url}}",
  "Retrying in {{seconds}} seconds\u2026 (attempt {{attempt}}/{{maxRetries}})": "Reintentant en {{seconds}} segons\u2026 (intent {{attempt}}/{{maxRetries}})",
  "Press Ctrl+Y to retry": "Premeu Ctrl+Y per reintentar",
  "No failed request to retry.": "No hi ha cap sol\xB7licitud fallida per reintentar.",
  "to retry last request": "per reintentar l'\xFAltima sol\xB7licitud",
  // ============================================================================
  // Autenticació de Coding Plan
  // ============================================================================
  "API key cannot be empty.": "La API Key no pot estar buida.",
  'Invalid API key. Coding Plan API keys start with "sk-sp-". Please check.': 'API Key no v\xE0lida. Les API Keys de Coding Plan comencen per "sk-sp-". Comproveu-la.',
  "You can get your Coding Plan API key here": "Podeu obtenir la vostra API Key de Coding Plan aqu\xED",
  "Failed to update Coding Plan configuration: {{message}}": "Error en actualitzar la configuraci\xF3 de Coding Plan: {{message}}",
  // ============================================================================
  // Configuració de API Key personalitzada
  // ============================================================================
  "You can configure your API key and models in settings.json": "Podeu configurar la vostra API Key i els models a settings.json",
  "Refer to the documentation for setup instructions": "Consulteu la documentaci\xF3 per a les instruccions de configuraci\xF3",
  // ============================================================================
  // Diàleg d'autenticació - Títols i etiquetes
  // ============================================================================
  "Coding Plan": "Coding Plan",
  Custom: "Personalitzat",
  "Select Region for Coding Plan": "Seleccioneu la regi\xF3 per a Coding Plan",
  "Choose based on where your account is registered": "Trieu en funci\xF3 d'on teniu registrat el compte",
  "Enter Coding Plan API Key": "Introdu\xEFu la API Key de Coding Plan",
  // ============================================================================
  // Actualitzacions internacionals de Coding Plan
  // ============================================================================
  "New model configurations are available for {{region}}. Update now?": "Hi ha noves configuracions de model disponibles per a {{region}}. Actualitzeu ara?",
  '{{region}} configuration updated successfully. Model switched to "{{model}}".': `La configuraci\xF3 de {{region}} s'ha actualitzat correctament. El model ha canviat a "{{model}}".`,
  // ============================================================================
  // Component d'ús del context
  // ============================================================================
  "Context Usage": "\xDAs del context",
  "% used": "% usat",
  "% context used": "% del context usat",
  "Context exceeds limit! Use /compress or /clear to reduce.": "El context supera el l\xEDmit! Useu /compress o /clear per reduir-lo.",
  "No API response yet. Send a message to see actual usage.": "Encara no hi ha cap resposta de l'API. Envieu un missatge per veure l'\xFAs real.",
  "Estimated pre-conversation overhead": "C\xE0rrega estimada pr\xE8via a la conversa",
  "Context window": "Finestra de context",
  tokens: "tokens",
  Used: "Usat",
  Free: "Lliure",
  "Autocompact buffer": "Mem\xF2ria interm\xE8dia de compactaci\xF3 autom\xE0tica",
  "Usage by category": "\xDAs per categoria",
  "System prompt": "Missatge del sistema",
  "Built-in tools": "Eines integrades",
  "MCP tools": "MCP tools",
  "Memory files": "Fitxers de mem\xF2ria",
  Skills: "Habilitats",
  Messages: "Missatges",
  "Run /context detail for per-item breakdown.": "Executeu /context detail per a un desglossament per element.",
  'Show context window usage breakdown. Use "/context detail" for per-item breakdown.': `Mostrar el desglossament de l'\xFAs de la finestra de context. Useu "/context detail" per a un desglossament per element.`,
  "body loaded": "cos carregat",
  memory: "mem\xF2ria",
  "{{region}} configuration updated successfully.": "La configuraci\xF3 de {{region}} s'ha actualitzat correctament.",
  "Authenticated successfully with {{region}}. API key and model configs saved to settings.json.": "S'ha autenticat correctament amb {{region}}. La API Key i les configuracions del model s'han desat a settings.json.",
  "Tip: Use /model to switch between available Coding Plan models.": "Consell: Useu /model per canviar entre els models de Coding Plan disponibles.",
  "Type something...": "Escriviu alguna cosa...",
  Submit: "Enviar",
  "Submit answers": "Enviar respostes",
  Cancel: "Cancel\xB7lar",
  "Your answers:": "Les vostres respostes:",
  "(not answered)": "(sense resposta)",
  "Ready to submit your answers?": "Preparats per enviar les vostres respostes?",
  "\u2191/\u2193: Navigate | \u2190/\u2192: Switch tabs | Enter: Select": "\u2191/\u2193: Navegar | \u2190/\u2192: Canviar pestanyes | Enter: Seleccionar",
  "\u2191/\u2193: Navigate | Enter: Select | Esc: Cancel": "\u2191/\u2193: Navegar | Enter: Seleccionar | Esc: Cancel\xB7lar",
  "Authenticate using Qwen OAuth": "Autenticar-se usant Qwen OAuth",
  "Authenticate using Alibaba Cloud Coding Plan": "Autenticar-se usant el Coding Plan d'Alibaba Cloud",
  "Region for Coding Plan (china/global)": "Regi\xF3 per a Coding Plan (china/global)",
  "API key for Coding Plan": "API Key per a Coding Plan",
  "Show current authentication status": "Mostrar l'estat d'autenticaci\xF3 actual",
  "Authentication completed successfully.": "L'autenticaci\xF3 s'ha completat correctament.",
  "Starting Qwen OAuth authentication...": "Iniciant l'autenticaci\xF3 Qwen OAuth...",
  "Successfully authenticated with Qwen OAuth.": "S'ha autenticat correctament amb Qwen OAuth.",
  "Failed to authenticate with Qwen OAuth: {{error}}": "Error en autenticar-se amb Qwen OAuth: {{error}}",
  "Processing Alibaba Cloud Coding Plan authentication...": "Processant l'autenticaci\xF3 de Coding Plan d'Alibaba Cloud...",
  "Successfully authenticated with Alibaba Cloud Coding Plan.": "S'ha autenticat correctament amb el Coding Plan d'Alibaba Cloud.",
  "Failed to authenticate with Coding Plan: {{error}}": "Error en autenticar-se amb el Coding Plan: {{error}}",
  "\u963F\u91CC\u4E91\u767E\u70BC (aliyun.com)": "\u963F\u91CC\u4E91\u767E\u70BC (aliyun.com)",
  Global: "Global",
  "Alibaba Cloud (alibabacloud.com)": "Alibaba Cloud (alibabacloud.com)",
  "Select region for Coding Plan:": "Seleccioneu la regi\xF3 per a Coding Plan:",
  "Enter your Coding Plan API key: ": "Introdu\xEFu la vostra API Key de Coding Plan: ",
  "Select authentication method:": "Seleccioneu el m\xE8tode d'autenticaci\xF3:",
  "\n=== Authentication Status ===\n": "\n=== Estat d'autenticaci\xF3 ===\n",
  "\u26A0\uFE0F  No authentication method configured.\n": "\u26A0\uFE0F  Cap m\xE8tode d'autenticaci\xF3 configurat.\n",
  "Run one of the following commands to get started:\n": "Executeu una de les ordres seg\xFCents per comen\xE7ar:\n",
  "  qwen auth qwen-oauth     - Authenticate with Qwen OAuth (discontinued)": "  qwen auth qwen-oauth     - Autenticar-se amb Qwen OAuth (descontinuat)",
  "Or simply run:": "O simplement executeu:",
  "  qwen auth                - Interactive authentication setup\n": "  qwen auth                - Configuraci\xF3 interactiva de l'autenticaci\xF3\n",
  "\u2713 Authentication Method: Qwen OAuth": "\u2713 M\xE8tode d'autenticaci\xF3: Qwen OAuth",
  "  Type: Free tier (discontinued 2026-04-15)": "  Tipus: Nivell gratu\xEFt (descontinuat el 15-04-2026)",
  "  Limit: No longer available": "  L\xEDmit: Ja no disponible",
  "Qwen OAuth free tier was discontinued on 2026-04-15. Run /auth to switch to Coding Plan, OpenRouter, Fireworks AI, or another provider.": "El nivell gratu\xEFt de Qwen OAuth es va descontinuar el 15-04-2026. Executeu /auth per canviar a Coding Plan, OpenRouter, Fireworks AI o un altre prove\xEFdor.",
  "\u2713 Authentication Method: Alibaba Cloud Coding Plan": "\u2713 M\xE8tode d'autenticaci\xF3: Coding Plan d'Alibaba Cloud",
  "Global - Alibaba Cloud": "Global - Alibaba Cloud",
  "  Region: {{region}}": "  Regi\xF3: {{region}}",
  "  Current Model: {{model}}": "  Model actual: {{model}}",
  "  Config Version: {{version}}": "  Versi\xF3 de configuraci\xF3: {{version}}",
  "  Status: API key configured\n": "  Estat: API Key configurada\n",
  "\u26A0\uFE0F  Authentication Method: Alibaba Cloud Coding Plan (Incomplete)": "\u26A0\uFE0F  M\xE8tode d'autenticaci\xF3: Coding Plan d'Alibaba Cloud (Incomplet)",
  "  Issue: API key not found in environment or settings\n": "  Problema: API Key no trobada a l'entorn o la configuraci\xF3\n",
  "  Run `qwen auth coding-plan` to re-configure.\n": "  Executeu `qwen auth coding-plan` per tornar a configurar.\n",
  "\u2713 Authentication Method: {{type}}": "\u2713 M\xE8tode d'autenticaci\xF3: {{type}}",
  "  Status: Configured\n": "  Estat: Configurat\n",
  "Failed to check authentication status: {{error}}": "Error en comprovar l'estat d'autenticaci\xF3: {{error}}",
  "Select an option:": "Seleccioneu una opci\xF3:",
  "Raw mode not available. Please run in an interactive terminal.": "El mode raw no est\xE0 disponible. Executeu en un terminal interactiu.",
  "(Use \u2191 \u2193 arrows to navigate, Enter to select, Ctrl+C to exit)\n": "(Useu les fletxes \u2191 \u2193 per navegar, Enter per seleccionar, Ctrl+C per sortir)\n",
  "Hide tool output and thinking for a cleaner view (toggle with Ctrl+O).": "Amagueu la sortida de les eines i el pensament per a una vista m\xE9s neta (canvieu amb Ctrl+O).",
  "Press Ctrl+O to show full tool output": "Premeu Ctrl+O per mostrar la sortida completa de les eines",
  "Switch to plan mode or exit plan mode": "Canviar al mode de planificaci\xF3 o sortir del mode de planificaci\xF3",
  "Exited plan mode. Previous approval mode restored.": "S'ha sortit del mode de planificaci\xF3. S'ha restaurat el mode d'aprovaci\xF3 anterior.",
  "Enabled plan mode. The agent will analyze and plan without executing tools.": "S'ha activat el mode de planificaci\xF3. L'agent analitzar\xE0 i planificar\xE0 sense executar eines.",
  'Already in plan mode. Use "/plan exit" to exit plan mode.': 'Ja esteu en mode de planificaci\xF3. Useu "/plan exit" per sortir del mode de planificaci\xF3.',
  'Not in plan mode. Use "/plan" to enter plan mode first.': 'No esteu en mode de planificaci\xF3. Useu "/plan" per entrar al mode de planificaci\xF3 primer.',
  "Set up Qwen Code's status line UI": "Configurar la interf\xEDcie de la barra d'estat de Qwen Code",
  // === Core: added from PR #3328 ===
  "Open the memory manager.": "Obrir el gestor de mem\xF2ria.",
  "Save a durable memory to the memory system.": "Desar una mem\xF2ria duradora al sistema de mem\xF2ria.",
  "Ask a quick side question without affecting the main conversation": "Fer una pregunta r\xE0pida sense afectar la conversa principal",
  "Browser-based authentication with third-party providers (e.g. OpenRouter, ModelScope)": "Autenticaci\xF3 basada en navegador amb prove\xEFdors de tercers (p. ex. OpenRouter, ModelScope)",
  "Manage Arena sessions": "Gestionar sessions d'Arena",
  "Start an Arena session with multiple models competing on the same task": "Iniciar una sessi\xF3 d'Arena amb m\xFAltiples models competint en la mateixa tasca",
  "Stop the current Arena session": "Aturar la sessi\xF3 d'Arena actual",
  "Show the current Arena session status": "Mostrar l'estat de la sessi\xF3 d'Arena actual",
  "Select a model result and merge its diff into the current workspace": "Seleccionar un resultat de model i fusionar-ne el diff a l'espai de treball actual",
  "No running Arena session found.": "No s'ha trobat cap sessi\xF3 d'Arena en execuci\xF3.",
  "No Arena session found. Start one with /arena start.": "No s'ha trobat cap sessi\xF3 d'Arena. Inicieu-ne una amb /arena start.",
  "Arena session is still running. Wait for it to complete or use /arena stop first.": "La sessi\xF3 d'Arena encara s'est\xE0 executant. Espereu que finalitzi o utilitzeu primer /arena stop.",
  "No successful agent results to select from. All agents failed or were cancelled.": "No hi ha resultats d'agent amb \xE8xit per seleccionar. Tots els agents han fallat o s'han cancel\xB7lat.",
  "Use /arena stop to end the session.": "Utilitzeu /arena stop per finalitzar la sessi\xF3.",
  'No idle agent found matching "{{name}}".': `No s'ha trobat cap agent inactiu que coincideixi amb "{{name}}".`,
  "Failed to apply changes from {{label}}: {{error}}": "No s'han pogut aplicar els canvis de {{label}}: {{error}}",
  "Applied changes from {{label}} to workspace. Arena session complete.": "S'han aplicat els canvis de {{label}} a l'espai de treball. Sessi\xF3 d'Arena completada.",
  "Discard all Arena results and clean up worktrees?": "Descartar tots els resultats d'Arena i netejar els arbres de treball?",
  "Arena results discarded. All worktrees cleaned up.": "Resultats d'Arena descartats. S'han netejat tots els arbres de treball.",
  "Arena is not supported in non-interactive mode. Use interactive mode to start an Arena session.": "Arena no \xE9s compatible amb el mode no interactiu. Utilitzeu el mode interactiu per iniciar una sessi\xF3 d'Arena.",
  "Arena is not supported in non-interactive mode. Use interactive mode to stop an Arena session.": "Arena no \xE9s compatible amb el mode no interactiu. Utilitzeu el mode interactiu per aturar una sessi\xF3 d'Arena.",
  "Arena is not supported in non-interactive mode.": "Arena no \xE9s compatible amb el mode no interactiu.",
  "An Arena session exists. Use /arena stop or /arena select to end it before starting a new one.": "Ja existeix una sessi\xF3 d'Arena. Utilitzeu /arena stop o /arena select per finalitzar-la abans d'iniciar-ne una de nova.",
  "Usage: /arena start --models model1,model2 <task>": "\xDAs: /arena start --models model1,model2 <tasca>",
  "Models to compete (required, at least 2)": "Models per competir (obligatori, almenys 2)",
  "Format: authType:modelId or just modelId": "Format: authType:modelId o nom\xE9s modelId",
  "Arena requires at least 2 models. Use --models model1,model2 to specify.": "Arena requereix almenys 2 models. Utilitzeu --models model1,model2 per especificar-los.",
  'Arena started with {{count}} agents on task: "{{task}}"\nModels:\n{{modelList}}': 'Arena iniciada amb {{count}} agents a la tasca: "{{task}}"\nModels:\n{{modelList}}',
  "Arena panes are running in tmux. Attach with: `{{command}}`": "Els panells d'Arena s'estan executant a tmux. Adjunteu-vos amb: `{{command}}`",
  "[{{label}}] failed: {{error}}": "[{{label}}] ha fallat: {{error}}",
  "Loading suggestions...": "S'estan carregant els suggeriments...",
  "Show per-item context usage breakdown.": "Mostrar el desglossament de l'\xFAs del context per element.",
  "Lock release warning": "Av\xEDs d'alliberament del bloqueig",
  "Metadata write warning": "Av\xEDs d'escriptura de metadades",
  "Subsequent dreams may be skipped as locked until the next session's staleness sweep cleans the file.": "Els dreams posteriors es poden ometre com a bloquejats fins que la propera neteja de sessions obsoletes elimini el fitxer.",
  "The scheduler gate did not see this dream's timestamp; the next dream cycle may re-fire sooner than usual.": "La porta del planificador no ha vist la marca de temps d'aquest dream; el proper cicle de dream es pot tornar a executar abans del normal.",
  "Manage extension settings": "Gestionar la configuraci\xF3 de les extensions",
  "Desc:": "Descripci\xF3:",
  "Ref:": "Refer\xE8ncia:",
  "\u4E2D\u56FD (China)": "Xina",
  "\u4E2D\u56FD (China) - \u963F\u91CC\u4E91\u767E\u70BC": "Xina - \u963F\u91CC\u4E91\u767E\u70BC",
  // Stats Dashboard — Category 2
  "Activity Heatmap": "Mapa d'activitat",
  Less: "Menys",
  More: "M\xE9s",
  Sessions: "Sessions",
  Duration: "Durada",
  Projects: "Projectes",
  "Loading stats...": "Carregant estad\xEDstiques...",
  "(no data)": "(sense dades)",
  d: "d",
  h: "h",
  m: "m",
  Input: "Entrada",
  Models: "Models",
  "All time": "Tot el temps",
  "Last 7 days": "\xDAltims 7 dies",
  "Last 30 days": "\xDAltims 30 dies",
  "Show usage statistics dashboard.": "Mostra el tauler d'estad\xEDstiques d'\xFAs.",
  // Stats Dashboard — keyboard hints (not translated)
  "tab \xB7 esc": "tab \xB7 esc",
  "tab \xB7 r dates \xB7 \u2190\u2192 month \xB7 esc": "tab \xB7 r dates \xB7 \u2190\u2192 month \xB7 esc",
  "tab \xB7 r dates \xB7 esc": "tab \xB7 r dates \xB7 esc",
  // Stats Dashboard — missing labels
  "API Requests": "Sol\xB7licituds d'API",
  "Tool Calls": "Crides d'eines",
  "Success rate": "Taxa d'\xE8xit",
  "Code Changes": "Canvis de codi",
  Tool: "Eina",
  reqs: "sol.",
  in: "ent.",
  out: "sort.",
  "In/Out": "Ent/Sort",
  // === History collapse/expand commands ===
  "Set history to collapse by default when resuming a session": "Establir la hist\xF2ria per reduir-se per defecte en reprendre una sessi\xF3",
  "Set history to expand by default when resuming a session": "Establir la hist\xF2ria per expandir-se per defecte en reprendre una sessi\xF3",
  "Expand the currently collapsed history transcript": "Expandir la transcripci\xF3 de la hist\xF2ria actualment redu\xEFda",
  "Control history display preferences and visibility": "Controlar les prefer\xE8ncies de visualitzaci\xF3 de la hist\xF2ria i la visibilitat",
  "History will be collapsed by default for future resumed sessions.": "La hist\xF2ria es reduir\xE0 per defecte per a futures sessions represes.",
  "History will be expanded by default for future resumed sessions.": "La hist\xF2ria s'expandir\xE0 per defecte per a futures sessions represes.",
  "History is already expanded in this session.": "La hist\xF2ria ja est\xE0 expandida en aquesta sessi\xF3.",
  "Usage: /history collapse-on-resume|expand-on-resume|expand-now": "\xDAs: /history collapse-on-resume|expand-on-resume|expand-now",
  "History collapsed: {{n}} messages hidden. Use /history expand-now to show.": "Hist\xF2ria redu\xEFda: {{n}} missatges ocults. Utilitzeu /history expand-now per mostrar."
};
export {
  ca_default as default
};
/**
 * @license
 * Copyright 2025 Qwen
 * SPDX-License-Identifier: Apache-2.0
 */
