// @qwen-code/qwen-code/chunks/chunk-BMFMJINR.js
// Force strict mode and setup for ESM
"use strict";
import {
  init_esbuild_shims
} from "./chunk-A4BMJM77.js";
import {
  __name
} from "./chunk-J2S4EL5Y.js";

// packages/cli/src/serve/workspace-service/types.ts
init_esbuild_shims();
var WorkspaceSettingsPartialPersistError = class extends Error {
  static {
    __name(this, "WorkspaceSettingsPartialPersistError");
  }
  committedWrites;
  cause;
  constructor(message, committedWrites, cause) {
    super(message);
    this.name = "WorkspaceSettingsPartialPersistError";
    this.committedWrites = committedWrites;
    this.cause = cause;
  }
};
var WorkspacePermissionRulesSessionRequiredError = class extends Error {
  static {
    __name(this, "WorkspacePermissionRulesSessionRequiredError");
  }
  constructor() {
    super(
      "setWorkspacePermissionRules requires a live ACP session to update active permission rules"
    );
    this.name = "WorkspacePermissionRulesSessionRequiredError";
  }
};

export {
  WorkspaceSettingsPartialPersistError,
  WorkspacePermissionRulesSessionRequiredError
};
/**
 * @license
 * Copyright 2025 Qwen Team
 * SPDX-License-Identifier: Apache-2.0
 */
