// @qwen-code/qwen-code/chunks/chunk-L7OKCI7T.js
// Force strict mode and setup for ESM
"use strict";
import {
  SkillError
} from "./chunk-YAINVQ3Q.js";
import {
  init_esbuild_shims
} from "./chunk-A4BMJM77.js";
import {
  __name
} from "./chunk-J2S4EL5Y.js";

// packages/acp-bridge/src/status.ts
init_esbuild_shims();
var STATUS_SCHEMA_VERSION = 1;
var BridgeTimeoutError = class extends Error {
  static {
    __name(this, "BridgeTimeoutError");
  }
  label;
  timeoutMs;
  constructor(label, timeoutMs) {
    super(`AcpSessionBridge ${label} timed out after ${timeoutMs}ms`);
    this.name = "BridgeTimeoutError";
    this.label = label;
    this.timeoutMs = timeoutMs;
  }
};
var BridgeChannelClosedError = class extends Error {
  static {
    __name(this, "BridgeChannelClosedError");
  }
  context;
  constructor(context) {
    super(`agent channel closed ${context}`);
    this.name = "BridgeChannelClosedError";
    this.context = context;
  }
};
var MissingCliEntryError = class extends Error {
  static {
    __name(this, "MissingCliEntryError");
  }
  constructor() {
    super(
      "Cannot determine CLI entry path for spawning the ACP child: process.argv[1] is empty and QWEN_CLI_ENTRY is unset. Set QWEN_CLI_ENTRY to the absolute path of the qwen entry script (e.g. `export QWEN_CLI_ENTRY=$(which qwen)`) to override."
    );
    this.name = "MissingCliEntryError";
  }
};
var SERVE_STATUS_EXT_METHODS = {
  workspaceMcp: "qwen/status/workspace/mcp",
  workspaceMcpTools: "qwen/status/workspace/mcp/tools",
  workspaceMcpResources: "qwen/status/workspace/mcp/resources",
  workspaceSkills: "qwen/status/workspace/skills",
  workspaceTools: "qwen/status/workspace/tools",
  workspaceProviders: "qwen/status/workspace/providers",
  workspaceMemory: "qwen/status/workspace/memory",
  workspaceAgents: "qwen/status/workspace/agents",
  workspacePreflight: "qwen/status/workspace/preflight",
  sessionContext: "qwen/status/session/context",
  sessionContextUsage: "qwen/status/session/context_usage",
  sessionSupportedCommands: "qwen/status/session/supported_commands",
  sessionTasks: "qwen/status/session/tasks",
  sessionStats: "qwen/status/session/stats",
  sessionLspStatus: "qwen/status/session/lsp",
  sessionRewindSnapshots: "qwen/status/session/rewind_snapshots",
  workspaceHooks: "qwen/status/workspace/hooks",
  sessionHooks: "qwen/status/session/hooks",
  workspaceExtensions: "qwen/status/workspace/extensions"
};
var SERVE_CONTROL_EXT_METHODS = {
  sessionClose: "qwen/control/session/close",
  sessionApprovalMode: "qwen/control/session/approval_mode",
  sessionBranch: "qwen/control/session/branch",
  sessionForkAgent: "qwen/control/session/fork_agent",
  sessionRecap: "qwen/control/session/recap",
  sessionBtw: "qwen/control/session/btw",
  sessionShellHistory: "qwen/control/session/shell_history",
  sessionLanguage: "qwen/control/session/language",
  sessionRewind: "qwen/control/session/rewind",
  sessionTitle: "qwen/control/session/title",
  workspaceMcpRestart: "qwen/control/workspace/mcp/restart",
  workspaceMcpManage: "qwen/control/workspace/mcp/manage",
  workspaceAgentGenerate: "qwen/control/workspace/agents/generate",
  // Runtime MCP server mutation ext-methods
  sessionTaskCancel: "qwen/control/session/task/cancel",
  sessionGoalClear: "qwen/control/session/goal/clear",
  workspaceMcpRuntimeAdd: "qwen/control/workspace/mcp/runtime-add",
  workspaceMcpRuntimeRemove: "qwen/control/workspace/mcp/runtime-remove",
  workspaceReload: "qwen/control/workspace/reload",
  workspaceExtensionsRefresh: "qwen/control/workspace/extensions/refresh",
  sessionCd: "qwen/control/session/cd"
};
var IDLE_HOOK_EVENTS = {
  PreToolUse: { description: "Before tool execution", matcherKind: "toolName" },
  PostToolUse: { description: "After tool execution", matcherKind: "toolName" },
  PostToolUseFailure: {
    description: "After tool execution fails",
    matcherKind: "toolName"
  },
  PostToolBatch: { description: "After a batch of tool calls resolves" },
  Notification: {
    description: "When notifications are sent",
    matcherKind: "notificationType"
  },
  UserPromptSubmit: { description: "When the user submits a prompt" },
  UserPromptExpansion: {
    description: "When a slash command expands into a prompt",
    matcherKind: "commandName"
  },
  SessionStart: {
    description: "When a new session is started",
    matcherKind: "sessionTrigger"
  },
  Stop: { description: "Right before Qwen Code concludes its response" },
  SubagentStart: {
    description: "When a subagent is started",
    matcherKind: "agentType"
  },
  SubagentStop: {
    description: "Right before a subagent concludes its response",
    matcherKind: "agentType"
  },
  PreCompact: {
    description: "Before conversation compaction",
    matcherKind: "trigger"
  },
  PostCompact: {
    description: "After conversation compaction",
    matcherKind: "trigger"
  },
  SessionEnd: {
    description: "When a session is ending",
    matcherKind: "sessionTrigger"
  },
  PermissionRequest: {
    description: "When a permission dialog is displayed",
    matcherKind: "toolName"
  },
  PermissionDenied: {
    description: "When a tool call is denied",
    matcherKind: "toolName"
  },
  StopFailure: {
    description: "When the turn ends due to an API error",
    matcherKind: "error"
  },
  TodoCreated: { description: "When a new todo item is created" },
  TodoCompleted: { description: "When a todo item is marked as completed" },
  InstructionsLoaded: {
    description: "When an instruction or context file is loaded",
    matcherKind: "filePath"
  }
};
function createIdleWorkspaceExtensionsStatus(workspaceCwd) {
  return {
    v: STATUS_SCHEMA_VERSION,
    workspaceCwd,
    initialized: false,
    extensions: []
  };
}
__name(createIdleWorkspaceExtensionsStatus, "createIdleWorkspaceExtensionsStatus");
function createIdleWorkspaceHooksStatus(workspaceCwd) {
  return {
    v: STATUS_SCHEMA_VERSION,
    workspaceCwd,
    initialized: false,
    disabled: false,
    hooks: [],
    events: IDLE_HOOK_EVENTS
  };
}
__name(createIdleWorkspaceHooksStatus, "createIdleWorkspaceHooksStatus");
function createIdleWorkspaceMemoryStatus(workspaceCwd) {
  return {
    v: STATUS_SCHEMA_VERSION,
    workspaceCwd,
    initialized: false,
    files: [],
    totalBytes: 0,
    fileCount: 0,
    ruleCount: 0
  };
}
__name(createIdleWorkspaceMemoryStatus, "createIdleWorkspaceMemoryStatus");
function createIdleWorkspaceMcpStatus(workspaceCwd) {
  return {
    v: STATUS_SCHEMA_VERSION,
    workspaceCwd,
    initialized: false,
    discoveryState: "not_started",
    servers: [],
    clientCount: 0,
    budgetMode: "off",
    budgets: []
  };
}
__name(createIdleWorkspaceMcpStatus, "createIdleWorkspaceMcpStatus");
function createIdleWorkspaceSkillsStatus(workspaceCwd) {
  return {
    v: STATUS_SCHEMA_VERSION,
    workspaceCwd,
    initialized: false,
    skills: []
  };
}
__name(createIdleWorkspaceSkillsStatus, "createIdleWorkspaceSkillsStatus");
function createIdleWorkspaceProvidersStatus(workspaceCwd) {
  return {
    v: STATUS_SCHEMA_VERSION,
    workspaceCwd,
    initialized: false,
    providers: []
  };
}
__name(createIdleWorkspaceProvidersStatus, "createIdleWorkspaceProvidersStatus");
function createIdleEnvStatus(workspaceCwd, acpChannelLive) {
  return {
    v: STATUS_SCHEMA_VERSION,
    workspaceCwd,
    initialized: true,
    acpChannelLive,
    cells: []
  };
}
__name(createIdleEnvStatus, "createIdleEnvStatus");
var ACP_PREFLIGHT_KINDS = [
  "auth",
  "mcp_discovery",
  "skills",
  "providers",
  "tool_registry",
  "egress"
];
function createIdleAcpPreflightCells() {
  return ACP_PREFLIGHT_KINDS.map((kind) => ({
    kind,
    status: "not_started",
    locality: "acp",
    hint: "spawn a session to populate"
  }));
}
__name(createIdleAcpPreflightCells, "createIdleAcpPreflightCells");
var SKILL_PARSE_CODES = /* @__PURE__ */ new Set([
  "PARSE_ERROR",
  "INVALID_CONFIG",
  "INVALID_NAME"
]);
var SKILL_FILE_CODES = /* @__PURE__ */ new Set([
  "FILE_ERROR",
  "NOT_FOUND"
]);
var FS_MISSING_CODES = /* @__PURE__ */ new Set([
  "ENOENT",
  "EACCES",
  "EPERM"
]);
var MODEL_CONFIG_ERROR_NAMES = /* @__PURE__ */ new Set([
  "StrictMissingCredentialsError",
  "StrictMissingModelIdError",
  "MissingApiKeyError",
  "MissingModelError",
  "MissingBaseUrlError",
  "MissingAnthropicBaseUrlEnvError"
]);
function mapDomainErrorToErrorKind(err) {
  if (err instanceof BridgeTimeoutError) return "init_timeout";
  if (err instanceof BridgeChannelClosedError) return "protocol_error";
  if (err instanceof MissingCliEntryError) return "missing_binary";
  if (err instanceof SkillError || err?.name === "SkillError") {
    const code2 = err.code;
    if (code2 && SKILL_PARSE_CODES.has(code2)) return "parse_error";
    if (code2 && SKILL_FILE_CODES.has(code2)) return "missing_file";
    return void 0;
  }
  if (err instanceof SyntaxError) return "parse_error";
  if (!(err instanceof Error)) return void 0;
  if (err.name === "TrustGateError") return "auth_env_error";
  if (MODEL_CONFIG_ERROR_NAMES.has(err.name)) return "auth_env_error";
  const code = err.code;
  if (typeof code === "string" && FS_MISSING_CODES.has(code)) {
    return "missing_file";
  }
  return void 0;
}
__name(mapDomainErrorToErrorKind, "mapDomainErrorToErrorKind");

export {
  STATUS_SCHEMA_VERSION,
  BridgeTimeoutError,
  BridgeChannelClosedError,
  MissingCliEntryError,
  SERVE_STATUS_EXT_METHODS,
  SERVE_CONTROL_EXT_METHODS,
  IDLE_HOOK_EVENTS,
  createIdleWorkspaceExtensionsStatus,
  createIdleWorkspaceHooksStatus,
  createIdleWorkspaceMemoryStatus,
  createIdleWorkspaceMcpStatus,
  createIdleWorkspaceSkillsStatus,
  createIdleWorkspaceProvidersStatus,
  createIdleEnvStatus,
  ACP_PREFLIGHT_KINDS,
  createIdleAcpPreflightCells,
  mapDomainErrorToErrorKind
};
/**
 * @license
 * Copyright 2025 Qwen Team
 * SPDX-License-Identifier: Apache-2.0
 */
