// @qwen-code/qwen-code/chunks/de-EOW5PVHF.js
// Force strict mode and setup for ESM
"use strict";
import {
  init_esbuild_shims
} from "./chunk-A4BMJM77.js";
import "./chunk-J2S4EL5Y.js";

// packages/cli/src/i18n/locales/de.js
init_esbuild_shims();
var de_default = {
  // ============================================================================
  // Help / UI Components
  // ============================================================================
  // Attachment hints
  "\u2191 to manage attachments": "\u2191 Anh\xE4nge verwalten",
  "\u2190 \u2192 select, Delete to remove, \u2193 to exit": "\u2190 \u2192 ausw\xE4hlen, Delete zum L\xF6schen, \u2193 beenden",
  "Attachments: ": "Anh\xE4nge: ",
  "Basics:": "Grundlagen:",
  "Add context": "Kontext hinzuf\xFCgen",
  "Use {{symbol}} to specify files for context (e.g., {{example}}) to target specific files or folders.": "Verwenden Sie {{symbol}}, um Dateien als Kontext anzugeben (z.B. {{example}}), um bestimmte Dateien oder Ordner auszuw\xE4hlen.",
  "@": "@",
  "@src/myFile.ts": "@src/myFile.ts",
  "Shell mode": "Shell-Modus",
  "YOLO mode": "YOLO-Modus",
  "Auto mode": "Auto-Modus",
  "plan mode": "Planungsmodus",
  "auto-accept edits": "\xC4nderungen automatisch akzeptieren",
  "Accepting edits": "\xC4nderungen werden akzeptiert",
  "(shift + tab to cycle)": "(Shift + Tab zum Wechseln)",
  "(tab to cycle)": "(Tab zum Wechseln)",
  "Execute shell commands via {{symbol}} (e.g., {{example1}}) or use natural language (e.g., {{example2}}).": "Shell-Befehle \xFCber {{symbol}} ausf\xFChren (z.B. {{example1}}) oder nat\xFCrliche Sprache verwenden (z.B. {{example2}}).",
  "!": "!",
  "!npm run start": "!npm run start",
  "start server": "Server starten",
  "Commands:": "Befehle:",
  "shell command": "Shell-Befehl",
  "Model Context Protocol command (from external servers)": "Model Context Protocol Befehl (von externen Servern)",
  "Keyboard Shortcuts:": "Tastenk\xFCrzel:",
  "Jump through words in the input": "W\xF6rter in der Eingabe \xFCberspringen",
  "Close dialogs, cancel requests, or quit application": "Dialoge schlie\xDFen, Anfragen abbrechen oder Anwendung beenden",
  "New line": "Neue Zeile",
  "New line (Alt+Enter works for certain linux distros)": "Neue Zeile (Alt+Enter funktioniert bei bestimmten Linux-Distributionen)",
  "Clear the screen": "Bildschirm l\xF6schen",
  "Open input in external editor": "Eingabe in externem Editor \xF6ffnen",
  "Send message": "Nachricht senden",
  "Initializing...": "Initialisierung...",
  "Connecting to MCP servers... ({{connected}}/{{total}})": "Verbindung zu MCP servers wird hergestellt... ({{connected}}/{{total}})",
  "Type your message or @path/to/file": "Nachricht eingeben oder @Pfad/zur/Datei",
  "Press 'i' for INSERT mode and 'Esc' for NORMAL mode.": "Dr\xFCcken Sie 'i' f\xFCr den EINF\xDCGE-Modus und 'Esc' f\xFCr den NORMAL-Modus.",
  "Cancel operation / Clear input (double press)": "Vorgang abbrechen / Eingabe l\xF6schen (doppelt dr\xFCcken)",
  "Cycle approval modes": "Genehmigungsmodi durchschalten",
  "Cycle through your prompt history": "Eingabeverlauf durchbl\xE4ttern",
  "For a full list of shortcuts, see {{docPath}}": "Eine vollst\xE4ndige Liste der Tastenk\xFCrzel finden Sie unter {{docPath}}",
  "docs/keyboard-shortcuts.md": "docs/keyboard-shortcuts.md",
  "for help on Qwen Code": "f\xFCr Hilfe zu Qwen Code",
  "show version info": "Versionsinformationen anzeigen",
  "submit a bug report": "Fehlerbericht einreichen",
  Status: "Status",
  // ============================================================================
  // System Information Fields
  // ============================================================================
  "Qwen Code": "Qwen Code",
  Runtime: "Laufzeit",
  OS: "Betriebssystem",
  Auth: "Authentifizierung",
  Model: "Modell",
  "Fast Model": "Schnelles Modell",
  Sandbox: "Sandbox",
  "Session ID": "Sitzungs-ID",
  "Base URL": "Base URL",
  Proxy: "Proxy",
  "Memory Usage": "Speichernutzung",
  "IDE Client": "IDE-Client",
  // ============================================================================
  // Commands - General
  // ============================================================================
  "Analyzes the project and creates a tailored QWEN.md file.": "Analysiert das Projekt und erstellt eine ma\xDFgeschneiderte QWEN.md-Datei.",
  "List available Qwen Code tools. Usage: /tools [desc]": "Verf\xFCgbare Qwen Code Werkzeuge auflisten. Verwendung: /tools [desc]",
  "Open the skills panel (browse, search, toggle, pick).": "Skills-Panel \xF6ffnen (durchsuchen, suchen, ein/aus, ausw\xE4hlen).",
  "Manage Skills": "Skills verwalten",
  "Skills configuration saved.": "Skills-Konfiguration gespeichert.",
  "Skills configuration saved, but refresh failed: {{error}}. Restart to ensure the new state is applied.": "Skills-Konfiguration gespeichert, aber Aktualisierung fehlgeschlagen: {{error}}. Bitte neu starten, um den neuen Zustand zu \xFCbernehmen.",
  "Workspace is untrusted; workspace settings are ignored by the merged config. Run /trust first to persist skills changes here, or edit ~/.qwen/settings.json directly to manage skills at user scope.": "Arbeitsbereich ist nicht vertrauensw\xFCrdig; Arbeitsbereichseinstellungen werden in der zusammengef\xFChrten Konfiguration ignoriert. F\xFChre zuerst /trust aus oder bearbeite ~/.qwen/settings.json direkt, um Skills auf Benutzerebene zu verwalten.",
  "SkillManager not available.": "SkillManager nicht verf\xFCgbar.",
  "Loading skills\u2026": "Skills werden geladen\u2026",
  "Failed to load skills: {{error}}": "Skills konnten nicht geladen werden: {{error}}",
  "Failed to save skills configuration: {{error}}": "Speichern der Skill-Konfiguration fehlgeschlagen: {{error}}",
  "All available skills are disabled. Edit ~/.qwen/settings.json or .qwen/settings.json (skills.disabled) to re-enable.": "Alle verf\xFCgbaren Skills sind deaktiviert. Bearbeite ~/.qwen/settings.json oder .qwen/settings.json (skills.disabled), um sie wieder zu aktivieren.",
  "Press esc to close.": "Esc dr\xFCcken, um zu schlie\xDFen.",
  "{{count}} skills \xB7 ": "{{count}} Skills \xB7 ",
  "{{matched}} / {{total}} skills \xB7 ": "{{matched}} / {{total}} Skills \xB7 ",
  "Space toggle \xB7 Enter pick (fill input) \xB7 Esc save & exit \xB7 workspace scope": "Leertaste umschalten \xB7 Enter ausw\xE4hlen (in Eingabe) \xB7 Esc speichern & beenden \xB7 Arbeitsbereich",
  "Search:": "Suche:",
  "type to filter\u2026": "Tippen zum Filtern\u2026",
  "No skills are currently available.": "Derzeit sind keine Skills verf\xFCgbar.",
  "All available skills are locked at a higher scope (see below).": "Alle verf\xFCgbaren Skills sind in einer h\xF6heren Ebene gesperrt (siehe unten).",
  "No skills match the search.": "Keine Skills passen zur Suche.",
  "Locked by higher-scope settings (cannot toggle here):": "Gesperrt durch Einstellungen einer h\xF6heren Ebene (kann hier nicht umgeschaltet werden):",
  "higher scope": "h\xF6here Ebene",
  "  {{name}} {{description}}  [locked: {{scope}}]": "  {{name}} {{description}}  [gesperrt: {{scope}}]",
  "\u2191/\u2193 navigate \xB7 backspace edits search": "\u2191/\u2193 navigieren \xB7 R\xFCcktaste bearbeitet Suche",
  Bundled: "Mitgeliefert",
  "Available Qwen Code CLI tools:": "Verf\xFCgbare Qwen Code CLI-Werkzeuge:",
  "No tools available": "Keine Werkzeuge verf\xFCgbar",
  "View or change the approval mode for tool usage": "Genehmigungsmodus f\xFCr Werkzeugnutzung anzeigen oder \xE4ndern",
  "View or change the language setting": "Spracheinstellung anzeigen oder \xE4ndern",
  "List background tasks (text dump \u2014 interactive dialog opens via the footer pill)": "Hintergrundaufgaben auflisten (Textausgabe; der interaktive Dialog l\xE4sst sich \xFCber die Schaltfl\xE4che in der Fu\xDFzeile \xF6ffnen)",
  "Delete a previous session": "Eine fr\xFChere Sitzung l\xF6schen",
  "Run installation and environment diagnostics": "Installations- und Umgebungsdiagnosen ausf\xFChren",
  "Browse dynamic model catalogs and choose which models stay enabled locally": "Dynamische Modellkataloge durchsuchen und ausw\xE4hlen, welche Modelle lokal aktiviert bleiben",
  "Generate a one-line session recap now": "Jetzt eine einzeilige Sitzungszusammenfassung erstellen",
  "Rename the current conversation. --auto lets the fast model pick a title.": "Die aktuelle Unterhaltung umbenennen. Mit --auto l\xE4sst du das schnelle Modell einen Titel w\xE4hlen.",
  "Rewind conversation to a previous turn": "Die Unterhaltung auf einen fr\xFCheren Gespr\xE4chsschritt zur\xFCcksetzen",
  "Rewind Conversation": "Unterhaltung zur\xFCckspulen",
  "No user turns to rewind to.": "Keine Benutzerrunden zum Zur\xFCckspulen.",
  "Rewind to: ": "Zur\xFCckspulen zu: ",
  "Restore code and conversation": "Code und Unterhaltung wiederherstellen",
  "Restore conversation only": "Nur Unterhaltung wiederherstellen",
  "Restore code only": "Nur Code wiederherstellen",
  "Never mind": "Egal",
  "Computing file changes...": "Datei\xE4nderungen werden berechnet...",
  "Restoring...": "Wiederherstellung l\xE4uft...",
  "Restored {{count}} file(s).": "{{count}} Datei(en) wiederhergestellt.",
  "Failed to restore files: {{error}}": "Fehler beim Wiederherstellen der Dateien: {{error}}",
  "Rewind failed: {{error}}": "Zur\xFCckspulen fehlgeschlagen: {{error}}",
  "Cannot rewind conversation: no active model client.": "Konversation kann nicht zur\xFCckgespult werden: kein aktiver Modell-Client.",
  "Code restored, but conversation could not be rewound (no active client).": "Code wiederhergestellt, aber Konversation konnte nicht zur\xFCckgespult werden (kein aktiver Client).",
  "Conversation rewound. Edit your prompt and press Enter to continue.": "Konversation zur\xFCckgespult. Bearbeite deinen Prompt und dr\xFCcke Enter, um fortzufahren.",
  "Rewinding does not affect files edited manually or via shell commands.": "Das Zur\xFCckspulen wirkt sich nicht auf Dateien aus, die manuell oder per Shell-Befehl ge\xE4ndert wurden.",
  "Cannot rewind to a turn that was compressed. Try a more recent turn.": "Zu einem komprimierten Turn kann nicht zur\xFCckgespult werden. Bitte einen aktuelleren Turn versuchen.",
  "File restore is unavailable for this turn (no captured file changes, or this turn predates the current session).": "Datei-Wiederherstellung ist f\xFCr diesen Turn nicht verf\xFCgbar (keine erfassten Datei\xE4nderungen, oder dieser Turn liegt vor der aktuellen Sitzung).",
  "(+{{insertions}} -{{deletions}} in {{count}} file)": "(+{{insertions}} -{{deletions}} in {{count}} Datei)",
  "(+{{insertions}} -{{deletions}} in {{count}} files)": "(+{{insertions}} -{{deletions}} in {{count}} Dateien)",
  "Failed to restore {{count}} file(s): {{files}}": "{{count}} Datei(en) konnten nicht wiederhergestellt werden: {{files}}",
  "Cannot restore files: this turn was created before file checkpointing was enabled.": "Dateien k\xF6nnen nicht wiederhergestellt werden: Dieser Turn wurde erstellt, bevor Datei-Checkpointing aktiviert war.",
  "No files needed to be restored.": "Keine Dateien mussten wiederhergestellt werden.",
  "\u2191\u2193 to navigate \xB7 Enter to select \xB7 Esc to go back": "\u2191\u2193 navigieren \xB7 Enter ausw\xE4hlen \xB7 Esc zur\xFCck",
  "\u2191\u2193 to navigate \xB7 Enter to select \xB7 Esc to cancel": "\u2191\u2193 navigieren \xB7 Enter ausw\xE4hlen \xB7 Esc abbrechen",
  "Enter/Y to confirm \xB7 Esc/N to go back": "Enter/Y best\xE4tigen \xB7 Esc/N zur\xFCck",
  "change the theme": "Design \xE4ndern",
  "Select Theme": "Design ausw\xE4hlen",
  Preview: "Vorschau",
  "(Use Enter to select, Tab to configure scope)": "(Enter zum Ausw\xE4hlen, Tab zum Konfigurieren des Bereichs)",
  "(Use Enter to apply scope, Tab to go back)": "(Enter zum Anwenden des Bereichs, Tab zum Zur\xFCckgehen)",
  "Theme configuration unavailable due to NO_COLOR env variable.": "Design-Konfiguration aufgrund der NO_COLOR-Umgebungsvariable nicht verf\xFCgbar.",
  'Theme "{{themeName}}" not found.': 'Design "{{themeName}}" nicht gefunden.',
  'Theme "{{themeName}}" not found in selected scope.': 'Design "{{themeName}}" im ausgew\xE4hlten Bereich nicht gefunden.',
  "Clear conversation history and free up context": "Gespr\xE4chsverlauf l\xF6schen und Kontext freigeben",
  "Compresses the context by replacing it with a summary.": "Komprimiert den Kontext durch Ersetzen mit einer Zusammenfassung.",
  "open full Qwen Code documentation in your browser": "Vollst\xE4ndige Qwen Code Dokumentation im Browser \xF6ffnen",
  "Configuration not available.": "Konfiguration nicht verf\xFCgbar.",
  "Connect an LLM provider": "LLM-Anbieter verbinden",
  "Copy the last AI response to clipboard (/copy N for Nth-latest)": "Letzte KI-Antwort in die Zwischenablage kopieren (/copy N f\xFCr die N-letzte)",
  // ============================================================================
  // Commands - Agents
  // ============================================================================
  "Manage subagents for specialized task delegation.": "Unteragenten f\xFCr spezialisierte Aufgabendelegation verwalten.",
  "Manage existing subagents (view, edit, delete).": "Bestehende Unteragenten verwalten (anzeigen, bearbeiten, l\xF6schen).",
  "Create a new subagent with guided setup.": "Neuen Unteragenten mit gef\xFChrter Einrichtung erstellen.",
  // ============================================================================
  // Agents - Management Dialog
  // ============================================================================
  Agents: "Agenten",
  "Choose Action": "Aktion w\xE4hlen",
  "Edit {{name}}": "{{name}} bearbeiten",
  "Edit Tools: {{name}}": "Werkzeuge bearbeiten: {{name}}",
  "Edit Color: {{name}}": "Farbe bearbeiten: {{name}}",
  "Delete {{name}}": "{{name}} l\xF6schen",
  "Unknown Step": "Unbekannter Schritt",
  "Esc to close": "Esc zum Schlie\xDFen",
  "Enter to select, \u2191\u2193 to navigate, Esc to close": "Enter zum Ausw\xE4hlen, \u2191\u2193 zum Navigieren, Esc zum Schlie\xDFen",
  "Esc to go back": "Esc zum Zur\xFCckgehen",
  "Enter to confirm, Esc to cancel": "Enter zum Best\xE4tigen, Esc zum Abbrechen",
  "Enter to select, \u2191\u2193 to navigate, Esc to go back": "Enter zum Ausw\xE4hlen, \u2191\u2193 zum Navigieren, Esc zum Zur\xFCckgehen",
  "Enter to submit, Esc to go back": "Enter zum Absenden, Esc zum Zur\xFCckgehen",
  "Invalid step: {{step}}": "Ung\xFCltiger Schritt: {{step}}",
  "No subagents found.": "Keine Unteragenten gefunden.",
  "Use '/agents create' to create your first subagent.": "Verwenden Sie '/agents create', um Ihren ersten Unteragenten zu erstellen.",
  "(built-in)": "(integriert)",
  "(overridden by project level agent)": "(\xFCberschrieben durch Projektagent)",
  "Project Level ({{path}})": "Projektebene ({{path}})",
  "User Level ({{path}})": "Benutzerebene ({{path}})",
  "Built-in Agents": "Integrierte Agenten",
  "Extension Agents": "Erweiterungs-Agenten",
  "Using: {{count}} agents": "Verwendet: {{count}} Agenten",
  "View Agent": "Agent anzeigen",
  "Edit Agent": "Agent bearbeiten",
  "Delete Agent": "Agent l\xF6schen",
  Back: "Zur\xFCck",
  "No agent selected": "Kein Agent ausgew\xE4hlt",
  "File Path: ": "Dateipfad: ",
  "Tools: ": "Werkzeuge: ",
  "Color: ": "Farbe: ",
  "Description:": "Beschreibung:",
  "System Prompt:": "System-Prompt:",
  "Open in editor": "Im Editor \xF6ffnen",
  "Edit tools": "Werkzeuge bearbeiten",
  "Edit color": "Farbe bearbeiten",
  "\u274C Error:": "\u274C Fehler:",
  'Are you sure you want to delete agent "{{name}}"?': 'Sind Sie sicher, dass Sie den Agenten "{{name}}" l\xF6schen m\xF6chten?',
  // ============================================================================
  // Agents - Creation Wizard
  // ============================================================================
  "Project Level (.qwen/agents/)": "Projektebene (.qwen/agents/)",
  "User Level (~/.qwen/agents/)": "Benutzerebene (~/.qwen/agents/)",
  "\u2705 Subagent Created Successfully!": "\u2705 Unteragent erfolgreich erstellt!",
  'Subagent "{{name}}" has been saved to {{level}} level.': 'Unteragent "{{name}}" wurde auf {{level}}-Ebene gespeichert.',
  "Name: ": "Name: ",
  "Location: ": "Speicherort: ",
  "\u274C Error saving subagent:": "\u274C Fehler beim Speichern des Unteragenten:",
  "Warnings:": "Warnungen:",
  'Name "{{name}}" already exists at {{level}} level - will overwrite existing subagent': 'Name "{{name}}" existiert bereits auf {{level}}-Ebene - bestehender Unteragent wird \xFCberschrieben',
  'Name "{{name}}" exists at user level - project level will take precedence': 'Name "{{name}}" existiert auf Benutzerebene - Projektebene hat Vorrang',
  'Name "{{name}}" exists at project level - existing subagent will take precedence': 'Name "{{name}}" existiert auf Projektebene - bestehender Unteragent hat Vorrang',
  "Description is over {{length}} characters": "Beschreibung ist \xFCber {{length}} Zeichen",
  "System prompt is over {{length}} characters": "System-Prompt ist \xFCber {{length}} Zeichen",
  // Agents - Creation Wizard Steps
  "Step {{n}}: Choose Location": "Schritt {{n}}: Speicherort w\xE4hlen",
  "Step {{n}}: Choose Generation Method": "Schritt {{n}}: Generierungsmethode w\xE4hlen",
  "Generate with Qwen Code (Recommended)": "Mit Qwen Code generieren (Empfohlen)",
  "Manual Creation": "Manuelle Erstellung",
  "Describe what this subagent should do and when it should be used. (Be comprehensive for best results)": "Beschreiben Sie, was dieser Unteragent tun soll und wann er verwendet werden soll. (Ausf\xFChrliche Beschreibung f\xFCr beste Ergebnisse)",
  "e.g., Expert code reviewer that reviews code based on best practices...": "z.B. Experte f\xFCr Code-Reviews, der Code nach Best Practices \xFCberpr\xFCft...",
  "Generating subagent configuration...": "Unteragent-Konfiguration wird generiert...",
  "Failed to generate subagent: {{error}}": "Fehler beim Generieren des Unteragenten: {{error}}",
  "Step {{n}}: Describe Your Subagent": "Schritt {{n}}: Unteragent beschreiben",
  "Step {{n}}: Enter Subagent Name": "Schritt {{n}}: Unteragent-Name eingeben",
  "Step {{n}}: Enter System Prompt": "Schritt {{n}}: System-Prompt eingeben",
  "Step {{n}}: Enter Description": "Schritt {{n}}: Beschreibung eingeben",
  // Agents - Tool Selection
  "Step {{n}}: Select Tools": "Schritt {{n}}: Werkzeuge ausw\xE4hlen",
  "All Tools (Default)": "Alle Werkzeuge (Standard)",
  "All Tools": "Alle Werkzeuge",
  "Read-only Tools": "Nur-Lese-Werkzeuge",
  "Read & Edit Tools": "Lese- und Bearbeitungswerkzeuge",
  "Read & Edit & Execution Tools": "Lese-, Bearbeitungs- und Ausf\xFChrungswerkzeuge",
  "All tools selected, including MCP tools": "Alle Tools ausgew\xE4hlt, einschlie\xDFlich MCP tools",
  "Selected tools:": "Ausgew\xE4hlte Werkzeuge:",
  "Read-only tools:": "Nur-Lese-Werkzeuge:",
  "Edit tools:": "Bearbeitungswerkzeuge:",
  "Execution tools:": "Ausf\xFChrungswerkzeuge:",
  "Step {{n}}: Choose Background Color": "Schritt {{n}}: Hintergrundfarbe w\xE4hlen",
  "Step {{n}}: Confirm and Save": "Schritt {{n}}: Best\xE4tigen und Speichern",
  // Agents - Navigation & Instructions
  "Esc to cancel": "Esc zum Abbrechen",
  "Press Enter to save, e to save and edit, Esc to go back": "Enter zum Speichern, e zum Speichern und Bearbeiten, Esc zum Zur\xFCckgehen",
  "Press Enter to continue, {{navigation}}Esc to {{action}}": "Enter zum Fortfahren, {{navigation}}Esc zum {{action}}",
  cancel: "Abbrechen",
  "go back": "Zur\xFCckgehen",
  "\u2191\u2193 to navigate, ": "\u2191\u2193 zum Navigieren, ",
  "Enter a clear, unique name for this subagent.": "Geben Sie einen eindeutigen Namen f\xFCr diesen Unteragenten ein.",
  "e.g., Code Reviewer": "z.B. Code-Reviewer",
  "Name cannot be empty.": "Name darf nicht leer sein.",
  "Write the system prompt that defines this subagent's behavior. Be comprehensive for best results.": "Schreiben Sie den System-Prompt, der das Verhalten dieses Unteragenten definiert. Ausf\xFChrlich f\xFCr beste Ergebnisse.",
  "e.g., You are an expert code reviewer...": "z.B. Sie sind ein Experte f\xFCr Code-Reviews...",
  "System prompt cannot be empty.": "System-Prompt darf nicht leer sein.",
  "Describe when and how this subagent should be used.": "Beschreiben Sie, wann und wie dieser Unteragent verwendet werden soll.",
  "e.g., Reviews code for best practices and potential bugs.": "z.B. \xDCberpr\xFCft Code auf Best Practices und m\xF6gliche Fehler.",
  "Description cannot be empty.": "Beschreibung darf nicht leer sein.",
  "Failed to launch editor: {{error}}": "Fehler beim Starten des Editors: {{error}}",
  "Failed to save and edit subagent: {{error}}": "Fehler beim Speichern und Bearbeiten des Unteragenten: {{error}}",
  // ============================================================================
  // Commands - General (continued)
  // ============================================================================
  "View and edit Qwen Code settings": "Qwen Code Einstellungen anzeigen und bearbeiten",
  Settings: "Einstellungen",
  "To see changes, Qwen Code must be restarted. Press r to exit and apply changes now.": "Um \xC4nderungen zu sehen, muss Qwen Code neu gestartet werden. Dr\xFCcken Sie r, um jetzt zu beenden und \xC4nderungen anzuwenden.",
  // ============================================================================
  // Settings Labels
  // ============================================================================
  "Vim Mode": "Vim-Modus",
  "Attribution: commit": "Attribution: Commit",
  "Terminal Bell Notification": "Terminal-Signalton",
  "Enable Usage Statistics": "Nutzungsstatistiken aktivieren",
  Theme: "Farbschema",
  "Preferred Editor": "Bevorzugter Editor",
  "Auto-connect to IDE": "Automatische Verbindung zur IDE",
  "Debug Keystroke Logging": "Debug-Protokollierung von Tastatureingaben",
  "Language: UI": "Sprache: Benutzeroberfl\xE4che",
  "Language: Model": "Sprache: Modell",
  "Output Format": "Ausgabeformat",
  "Hide Window Title": "Fenstertitel ausblenden",
  "Show Status in Title": "Status im Titel anzeigen",
  "Hide Tips": "Tipps ausblenden",
  "Show Line Numbers in Code": "Zeilennummern im Code anzeigen",
  "Show Citations": "Quellenangaben anzeigen",
  "Custom Witty Phrases": "Benutzerdefinierte Witzige Spr\xFCche",
  "Show Welcome Back Dialog": "Willkommen-zur\xFCck-Dialog anzeigen",
  "Enable User Feedback": "Benutzerfeedback aktivieren",
  "How is Qwen doing this session? (optional)": "Wie macht sich Qwen in dieser Sitzung? (optional)",
  Bad: "Schlecht",
  Fine: "In Ordnung",
  Good: "Gut",
  Dismiss: "Ignorieren",
  "Screen Reader Mode": "Bildschirmleser-Modus",
  "Max Session Turns": "Maximale Sitzungsrunden",
  "Skip Next Speaker Check": "N\xE4chste-Sprecher-Pr\xFCfung \xFCberspringen",
  "Skip Loop Detection": "Schleifenerkennung \xFCberspringen",
  "Skip Startup Context": "Startkontext \xFCberspringen",
  "Enable OpenAI Logging": "OpenAI-Protokollierung aktivieren",
  "OpenAI Logging Directory": "OpenAI-Protokollierungsverzeichnis",
  Timeout: "Zeitlimit",
  "Max Retries": "Maximale Wiederholungen",
  "Load Memory From Include Directories": "Speicher aus Include-Verzeichnissen laden",
  "Respect .gitignore": ".gitignore beachten",
  "Respect .qwenignore": ".qwenignore beachten",
  "Enable Recursive File Search": "Rekursive Dateisuche aktivieren",
  "Interactive Shell (PTY)": "Interaktive Shell (PTY)",
  "Show Color": "Farbe anzeigen",
  "Auto Accept": "Automatisch akzeptieren",
  "Use Ripgrep": "Ripgrep verwenden",
  "Use Builtin Ripgrep": "Integriertes Ripgrep verwenden",
  "Tool Output Truncation Threshold": "Schwellenwert f\xFCr Werkzeugausgabe-K\xFCrzung",
  "Tool Output Truncation Lines": "Zeilen f\xFCr Werkzeugausgabe-K\xFCrzung",
  "Folder Trust": "Ordnervertrauen",
  "Tool Schema Compliance": "Tool Schema-Konformit\xE4t",
  // Settings enum options
  "Auto (detect from system)": "Automatisch (vom System erkennen)",
  "Auto (detect terminal theme)": "Automatisch (Terminal-Theme erkennen)",
  Auto: "Automatisch",
  Text: "Text",
  JSON: "JSON",
  Plan: "Plan",
  "Ask permissions": "Berechtigung anfragen",
  "Auto Edit": "Automatisch bearbeiten",
  YOLO: "YOLO",
  "toggle vim mode on/off": "Vim-Modus ein-/ausschalten",
  "Show model-specific usage statistics.": "Modellspezifische Nutzungsstatistiken anzeigen.",
  "Show tool-specific usage statistics.": "Werkzeugspezifische Nutzungsstatistiken anzeigen.",
  "Show daily token usage statistics.": "T\xE4gliche Token-Nutzungsstatistiken anzeigen.",
  "Show monthly token usage statistics.": "Monatliche Token-Nutzungsstatistiken anzeigen.",
  "Export token usage statistics to CSV or JSON.": "Token-Nutzungsstatistiken als CSV oder JSON exportieren.",
  "No usage data.": "Keine Nutzungsdaten.",
  "{{label}}: {{tokens}} tokens ({{requests}} requests)": "{{label}}: {{tokens}} Tokens ({{requests}} Anfragen)",
  "Daily token usage for {{value}}": "T\xE4gliche Token-Nutzung f\xFCr {{value}}",
  "Monthly token usage for {{value}}": "Monatliche Token-Nutzung f\xFCr {{value}}",
  "Total: {{tokens}} tokens": "Gesamt: {{tokens}} Tokens",
  "Requests: {{requests}}": "Anfragen: {{requests}}",
  "Breakdown:": "Aufschl\xFCsselung:",
  "Input: {{tokens}}": "Eingabe: {{tokens}}",
  "Output: {{tokens}}": "Ausgabe: {{tokens}}",
  "Cached (included in Input): {{tokens}}": "Cache (in Eingabe enthalten): {{tokens}}",
  "Thoughts: {{tokens}}": "Gedanken: {{tokens}}",
  "By model:": "Nach Modell:",
  "By auth type:": "Nach Authentifizierungstyp:",
  "By model/auth type:": "Nach Modell/Authentifizierungstyp:",
  "By source:": "Nach Quelle:",
  "Failed to load token usage stats: {{error}}": "Token-Nutzungsstatistiken konnten nicht geladen werden: {{error}}",
  "Expected --format csv or --format json.": "--format csv oder --format json erwartet.",
  "Expected a file path after --output.": "Nach --output wird ein Dateipfad erwartet.",
  "Unexpected argument: {{argument}}": "Unerwartetes Argument: {{argument}}",
  "Usage: /stats export <daily|monthly> [YYYY-MM-DD|YYYY-MM] [--format csv|json] [--output path]": "Verwendung: /stats export <daily|monthly> [YYYY-MM-DD|YYYY-MM] [--format csv|json] [--output path]",
  "Token usage export path must be within the project working directory.": "Der Exportpfad f\xFCr die Token-Nutzung muss im Arbeitsverzeichnis des Projekts liegen.",
  "Export target does not exist: {{path}}": "Exportziel existiert nicht: {{path}}",
  "Cannot resolve export path within the working directory.": "Der Exportpfad kann nicht innerhalb des Arbeitsverzeichnisses aufgel\xF6st werden.",
  "Could not create a temporary export file.": "Tempor\xE4re Exportdatei konnte nicht erstellt werden.",
  "Token usage exported to {{format}}: {{path}}": "Token-Nutzung nach {{format}} exportiert: {{path}}",
  "Failed to export token usage stats: {{error}}": "Token-Nutzungsstatistiken konnten nicht exportiert werden: {{error}}",
  "Unclosed quote in arguments.": "Nicht geschlossenes Anf\xFChrungszeichen in Argumenten.",
  "Note: generation timing (TTFT/TPS) belongs to generation metrics.": "Hinweis: Generierungszeiten (TTFT/TPS) geh\xF6ren zu den Generierungsmetriken.",
  "exit the cli": "CLI beenden",
  "Manage workspace directories": "Arbeitsbereichsverzeichnisse verwalten",
  "Add directories to the workspace. Use comma to separate multiple paths": "Verzeichnisse zum Arbeitsbereich hinzuf\xFCgen. Komma zum Trennen mehrerer Pfade verwenden",
  "Show all directories in the workspace": "Alle Verzeichnisse im Arbeitsbereich anzeigen",
  "set external editor preference": "Externen Editor festlegen",
  "Select Editor": "Editor ausw\xE4hlen",
  "Editor Preference": "Editor-Einstellung",
  "These editors are currently supported. Please note that some editors cannot be used in sandbox mode.": "Diese Editoren werden derzeit unterst\xFCtzt. Bitte beachten Sie, dass einige Editoren nicht im Sandbox-Modus verwendet werden k\xF6nnen.",
  "Your preferred editor is:": "Ihr bevorzugter Editor ist:",
  "Manage extensions": "Erweiterungen verwalten",
  "Manage installed extensions": "Installierte Erweiterungen verwalten",
  "Disable an extension": "Erweiterung deaktivieren",
  "Enable an extension": "Erweiterung aktivieren",
  "Install an extension from a git repo or local path": "Erweiterung aus Git-Repository oder lokalem Pfad installieren",
  "Uninstall an extension": "Erweiterung deinstallieren",
  "No extensions installed.": "Keine Erweiterungen installiert.",
  'Extension "{{name}}" not found.': 'Erweiterung "{{name}}" nicht gefunden.',
  "No extensions to update.": "Keine Erweiterungen zum Aktualisieren.",
  "Usage: /extensions install <source>": "Verwendung: /extensions install <Quelle>",
  'Installing extension from "{{source}}"...': 'Installiere Erweiterung von "{{source}}"...',
  'Extension "{{name}}" installed successfully.': 'Erweiterung "{{name}}" erfolgreich installiert.',
  'Failed to install extension from "{{source}}": {{error}}': 'Fehler beim Installieren der Erweiterung von "{{source}}": {{error}}',
  "Do you want to continue? [Y/n]: ": "M\xF6chten Sie fortfahren? [Y/n]: ",
  "Do you want to continue?": "M\xF6chten Sie fortfahren?",
  'Installing extension "{{name}}".': 'Erweiterung "{{name}}" wird installiert.',
  "**Extensions may introduce unexpected behavior. Ensure you have investigated the extension source and trust the author.**": "**Erweiterungen k\xF6nnen unerwartetes Verhalten verursachen. Stellen Sie sicher, dass Sie die Erweiterungsquelle untersucht haben und dem Autor vertrauen.**",
  "This extension will run the following MCP servers:": "Diese Erweiterung wird folgende MCP servers ausf\xFChren:",
  local: "lokal",
  "This extension will add the following commands: {{commands}}.": "Diese Erweiterung wird folgende Befehle hinzuf\xFCgen: {{commands}}.",
  "This extension will append info to your QWEN.md context using {{fileName}}": "Diese Erweiterung wird Informationen zu Ihrem QWEN.md-Kontext mit {{fileName}} hinzuf\xFCgen",
  "This extension will install the following skills:": "Diese Erweiterung wird folgende F\xE4higkeiten installieren:",
  "This extension will install the following subagents:": "Diese Erweiterung wird folgende Unteragenten installieren:",
  'Installation cancelled for "{{name}}".': 'Installation von "{{name}}" abgebrochen.',
  "You are installing an extension from {{originSource}}. Some features may not work perfectly with Qwen Code.": "Sie installieren eine Erweiterung von {{originSource}}. Einige Funktionen funktionieren m\xF6glicherweise nicht perfekt mit Qwen Code.",
  "--ref and --auto-update are not applicable for marketplace extensions.": "--ref und --auto-update sind nicht anwendbar f\xFCr Marketplace-Erweiterungen.",
  'Extension "{{name}}" installed successfully and enabled.': 'Erweiterung "{{name}}" erfolgreich installiert und aktiviert.',
  "The github URL, local path, or marketplace source (marketplace-url:plugin-name) of the extension to install.": "Die GitHub-URL, der lokale Pfad oder die Marketplace-Quelle (marketplace-url:plugin-name) der zu installierenden Erweiterung.",
  "The git ref to install from.": "Die Git-Referenz f\xFCr die Installation.",
  "Enable auto-update for this extension.": "Automatisches Update f\xFCr diese Erweiterung aktivieren.",
  "Enable pre-release versions for this extension.": "Pre-Release-Versionen f\xFCr diese Erweiterung aktivieren.",
  "Acknowledge the security risks of installing an extension and skip the confirmation prompt.": "Sicherheitsrisiken der Erweiterungsinstallation best\xE4tigen und Best\xE4tigungsaufforderung \xFCberspringen.",
  "The source argument must be provided.": "Das Quellargument muss angegeben werden.",
  'Extension "{{name}}" successfully uninstalled.': 'Erweiterung "{{name}}" erfolgreich deinstalliert.',
  "Uninstalls an extension.": "Deinstalliert eine Erweiterung.",
  "The name or source path of the extension to uninstall.": "Der Name oder Quellpfad der zu deinstallierenden Erweiterung.",
  "Please include the name of the extension to uninstall as a positional argument.": "Bitte geben Sie den Namen der zu deinstallierenden Erweiterung als Positionsargument an.",
  "Enables an extension.": "Aktiviert eine Erweiterung.",
  "The name of the extension to enable.": "Der Name der zu aktivierenden Erweiterung.",
  "The scope to enable the extenison in. If not set, will be enabled in all scopes.": "Der Bereich, in dem die Erweiterung aktiviert werden soll. Wenn nicht gesetzt, wird sie in allen Bereichen aktiviert.",
  'Extension "{{name}}" successfully enabled for scope "{{scope}}".': 'Erweiterung "{{name}}" erfolgreich f\xFCr Bereich "{{scope}}" aktiviert.',
  'Extension "{{name}}" successfully enabled in all scopes.': 'Erweiterung "{{name}}" erfolgreich in allen Bereichen aktiviert.',
  "Invalid scope: {{scope}}. Please use one of {{scopes}}.": "Ung\xFCltiger Bereich: {{scope}}. Bitte verwenden Sie einen von {{scopes}}.",
  "Disables an extension.": "Deaktiviert eine Erweiterung.",
  "The name of the extension to disable.": "Der Name der zu deaktivierenden Erweiterung.",
  "The scope to disable the extenison in.": "Der Bereich, in dem die Erweiterung deaktiviert werden soll.",
  'Extension "{{name}}" successfully disabled for scope "{{scope}}".': 'Erweiterung "{{name}}" erfolgreich f\xFCr Bereich "{{scope}}" deaktiviert.',
  'Extension "{{name}}" successfully updated: {{oldVersion}} \u2192 {{newVersion}}.': 'Erweiterung "{{name}}" erfolgreich aktualisiert: {{oldVersion}} \u2192 {{newVersion}}.',
  'Unable to install extension "{{name}}" due to missing install metadata': 'Erweiterung "{{name}}" kann aufgrund fehlender Installationsmetadaten nicht installiert werden',
  'Extension "{{name}}" is already up to date.': 'Erweiterung "{{name}}" ist bereits aktuell.',
  "Updates all extensions or a named extension to the latest version.": "Aktualisiert alle Erweiterungen oder eine benannte Erweiterung auf die neueste Version.",
  "The name of the extension to update.": "Der Name der zu aktualisierenden Erweiterung.",
  "Update all extensions.": "Alle Erweiterungen aktualisieren.",
  "Either an extension name or --all must be provided": "Entweder ein Erweiterungsname oder --all muss angegeben werden",
  "Lists installed extensions.": "Listet installierte Erweiterungen auf.",
  "Path:": "Pfad:",
  "Source:": "Quelle:",
  "Type:": "Typ:",
  "Release tag:": "Release-Tag:",
  "Enabled (User):": "Aktiviert (Benutzer):",
  "Enabled (Workspace):": "Aktiviert (Arbeitsbereich):",
  "Context files:": "Kontextdateien:",
  "MCP servers:": "MCP servers:",
  "Link extension failed to install.": "Verkn\xFCpfte Erweiterung konnte nicht installiert werden.",
  'Extension "{{name}}" linked successfully and enabled.': 'Erweiterung "{{name}}" erfolgreich verkn\xFCpft und aktiviert.',
  "Links an extension from a local path. Updates made to the local path will always be reflected.": "Verkn\xFCpft eine Erweiterung von einem lokalen Pfad. \xC4nderungen am lokalen Pfad werden immer widergespiegelt.",
  "The name of the extension to link.": "Der Name der zu verkn\xFCpfenden Erweiterung.",
  "Set a specific setting for an extension.": "Legt eine bestimmte Einstellung f\xFCr eine Erweiterung fest.",
  "Name of the extension to configure.": "Name der zu konfigurierenden Erweiterung.",
  "The setting to configure (name or env var).": "Die zu konfigurierende Einstellung (Name oder Umgebungsvariable).",
  "The scope to set the setting in.": "Der Bereich, in dem die Einstellung gesetzt werden soll.",
  "List all settings for an extension.": "Listet alle Einstellungen einer Erweiterung auf.",
  "Name of the extension.": "Name der Erweiterung.",
  'Extension "{{name}}" has no settings to configure.': 'Erweiterung "{{name}}" hat keine zu konfigurierenden Einstellungen.',
  'Settings for "{{name}}":': 'Einstellungen f\xFCr "{{name}}":',
  "(workspace)": "(Arbeitsbereich)",
  "(user)": "(Benutzer)",
  "[not set]": "[nicht gesetzt]",
  "[value stored in keychain]": "[Wert in Schl\xFCsselbund gespeichert]",
  "Manage extension settings.": "Erweiterungseinstellungen verwalten.",
  "You need to specify a command (set or list).": "Sie m\xFCssen einen Befehl angeben (set oder list).",
  // ============================================================================
  // Plugin Choice / Marketplace
  // ============================================================================
  "No plugins available in this marketplace.": "In diesem Marktplatz sind keine Plugins verf\xFCgbar.",
  'Select a plugin to install from marketplace "{{name}}":': 'W\xE4hlen Sie ein Plugin zur Installation aus Marktplatz "{{name}}":',
  "Plugin selection cancelled.": "Plugin-Auswahl abgebrochen.",
  'Select a plugin from "{{name}}"': 'Plugin aus "{{name}}" ausw\xE4hlen',
  "Use \u2191\u2193 or j/k to navigate, Enter to select, Escape to cancel": "Verwenden Sie \u2191\u2193 oder j/k zum Navigieren, Enter zum Ausw\xE4hlen, Escape zum Abbrechen",
  "{{count}} more above": "{{count}} weitere oben",
  "{{count}} more below": "{{count}} weitere unten",
  "manage IDE integration": "IDE-Integration verwalten",
  "check status of IDE integration": "Status der IDE-Integration pr\xFCfen",
  "install required IDE companion for {{ideName}}": "Erforderlichen IDE-Begleiter f\xFCr {{ideName}} installieren",
  "enable IDE integration": "IDE-Integration aktivieren",
  "disable IDE integration": "IDE-Integration deaktivieren",
  "IDE integration is not supported in your current environment. To use this feature, run Qwen Code in one of these supported IDEs: VS Code or VS Code forks.": "IDE-Integration wird in Ihrer aktuellen Umgebung nicht unterst\xFCtzt. Um diese Funktion zu nutzen, f\xFChren Sie Qwen Code in einer dieser unterst\xFCtzten IDEs aus: VS Code oder VS Code-Forks.",
  "Set up GitHub Actions": "GitHub Actions einrichten",
  "Configure terminal keybindings for multiline input (VS Code, Cursor, Windsurf, Trae)": "Terminal-Tastenbelegungen f\xFCr mehrzeilige Eingabe konfigurieren (VS Code, Cursor, Windsurf, Trae)",
  "Please restart your terminal for the changes to take effect.": "Bitte starten Sie Ihr Terminal neu, damit die \xC4nderungen wirksam werden.",
  "Failed to configure terminal: {{error}}": "Fehler beim Konfigurieren des Terminals: {{error}}",
  "Could not determine {{terminalName}} config path on Windows: APPDATA environment variable is not set.": "Konnte {{terminalName}}-Konfigurationspfad unter Windows nicht ermitteln: APPDATA-Umgebungsvariable ist nicht gesetzt.",
  "{{terminalName}} keybindings.json exists but is not a valid JSON array. Please fix the file manually or delete it to allow automatic configuration.": "{{terminalName}} keybindings.json existiert, ist aber kein g\xFCltiges JSON-Array. Bitte korrigieren Sie die Datei manuell oder l\xF6schen Sie sie, um automatische Konfiguration zu erm\xF6glichen.",
  "File: {{file}}": "Datei: {{file}}",
  "Failed to parse {{terminalName}} keybindings.json. The file contains invalid JSON. Please fix the file manually or delete it to allow automatic configuration.": "Fehler beim Parsen von {{terminalName}} keybindings.json. Die Datei enth\xE4lt ung\xFCltiges JSON. Bitte korrigieren Sie die Datei manuell oder l\xF6schen Sie sie, um automatische Konfiguration zu erm\xF6glichen.",
  "Error: {{error}}": "Fehler: {{error}}",
  "Shift+Enter binding already exists": "Shift+Enter-Belegung existiert bereits",
  "Ctrl+Enter binding already exists": "Ctrl+Enter-Belegung existiert bereits",
  "Existing keybindings detected. Will not modify to avoid conflicts.": "Bestehende Tastenbelegungen erkannt. Keine \xC4nderungen, um Konflikte zu vermeiden.",
  "Please check and modify manually if needed: {{file}}": "Bitte pr\xFCfen und bei Bedarf manuell \xE4ndern: {{file}}",
  "Added Shift+Enter and Ctrl+Enter keybindings to {{terminalName}}.": "Shift+Enter und Ctrl+Enter Tastenbelegungen zu {{terminalName}} hinzugef\xFCgt.",
  "Modified: {{file}}": "Ge\xE4ndert: {{file}}",
  "{{terminalName}} keybindings already configured.": "{{terminalName}}-Tastenbelegungen bereits konfiguriert.",
  "Failed to configure {{terminalName}}.": "Fehler beim Konfigurieren von {{terminalName}}.",
  "Your terminal is already configured for an optimal experience with multiline input (Shift+Enter and Ctrl+Enter).": "Ihr Terminal ist bereits f\xFCr optimale Erfahrung mit mehrzeiliger Eingabe konfiguriert (Shift+Enter und Ctrl+Enter).",
  // ============================================================================
  // Commands - Hooks
  // ============================================================================
  "Manage Qwen Code hooks": "Qwen Code-Hooks verwalten",
  "List all configured hooks": "Alle konfigurierten Hooks auflisten",
  // Hooks - Dialog
  Hooks: "Hooks",
  "Loading hooks...": "Hooks werden geladen...",
  "Error loading hooks:": "Fehler beim Laden der Hooks:",
  "Press Escape to close": "Escape zum Schlie\xDFen dr\xFCcken",
  "Press Escape, Ctrl+C, or Ctrl+D to cancel": "Escape, Ctrl+C oder Ctrl+D zum Abbrechen",
  "Press Space, Enter, or Escape to dismiss": "Space, Enter oder Escape zum Schlie\xDFen",
  "No hook selected": "Kein Hook ausgew\xE4hlt",
  // Hooks - List Step
  "No hook events found.": "Keine Hook-Ereignisse gefunden.",
  "{{count}} hook configured": "{{count}} Hook konfiguriert",
  "{{count}} hooks configured": "{{count}} Hooks konfiguriert",
  "This menu is read-only. To add or modify hooks, edit settings.json directly or ask Qwen Code.": "Dieses Men\xFC ist schreibgesch\xFCtzt. Um Hooks hinzuzuf\xFCgen oder zu \xE4ndern, bearbeiten Sie settings.json direkt oder fragen Sie Qwen Code.",
  "Enter to select \xB7 Esc to cancel": "Enter zum Ausw\xE4hlen \xB7 Esc zum Abbrechen",
  // Hooks - Detail Step
  "Exit codes:": "Exit-Codes:",
  "Configured hooks:": "Konfigurierte Hooks:",
  "No hooks configured for this event.": "F\xFCr dieses Ereignis sind keine Hooks konfiguriert.",
  "To add hooks, edit settings.json directly or ask Qwen.": "Um Hooks hinzuzuf\xFCgen, bearbeiten Sie settings.json direkt oder fragen Sie Qwen.",
  "Enter to select \xB7 Esc to go back": "Enter zum Ausw\xE4hlen \xB7 Esc zum Zur\xFCck",
  // Hooks - Config Detail Step
  "Hook details": "Hook-Details",
  "Event:": "Ereignis:",
  "Extension:": "Erweiterung:",
  "Desc:": "Beschreibung:",
  "No hook config selected": "Keine Hook-Konfiguration ausgew\xE4hlt",
  "To modify or remove this hook, edit settings.json directly or ask Qwen to help.": "Um diesen Hook zu \xE4ndern oder zu entfernen, bearbeiten Sie settings.json direkt oder fragen Sie Qwen.",
  // Hooks - Disabled Step
  "Hook Configuration - Disabled": "Hook-Konfiguration - Deaktiviert",
  "All hooks are currently disabled. You have {{count}} that are not running.": "Alle Hooks sind derzeit deaktiviert. Sie haben {{count}} die nicht ausgef\xFChrt werden.",
  "{{count}} configured hook": "{{count}} konfigurierter Hook",
  "{{count}} configured hooks": "{{count}} konfigurierte Hooks",
  "When hooks are disabled:": "Wenn Hooks deaktiviert sind:",
  "No hook commands will execute": "Keine Hook-Befehle werden ausgef\xFChrt",
  "StatusLine will not be displayed": "StatusLine wird nicht angezeigt",
  "Tool operations will proceed without hook validation": "Tool-Operationen werden ohne Hook-Validierung fortgesetzt",
  'To re-enable hooks, remove "disableAllHooks" from settings.json or ask Qwen Code.': 'Um Hooks wieder zu aktivieren, entfernen Sie "disableAllHooks" aus settings.json oder fragen Sie Qwen Code.',
  // Hooks - Source
  Project: "Projekt",
  User: "Benutzer",
  Skill: "Skill",
  System: "System",
  Extension: "Erweiterung",
  "Local Settings": "Lokale Einstellungen",
  "User Settings": "Benutzereinstellungen",
  "System Settings": "Systemeinstellungen",
  Extensions: "Erweiterungen",
  "Session (temporary)": "Sitzung (tempor\xE4r)",
  // Hooks - Event Descriptions (short)
  "Before tool execution": "Vor der Tool-Ausf\xFChrung",
  "After tool execution": "Nach der Tool-Ausf\xFChrung",
  "After tool execution fails": "Wenn die Tool-Ausf\xFChrung fehlschl\xE4gt",
  "When notifications are sent": "Wenn Benachrichtigungen gesendet werden",
  "When the user submits a prompt": "Wenn der Benutzer einen Prompt absendet",
  "When a slash command expands into a prompt": "Wenn ein Slash-Befehl zu einem Prompt erweitert wird",
  "When a new session is started": "Wenn eine neue Sitzung gestartet wird",
  "Right before Qwen Code concludes its response": "Direkt bevor Qwen Code seine Antwort abschlie\xDFt",
  "When a subagent (Agent tool call) is started": "Wenn ein Subagent (Agent-Tool-Aufruf) gestartet wird",
  "Right before a subagent concludes its response": "Direkt bevor ein Subagent seine Antwort abschlie\xDFt",
  "Before conversation compaction": "Vor der Gespr\xE4chskomprimierung",
  "When a session is ending": "Wenn eine Sitzung endet",
  "When a permission dialog is displayed": "Wenn ein Berechtigungsdialog angezeigt wird",
  "When a new todo item is created": "Wenn ein neues Todo-Element erstellt wird",
  "When a todo item is marked as completed": "Wenn ein Todo-Element als erledigt markiert wird",
  // Hooks - Event Descriptions (detailed)
  "Input to command is JSON of tool call arguments.": "Die Eingabe an den Befehl ist JSON der Tool-Aufruf-Argumente.",
  'Input to command is JSON with fields "inputs" (tool call arguments) and "response" (tool call response).': 'Die Eingabe an den Befehl ist JSON mit den Feldern "inputs" (Tool-Aufruf-Argumente) und "response" (Tool-Aufruf-Antwort).',
  "Input to command is JSON with tool_name, tool_input, tool_use_id, error, error_type, is_interrupt, and is_timeout.": "Die Eingabe an den Befehl ist JSON mit tool_name, tool_input, tool_use_id, error, error_type, is_interrupt und is_timeout.",
  "Input to command is JSON with notification message and type.": "Die Eingabe an den Befehl ist JSON mit Benachrichtigungsnachricht und -typ.",
  "Input to command is JSON with original user prompt text.": "Die Eingabe an den Befehl ist JSON mit dem urspr\xFCnglichen Benutzer-Prompt-Text.",
  "Input to command is JSON with command_name, command_args, and expanded prompt text.": "Die Eingabe an den Befehl ist JSON mit command_name, command_args und erweitertem Prompt-Text.",
  "Input to command is JSON with session start source.": "Die Eingabe an den Befehl ist JSON mit der Sitzungsstart-Quelle.",
  "Input to command is JSON with session end reason.": "Die Eingabe an den Befehl ist JSON mit dem Sitzungsende-Grund.",
  "Input to command is JSON with agent_id and agent_type.": "Die Eingabe an den Befehl ist JSON mit agent_id und agent_type.",
  "Input to command is JSON with agent_id, agent_type, and agent_transcript_path.": "Die Eingabe an den Befehl ist JSON mit agent_id, agent_type und agent_transcript_path.",
  "Input to command is JSON with compaction details.": "Die Eingabe an den Befehl ist JSON mit Komprimierungsdetails.",
  "Input to command is JSON with tool_name, tool_input, and tool_use_id. Output JSON with hookSpecificOutput containing decision to allow or deny.": "Die Eingabe an den Befehl ist JSON mit tool_name, tool_input und tool_use_id. Ausgabe ist JSON mit hookSpecificOutput, das die Entscheidung zum Zulassen oder Ablehnen enth\xE4lt.",
  "Input to command is JSON with todo_id, todo_content, todo_status, all_todos, and phase. In validation, output JSON with decision (allow/block/deny) and reason. In postWrite, block/deny is ignored.": "Die Eingabe an den Befehl ist JSON mit todo_id, todo_content, todo_status, all_todos und phase. In validation ist die Ausgabe JSON mit decision (allow/block/deny) und reason. In postWrite wird block/deny ignoriert.",
  "Input to command is JSON with todo_id, todo_content, previous_status, all_todos, and phase. In validation, output JSON with decision (allow/block/deny) and reason. In postWrite, block/deny is ignored.": "Die Eingabe an den Befehl ist JSON mit todo_id, todo_content, previous_status, all_todos und phase. In validation ist die Ausgabe JSON mit decision (allow/block/deny) und reason. In postWrite wird block/deny ignoriert.",
  // Hooks - Exit Code Descriptions
  "stdout/stderr not shown": "stdout/stderr nicht angezeigt",
  "show stderr to model and continue conversation": "stderr dem Modell anzeigen und Konversation fortsetzen",
  "show stderr to user only": "stderr nur dem Benutzer anzeigen",
  "stdout shown in transcript mode (ctrl+o)": "stdout im Transkriptmodus angezeigt (ctrl+o)",
  "show stderr to model immediately": "stderr sofort dem Modell anzeigen",
  "show stderr to user only but continue with tool call": "stderr nur dem Benutzer anzeigen, aber mit Tool-Aufruf fortfahren",
  "block processing, erase original prompt, and show stderr to user only": "Verarbeitung blockieren, urspr\xFCnglichen Prompt l\xF6schen und stderr nur dem Benutzer anzeigen",
  "block expanded prompt submission and show stderr to user only": "Einreichen des erweiterten Prompts blockieren und stderr nur dem Benutzer anzeigen",
  "stdout shown to Qwen": "stdout dem Qwen anzeigen",
  "show stderr to user only (blocking errors ignored)": "stderr nur dem Benutzer anzeigen (Blockierungsfehler ignoriert)",
  "command completes successfully": "Befehl erfolgreich abgeschlossen",
  "stdout shown to subagent": "stdout dem Subagenten anzeigen",
  "show stderr to subagent and continue having it run": "stderr dem Subagenten anzeigen und ihn weiterlaufen lassen",
  "stdout appended as custom compact instructions": "stdout als benutzerdefinierte Komprimierungsanweisungen angeh\xE4ngt",
  "block compaction": "Komprimierung blockieren",
  "show stderr to user only but continue with compaction": "stderr nur dem Benutzer anzeigen, aber mit Komprimierung fortfahren",
  "use hook decision if provided": "Hook-Entscheidung verwenden, falls bereitgestellt",
  "allow todo creation": "Todo-Erstellung zulassen",
  "block todo creation and show reason to model": "Todo-Erstellung blockieren und Grund dem Modell anzeigen",
  "allow todo completion": "Todo-Abschluss zulassen",
  "block todo completion and show reason to model": "Todo-Abschluss blockieren und Grund dem Modell anzeigen",
  // Hooks - Messages
  "Config not loaded.": "Konfiguration nicht geladen.",
  "Hooks are not enabled. Enable hooks in settings to use this feature.": "Hooks sind nicht aktiviert. Aktivieren Sie Hooks in den Einstellungen, um diese Funktion zu nutzen.",
  // ============================================================================
  // Commands - Session Export
  // ============================================================================
  "Export current session message history to a file": "Den Nachrichtenverlauf der aktuellen Sitzung in eine Datei exportieren",
  "Export session to HTML format": "Sitzung in das HTML-Format exportieren",
  "Export session to JSON format": "Sitzung in das JSON-Format exportieren",
  "Export session to JSONL format (one message per line)": "Sitzung in das JSONL-Format exportieren (eine Nachricht pro Zeile)",
  "Export session to markdown format": "Sitzung in das Markdown-Format exportieren",
  // ============================================================================
  // Commands - Insights
  // ============================================================================
  "generate personalized programming insights from your chat history": "Personalisierte Programmier-Einblicke aus Ihrem Chatverlauf generieren",
  // ============================================================================
  // Commands - Session History
  // ============================================================================
  "Resume a previous session": "Eine vorherige Sitzung fortsetzen",
  "Fork the current conversation into a new session": "Die aktuelle Unterhaltung in eine neue Sitzung verzweigen",
  "Spawn a background agent that inherits the full conversation": "Einen Hintergrund-Agenten starten, der die gesamte Unterhaltung \xFCbernimmt",
  "Please provide a directive. Usage: /fork <directive>": "Bitte geben Sie eine Anweisung an. Verwendung: /fork <Anweisung>",
  "Cannot fork while a response or tool call is in progress. Wait for it to finish or resolve the pending tool call.": "W\xE4hrend eine Antwort oder ein Tool-Aufruf l\xE4uft, kann kein Hintergrund-Fork erstellt werden. Warten Sie, bis der Vorgang abgeschlossen ist, oder bearbeiten Sie den ausstehenden Tool-Aufruf.",
  "Cannot fork before the first conversation turn.": "Vor der ersten Gespr\xE4chsrunde kann kein Fork erstellt werden.",
  "The agent tool is unavailable; cannot fork.": "Das Agent-Tool ist nicht verf\xFCgbar; Fork kann nicht gestartet werden.",
  "Failed to launch fork: {{error}}": "Fork konnte nicht gestartet werden: {{error}}",
  "User launched a background fork via /fork: {{directive}}": "Benutzer hat \xFCber /fork einen Hintergrund-Fork gestartet: {{directive}}",
  "Forked into a background agent. It inherits this conversation and runs without blocking \u2014 track it in the background tasks panel; it reports back when done.": "In einen Hintergrund-Agenten verzweigt. Er \xFCbernimmt diese Unterhaltung und l\xE4uft ohne zu blockieren \u2014 verfolgen Sie ihn im Hintergrundaufgaben-Panel; er meldet sich nach Abschluss zur\xFCck.",
  "Cannot branch while a response or tool call is in progress. Wait for it to finish or resolve the pending tool call.": "W\xE4hrend eine Antwort oder ein Tool-Aufruf l\xE4uft, kann keine Verzweigung erstellt werden. Warten Sie, bis der Vorgang abgeschlossen ist, oder bearbeiten Sie den ausstehenden Tool-Aufruf.",
  "No conversation to branch.": "Keine Unterhaltung zum Verzweigen vorhanden.",
  "Restore a tool call. This will reset the conversation and file history to the state it was in when the tool call was suggested": "Einen Tool-Aufruf wiederherstellen. Dadurch werden Konversations- und Dateiverlauf auf den Zustand zur\xFCckgesetzt, in dem der Tool-Aufruf vorgeschlagen wurde",
  "Could not detect terminal type. Supported terminals: VS Code, Cursor, Windsurf, and Trae.": "Terminal-Typ konnte nicht erkannt werden. Unterst\xFCtzte Terminals: VS Code, Cursor, Windsurf und Trae.",
  'Terminal "{{terminal}}" is not supported yet.': 'Terminal "{{terminal}}" wird noch nicht unterst\xFCtzt.',
  // ============================================================================
  // Commands - Language
  // ============================================================================
  "Invalid language. Available: {{options}}": "Ung\xFCltige Sprache. Verf\xFCgbar: {{options}}",
  "Language subcommands do not accept additional arguments.": "Sprach-Unterbefehle akzeptieren keine zus\xE4tzlichen Argumente.",
  "Current UI language: {{lang}}": "Aktuelle UI-Sprache: {{lang}}",
  "Current LLM output language: {{lang}}": "Aktuelle LLM-Ausgabesprache: {{lang}}",
  "Set UI language": "UI-Sprache festlegen",
  "Set LLM output language": "LLM-Ausgabesprache festlegen",
  "Usage: /language ui [{{options}}]": "Verwendung: /language ui [{{options}}]",
  "Usage: /language output <language>": "Verwendung: /language output <Sprache>",
  "Example: /language output \u4E2D\u6587": "Beispiel: /language output Deutsch",
  "Example: /language output English": "Beispiel: /language output Englisch",
  "Example: /language output \u65E5\u672C\u8A9E": "Beispiel: /language output Japanisch",
  "UI language changed to {{lang}}": "UI-Sprache ge\xE4ndert zu {{lang}}",
  "LLM output language set to {{lang}}": "LLM-Ausgabesprache auf {{lang}} gesetzt",
  "Please restart the application for the changes to take effect.": "Bitte starten Sie die Anwendung neu, damit die \xC4nderungen wirksam werden.",
  "Failed to generate LLM output language rule file: {{error}}": "Fehler beim Generieren der LLM-Ausgabesprach-Regeldatei: {{error}}",
  "Invalid command. Available subcommands:": "Ung\xFCltiger Befehl. Verf\xFCgbare Unterbefehle:",
  "Available subcommands:": "Verf\xFCgbare Unterbefehle:",
  "To request additional UI language packs, please open an issue on GitHub.": "Um zus\xE4tzliche UI-Sprachpakete anzufordern, \xF6ffnen Sie bitte ein Issue auf GitHub.",
  "Available options:": "Verf\xFCgbare Optionen:",
  "Set UI language to {{name}}": "UI-Sprache auf {{name}} setzen",
  // ============================================================================
  // Commands - Approval Mode
  // ============================================================================
  "Tool Approval Mode": "Werkzeug-Genehmigungsmodus",
  "Analyze only, do not modify files or execute commands": "Nur analysieren, keine Dateien \xE4ndern oder Befehle ausf\xFChren",
  "Require approval for file edits or shell commands": "Genehmigung f\xFCr Dateibearbeitungen oder Shell-Befehle erforderlich",
  "Automatically approve file edits": "Dateibearbeitungen automatisch genehmigen",
  "Use classifier to automatically approve safe tool calls": "Klassifikator verwenden, um sichere Werkzeugaufrufe automatisch zu genehmigen",
  "Automatically approve all tools": "Alle Werkzeuge automatisch genehmigen",
  "Workspace approval mode exists and takes priority. User-level change will have no effect.": "Arbeitsbereich-Genehmigungsmodus existiert und hat Vorrang. Benutzerebene-\xC4nderung hat keine Wirkung.",
  "Apply To": "Anwenden auf",
  "Workspace Settings": "Arbeitsbereich-Einstellungen",
  "Open auto-memory folder": "Auto-Speicher-Ordner \xF6ffnen",
  "Auto-memory: {{status}}": "Auto-Speicher: {{status}}",
  "Auto-dream: {{status}} \xB7 {{lastDream}} \xB7 /dream to run": "Auto-Konsolidierung: {{status}} \xB7 {{lastDream}} \xB7 /dream zum Ausf\xFChren",
  "Auto-skill: {{status}}": "Auto-Skill: {{status}}",
  never: "nie",
  on: "ein",
  off: "aus",
  "Remove matching entries from managed auto-memory.": "Passende Eintr\xE4ge aus dem verwalteten Auto-Speicher entfernen.",
  "Usage: /forget <memory text to remove>": "Verwendung: /forget <zu entfernender Erinnerungstext>",
  "No managed auto-memory entries matched: {{query}}": "Keine verwalteten Auto-Speicher-Eintr\xE4ge gefunden: {{query}}",
  "Consolidate managed auto-memory topic files.": "Verwaltete Auto-Speicher-Themendateien konsolidieren.",
  "Could not retrieve tool registry.": "Werkzeugregister konnte nicht abgerufen werden.",
  "Successfully authenticated and refreshed tools for '{{name}}'.": "Erfolgreich authentifiziert und Werkzeuge f\xFCr '{{name}}' aktualisiert.",
  "Re-discovering tools from '{{name}}'...": "Werkzeuge von '{{name}}' werden neu erkannt...",
  "Discovered {{count}} tool(s) from '{{name}}'.": "{{count}} Werkzeug(e) von '{{name}}' entdeckt.",
  "Authentication complete. Returning to server details...": "Authentifizierung abgeschlossen. Zur\xFCck zu den Serverdetails...",
  "Authentication successful.": "Authentifizierung erfolgreich.",
  // =========================================================
  // Commands - Summary
  // ============================================================================
  "Generate a project summary and save it to .qwen/PROJECT_SUMMARY.md": "Projektzusammenfassung generieren und in .qwen/PROJECT_SUMMARY.md speichern",
  "No chat client available to generate summary.": "Kein Chat-Client verf\xFCgbar, um Zusammenfassung zu generieren.",
  "Already generating summary, wait for previous request to complete": "Zusammenfassung wird bereits generiert, warten Sie auf Abschluss der vorherigen Anfrage",
  "No conversation found to summarize.": "Kein Gespr\xE4ch zum Zusammenfassen gefunden.",
  "Failed to generate project context summary: {{error}}": "Fehler beim Generieren der Projektkontextzusammenfassung: {{error}}",
  "Saved project summary to {{filePathForDisplay}}.": "Projektzusammenfassung gespeichert unter {{filePathForDisplay}}.",
  "Saving project summary...": "Projektzusammenfassung wird gespeichert...",
  "Generating project summary...": "Projektzusammenfassung wird generiert...",
  "Processing summary...": "Projektzusammenfassung wird verarbeitet...",
  "Project summary generated and saved successfully!": "Projektzusammenfassung wurde erfolgreich erstellt und gespeichert!",
  "Saved to: {{filePath}}": "Gespeichert unter: {{filePath}}",
  "Stopped because": "Angehalten, weil",
  "Failed to generate summary - no text content received from LLM response": "Fehler beim Generieren der Zusammenfassung - kein Textinhalt von LLM-Antwort erhalten",
  // ============================================================================
  // Commands - Model
  // ============================================================================
  "Switch the model for this session (--fast for suggestion model, [model-id] to switch immediately).": "Modell f\xFCr diese Sitzung wechseln (--fast f\xFCr Vorschlagsmodell)",
  "Set a lighter model for prompt suggestions and speculative execution": "Leichteres Modell f\xFCr Eingabevorschl\xE4ge und spekulative Ausf\xFChrung festlegen",
  "Content generator configuration not available.": "Inhaltsgenerator-Konfiguration nicht verf\xFCgbar.",
  "Authentication type not available.": "Authentifizierungstyp nicht verf\xFCgbar.",
  "No models available for the current authentication type ({{authType}}).": "Keine Modelle f\xFCr den aktuellen Authentifizierungstyp ({{authType}}) verf\xFCgbar.",
  // Needs translation
  " (not in model registry)": " (not in model registry)",
  // ============================================================================
  // Commands - Clear
  // ============================================================================
  "Starting a new session, resetting chat, and clearing terminal.": "Neue Sitzung wird gestartet, Chat wird zur\xFCckgesetzt und Terminal wird gel\xF6scht.",
  "Starting a new session and clearing.": "Neue Sitzung wird gestartet und gel\xF6scht.",
  // ============================================================================
  // Commands - Compress
  // ============================================================================
  "Already compressing, wait for previous request to complete": "Komprimierung l\xE4uft bereits, warten Sie auf Abschluss der vorherigen Anfrage",
  "Failed to compress chat history.": "Fehler beim Komprimieren des Chatverlaufs.",
  "Failed to compress chat history: {{error}}": "Fehler beim Komprimieren des Chatverlaufs: {{error}}",
  "Compressing chat history": "Chatverlauf wird komprimiert",
  "Chat history compressed from {{originalTokens}} to {{newTokens}} tokens.": "Chatverlauf komprimiert von {{originalTokens}} auf {{newTokens}} Token.",
  "Compression was not beneficial for this history size.": "Komprimierung war f\xFCr diese Verlaufsgr\xF6\xDFe nicht vorteilhaft.",
  "Chat history compression did not reduce size. This may indicate issues with the compression prompt.": "Chatverlauf-Komprimierung hat die Gr\xF6\xDFe nicht reduziert. Dies kann auf Probleme mit dem Komprimierungs-Prompt hindeuten.",
  "Could not compress chat history due to a token counting error.": "Chatverlauf konnte aufgrund eines Token-Z\xE4hlfehlers nicht komprimiert werden.",
  // ============================================================================
  // Commands - Directory
  // ============================================================================
  "Configuration is not available.": "Konfiguration ist nicht verf\xFCgbar.",
  "Please provide at least one path to add.": "Bitte geben Sie mindestens einen Pfad zum Hinzuf\xFCgen an.",
  "The /directory add command is not supported in restrictive sandbox profiles. Please use --include-directories when starting the session instead.": "Der Befehl /directory add wird in restriktiven Sandbox-Profilen nicht unterst\xFCtzt. Bitte verwenden Sie --include-directories beim Starten der Sitzung.",
  "Error adding '{{path}}': {{error}}": "Fehler beim Hinzuf\xFCgen von '{{path}}': {{error}}",
  "Successfully added QWEN.md files from the following directories if there are:\n- {{directories}}": "QWEN.md-Dateien aus folgenden Verzeichnissen erfolgreich hinzugef\xFCgt, falls vorhanden:\n- {{directories}}",
  "Error refreshing memory: {{error}}": "Fehler beim Aktualisieren des Speichers: {{error}}",
  "Successfully added directories:\n- {{directories}}": "Verzeichnisse erfolgreich hinzugef\xFCgt:\n- {{directories}}",
  "Current workspace directories:\n{{directories}}": "Aktuelle Arbeitsbereichsverzeichnisse:\n{{directories}}",
  // ============================================================================
  // Commands - Docs
  // ============================================================================
  "Please open the following URL in your browser to view the documentation:\n{{url}}": "Bitte \xF6ffnen Sie folgende URL in Ihrem Browser, um die Dokumentation anzusehen:\n{{url}}",
  "Opening documentation in your browser: {{url}}": "Dokumentation wird in Ihrem Browser ge\xF6ffnet: {{url}}",
  // ============================================================================
  // Dialogs - Tool Confirmation
  // ============================================================================
  "Do you want to proceed?": "M\xF6chten Sie fortfahren?",
  "Yes, allow once": "Ja, einmal erlauben",
  "Allow always": "Immer erlauben",
  Yes: "Ja",
  No: "Nein",
  "No (esc)": "Nein (Esc)",
  // MCP Management Dialog (translations for MCP UI components)
  "Manage MCP servers": "MCP servers verwalten",
  "Server Detail": "Serverdetails",
  Tools: "Werkzeuge",
  "Tool Detail": "Werkzeugdetails",
  "Loading...": "L\xE4dt...",
  "Unknown step": "Unbekannter Schritt",
  "Esc to back": "Esc zur\xFCck",
  "\u2191\u2193 to navigate \xB7 Enter to select \xB7 Esc to close": "\u2191\u2193 navigieren \xB7 Enter ausw\xE4hlen \xB7 Esc schlie\xDFen",
  "\u2191\u2193 to navigate \xB7 Enter to select \xB7 Esc to back": "\u2191\u2193 navigieren \xB7 Enter ausw\xE4hlen \xB7 Esc zur\xFCck",
  "\u2191\u2193 to navigate \xB7 Enter to confirm \xB7 Esc to back": "\u2191\u2193 navigieren \xB7 Enter best\xE4tigen \xB7 Esc zur\xFCck",
  "User Settings (global)": "Benutzereinstellungen (global)",
  "Workspace Settings (project-specific)": "Arbeitsbereichseinstellungen (projektspezifisch)",
  "Disable server:": "Server deaktivieren:",
  "Select where to add the server to the exclude list:": "W\xE4hlen Sie, wo der Server zur Ausschlussliste hinzugef\xFCgt werden soll:",
  "Press Enter to confirm, Esc to cancel": "Enter zum Best\xE4tigen, Esc zum Abbrechen",
  Disable: "Deaktivieren",
  Enable: "Aktivieren",
  Authenticate: "Authentifizieren",
  "Re-authenticate": "Erneut authentifizieren",
  "Clear Authentication": "Authentifizierung l\xF6schen",
  disabled: "deaktiviert",
  enabled: "aktiviert",
  "Server:": "Server:",
  Reconnect: "Neu verbinden",
  "View tools": "Werkzeuge anzeigen",
  "Status:": "Status:",
  "Command:": "Befehl:",
  "Working Directory:": "Arbeitsverzeichnis:",
  "No server selected": "Kein Server ausgew\xE4hlt",
  "Error:": "Fehler:",
  tool: "Werkzeug",
  tools: "Werkzeuge",
  connected: "verbunden",
  connecting: "verbindet",
  disconnected: "getrennt",
  error: "Fehler",
  // MCP Server List
  "User MCPs": "Benutzer-MCPs",
  "Project MCPs": "Projekt-MCPs",
  "Extension MCPs": "Erweiterungs-MCPs",
  server: "Server",
  servers: "Server",
  "Add MCP servers to your settings to get started.": "F\xFCgen Sie MCP servers zu Ihren Einstellungen hinzu, um zu beginnen.",
  "Run qwen --debug to see error logs": "F\xFChren Sie qwen --debug aus, um Fehlerprotokolle anzuzeigen",
  // MCP OAuth Authentication
  "OAuth Authentication": "OAuth-Authentifizierung",
  "Authenticating... Please complete the login in your browser.": "Authentifizierung l\xE4uft... Bitte schlie\xDFen Sie die Anmeldung in Ihrem Browser ab.",
  // MCP Tool List
  "No tools available for this server.": "Keine Werkzeuge f\xFCr diesen Server verf\xFCgbar.",
  destructive: "destruktiv",
  "read-only": "schreibgesch\xFCtzt",
  "open-world": "offene Welt",
  idempotent: "idempotent",
  "Tools for {{serverName}}": "Werkzeuge f\xFCr {{serverName}}",
  "{{current}}/{{total}}": "{{current}}/{{total}}",
  // MCP Tool Detail
  required: "erforderlich",
  Parameters: "Parameter",
  "No tool selected": "Kein Werkzeug ausgew\xE4hlt",
  Server: "Server",
  // Invalid tool related translations
  "{{count}} invalid tools": "{{count}} ung\xFCltige Werkzeuge",
  invalid: "ung\xFCltig",
  "invalid: {{reason}}": "ung\xFCltig: {{reason}}",
  "missing name": "Name fehlt",
  "missing description": "Beschreibung fehlt",
  "(unnamed)": "(unbenannt)",
  "Warning: This tool cannot be called by the LLM": "Warnung: Dieses Werkzeug kann nicht vom LLM aufgerufen werden",
  Reason: "Grund",
  "Tools must have both name and description to be used by the LLM.": "Werkzeuge m\xFCssen sowohl einen Namen als auch eine Beschreibung haben, um vom LLM verwendet zu werden.",
  "Modify in progress:": "\xC4nderung in Bearbeitung:",
  "Save and close external editor to continue": "Speichern und externen Editor schlie\xDFen, um fortzufahren",
  "Apply this change?": "Diese \xC4nderung anwenden?",
  "Yes, allow always": "Ja, immer erlauben",
  "Modify with external editor": "Mit externem Editor bearbeiten",
  "No, suggest changes (esc)": "Nein, \xC4nderungen vorschlagen (Esc)",
  "Allow execution of: '{{command}}'?": "Ausf\xFChrung erlauben von: '{{command}}'?",
  "Always allow in this project": "In diesem Projekt immer erlauben",
  "Always allow {{action}} in this project": "{{action}} in diesem Projekt immer erlauben",
  "Always allow for this user": "F\xFCr diesen Benutzer immer erlauben",
  "Always allow {{action}} for this user": "{{action}} f\xFCr diesen Benutzer immer erlauben",
  "Yes, restore previous mode ({{mode}})": "Ja, vorherigen Modus wiederherstellen ({{mode}})",
  "Yes, and auto-accept edits": "Ja, und \xC4nderungen automatisch akzeptieren",
  "Yes, and manually approve edits": "Ja, und \xC4nderungen manuell genehmigen",
  "No, keep planning (esc)": "Nein, weiter planen (Esc)",
  "URLs to fetch:": "Abzurufende URLs:",
  "MCP Server: {{server}}": "MCP Server: {{server}}",
  "Tool: {{tool}}": "Werkzeug: {{tool}}",
  'Allow execution of MCP tool "{{tool}}" from server "{{server}}"?': 'Ausf\xFChrung von MCP tool "{{tool}}" von MCP server "{{server}}" erlauben?',
  // ============================================================================
  // Dialogs - Shell Confirmation
  // ============================================================================
  "Shell Command Execution": "Shell-Befehlsausf\xFChrung",
  "A custom command wants to run the following shell commands:": "Ein benutzerdefinierter Befehl m\xF6chte folgende Shell-Befehle ausf\xFChren:",
  // ============================================================================
  // Dialogs - Welcome Back
  // ============================================================================
  "Current Plan:": "Aktueller Plan:",
  "Progress: {{done}}/{{total}} tasks completed": "Fortschritt: {{done}}/{{total}} Aufgaben abgeschlossen",
  ", {{inProgress}} in progress": ", {{inProgress}} in Bearbeitung",
  "Pending Tasks:": "Ausstehende Aufgaben:",
  "What would you like to do?": "Was m\xF6chten Sie tun?",
  "Choose how to proceed with your session:": "W\xE4hlen Sie, wie Sie mit Ihrer Sitzung fortfahren m\xF6chten:",
  "Start new chat session": "Neue Chat-Sitzung starten",
  "Continue previous conversation": "Vorheriges Gespr\xE4ch fortsetzen",
  "\u{1F44B} Welcome back! (Last updated: {{timeAgo}})": "\u{1F44B} Willkommen zur\xFCck! (Zuletzt aktualisiert: {{timeAgo}})",
  "\u{1F3AF} Overall Goal:": "\u{1F3AF} Gesamtziel:",
  "Connect a Provider": "Anbieter verbinden",
  "You must connect a provider to proceed. Press Ctrl+C again to exit.": "Sie m\xFCssen einen Anbieter verbinden, um fortzufahren. Dr\xFCcken Sie erneut Ctrl+C zum Beenden.",
  "Terms of Services and Privacy Notice": "Nutzungsbedingungen und Datenschutzhinweis",
  "Qwen OAuth": "Qwen OAuth",
  "Discontinued \u2014 switch to Coding Plan or API Key": "Eingestellt \u2014 wechseln Sie zu Coding Plan oder API Key",
  "Qwen OAuth free tier was discontinued on 2026-04-15. Please select Coding Plan or API Key instead.": "Das kostenlose Qwen OAuth-Kontingent wurde am 2026-04-15 eingestellt. Bitte w\xE4hlen Sie Coding Plan oder API Key.",
  "Qwen OAuth free tier was discontinued on 2026-04-15. Please select a model from another provider or run /auth to switch.": "Das kostenlose Qwen OAuth-Angebot wurde am 2026-04-15 eingestellt. Bitte w\xE4hlen Sie ein Modell eines anderen Anbieter oder f\xFChren Sie /auth aus, um zu wechseln.",
  "\n\u26A0 Qwen OAuth free tier was discontinued on 2026-04-15. Please select another option.\n": "\n\u26A0 Das kostenlose Qwen OAuth-Kontingent wurde am 2026-04-15 eingestellt. Bitte w\xE4hlen Sie eine andere Option.\n",
  "Paid \xB7 Up to 6,000 requests/5 hrs \xB7 All Alibaba Cloud Coding Plan Models": "Kostenpflichtig \xB7 Bis zu 6.000 Anfragen/5 Std. \xB7 Alle Alibaba Cloud Coding Plan Modelle",
  "Alibaba Cloud Coding Plan": "Alibaba Cloud Coding Plan",
  "Bring your own API key": "Eigenen API Key verwenden",
  "Browser-based authentication with third-party providers (e.g. OpenRouter, ModelScope)": "Browserbasierte Authentifizierung mit externen Anbietern (z. B. OpenRouter, ModelScope)",
  "Authentication is enforced to be {{enforcedType}}, but you are currently using {{currentType}}.": "Authentifizierung ist auf {{enforcedType}} festgelegt, aber Sie verwenden derzeit {{currentType}}.",
  "Qwen OAuth Authentication": "Qwen OAuth-Authentifizierung",
  "Please visit this URL to authorize:": "Bitte besuchen Sie diese URL zur Autorisierung:",
  "Waiting for authorization": "Warten auf Autorisierung",
  "Time remaining:": "Verbleibende Zeit:",
  "Qwen OAuth Authentication Timeout": "Qwen OAuth-Authentifizierung abgelaufen",
  "OAuth token expired (over {{seconds}} seconds). Please select authentication method again.": "OAuth-Token abgelaufen (\xFCber {{seconds}} Sekunden). Bitte w\xE4hlen Sie erneut eine Authentifizierungsmethode.",
  "Press any key to return to authentication type selection.": "Dr\xFCcken Sie eine beliebige Taste, um zur Authentifizierungstypauswahl zur\xFCckzukehren.",
  "Waiting for Qwen OAuth authentication...": "Warten auf Qwen OAuth-Authentifizierung...",
  "Authentication timed out. Please try again.": "Authentifizierung abgelaufen. Bitte versuchen Sie es erneut.",
  "Waiting for auth... (Press ESC or CTRL+C to cancel)": "Warten auf Authentifizierung... (ESC oder CTRL+C zum Abbrechen dr\xFCcken)",
  "Missing API key for OpenAI-compatible auth. Set settings.security.auth.apiKey, or set the {{envKeyHint}} environment variable.": "API Key f\xFCr OpenAI-kompatible Authentifizierung fehlt. Setzen Sie settings.security.auth.apiKey oder die Umgebungsvariable {{envKeyHint}}.",
  "{{envKeyHint}} environment variable not found. Please set it in your .env file or environment variables.": "Umgebungsvariable {{envKeyHint}} wurde nicht gefunden. Bitte legen Sie sie in Ihrer .env-Datei oder den Systemumgebungsvariablen fest.",
  "{{envKeyHint}} environment variable not found (or set settings.security.auth.apiKey). Please set it in your .env file or environment variables.": "Umgebungsvariable {{envKeyHint}} wurde nicht gefunden (oder setzen Sie settings.security.auth.apiKey). Bitte legen Sie sie in Ihrer .env-Datei oder den Systemumgebungsvariablen fest.",
  "Missing API key for OpenAI-compatible auth. Set the {{envKeyHint}} environment variable.": "API Key f\xFCr OpenAI-kompatible Authentifizierung fehlt. Setzen Sie die Umgebungsvariable {{envKeyHint}}.",
  "Anthropic provider missing required baseUrl in modelProviders[].baseUrl.": "Anthropic-Anbieter fehlt erforderliche baseUrl in modelProviders[].baseUrl.",
  "ANTHROPIC_BASE_URL environment variable not found.": "Umgebungsvariable ANTHROPIC_BASE_URL wurde nicht gefunden.",
  "Invalid auth method selected.": "Ung\xFCltige Authentifizierungsmethode ausgew\xE4hlt.",
  "Failed to authenticate. Message: {{message}}": "Authentifizierung fehlgeschlagen. Meldung: {{message}}",
  "Authenticated successfully with {{authType}} credentials.": "Erfolgreich mit {{authType}}-Anmeldedaten authentifiziert.",
  'Invalid QWEN_DEFAULT_AUTH_TYPE value: "{{value}}". Valid values are: {{validValues}}': 'Ung\xFCltiger QWEN_DEFAULT_AUTH_TYPE-Wert: "{{value}}". G\xFCltige Werte sind: {{validValues}}',
  // ============================================================================
  // Dialogs - Model
  // ============================================================================
  "Select Model": "Modell ausw\xE4hlen",
  "API Key": "API Key",
  "(default)": "(Standard)",
  "(not set)": "(nicht gesetzt)",
  Modality: "Modalit\xE4t",
  "Context Window": "Kontextfenster",
  text: "Text",
  "text-only": "nur Text",
  image: "Bild",
  pdf: "PDF",
  audio: "Audio",
  video: "Video",
  "not set": "nicht gesetzt",
  none: "keine",
  unknown: "unbekannt",
  // ============================================================================
  // Dialogs - Permissions
  // ============================================================================
  "Manage folder trust settings": "Ordnervertrauenseinstellungen verwalten",
  "Manage permission rules": "permission rules verwalten",
  Allow: "Erlauben",
  Ask: "Fragen",
  Deny: "Verweigern",
  Workspace: "Arbeitsbereich",
  "Qwen Code won't ask before using allowed tools.": "Qwen Code fragt nicht, bevor erlaubte Tools verwendet werden.",
  "Qwen Code will ask before using these tools.": "Qwen Code fragt, bevor diese Tools verwendet werden.",
  "Qwen Code is not allowed to use denied tools.": "Qwen Code darf verweigerte Tools nicht verwenden.",
  "Manage trusted directories for this workspace.": "Vertrauensw\xFCrdige Verzeichnisse f\xFCr diesen Arbeitsbereich verwalten.",
  "Any use of the {{tool}} tool": "Jede Verwendung des {{tool}}-Tools",
  "{{tool}} commands matching '{{pattern}}'": "{{tool}}-Befehle, die '{{pattern}}' entsprechen",
  "From user settings": "Aus Benutzereinstellungen",
  "From project settings": "Aus Projekteinstellungen",
  "From session": "Aus Sitzung",
  "Project settings": "Projekteinstellungen",
  "Checked in at .qwen/settings.json": "Eingecheckt in .qwen/settings.json",
  "User settings": "Benutzereinstellungen",
  "Saved in at ~/.qwen/settings.json": "Gespeichert in ~/.qwen/settings.json",
  "Add a new rule\u2026": "Neue Regel hinzuf\xFCgen\u2026",
  "Add {{type}} permission rule": "{{type}} permission rule hinzuf\xFCgen",
  "Permission rules are a tool name, optionally followed by a specifier in parentheses.": "permission rules sind ein Toolname, optional gefolgt von einem Bezeichner in Klammern.",
  "e.g.,": "z.B.",
  or: "oder",
  "Enter permission rule\u2026": "permission rule eingeben\u2026",
  "Enter to submit \xB7 Esc to cancel": "Enter zum Absenden \xB7 Esc zum Abbrechen",
  "Where should this rule be saved?": "Wo soll diese Regel gespeichert werden?",
  "Enter to confirm \xB7 Esc to cancel": "Enter zum Best\xE4tigen \xB7 Esc zum Abbrechen",
  "Delete {{type}} rule?": "{{type}}-Regel l\xF6schen?",
  "Are you sure you want to delete this permission rule?": "Sind Sie sicher, dass Sie diese permission rule l\xF6schen m\xF6chten?",
  "Permissions:": "Berechtigungen:",
  "(\u2190/\u2192 or tab to cycle)": "(\u2190/\u2192 oder Tab zum Wechseln)",
  "Press \u2191\u2193 to navigate \xB7 Enter to select \xB7 Type to search \xB7 Esc to cancel": "\u2191\u2193 navigieren \xB7 Enter ausw\xE4hlen \xB7 Tippen suchen \xB7 Esc abbrechen",
  "Search\u2026": "Suche\u2026",
  // Workspace directory management
  "Add directory\u2026": "Verzeichnis hinzuf\xFCgen\u2026",
  "Add directory to workspace": "Verzeichnis zum Arbeitsbereich hinzuf\xFCgen",
  "Qwen Code can read files in the workspace, and make edits when auto-accept edits is on.": "Qwen Code kann Dateien im Arbeitsbereich lesen und Bearbeitungen vornehmen, wenn die automatische Akzeptierung aktiviert ist.",
  "Qwen Code will be able to read files in this directory and make edits when auto-accept edits is on.": "Qwen Code kann Dateien in diesem Verzeichnis lesen und Bearbeitungen vornehmen, wenn die automatische Akzeptierung aktiviert ist.",
  "Enter the path to the directory:": "Pfad zum Verzeichnis eingeben:",
  "Enter directory path\u2026": "Verzeichnispfad eingeben\u2026",
  "Tab to complete \xB7 Enter to add \xB7 Esc to cancel": "Tab zum Vervollst\xE4ndigen \xB7 Enter zum Hinzuf\xFCgen \xB7 Esc zum Abbrechen",
  "Remove directory?": "Verzeichnis entfernen?",
  "Are you sure you want to remove this directory from the workspace?": "M\xF6chten Sie dieses Verzeichnis wirklich aus dem Arbeitsbereich entfernen?",
  "  (Original working directory)": "  (Urspr\xFCngliches Arbeitsverzeichnis)",
  "  (from settings)": "  (aus Einstellungen)",
  "Directory does not exist.": "Verzeichnis existiert nicht.",
  "Path is not a directory.": "Pfad ist kein Verzeichnis.",
  "This directory is already in the workspace.": "Dieses Verzeichnis ist bereits im Arbeitsbereich.",
  "Already covered by existing directory: {{dir}}": "Bereits durch vorhandenes Verzeichnis abgedeckt: {{dir}}",
  // ============================================================================
  // Status Bar
  // ============================================================================
  "Using:": "Verwendet:",
  "{{count}} open file": "{{count}} ge\xF6ffnete Datei",
  "{{count}} open files": "{{count}} ge\xF6ffnete Dateien",
  "(ctrl+g to view)": "(Ctrl+G zum Anzeigen)",
  "{{count}} {{name}} file": "{{count}} {{name}}-Datei",
  "{{count}} {{name}} files": "{{count}} {{name}}-Dateien",
  "{{count}} MCP server": "{{count}} MCP server",
  "{{count}} MCP servers": "{{count}} MCP servers",
  "{{count}} Blocked": "{{count}} blockiert",
  "(ctrl+t to view)": "(Ctrl+T zum Anzeigen)",
  "(ctrl+t to toggle)": "(Ctrl+T zum Umschalten)",
  "Press Ctrl+C again to exit.": "Dr\xFCcken Sie erneut Ctrl+C zum Beenden.",
  "Press Ctrl+D again to exit.": "Dr\xFCcken Sie erneut Ctrl+D zum Beenden.",
  "Press Esc again to clear.": "Dr\xFCcken Sie erneut Esc zum L\xF6schen.",
  "Press \u2191 to edit queued messages": "Dr\xFCcken Sie \u2191, um Nachrichten in der Warteschlange zu bearbeiten",
  // ============================================================================
  // MCP Status
  // ============================================================================
  "No MCP servers configured.": "Keine MCP servers konfiguriert.",
  "\u23F3 MCP servers are starting up ({{count}} initializing)...": "\u23F3 MCP servers werden gestartet ({{count}} werden initialisiert)...",
  "Note: First startup may take longer. Tool availability will update automatically.": "Hinweis: Der erste Start kann l\xE4nger dauern. Werkzeugverf\xFCgbarkeit wird automatisch aktualisiert.",
  "Configured MCP servers:": "Konfigurierte MCP servers:",
  Ready: "Bereit",
  "Starting... (first startup may take longer)": "Wird gestartet... (erster Start kann l\xE4nger dauern)",
  Disconnected: "Getrennt",
  "{{count}} tool": "{{count}} Werkzeug",
  "{{count}} tools": "{{count}} Werkzeuge",
  "{{count}} prompt": "{{count}} Prompt",
  "{{count}} prompts": "{{count}} Prompts",
  "(from {{extensionName}})": "(von {{extensionName}})",
  OAuth: "OAuth",
  "OAuth expired": "OAuth abgelaufen",
  "OAuth not authenticated": "OAuth nicht authentifiziert",
  "tools and prompts will appear when ready": "Werkzeuge und Prompts werden angezeigt, wenn bereit",
  "{{count}} tools cached": "{{count}} Werkzeuge zwischengespeichert",
  "Tools:": "Werkzeuge:",
  "Parameters:": "Parameter:",
  Blocked: "Blockiert",
  "\u{1F4A1} Tips:": "\u{1F4A1} Tipps:",
  Use: "Verwenden",
  "to show server and tool descriptions": "um Server- und Werkzeugbeschreibungen anzuzeigen",
  "to show tool parameter schemas": "um tool parameter schemas anzuzeigen",
  "to hide descriptions": "um Beschreibungen auszublenden",
  "to authenticate with OAuth-enabled servers": "um sich bei OAuth-f\xE4higen Servern zu authentifizieren",
  Press: "Dr\xFCcken Sie",
  "to toggle tool descriptions on/off": "um Werkzeugbeschreibungen ein-/auszuschalten",
  "Starting OAuth authentication for MCP server '{{name}}'...": "OAuth-Authentifizierung f\xFCr MCP server '{{name}}' wird gestartet...",
  // ============================================================================
  // Startup Tips
  // ============================================================================
  // ============================================================================
  // Exit Screen / Stats
  // ============================================================================
  "Agent powering down. Goodbye!": "Agent wird heruntergefahren. Auf Wiedersehen!",
  "To continue this session, run": "Um diese Sitzung fortzusetzen, f\xFChren Sie aus",
  "Interaction Summary": "Interaktionszusammenfassung",
  "Session ID:": "Sitzungs-ID:",
  "Tool Calls:": "Werkzeugaufrufe:",
  "Success Rate:": "Erfolgsrate:",
  "User Agreement:": "Benutzerzustimmung:",
  reviewed: "\xFCberpr\xFCft",
  "Code Changes:": "Code\xE4nderungen:",
  Performance: "Leistung",
  "Wall Time:": "Gesamtzeit:",
  "Agent Active:": "Agent aktiv:",
  "API Time:": "API-Zeit:",
  "Tool Time:": "Werkzeugzeit:",
  "Session Stats": "Sitzungsstatistiken",
  "Model Usage": "Modellnutzung",
  Reqs: "Anfragen",
  "Input Tokens": "Eingabe-Token",
  "Output Tokens": "Ausgabe-Token",
  "Savings Highlight:": "Einsparungen:",
  "of input tokens were served from the cache, reducing costs.": "der Eingabe-Token wurden aus dem Cache bedient, was die Kosten reduziert.",
  "Tip: For a full token breakdown, run `/stats model`.": "Tipp: F\xFCr eine vollst\xE4ndige Token-Aufschl\xFCsselung f\xFChren Sie `/stats model` aus.",
  "Model Stats For Nerds": "Modellstatistiken f\xFCr Nerds",
  "Tool Stats For Nerds": "Werkzeugstatistiken f\xFCr Nerds",
  Metric: "Metrik",
  API: "API",
  Requests: "Anfragen",
  Errors: "Fehler",
  "Avg Latency": "Durchschn. Latenz",
  Tokens: "Token",
  Total: "Gesamt",
  Cached: "Zwischengespeichert",
  Thoughts: "Gedanken",
  Output: "Ausgabe",
  "No API calls have been made in this session.": "In dieser Sitzung wurden keine API-Aufrufe gemacht.",
  "Tool Name": "Werkzeugname",
  Calls: "Aufrufe",
  "Success Rate": "Erfolgsrate",
  "Avg Duration": "Durchschn. Dauer",
  "User Decision Summary": "Benutzerentscheidungs-Zusammenfassung",
  "Total Reviewed Suggestions:": "Insgesamt \xFCberpr\xFCfter Vorschl\xE4ge:",
  " \xBB Accepted:": " \xBB Akzeptiert:",
  " \xBB Rejected:": " \xBB Abgelehnt:",
  " \xBB Modified:": " \xBB Ge\xE4ndert:",
  " Overall Agreement Rate:": " Gesamtzustimmungsrate:",
  "No tool calls have been made in this session.": "In dieser Sitzung wurden keine Werkzeugaufrufe gemacht.",
  "Session start time is unavailable, cannot calculate stats.": "Sitzungsstartzeit nicht verf\xFCgbar, Statistiken k\xF6nnen nicht berechnet werden.",
  Activity: "Aktivit\xE4t",
  Efficiency: "Effizienz",
  Today: "Heute",
  "Token Trend": "Token-Trend",
  "Cache Hit Rate": "Cache-Trefferquote",
  "Tool Success": "Tool-Erfolgsrate",
  "Tool Leaderboard": "Tool-Rangliste",
  Time: "Zeit",
  Success: "Erfolg",
  Cache: "Cache",
  Latency: "Latenz",
  "Code Impact": "Code-\xC4nderungen",
  net: "netto",
  streak: "Serie",
  best: "Rekord",
  // ============================================================================
  // Command Format Migration
  // ============================================================================
  "Command Format Migration": "Befehlsformat-Migration",
  "Found {{count}} TOML command file:": "{{count}} TOML-Befehlsdatei gefunden:",
  "Found {{count}} TOML command files:": "{{count}} TOML-Befehlsdateien gefunden:",
  "Current tasks": "Aktuelle Aufgaben",
  "Background tasks": "Hintergrundaufgaben",
  "No tasks currently running": "Derzeit laufen keine Aufgaben",
  "No entry to show.": "Kein Eintrag zum Anzeigen.",
  "needs approval": "wartet auf Genehmigung",
  "rejected \u2014 edit config to re-approve": "abgelehnt \u2014 Konfiguration bearbeiten, um erneut zu genehmigen",
  "Background agent needs approval": "Hintergrund-Agent wartet auf Genehmigung",
  "Approve or deny the request above": "Genehmigen oder lehnen Sie die obige Anfrage ab",
  Running: "L\xE4uft",
  Paused: "Pausiert",
  Completed: "Abgeschlossen",
  Failed: "Fehlgeschlagen",
  Stopped: "Gestoppt",
  Shell: "Shell",
  Monitor: "Monitor",
  Command: "Befehl",
  Dream: "Dream",
  "[dream] memory consolidation": "[dream] Speicher-Konsolidierung",
  "[dream] memory consolidation (reviewing {{count}} session)": "[dream] Speicher-Konsolidierung (pr\xFCft {{count}} Sitzung)",
  "[dream] memory consolidation (reviewing {{count}} sessions)": "[dream] Speicher-Konsolidierung (pr\xFCft {{count}} Sitzungen)",
  "... and {{count}} more": "... und {{count}} weitere",
  "The TOML format is deprecated. Would you like to migrate them to Markdown format?": "Das TOML-Format ist veraltet. M\xF6chten Sie sie ins Markdown-Format migrieren?",
  "(Backups will be created and original files will be preserved)": "(Backups werden erstellt und Originaldateien werden beibehalten)",
  // ============================================================================
  // Loading Phrases
  // ============================================================================
  "Waiting for user confirmation...": "Warten auf Benutzerbest\xE4tigung...",
  // ============================================================================
  // Loading Phrases
  // ============================================================================
  WITTY_LOADING_PHRASES: [
    "Auf gut Gl\xFCck!",
    "Genialit\xE4t wird ausgeliefert...",
    "Die Serifen werden aufgemalt...",
    "Durch den Schleimpilz navigieren...",
    "Die digitalen Geister werden befragt...",
    "Splines werden retikuliert...",
    "Die KI-Hamster werden aufgew\xE4rmt...",
    "Die Zaubermuschel wird befragt...",
    "Witzige Erwiderung wird generiert...",
    "Die Algorithmen werden poliert...",
    "Perfektion braucht Zeit (mein Code auch)...",
    "Frische Bytes werden gebr\xFCht...",
    "Elektronen werden gez\xE4hlt...",
    "Kognitive Prozessoren werden aktiviert...",
    "Auf Syntaxfehler im Universum wird gepr\xFCft...",
    "Einen Moment, Humor wird optimiert...",
    "Pointen werden gemischt...",
    "Neuronale Netze werden entwirrt...",
    "Brillanz wird kompiliert...",
    "wit.exe wird geladen...",
    "Die Wolke der Weisheit wird beschworen...",
    "Eine witzige Antwort wird vorbereitet...",
    "Einen Moment, ich debugge die Realit\xE4t...",
    "Die Optionen werden verwirrt...",
    "Kosmische Frequenzen werden eingestellt...",
    "Eine Antwort wird erstellt, die Ihrer Geduld w\xFCrdig ist...",
    "Die Einsen und Nullen werden kompiliert...",
    "Abh\xE4ngigkeiten werden aufgel\xF6st... und existenzielle Krisen...",
    "Erinnerungen werden defragmentiert... sowohl RAM als auch pers\xF6nliche...",
    "Das Humor-Modul wird neu gestartet...",
    "Das Wesentliche wird zwischengespeichert (haupts\xE4chlich Katzen-Memes)...",
    "F\xFCr l\xE4cherliche Geschwindigkeit wird optimiert",
    "Bits werden getauscht... sagen Sie es nicht den Bytes...",
    "Garbage Collection l\xE4uft... bin gleich zur\xFCck...",
    "Das Internet wird zusammengebaut...",
    "Kaffee wird in Code umgewandelt...",
    "Die Syntax der Realit\xE4t wird aktualisiert...",
    "Die Synapsen werden neu verdrahtet...",
    "Ein verlegtes Semikolon wird gesucht...",
    "Die Zahnr\xE4der werden geschmiert...",
    "Die Server werden vorgeheizt...",
    "Der Fluxkompensator wird kalibriert...",
    "Der Unwahrscheinlichkeitsantrieb wird aktiviert...",
    "Die Macht wird kanalisiert...",
    "Die Sterne werden f\xFCr optimale Antwort ausgerichtet...",
    "So sagen wir alle...",
    "Die n\xE4chste gro\xDFe Idee wird geladen...",
    "Einen Moment, ich bin in der Zone...",
    "Bereite mich vor, Sie mit Brillanz zu blenden...",
    "Einen Augenblick, ich poliere meinen Witz...",
    "Halten Sie durch, ich erschaffe ein Meisterwerk...",
    "Einen Moment, ich debugge das Universum...",
    "Einen Moment, ich richte die Pixel aus...",
    "Einen Moment, ich optimiere den Humor...",
    "Einen Moment, ich tune die Algorithmen...",
    "Warp-Geschwindigkeit aktiviert...",
    "Mehr Dilithium-Kristalle werden gesucht...",
    "Keine Panik...",
    "Dem wei\xDFen Kaninchen wird gefolgt...",
    "Die Wahrheit ist hier drin... irgendwo...",
    "Auf die Kassette wird gepustet...",
    "Ladevorgang... Machen Sie eine Fassrolle!",
    "Auf den Respawn wird gewartet...",
    "Der Kessel-Flug wird in weniger als 12 Parsec beendet...",
    "Der Kuchen ist keine L\xFCge, er l\xE4dt nur noch...",
    "Am Charaktererstellungsbildschirm wird herumgefummelt...",
    "Einen Moment, ich suche das richtige Meme...",
    "'A' wird zum Fortfahren gedr\xFCckt...",
    "Digitale Katzen werden geh\xFCtet...",
    "Die Pixel werden poliert...",
    "Ein passender Ladebildschirm-Witz wird gesucht...",
    "Ich lenke Sie mit diesem witzigen Spruch ab...",
    "Fast da... wahrscheinlich...",
    "Unsere Hamster arbeiten so schnell sie k\xF6nnen...",
    "Cloudy wird am Kopf gestreichelt...",
    "Die Katze wird gestreichelt...",
    "Meinen Chef rickrollen...",
    "Never gonna give you up, never gonna let you down...",
    "Auf den Bass wird geschlagen...",
    "Die Schnozbeeren werden probiert...",
    "I'm going the distance, I'm going for speed...",
    "Ist dies das wahre Leben? Ist dies nur Fantasie?...",
    "Ich habe ein gutes Gef\xFChl dabei...",
    "Den B\xE4ren wird gestupst...",
    "Recherche zu den neuesten Memes...",
    "\xDCberlege, wie ich das witziger machen kann...",
    "Hmmm... lassen Sie mich nachdenken...",
    "Wie nennt man einen Fisch ohne Augen? Ein Fsh...",
    "Warum ging der Computer zur Therapie? Er hatte zu viele Bytes...",
    "Warum m\xF6gen Programmierer keine Natur? Sie hat zu viele Bugs...",
    "Warum bevorzugen Programmierer den Dunkelmodus? Weil Licht Bugs anzieht...",
    "Warum ging der Entwickler pleite? Er hat seinen ganzen Cache aufgebraucht...",
    "Was kann man mit einem kaputten Bleistift machen? Nichts, er ist sinnlos...",
    "Perkussive Wartung wird angewendet...",
    "Die richtige USB-Ausrichtung wird gesucht...",
    "Es wird sichergestellt, dass der magische Rauch in den Kabeln bleibt...",
    "Versuche Vim zu beenden...",
    "Das Hamsterrad wird angeworfen...",
    "Das ist kein Bug, das ist ein undokumentiertes Feature...",
    "Engage.",
    "Ich komme wieder... mit einer Antwort.",
    "Mein anderer Prozess ist eine TARDIS...",
    "Mit dem Maschinengeist wird kommuniziert...",
    "Die Gedanken marinieren lassen...",
    "Gerade erinnert, wo ich meine Schl\xFCssel hingelegt habe...",
    "\xDCber die Kugel wird nachgedacht...",
    "Ich habe Dinge gesehen, die Sie nicht glauben w\xFCrden... wie einen Benutzer, der Lademeldungen liest.",
    "Nachdenklicher Blick wird initiiert...",
    "Was ist der Lieblingssnack eines Computers? Mikrochips.",
    "Warum tragen Java-Entwickler Brillen? Weil sie nicht C#.",
    "Der Laser wird aufgeladen... pew pew!",
    "Durch Null wird geteilt... nur Spa\xDF!",
    "Suche nach einem erwachsenen Aufseh... ich meine, Verarbeitung.",
    "Es piept und boopt.",
    "Pufferung... weil auch KIs einen Moment brauchen.",
    "Quantenteilchen werden f\xFCr schnellere Antwort verschr\xE4nkt...",
    "Das Chrom wird poliert... an den Algorithmen.",
    "Sind Sie nicht unterhalten? (Arbeite daran!)",
    "Die Code-Gremlins werden beschworen... zum Helfen, nat\xFCrlich.",
    "Warte nur auf das Einwahlton-Ende...",
    "Das Humor-O-Meter wird neu kalibriert.",
    "Mein anderer Ladebildschirm ist noch lustiger.",
    "Ziemlich sicher, dass irgendwo eine Katze \xFCber die Tastatur l\xE4uft...",
    "Verbessern... Verbessern... L\xE4dt noch.",
    "Das ist kein Bug, das ist ein Feature... dieses Ladebildschirms.",
    "Haben Sie versucht, es aus- und wieder einzuschalten? (Den Ladebildschirm, nicht mich.)",
    "Zus\xE4tzliche Pylonen werden gebaut..."
  ],
  // ============================================================================
  // Extension Settings Input
  // ============================================================================
  "Enter value...": "Wert eingeben...",
  "Enter sensitive value...": "Sensiblen Wert eingeben...",
  "Press Enter to submit, Escape to cancel": "Enter zum Absenden, Escape zum Abbrechen dr\xFCcken",
  // ============================================================================
  // Command Migration Tool
  // ============================================================================
  "Markdown file already exists: {{filename}}": "Markdown-Datei existiert bereits: {{filename}}",
  "TOML Command Format Deprecation Notice": "TOML-Befehlsformat Veraltet-Hinweis",
  "Found {{count}} command file(s) in TOML format:": "{{count}} Befehlsdatei(en) im TOML-Format gefunden:",
  "The TOML format for commands is being deprecated in favor of Markdown format.": "Das TOML-Format f\xFCr Befehle wird zugunsten des Markdown-Formats eingestellt.",
  "Markdown format is more readable and easier to edit.": "Das Markdown-Format ist lesbarer und einfacher zu bearbeiten.",
  "You can migrate these files automatically using:": "Sie k\xF6nnen diese Dateien automatisch migrieren mit:",
  "Or manually convert each file:": "Oder jede Datei manuell konvertieren:",
  'TOML: prompt = "..." / description = "..."': 'TOML: prompt = "..." / description = "..."',
  "Markdown: YAML frontmatter + content": "Markdown: YAML-Frontmatter + Inhalt",
  "The migration tool will:": "Das Migrationstool wird:",
  "Convert TOML files to Markdown": "TOML-Dateien in Markdown konvertieren",
  "Create backups of original files": "Sicherungen der Originaldateien erstellen",
  "Preserve all command functionality": "Alle Befehlsfunktionen beibehalten",
  "TOML format will continue to work for now, but migration is recommended.": "Das TOML-Format funktioniert vorerst weiter, aber eine Migration wird empfohlen.",
  // ============================================================================
  // Extensions - Explore Command
  // ============================================================================
  "Open extensions page in your browser": "Erweiterungsseite im Browser \xF6ffnen",
  "Unknown extensions source: {{source}}.": "Unbekannte Erweiterungsquelle: {{source}}.",
  "Would open extensions page in your browser: {{url}} (skipped in test environment)": "W\xFCrde Erweiterungsseite im Browser \xF6ffnen: {{url}} (\xFCbersprungen in Testumgebung)",
  "View available extensions at {{url}}": "Verf\xFCgbare Erweiterungen ansehen unter {{url}}",
  "Opening extensions page in your browser: {{url}}": "Erweiterungsseite wird im Browser ge\xF6ffnet: {{url}}",
  "Failed to open browser. Check out the extensions gallery at {{url}}": "Browser konnte nicht ge\xF6ffnet werden. Besuchen Sie die Erweiterungsgalerie unter {{url}}",
  "Use /compress when the conversation gets long to summarize history and free up context.": "Verwenden Sie /compress, wenn die Unterhaltung lang wird, um den Verlauf zusammenzufassen und Kontext freizugeben.",
  "Start a fresh idea with /clear or /new; the previous session stays available in history.": "Starten Sie eine neue Idee mit /clear oder /new; die vorherige Sitzung bleibt im Verlauf verf\xFCgbar.",
  "Use /bug to submit issues to the maintainers when something goes off.": "Verwenden Sie /bug, um Probleme an die Betreuer zu melden, wenn etwas schiefgeht.",
  "Switch auth type quickly with /auth.": "Wechseln Sie den Authentifizierungstyp schnell mit /auth.",
  "You can run any shell commands from Qwen Code using ! (e.g. !ls).": "Sie k\xF6nnen beliebige Shell-Befehle in Qwen Code mit ! ausf\xFChren (z. B. !ls).",
  "Type / to open the command popup; Tab autocompletes slash commands and saved prompts.": "Geben Sie / ein, um das Befehlsmen\xFC zu \xF6ffnen; Tab vervollst\xE4ndigt Slash-Befehle und gespeicherte Prompts.",
  "You can resume a previous conversation by running qwen --continue or qwen --resume.": "Sie k\xF6nnen eine fr\xFChere Unterhaltung mit qwen --continue oder qwen --resume fortsetzen.",
  "You can switch permission mode quickly with Shift+Tab or /approval-mode.": "Sie k\xF6nnen den Berechtigungsmodus schnell mit Shift+Tab oder /approval-mode wechseln.",
  "You can switch permission mode quickly with Tab or /approval-mode.": "Sie k\xF6nnen den Berechtigungsmodus schnell mit Tab oder /approval-mode wechseln.",
  "Try /insight to generate personalized insights from your chat history.": "Probieren Sie /insight, um personalisierte Erkenntnisse aus Ihrem Chatverlauf zu erstellen.",
  "Press Ctrl+O to toggle compact mode \u2014 hide tool output and thinking for a cleaner view.": "Ctrl+O dr\xFCcken, um den Kompaktmodus umzuschalten \u2014 Tool-Ausgabe und Denkprozess ausblenden.",
  "Add a QWEN.md file to give Qwen Code persistent project context.": "F\xFCgen Sie eine QWEN.md-Datei hinzu, um Qwen Code dauerhaften Projektkontext zu geben.",
  "Use /btw to ask a quick side question without disrupting the conversation.": "Verwenden Sie /btw, um eine kurze Nebenfrage zu stellen, ohne die Unterhaltung zu unterbrechen.",
  "Context is almost full! Run /compress now or start /new to continue.": "Der Kontext ist fast voll! F\xFChren Sie jetzt /compress aus oder starten Sie /new, um fortzufahren.",
  "Context is getting full. Use /compress to free up space.": "Der Kontext f\xFCllt sich. Verwenden Sie /compress, um Platz freizugeben.",
  "Long conversation? /compress summarizes history to free context.": "Lange Unterhaltung? /compress fasst den Verlauf zusammen, um Kontext freizugeben.",
  // ============================================================================
  // Custom API Key Configuration
  // ============================================================================
  "You can configure your API key and models in settings.json": "Sie k\xF6nnen Ihren API Key und Modelle in settings.json konfigurieren",
  "Refer to the documentation for setup instructions": "Einrichtungsanweisungen finden Sie in der Dokumentation",
  // ============================================================================
  // Coding Plan Authentication
  // ============================================================================
  "API key cannot be empty.": "API Key darf nicht leer sein.",
  "You can get your Coding Plan API key here": "Sie k\xF6nnen Ihren Coding Plan API Key hier erhalten",
  "Failed to update Coding Plan configuration: {{message}}": "Fehler beim Aktualisieren der Coding Plan-Konfiguration: {{message}}",
  // ============================================================================
  // Auth Dialog - View Titles and Labels
  // ============================================================================
  "Coding Plan": "Coding Plan",
  Custom: "Benutzerdefiniert",
  "Select Region for Coding Plan": "Region f\xFCr Coding Plan ausw\xE4hlen",
  "Choose based on where your account is registered": "W\xE4hlen Sie basierend auf dem Registrierungsort Ihres Kontos",
  "Enter Coding Plan API Key": "Coding Plan API Key eingeben",
  // ============================================================================
  // Coding Plan International Updates
  // ============================================================================
  "New model configurations are available for {{region}}. Update now?": "Neue Modellkonfigurationen sind f\xFCr {{region}} verf\xFCgbar. Jetzt aktualisieren?",
  '{{region}} configuration updated successfully. Model switched to "{{model}}".': '{{region}}-Konfiguration erfolgreich aktualisiert. Modell auf "{{model}}" umgeschaltet.',
  // ============================================================================
  // Context Usage Component
  // ============================================================================
  "Context Usage": "Kontextnutzung",
  "% used": "% verwendet",
  "% context used": "% Kontext verwendet",
  "Context exceeds limit! Use /compress or /clear to reduce.": "Kontext \xFCberschreitet Limit! Verwenden Sie /compress oder /clear zum Reduzieren.",
  "No API response yet. Send a message to see actual usage.": "Noch keine API-Antwort. Senden Sie eine Nachricht, um die tats\xE4chliche Nutzung anzuzeigen.",
  "Estimated pre-conversation overhead": "Gesch\xE4tzte Vorabkosten vor der Unterhaltung",
  "Context window": "Kontextfenster",
  tokens: "Tokens",
  Used: "Verwendet",
  Free: "Frei",
  "Autocompact buffer": "Autokomprimierungs-Puffer",
  "Usage by category": "Verwendung nach Kategorie",
  "System prompt": "System-Prompt",
  "Built-in tools": "Integrierte Tools",
  "MCP tools": "MCP tools",
  "Memory files": "Speicherdateien",
  Skills: "F\xE4higkeiten",
  Messages: "Nachrichten",
  "Run /context detail for per-item breakdown.": "F\xFChren Sie /context detail f\xFCr eine Aufschl\xFCsselung nach Elementen aus.",
  active: "aktiv",
  "body loaded": "Inhalt geladen",
  memory: "Speicher",
  "{{region}} configuration updated successfully.": "{{region}}-Konfiguration erfolgreich aktualisiert.",
  "Authenticated successfully with {{region}}. API key and model configs saved to settings.json.": "Erfolgreich mit {{region}} authentifiziert. API Key und Modellkonfigurationen wurden in settings.json gespeichert.",
  "Tip: Use /model to switch between available Coding Plan models.": "Tipp: Verwenden Sie /model, um zwischen verf\xFCgbaren Coding Plan-Modellen zu wechseln.",
  "Type something...": "Etwas eingeben...",
  Submit: "Senden",
  "Submit answers": "Antworten senden",
  Cancel: "Abbrechen",
  "Your answers:": "Ihre Antworten:",
  "(not answered)": "(nicht beantwortet)",
  "Ready to submit your answers?": "Bereit, Ihre Antworten zu senden?",
  "\u2191/\u2193: Navigate | \u2190/\u2192: Switch tabs | Enter: Select": "\u2191/\u2193: Navigieren | \u2190/\u2192: Tabs wechseln | Enter: Ausw\xE4hlen",
  "\u2191/\u2193: Navigate | Enter: Select | Esc: Cancel": "\u2191/\u2193: Navigieren | Enter: Ausw\xE4hlen | Esc: Abbrechen",
  "Authenticate using Qwen OAuth": "Mit Qwen OAuth authentifizieren",
  "Authenticate using Alibaba Cloud Coding Plan": "Mit Alibaba Cloud Coding Plan authentifizieren",
  "Region for Coding Plan (china/global)": "Region f\xFCr Coding Plan (china/global)",
  "API key for Coding Plan": "API Key f\xFCr Coding Plan",
  "Show current authentication status": "Aktuellen Authentifizierungsstatus anzeigen",
  "Authentication completed successfully.": "Authentifizierung erfolgreich abgeschlossen.",
  "Starting Qwen OAuth authentication...": "Qwen OAuth-Authentifizierung wird gestartet...",
  "Successfully authenticated with Qwen OAuth.": "Erfolgreich mit Qwen OAuth authentifiziert.",
  "Failed to authenticate with Qwen OAuth: {{error}}": "Authentifizierung mit Qwen OAuth fehlgeschlagen: {{error}}",
  "Processing Alibaba Cloud Coding Plan authentication...": "Alibaba Cloud Coding Plan-Authentifizierung wird verarbeitet...",
  "Successfully authenticated with Alibaba Cloud Coding Plan.": "Erfolgreich mit Alibaba Cloud Coding Plan authentifiziert.",
  "Failed to authenticate with Coding Plan: {{error}}": "Authentifizierung mit Coding Plan fehlgeschlagen: {{error}}",
  "\u963F\u91CC\u4E91\u767E\u70BC (aliyun.com)": "\u963F\u91CC\u4E91\u767E\u70BC (aliyun.com)",
  Global: "Global",
  "Alibaba Cloud (alibabacloud.com)": "Alibaba Cloud (alibabacloud.com)",
  "Select region for Coding Plan:": "Region f\xFCr Coding Plan ausw\xE4hlen:",
  "Enter your Coding Plan API key: ": "Geben Sie Ihren Coding Plan API Key ein: ",
  "Select authentication method:": "Authentifizierungsmethode ausw\xE4hlen:",
  "\n=== Authentication Status ===\n": "\n=== Authentifizierungsstatus ===\n",
  "\u26A0\uFE0F  No authentication method configured.\n": "\u26A0\uFE0F  Keine Authentifizierungsmethode konfiguriert.\n",
  "Run one of the following commands to get started:\n": "F\xFChren Sie einen der folgenden Befehle aus, um zu beginnen:\n",
  "  qwen auth qwen-oauth     - Authenticate with Qwen OAuth (discontinued)": "  qwen auth qwen-oauth     - Mit Qwen OAuth authentifizieren (eingestellt)",
  "Or simply run:": "Oder einfach ausf\xFChren:",
  "  qwen auth                - Interactive authentication setup\n": "  qwen auth                - Interaktive Authentifizierungseinrichtung\n",
  "\u2713 Authentication Method: Qwen OAuth": "\u2713 Authentifizierungsmethode: Qwen OAuth",
  "  Type: Free tier (discontinued 2026-04-15)": "  Typ: Kostenloses Kontingent (eingestellt 2026-04-15)",
  "  Limit: No longer available": "  Limit: Nicht mehr verf\xFCgbar",
  "Qwen OAuth free tier was discontinued on 2026-04-15. Run /auth to switch to Coding Plan, OpenRouter, Fireworks AI, or another provider.": "Das kostenlose Qwen OAuth-Kontingent wurde am 2026-04-15 eingestellt. F\xFChren Sie /auth aus, um zu Coding Plan, OpenRouter, Fireworks AI oder einem anderen Anbieter zu wechseln.",
  "\u2713 Authentication Method: Alibaba Cloud Coding Plan": "\u2713 Authentifizierungsmethode: Alibaba Cloud Coding Plan",
  "Global - Alibaba Cloud": "Global - Alibaba Cloud",
  "  Region: {{region}}": "  Region: {{region}}",
  "  Current Model: {{model}}": "  Aktuelles Modell: {{model}}",
  "  Config Version: {{version}}": "  Konfigurationsversion: {{version}}",
  "  Status: API key configured\n": "  Status: API Key konfiguriert\n",
  "\u26A0\uFE0F  Authentication Method: Alibaba Cloud Coding Plan (Incomplete)": "\u26A0\uFE0F  Authentifizierungsmethode: Alibaba Cloud Coding Plan (Unvollst\xE4ndig)",
  "  Issue: API key not found in environment or settings\n": "  Problem: API Key nicht in Umgebung oder Einstellungen gefunden\n",
  "  Run `qwen auth coding-plan` to re-configure.\n": "  F\xFChren Sie `qwen auth coding-plan` aus, um neu zu konfigurieren.\n",
  "\u2713 Authentication Method: {{type}}": "\u2713 Authentifizierungsmethode: {{type}}",
  "  Status: Configured\n": "  Status: Konfiguriert\n",
  "Failed to check authentication status: {{error}}": "Authentifizierungsstatus konnte nicht \xFCberpr\xFCft werden: {{error}}",
  "Select an option:": "Option ausw\xE4hlen:",
  "Raw mode not available. Please run in an interactive terminal.": "Raw-Modus nicht verf\xFCgbar. Bitte in einem interaktiven Terminal ausf\xFChren.",
  "(Use \u2191 \u2193 arrows to navigate, Enter to select, Ctrl+C to exit)\n": "(\u2191 \u2193 Pfeiltasten zum Navigieren, Enter zum Ausw\xE4hlen, Ctrl+C zum Beenden)\n",
  "to toggle compact mode": "Kompaktmodus umschalten",
  "Hide tool output and thinking for a cleaner view (toggle with Ctrl+O).": "Tool-Ausgabe und Denkprozess ausblenden f\xFCr eine \xFCbersichtlichere Ansicht (mit Ctrl+O umschalten).",
  "Press Ctrl+O to show full tool output": "Ctrl+O f\xFCr vollst\xE4ndige Tool-Ausgabe dr\xFCcken",
  "Switch to plan mode or exit plan mode": "In den Plan-Modus wechseln oder den Plan-Modus verlassen",
  "Exited plan mode. Previous approval mode restored.": "Plan-Modus verlassen. Vorheriger Genehmigungsmodus wiederhergestellt.",
  "Enabled plan mode. The agent will analyze and plan without executing tools.": "Plan-Modus aktiviert. Der Agent analysiert und plant, ohne Werkzeuge auszuf\xFChren.",
  'Already in plan mode. Use "/plan exit" to exit plan mode.': 'Bereits im Plan-Modus. Verwenden Sie "/plan exit", um den Plan-Modus zu verlassen.',
  'Not in plan mode. Use "/plan" to enter plan mode first.': 'Nicht im Plan-Modus. Verwenden Sie "/plan", um zuerst in den Plan-Modus zu gelangen.',
  "Set up Qwen Code's status line UI": "Qwen Codes Statusleisten-UI einrichten",
  // === Core: added from PR #3328 ===
  "Open the memory manager.": "Den Speicher-Manager \xF6ffnen.",
  "Save a durable memory to the memory system.": "Eine dauerhafte Erinnerung im Speichersystem speichern.",
  "Open MCP management dialog": "MCP-Verwaltungsdialog \xF6ffnen",
  "Manage extension settings": "Erweiterungseinstellungen verwalten",
  prompts: "Eingabeaufforderungen",
  "Manage Extensions": "Erweiterungen verwalten",
  "Extension Details": "Erweiterungsdetails",
  "View Extension": "Erweiterung anzeigen",
  "Update Extension": "Erweiterung aktualisieren",
  "Disable Extension": "Erweiterung deaktivieren",
  "Enable Extension": "Erweiterung aktivieren",
  "Uninstall Extension": "Erweiterung deinstallieren",
  "Select Scope": "Bereich ausw\xE4hlen",
  "User Scope": "Benutzerbereich",
  "Workspace Scope": "Arbeitsbereich",
  "No extensions found.": "Keine Erweiterungen gefunden.",
  "Toggle this help display": "Diese Hilfe ein- oder ausblenden",
  "Toggle shell mode": "Shell-Modus umschalten",
  "Open command menu": "Befehlsmen\xFC \xF6ffnen",
  "Add file context": "Dateikontext hinzuf\xFCgen",
  "Accept suggestion / Autocomplete": "Vorschlag akzeptieren / automatisch vervollst\xE4ndigen",
  "Reverse search history": "Verlauf r\xFCckw\xE4rts durchsuchen",
  "Press ? again to close": "Erneut ? dr\xFCcken, um zu schlie\xDFen",
  "? for shortcuts": "? f\xFCr Tastenk\xFCrzel",
  'Invalid approval mode "{{arg}}". Valid modes: {{modes}}': 'Ung\xFCltiger Freigabemodus "{{arg}}". G\xFCltige Modi: {{modes}}',
  'Approval mode set to "{{mode}}"': 'Freigabemodus auf "{{mode}}" gesetzt',
  'Are you sure you want to uninstall extension "{{name}}"?': 'Sind Sie sicher, dass Sie die Erweiterung "{{name}}" deinstallieren m\xF6chten?',
  "This action cannot be undone.": "Diese Aktion kann nicht r\xFCckg\xE4ngig gemacht werden.",
  'Extension "{{name}}" updated successfully.': 'Erweiterung "{{name}}" erfolgreich aktualisiert.',
  "Name:": "Name:",
  "MCP Servers:": "MCP Servers:",
  "Settings:": "Einstellungen:",
  "View Details": "Details anzeigen",
  "Update failed:": "Aktualisierung fehlgeschlagen:",
  "Updating {{name}}...": "{{name}} wird aktualisiert...",
  "Update complete!": "Aktualisierung abgeschlossen!",
  "User (global)": "Benutzer (global)",
  "Workspace (project-specific)": "Arbeitsbereich (projektspezifisch)",
  'Disable "{{name}}" - Select Scope': '"{{name}}" deaktivieren - Bereich ausw\xE4hlen',
  'Enable "{{name}}" - Select Scope': '"{{name}}" aktivieren - Bereich ausw\xE4hlen',
  "No extension selected": "Keine Erweiterung ausgew\xE4hlt",
  "{{count}} extensions installed": "{{count}} Erweiterungen installiert",
  "up to date": "aktuell",
  "update available": "Update verf\xFCgbar",
  "checking...": "wird gepr\xFCft...",
  "not updatable": "nicht aktualisierbar",
  "Ask a quick side question without affecting the main conversation": "Eine kurze Nebenfrage stellen, ohne die Hauptunterhaltung zu beeinflussen",
  "Manage Arena sessions": "Arena-Sitzungen verwalten",
  "Start an Arena session with multiple models competing on the same task": "Eine Arena-Sitzung starten, in der mehrere Modelle dieselbe Aufgabe bearbeiten",
  "Stop the current Arena session": "Die aktuelle Arena-Sitzung beenden",
  "Show the current Arena session status": "Den Status der aktuellen Arena-Sitzung anzeigen",
  "Select a model result and merge its diff into the current workspace": "Ein Modellergebnis ausw\xE4hlen und dessen Diff in den aktuellen Arbeitsbereich \xFCbernehmen",
  "No running Arena session found.": "Keine laufende Arena-Sitzung gefunden.",
  "No Arena session found. Start one with /arena start.": "Keine Arena-Sitzung gefunden. Starten Sie eine mit /arena start.",
  "Arena session is still running. Wait for it to complete or use /arena stop first.": "Die Arena-Sitzung l\xE4uft noch. Warten Sie, bis sie abgeschlossen ist, oder verwenden Sie zuerst /arena stop.",
  "No successful agent results to select from. All agents failed or were cancelled.": "Keine erfolgreichen Agent-Ergebnisse zur Auswahl. Alle Agents sind fehlgeschlagen oder wurden abgebrochen.",
  "Use /arena stop to end the session.": "Verwenden Sie /arena stop, um die Sitzung zu beenden.",
  'No idle agent found matching "{{name}}".': 'Kein inaktiver Agent gefunden, der "{{name}}" entspricht.',
  "Failed to apply changes from {{label}}: {{error}}": "Anwenden der \xC4nderungen von {{label}} fehlgeschlagen: {{error}}",
  "Applied changes from {{label}} to workspace. Arena session complete.": "\xC4nderungen von {{label}} auf den Arbeitsbereich angewendet. Arena-Sitzung abgeschlossen.",
  "Discard all Arena results and clean up worktrees?": "Alle Arena-Ergebnisse verwerfen und Arbeitsb\xE4ume bereinigen?",
  "Arena results discarded. All worktrees cleaned up.": "Arena-Ergebnisse verworfen. Alle Arbeitsb\xE4ume wurden bereinigt.",
  "Arena is not supported in non-interactive mode. Use interactive mode to start an Arena session.": "Arena wird im nicht-interaktiven Modus nicht unterst\xFCtzt. Verwenden Sie den interaktiven Modus, um eine Arena-Sitzung zu starten.",
  "Arena is not supported in non-interactive mode. Use interactive mode to stop an Arena session.": "Arena wird im nicht-interaktiven Modus nicht unterst\xFCtzt. Verwenden Sie den interaktiven Modus, um eine Arena-Sitzung zu beenden.",
  "Arena is not supported in non-interactive mode.": "Arena wird im nicht-interaktiven Modus nicht unterst\xFCtzt.",
  "An Arena session exists. Use /arena stop or /arena select to end it before starting a new one.": "Es existiert bereits eine Arena-Sitzung. Verwenden Sie /arena stop oder /arena select, um sie zu beenden, bevor Sie eine neue starten.",
  "Usage: /arena start --models model1,model2 <task>": "Verwendung: /arena start --models model1,model2 <Aufgabe>",
  "Models to compete (required, at least 2)": "Wettbewerbsmodelle (erforderlich, mindestens 2)",
  "Format: authType:modelId or just modelId": "Format: authType:modelId oder nur modelId",
  "Arena requires at least 2 models. Use --models model1,model2 to specify.": "Arena ben\xF6tigt mindestens 2 Modelle. Verwenden Sie --models model1,model2 zur Angabe.",
  'Arena started with {{count}} agents on task: "{{task}}"\nModels:\n{{modelList}}': 'Arena mit {{count}} Agents f\xFCr Aufgabe "{{task}}" gestartet\nModelle:\n{{modelList}}',
  "Arena panes are running in tmux. Attach with: `{{command}}`": "Arena-Panels laufen in tmux. Verbinden mit: `{{command}}`",
  "[{{label}}] failed: {{error}}": "[{{label}}] fehlgeschlagen: {{error}}",
  "Loading suggestions...": "Vorschl\xE4ge werden geladen...",
  'Show context window usage breakdown. Use "/context detail" for per-item breakdown.': 'Die Aufschl\xFCsselung der Nutzung des Kontextfensters anzeigen. F\xFCr Details pro Element "/context detail" verwenden.',
  "Show per-item context usage breakdown.": "Die Aufschl\xFCsselung der Kontextnutzung pro Element anzeigen.",
  // === Missing key backfill ===
  "for shell mode": "f\xFCr Shell-Modus",
  "for commands": "f\xFCr Befehle",
  "for file paths": "f\xFCr Dateipfade",
  "to clear input": "zum Leeren der Eingabe",
  "to cycle approvals": "zum Wechseln der Freigaben",
  "to quit": "zum Beenden",
  "for newline": "f\xFCr Zeilenumbruch",
  "to clear screen": "zum Leeren des Bildschirms",
  "to search history": "zum Durchsuchen des Verlaufs",
  "to paste images": "zum Einf\xFCgen von Bildern",
  "for external editor": "f\xFCr externen Editor",
  "Updating...": "Wird aktualisiert...",
  Unknown: "Unbekannt",
  Error: "Fehler",
  "Version:": "Version:",
  "Use '/extensions install' to install your first extension.": "Verwenden Sie '/extensions install', um Ihre erste Erweiterung zu installieren.",
  "Value:": "Wert:",
  "Press c to copy the authorization URL to your clipboard.": "Dr\xFCcken Sie c, um die Autorisierungs-URL in die Zwischenablage zu kopieren.",
  "Copy request sent to your terminal. If paste is empty, copy the URL above manually.": "Kopieranfrage an Ihr Terminal gesendet. Wenn das Einf\xFCgen leer ist, kopieren Sie die URL oben manuell.",
  "Cannot write to terminal \u2014 copy the URL above manually.": "Schreiben ins Terminal nicht m\xF6glich \u2014 kopieren Sie die URL oben manuell.",
  "Tips:": "Tipps:",
  "Retrying in {{seconds}} seconds\u2026 (attempt {{attempt}}/{{maxRetries}})": "Erneuter Versuch in {{seconds}} Sekunden\u2026 (Versuch {{attempt}}/{{maxRetries}})",
  "Press Ctrl+Y to retry": "Dr\xFCcken Sie Ctrl+Y, um es erneut zu versuchen",
  "No failed request to retry.": "Keine fehlgeschlagene Anfrage zum Wiederholen.",
  "to retry last request": "um die letzte Anfrage erneut zu versuchen",
  'Invalid API key. Coding Plan API keys start with "sk-sp-". Please check.': 'Ung\xFCltiger API Key. Coding Plan API Keys beginnen mit "sk-sp-". Bitte pr\xFCfen.',
  "Lock release warning": "Warnung zur Sperrfreigabe",
  "Metadata write warning": "Warnung beim Schreiben der Metadaten",
  "Subsequent dreams may be skipped as locked until the next session's staleness sweep cleans the file.": "Weitere Dream-L\xE4ufe k\xF6nnen als gesperrt \xFCbersprungen werden, bis der n\xE4chste Stale-Sweep der Sitzung die Datei bereinigt.",
  "The scheduler gate did not see this dream's timestamp; the next dream cycle may re-fire sooner than usual.": "Das Scheduler-Gate hat den Zeitstempel dieses Dream-Laufs nicht gesehen; der n\xE4chste Dream-Zyklus kann fr\xFCher als \xFCblich erneut starten.",
  // === History collapse/expand commands ===
  "Set history to collapse by default when resuming a session": "Set history to collapse by default when resuming a session",
  "Set history to expand by default when resuming a session": "Set history to expand by default when resuming a session",
  "Expand the currently collapsed history transcript": "Expand the currently collapsed history transcript",
  "Control history display preferences and visibility": "Control history display preferences and visibility",
  "History will be collapsed by default for future resumed sessions.": "History will be collapsed by default for future resumed sessions.",
  "History will be expanded by default for future resumed sessions.": "History will be expanded by default for future resumed sessions.",
  "History is already expanded in this session.": "History is already expanded in this session.",
  "Usage: /history collapse-on-resume|expand-on-resume|expand-now": "Usage: /history collapse-on-resume|expand-on-resume|expand-now",
  "History collapsed: {{n}} messages hidden. Use /history expand-now to show.": "Geschichte eingeklappt: {{n}} Nachrichten ausgeblendet. Verwenden Sie /history expand-now zum Anzeigen.",
  // === Same-as-English optimization ===
  "Agents:": "Agenten:",
  Prompt: "Eingabe",
  "Prompts:": "Eingaben:",
  "Resources:": "Ressourcen:",
  "Ref:": "Referenz:",
  "Skills:": "F\xE4higkeiten:",
  remote: "entfernt",
  "\u4E2D\u56FD (China)": "China",
  "\u4E2D\u56FD (China) - \u963F\u91CC\u4E91\u767E\u70BC": "China - \u963F\u91CC\u4E91\u767E\u70BC",
  // Stats Dashboard — Category 2
  "Activity Heatmap": "Aktivit\xE4ts-Heatmap",
  Less: "Weniger",
  More: "Mehr",
  Sessions: "Sitzungen",
  Duration: "Dauer",
  Projects: "Projekte",
  "Loading stats...": "Statistiken werden geladen...",
  "(no data)": "(keine Daten)",
  d: "d",
  h: "h",
  m: "m",
  Input: "Eingabe",
  Models: "Modelle",
  "All time": "Gesamtzeitraum",
  "Last 7 days": "Letzte 7 Tage",
  "Last 30 days": "Letzte 30 Tage",
  "Show usage statistics dashboard.": "Nutzungsstatistik-Dashboard anzeigen.",
  // Stats Dashboard — keyboard hints (not translated)
  "tab \xB7 esc": "tab \xB7 esc",
  "tab \xB7 r dates \xB7 \u2190\u2192 month \xB7 esc": "tab \xB7 r dates \xB7 \u2190\u2192 month \xB7 esc",
  "tab \xB7 r dates \xB7 esc": "tab \xB7 r dates \xB7 esc",
  // Stats Dashboard — missing labels
  "API Requests": "API-Anfragen",
  "Tool Calls": "Tool-Aufrufe",
  "Success rate": "Erfolgsrate",
  "Code Changes": "Code-\xC4nderungen",
  Tool: "Tool",
  reqs: "Anfr.",
  in: "ein",
  out: "aus",
  "In/Out": "Ein/Aus"
};
export {
  de_default as default
};
/**
 * @license
 * Copyright 2025 Qwen
 * SPDX-License-Identifier: Apache-2.0
 */
