// @qwen-code/qwen-code/chunks/pt-TNZRNLQB.js
// Force strict mode and setup for ESM
"use strict";
import {
  init_esbuild_shims
} from "./chunk-A4BMJM77.js";
import "./chunk-J2S4EL5Y.js";

// packages/cli/src/i18n/locales/pt.js
init_esbuild_shims();
var pt_default = {
  // ============================================================================
  // Help / UI Components
  // ============================================================================
  "Basics:": "No\xE7\xF5es b\xE1sicas:",
  "Add context": "Adicionar contexto",
  "Use {{symbol}} to specify files for context (e.g., {{example}}) to target specific files or folders.": "Use {{symbol}} para especificar arquivos para o contexto (ex: {{example}}) para atingir arquivos ou pastas espec\xEDficos.",
  "@": "@",
  "@src/myFile.ts": "@src/myFile.ts",
  "Shell mode": "Modo shell",
  "YOLO mode": "Modo YOLO",
  "Auto mode": "Modo auto",
  "plan mode": "modo planejamento",
  "auto-accept edits": "aceitar edi\xE7\xF5es automaticamente",
  "Accepting edits": "Aceitando edi\xE7\xF5es",
  "(shift + tab to cycle)": "(Shift + Tab para alternar)",
  "Execute shell commands via {{symbol}} (e.g., {{example1}}) or use natural language (e.g., {{example2}}).": "Execute comandos shell via {{symbol}} (ex: {{example1}}) ou use linguagem natural (ex: {{example2}}).",
  "!": "!",
  "!npm run start": "!npm run start",
  "start server": "iniciar servidor",
  "Commands:": "Comandos:",
  "shell command": "comando shell",
  "Model Context Protocol command (from external servers)": "Comando Model Context Protocol (de servidores externos)",
  "Keyboard Shortcuts:": "Atalhos de teclado:",
  "Toggle this help display": "Alternar exibi\xE7\xE3o desta ajuda",
  "Toggle shell mode": "Alternar modo shell",
  "Open command menu": "Abrir menu de comandos",
  "Add file context": "Adicionar contexto de arquivo",
  "Accept suggestion / Autocomplete": "Aceitar sugest\xE3o / Autocompletar",
  "Reverse search history": "Pesquisa reversa no hist\xF3rico",
  "Press ? again to close": "Pressione ? novamente para fechar",
  // Keyboard shortcuts panel descriptions
  "for shell mode": "para modo shell",
  "for commands": "para comandos",
  "for file paths": "para caminhos de arquivo",
  "to clear input": "para limpar entrada",
  "to cycle approvals": "para alternar aprova\xE7\xF5es",
  "to quit": "para sair",
  "for newline": "para nova linha",
  "to clear screen": "para limpar a tela",
  "to search history": "para pesquisar no hist\xF3rico",
  "to paste images": "para colar imagens",
  "for external editor": "para editor externo",
  "to toggle compact mode": "alternar modo compacto",
  "Jump through words in the input": "Pular palavras na entrada",
  "Close dialogs, cancel requests, or quit application": "Fechar di\xE1logos, cancelar solicita\xE7\xF5es ou sair do aplicativo",
  "New line": "Nova linha",
  "New line (Alt+Enter works for certain linux distros)": "Nova linha (Alt+Enter funciona em certas distros linux)",
  "Clear the screen": "Limpar a tela",
  "Open input in external editor": "Abrir entrada no editor externo",
  "Send message": "Enviar mensagem",
  "Initializing...": "Inicializando...",
  "Connecting to MCP servers... ({{connected}}/{{total}})": "Conectando aos MCP servers... ({{connected}}/{{total}})",
  "Type your message or @path/to/file": "Digite sua mensagem ou @caminho/do/arquivo",
  "? for shortcuts": "? para atalhos",
  "Press 'i' for INSERT mode and 'Esc' for NORMAL mode.": "Pressione 'i' para modo INSER\xC7\xC3O e 'Esc' para modo NORMAL.",
  "Cancel operation / Clear input (double press)": "Cancelar opera\xE7\xE3o / Limpar entrada (pressionar duas vezes)",
  "Cycle approval modes": "Alternar modos de aprova\xE7\xE3o",
  "Cycle through your prompt history": "Alternar hist\xF3rico de prompts",
  "For a full list of shortcuts, see {{docPath}}": "Para uma lista completa de atalhos, consulte {{docPath}}",
  "docs/keyboard-shortcuts.md": "docs/keyboard-shortcuts.md",
  "for help on Qwen Code": "para ajuda sobre o Qwen Code",
  "show version info": "mostrar informa\xE7\xF5es de vers\xE3o",
  "submit a bug report": "enviar um relat\xF3rio de erro",
  // ============================================================================
  // System Information Fields
  // ============================================================================
  "Qwen Code": "Qwen Code",
  OS: "SO",
  Auth: "Autentica\xE7\xE3o",
  Model: "Modelo",
  "Fast Model": "Modelo R\xE1pido",
  Sandbox: "Sandbox",
  "Session ID": "ID da Sess\xE3o",
  "Base URL": "Base URL",
  Proxy: "Proxy",
  "Memory Usage": "Uso de Mem\xF3ria",
  "IDE Client": "Cliente IDE",
  // ============================================================================
  // Commands - General
  // ============================================================================
  "Analyzes the project and creates a tailored QWEN.md file.": "Analisa o projeto e cria um arquivo QWEN.md personalizado.",
  "List available Qwen Code tools. Usage: /tools [desc]": "Listar ferramentas Qwen Code dispon\xEDveis. Uso: /tools [desc]",
  "Open the skills panel (browse, search, toggle, pick).": "Abrir o painel de habilidades (explorar, pesquisar, ativar, selecionar).",
  "Manage Skills": "Gerenciar Habilidades",
  "Skills configuration saved.": "Configura\xE7\xE3o de habilidades salva.",
  "Skills configuration saved, but refresh failed: {{error}}. Restart to ensure the new state is applied.": "Configura\xE7\xE3o de habilidades salva, mas a atualiza\xE7\xE3o falhou: {{error}}. Reinicie para garantir que o novo estado seja aplicado.",
  "Workspace is untrusted; workspace settings are ignored by the merged config. Run /trust first to persist skills changes here, or edit ~/.qwen/settings.json directly to manage skills at user scope.": "O espa\xE7o de trabalho n\xE3o \xE9 confi\xE1vel; as configura\xE7\xF5es do espa\xE7o de trabalho s\xE3o ignoradas pela configura\xE7\xE3o combinada. Execute /trust primeiro, ou edite ~/.qwen/settings.json diretamente para gerenciar habilidades no escopo do usu\xE1rio.",
  "SkillManager not available.": "SkillManager indispon\xEDvel.",
  "Loading skills\u2026": "Carregando habilidades\u2026",
  "Failed to load skills: {{error}}": "Falha ao carregar habilidades: {{error}}",
  "Failed to save skills configuration: {{error}}": "Falha ao salvar a configura\xE7\xE3o de habilidades: {{error}}",
  "All available skills are disabled. Edit ~/.qwen/settings.json or .qwen/settings.json (skills.disabled) to re-enable.": "Todas as habilidades dispon\xEDveis est\xE3o desativadas. Edite ~/.qwen/settings.json ou .qwen/settings.json (skills.disabled) para reativ\xE1-las.",
  "Press esc to close.": "Pressione Esc para fechar.",
  "{{count}} skills \xB7 ": "{{count}} habilidades \xB7 ",
  "{{matched}} / {{total}} skills \xB7 ": "{{matched}} / {{total}} habilidades \xB7 ",
  "Space toggle \xB7 Enter pick (fill input) \xB7 Esc save & exit \xB7 workspace scope": "Espa\xE7o alternar \xB7 Enter selecionar (preencher entrada) \xB7 Esc salvar & sair \xB7 escopo do espa\xE7o de trabalho",
  "Search:": "Pesquisar:",
  "type to filter\u2026": "digite para filtrar\u2026",
  "No skills are currently available.": "Nenhuma habilidade est\xE1 dispon\xEDvel no momento.",
  "All available skills are locked at a higher scope (see below).": "Todas as habilidades dispon\xEDveis est\xE3o bloqueadas em um escopo superior (veja abaixo).",
  "No skills match the search.": "Nenhuma habilidade corresponde \xE0 pesquisa.",
  "Locked by higher-scope settings (cannot toggle here):": "Bloqueado por configura\xE7\xF5es de escopo superior (n\xE3o \xE9 poss\xEDvel alternar aqui):",
  "higher scope": "escopo superior",
  "  {{name}} {{description}}  [locked: {{scope}}]": "  {{name}} {{description}}  [bloqueado: {{scope}}]",
  "\u2191/\u2193 navigate \xB7 backspace edits search": "\u2191/\u2193 navegar \xB7 Backspace edita a pesquisa",
  Bundled: "Integrada",
  "Available Qwen Code CLI tools:": "Ferramentas CLI do Qwen Code dispon\xEDveis:",
  "No tools available": "Nenhuma ferramenta dispon\xEDvel",
  "View or change the approval mode for tool usage": "Ver ou alterar o modo de aprova\xE7\xE3o para uso de ferramentas",
  'Invalid approval mode "{{arg}}". Valid modes: {{modes}}': 'Modo de aprova\xE7\xE3o inv\xE1lido "{{arg}}". Modos v\xE1lidos: {{modes}}',
  'Approval mode set to "{{mode}}"': 'Modo de aprova\xE7\xE3o definido como "{{mode}}"',
  "View or change the language setting": "Ver ou alterar a configura\xE7\xE3o de idioma",
  "List background tasks (text dump \u2014 interactive dialog opens via the footer pill)": "Listar tarefas em segundo plano (sa\xEDda em texto; a caixa de di\xE1logo interativa pode ser aberta pelo indicador no rodap\xE9)",
  "Delete a previous session": "Excluir uma sess\xE3o anterior",
  "Run installation and environment diagnostics": "Executar diagn\xF3sticos de instala\xE7\xE3o e ambiente",
  "Browse dynamic model catalogs and choose which models stay enabled locally": "Navegar pelos cat\xE1logos din\xE2micos de modelos e escolher quais modelos permanecem ativados localmente",
  "Generate a one-line session recap now": "Gerar agora um resumo da sess\xE3o em uma linha",
  "Rename the current conversation. --auto lets the fast model pick a title.": "Renomear a conversa atual. --auto permite que o modelo r\xE1pido escolha um t\xEDtulo.",
  "Rewind conversation to a previous turn": "Voltar a conversa para um turno anterior",
  "Rewind Conversation": "Rebobinar conversa",
  "No user turns to rewind to.": "Nenhum turno de usu\xE1rio para rebobinar.",
  "Rewind to: ": "Rebobinar para: ",
  "Restore code and conversation": "Restaurar c\xF3digo e conversa",
  "Restore conversation only": "Restaurar apenas a conversa",
  "Restore code only": "Restaurar apenas o c\xF3digo",
  "Never mind": "Deixa pra l\xE1",
  "Computing file changes...": "Calculando altera\xE7\xF5es de arquivo...",
  "Restoring...": "Restaurando...",
  "Restored {{count}} file(s).": "{{count}} arquivo(s) restaurado(s).",
  "Failed to restore files: {{error}}": "Falha ao restaurar arquivos: {{error}}",
  "Rewind failed: {{error}}": "Falha ao retroceder: {{error}}",
  "Cannot rewind conversation: no active model client.": "N\xE3o \xE9 poss\xEDvel retroceder a conversa: nenhum cliente de modelo ativo.",
  "Code restored, but conversation could not be rewound (no active client).": "C\xF3digo restaurado, mas a conversa n\xE3o p\xF4de ser retrocedida (sem cliente ativo).",
  "Conversation rewound. Edit your prompt and press Enter to continue.": "Conversa retrocedida. Edite seu prompt e pressione Enter para continuar.",
  "Rewinding does not affect files edited manually or via shell commands.": "O retrocesso n\xE3o afeta arquivos editados manualmente ou por meio de comandos shell.",
  "Cannot rewind to a turn that was compressed. Try a more recent turn.": "N\xE3o \xE9 poss\xEDvel retroceder para um turno que foi compactado. Tente um turno mais recente.",
  "File restore is unavailable for this turn (no captured file changes, or this turn predates the current session).": "A restaura\xE7\xE3o de arquivos n\xE3o est\xE1 dispon\xEDvel para este turno (sem altera\xE7\xF5es capturadas, ou o turno \xE9 anterior \xE0 sess\xE3o atual).",
  "(+{{insertions}} -{{deletions}} in {{count}} file)": "(+{{insertions}} -{{deletions}} em {{count}} arquivo)",
  "(+{{insertions}} -{{deletions}} in {{count}} files)": "(+{{insertions}} -{{deletions}} em {{count}} arquivos)",
  "Failed to restore {{count}} file(s): {{files}}": "Falha ao restaurar {{count}} arquivo(s): {{files}}",
  "Cannot restore files: this turn was created before file checkpointing was enabled.": "N\xE3o \xE9 poss\xEDvel restaurar arquivos: este turno foi criado antes do checkpoint de arquivos ser ativado.",
  "No files needed to be restored.": "Nenhum arquivo precisou ser restaurado.",
  "\u2191\u2193 to navigate \xB7 Enter to select \xB7 Esc to go back": "\u2191\u2193 navegar \xB7 Enter selecionar \xB7 Esc voltar",
  "\u2191\u2193 to navigate \xB7 Enter to select \xB7 Esc to cancel": "\u2191\u2193 navegar \xB7 Enter selecionar \xB7 Esc cancelar",
  "Enter/Y to confirm \xB7 Esc/N to go back": "Enter/Y confirmar \xB7 Esc/N voltar",
  "change the theme": "alterar o tema",
  "Select Theme": "Selecionar Tema",
  Preview: "Visualizar",
  "(Use Enter to select, Tab to configure scope)": "(Use Enter para selecionar, Tab para configurar o escopo)",
  "(Use Enter to apply scope, Tab to go back)": "(Use Enter para aplicar o escopo, Tab para voltar)",
  "Theme configuration unavailable due to NO_COLOR env variable.": "Configura\xE7\xE3o de tema indispon\xEDvel devido \xE0 vari\xE1vel de ambiente NO_COLOR.",
  'Theme "{{themeName}}" not found.': 'Tema "{{themeName}}" n\xE3o encontrado.',
  'Theme "{{themeName}}" not found in selected scope.': 'Tema "{{themeName}}" n\xE3o encontrado no escopo selecionado.',
  "Clear conversation history and free up context": "Limpar hist\xF3rico de conversa e liberar contexto",
  "Compresses the context by replacing it with a summary.": "Comprime o contexto substituindo-o por um resumo.",
  "open full Qwen Code documentation in your browser": "abrir documenta\xE7\xE3o completa do Qwen Code no seu navegador",
  "Configuration not available.": "Configura\xE7\xE3o n\xE3o dispon\xEDvel.",
  "Connect an LLM provider": "Conectar a um provedor LLM",
  "Copy the last AI response to clipboard (/copy N for Nth-latest)": "Copiar a \xFAltima resposta da IA para a \xE1rea de transfer\xEAncia (/copy N para a N-\xE9sima)",
  // ============================================================================
  // Commands - Agents
  // ============================================================================
  "Manage subagents for specialized task delegation.": "Gerenciar subagentes para delega\xE7\xE3o de tarefas especializadas.",
  "Manage existing subagents (view, edit, delete).": "Gerenciar subagentes existentes (ver, editar, excluir).",
  "Create a new subagent with guided setup.": "Criar um novo subagente com configura\xE7\xE3o guiada.",
  // ============================================================================
  // Agents - Management Dialog
  // ============================================================================
  Agents: "Agentes",
  "Choose Action": "Escolher A\xE7\xE3o",
  "Edit {{name}}": "Editar {{name}}",
  "Edit Tools: {{name}}": "Editar Ferramentas: {{name}}",
  "Edit Color: {{name}}": "Editar Cor: {{name}}",
  "Delete {{name}}": "Excluir {{name}}",
  "Unknown Step": "Etapa Desconhecida",
  "Esc to close": "Esc para fechar",
  "Enter to select, \u2191\u2193 to navigate, Esc to close": "Enter para selecionar, \u2191\u2193 para navegar, Esc para fechar",
  "Esc to go back": "Esc para voltar",
  "Enter to confirm, Esc to cancel": "Enter para confirmar, Esc para cancelar",
  "Enter to select, \u2191\u2193 to navigate, Esc to go back": "Enter para selecionar, \u2191\u2193 para navegar, Esc para voltar",
  "Enter to submit, Esc to go back": "Enter para enviar, Esc para voltar",
  "Invalid step: {{step}}": "Etapa inv\xE1lida: {{step}}",
  "No subagents found.": "Nenhum subagente encontrado.",
  "Use '/agents create' to create your first subagent.": "Use '/agents create' para criar seu primeiro subagente.",
  "(built-in)": "(integrado)",
  "(overridden by project level agent)": "(substitu\xEDdo por agente de n\xEDvel de projeto)",
  "Project Level ({{path}})": "N\xEDvel de Projeto ({{path}})",
  "User Level ({{path}})": "N\xEDvel de Usu\xE1rio ({{path}})",
  "Built-in Agents": "Agentes Integrados",
  "Extension Agents": "Agentes de Extens\xE3o",
  "Using: {{count}} agents": "Usando: {{count}} agentes",
  "View Agent": "Ver Agente",
  "Edit Agent": "Editar Agente",
  "Delete Agent": "Excluir Agente",
  Back: "Voltar",
  "No agent selected": "Nenhum agente selecionado",
  "File Path: ": "Caminho do Arquivo: ",
  "Tools: ": "Ferramentas: ",
  "Color: ": "Cor: ",
  "Description:": "Descri\xE7\xE3o:",
  "System Prompt:": "Prompt do Sistema:",
  "Open in editor": "Abrir no editor",
  "Edit tools": "Editar ferramentas",
  "Edit color": "Editar cor",
  "\u274C Error:": "\u274C Erro:",
  'Are you sure you want to delete agent "{{name}}"?': 'Tem certeza que deseja excluir o agente "{{name}}"?',
  // ============================================================================
  // Agents - Creation Wizard
  // ============================================================================
  "Project Level (.qwen/agents/)": "N\xEDvel de Projeto (.qwen/agents/)",
  "User Level (~/.qwen/agents/)": "N\xEDvel de Usu\xE1rio (~/.qwen/agents/)",
  "\u2705 Subagent Created Successfully!": "\u2705 Subagente criado com sucesso!",
  'Subagent "{{name}}" has been saved to {{level}} level.': 'O subagente "{{name}}" foi salvo no n\xEDvel {{level}}.',
  "Name: ": "Nome: ",
  "Location: ": "Localiza\xE7\xE3o: ",
  "\u274C Error saving subagent:": "\u274C Erro ao salvar subagente:",
  "Warnings:": "Avisos:",
  'Name "{{name}}" already exists at {{level}} level - will overwrite existing subagent': 'O nome "{{name}}" j\xE1 existe no n\xEDvel {{level}} - o subagente existente ser\xE1 substitu\xEDdo',
  'Name "{{name}}" exists at user level - project level will take precedence': 'O nome "{{name}}" existe no n\xEDvel de usu\xE1rio - o n\xEDvel de projeto ter\xE1 preced\xEAncia',
  'Name "{{name}}" exists at project level - existing subagent will take precedence': 'O nome "{{name}}" existe no n\xEDvel de projeto - o subagente existente ter\xE1 preced\xEAncia',
  "Description is over {{length}} characters": "A descri\xE7\xE3o tem mais de {{length}} caracteres",
  "System prompt is over {{length}} characters": "O prompt do sistema tem mais de {{length}} caracteres",
  // ============================================================================
  // Agents - Creation Wizard Steps
  // ============================================================================
  "Step {{n}}: Choose Location": "Etapa {{n}}: Escolher Localiza\xE7\xE3o",
  "Step {{n}}: Choose Generation Method": "Etapa {{n}}: Escolher M\xE9todo de Gera\xE7\xE3o",
  "Generate with Qwen Code (Recommended)": "Gerar com Qwen Code (Recomendado)",
  "Manual Creation": "Cria\xE7\xE3o Manual",
  "Describe what this subagent should do and when it should be used. (Be comprehensive for best results)": "Descreva o que este subagente deve fazer e quando deve ser usado. (Seja abrangente para melhores resultados)",
  "e.g., Expert code reviewer that reviews code based on best practices...": "ex: Revisor de c\xF3digo especialista que revisa c\xF3digo com base em melhores pr\xE1ticas...",
  "Generating subagent configuration...": "Gerando configura\xE7\xE3o do subagente...",
  "Failed to generate subagent: {{error}}": "Falha ao gerar subagente: {{error}}",
  "Step {{n}}: Describe Your Subagent": "Etapa {{n}}: Descreva Seu Subagente",
  "Step {{n}}: Enter Subagent Name": "Etapa {{n}}: Digite o Nome do Subagente",
  "Step {{n}}: Enter System Prompt": "Etapa {{n}}: Digite o Prompt do Sistema",
  "Step {{n}}: Enter Description": "Etapa {{n}}: Digite a Descri\xE7\xE3o",
  // ============================================================================
  // Agents - Tool Selection
  // ============================================================================
  "Step {{n}}: Select Tools": "Etapa {{n}}: Selecionar Ferramentas",
  "All Tools (Default)": "Todas as Ferramentas (Padr\xE3o)",
  "All Tools": "Todas as Ferramentas",
  "Read-only Tools": "Ferramentas de Somente Leitura",
  "Read & Edit Tools": "Ferramentas de Leitura e Edi\xE7\xE3o",
  "Read & Edit & Execution Tools": "Ferramentas de Leitura, Edi\xE7\xE3o e Execu\xE7\xE3o",
  "All tools selected, including MCP tools": "Todas as ferramentas selecionadas, incluindo MCP tools",
  "Selected tools:": "Ferramentas selecionadas:",
  "Read-only tools:": "Ferramentas de somente leitura:",
  "Edit tools:": "Ferramentas de edi\xE7\xE3o:",
  "Execution tools:": "Ferramentas de execu\xE7\xE3o:",
  "Step {{n}}: Choose Background Color": "Etapa {{n}}: Escolher Cor de Fundo",
  "Step {{n}}: Confirm and Save": "Etapa {{n}}: Confirmar e Salvar",
  // ============================================================================
  // Agents - Navigation & Instructions
  // ============================================================================
  "Esc to cancel": "Esc para cancelar",
  "Press Enter to save, e to save and edit, Esc to go back": "Pressione Enter para salvar, e para salvar e editar, Esc para voltar",
  "Press Enter to continue, {{navigation}}Esc to {{action}}": "Pressione Enter para continuar, {{navigation}}Esc para {{action}}",
  cancel: "cancelar",
  "go back": "voltar",
  "\u2191\u2193 to navigate, ": "\u2191\u2193 para navegar, ",
  "Enter a clear, unique name for this subagent.": "Digite um nome claro e \xFAnico para este subagente.",
  "e.g., Code Reviewer": "ex: Revisor de C\xF3digo",
  "Name cannot be empty.": "O nome n\xE3o pode estar vazio.",
  "Write the system prompt that defines this subagent's behavior. Be comprehensive for best results.": "Escreva o prompt do sistema que define o comportamento deste subagente. Seja abrangente para melhores resultados.",
  "e.g., You are an expert code reviewer...": "ex: Voc\xEA \xE9 um revisor de c\xF3digo especialista...",
  "System prompt cannot be empty.": "O prompt do sistema n\xE3o pode estar vazio.",
  "Describe when and how this subagent should be used.": "Descreva quando e como este subagente deve ser usado.",
  "e.g., Reviews code for best practices and potential bugs.": "ex: Revisa o c\xF3digo em busca de melhores pr\xE1ticas e erros potenciais.",
  "Description cannot be empty.": "A descri\xE7\xE3o n\xE3o pode estar vazia.",
  "Failed to launch editor: {{error}}": "Falha ao iniciar editor: {{error}}",
  "Failed to save and edit subagent: {{error}}": "Falha ao salvar e editar subagente: {{error}}",
  // ============================================================================
  // Commands - General (continued)
  // ============================================================================
  "View and edit Qwen Code settings": "Ver e editar configura\xE7\xF5es do Qwen Code",
  Settings: "Configura\xE7\xF5es",
  "To see changes, Qwen Code must be restarted. Press r to exit and apply changes now.": "Para ver as altera\xE7\xF5es, o Qwen Code deve ser reiniciado. Pressione r para sair e aplicar as altera\xE7\xF5es agora.",
  // ============================================================================
  // Settings Labels
  // ============================================================================
  "Vim Mode": "Modo Vim",
  "Attribution: commit": "Atribui\xE7\xE3o: commit",
  "Terminal Bell Notification": "Notifica\xE7\xE3o Sonora do Terminal",
  "Enable Usage Statistics": "Ativar Estat\xEDsticas de Uso",
  Theme: "Tema",
  "Preferred Editor": "Editor Preferido",
  "Auto-connect to IDE": "Conex\xE3o Autom\xE1tica com IDE",
  "Debug Keystroke Logging": "Log de Depura\xE7\xE3o de Teclas",
  "Language: UI": "Idioma: Interface",
  "Language: Model": "Idioma: Modelo",
  "Output Format": "Formato de Sa\xEDda",
  "Hide Window Title": "Ocultar T\xEDtulo da Janela",
  "Show Status in Title": "Mostrar Status no T\xEDtulo",
  "Hide Tips": "Ocultar Dicas",
  "Show Line Numbers in Code": "Mostrar N\xFAmeros de Linhas no C\xF3digo",
  "Show Citations": "Mostrar Cita\xE7\xF5es",
  "Custom Witty Phrases": "Frases de Efeito Personalizadas",
  "Show Welcome Back Dialog": "Mostrar Di\xE1logo de Bem-vindo de Volta",
  "Enable User Feedback": "Ativar Feedback do Usu\xE1rio",
  "How is Qwen doing this session? (optional)": "Como o Qwen est\xE1 se saindo nesta sess\xE3o? (opcional)",
  Bad: "Ruim",
  Fine: "Bom",
  Good: "\xD3timo",
  Dismiss: "Ignorar",
  "Screen Reader Mode": "Modo de Leitor de Tela",
  "Max Session Turns": "M\xE1ximo de Turnos da Sess\xE3o",
  "Skip Next Speaker Check": "Pular Verifica\xE7\xE3o do Pr\xF3ximo Falante",
  "Skip Loop Detection": "Pular Detec\xE7\xE3o de Loop",
  "Skip Startup Context": "Pular Contexto de Inicializa\xE7\xE3o",
  "Enable OpenAI Logging": "Ativar Log do OpenAI",
  "OpenAI Logging Directory": "Diret\xF3rio de Log do OpenAI",
  Timeout: "Tempo Limite",
  "Max Retries": "M\xE1ximo de Tentativas",
  "Load Memory From Include Directories": "Carregar Mem\xF3ria de Diret\xF3rios Inclu\xEDdos",
  "Respect .gitignore": "Respeitar .gitignore",
  "Respect .qwenignore": "Respeitar .qwenignore",
  "Enable Recursive File Search": "Ativar Pesquisa Recursiva de Arquivos",
  "Interactive Shell (PTY)": "Shell Interativo (PTY)",
  "Show Color": "Mostrar Cores",
  "Auto Accept": "Aceitar Automaticamente",
  "Use Ripgrep": "Usar Ripgrep",
  "Use Builtin Ripgrep": "Usar Ripgrep Integrado",
  "Tool Output Truncation Threshold": "Limite de Truncamento de Sa\xEDda de Ferramenta",
  "Tool Output Truncation Lines": "Linhas de Truncamento de Sa\xEDda de Ferramenta",
  "Folder Trust": "Confian\xE7a de Pasta",
  "Tool Schema Compliance": "Conformidade de Tool Schema",
  // Settings enum options
  "Auto (detect from system)": "Autom\xE1tico (detectar do sistema)",
  "Auto (detect terminal theme)": "Autom\xE1tico (detectar tema do terminal)",
  Auto: "Autom\xE1tico",
  Text: "Texto",
  JSON: "JSON",
  Plan: "Planejamento",
  "Ask permissions": "Pedir permiss\xE3o",
  "Auto Edit": "Edi\xE7\xE3o Autom\xE1tica",
  YOLO: "YOLO",
  "toggle vim mode on/off": "alternar modo vim ligado/desligado",
  "Show model-specific usage statistics.": "Mostrar estat\xEDsticas de uso espec\xEDficas do modelo.",
  "Show tool-specific usage statistics.": "Mostrar estat\xEDsticas de uso espec\xEDficas da ferramenta.",
  "Show daily token usage statistics.": "Mostrar estat\xEDsticas di\xE1rias de uso de tokens.",
  "Show monthly token usage statistics.": "Mostrar estat\xEDsticas mensais de uso de tokens.",
  "Export token usage statistics to CSV or JSON.": "Exportar estat\xEDsticas de uso de tokens para CSV ou JSON.",
  "No usage data.": "Nenhum dado de uso.",
  "{{label}}: {{tokens}} tokens ({{requests}} requests)": "{{label}}: {{tokens}} tokens ({{requests}} requisi\xE7\xF5es)",
  "Daily token usage for {{value}}": "Uso di\xE1rio de tokens para {{value}}",
  "Monthly token usage for {{value}}": "Uso mensal de tokens para {{value}}",
  "Total: {{tokens}} tokens": "Total: {{tokens}} tokens",
  "Requests: {{requests}}": "Requisi\xE7\xF5es: {{requests}}",
  "Breakdown:": "Detalhamento:",
  "Input: {{tokens}}": "Entrada: {{tokens}}",
  "Output: {{tokens}}": "Sa\xEDda: {{tokens}}",
  "Cached (included in Input): {{tokens}}": "Cache (inclu\xEDdo na entrada): {{tokens}}",
  "Thoughts: {{tokens}}": "Racioc\xEDnio: {{tokens}}",
  "By model:": "Por modelo:",
  "By auth type:": "Por tipo de autentica\xE7\xE3o:",
  "By model/auth type:": "Por modelo/tipo de autentica\xE7\xE3o:",
  "By source:": "Por origem:",
  "Failed to load token usage stats: {{error}}": "Falha ao carregar estat\xEDsticas de uso de tokens: {{error}}",
  "Expected --format csv or --format json.": "Esperado --format csv ou --format json.",
  "Expected a file path after --output.": "Esperado um caminho de arquivo ap\xF3s --output.",
  "Unexpected argument: {{argument}}": "Argumento inesperado: {{argument}}",
  "Usage: /stats export <daily|monthly> [YYYY-MM-DD|YYYY-MM] [--format csv|json] [--output path]": "Uso: /stats export <daily|monthly> [YYYY-MM-DD|YYYY-MM] [--format csv|json] [--output path]",
  "Token usage export path must be within the project working directory.": "O caminho de exporta\xE7\xE3o do uso de tokens deve estar dentro do diret\xF3rio de trabalho do projeto.",
  "Export target does not exist: {{path}}": "O destino da exporta\xE7\xE3o n\xE3o existe: {{path}}",
  "Cannot resolve export path within the working directory.": "N\xE3o foi poss\xEDvel resolver o caminho de exporta\xE7\xE3o dentro do diret\xF3rio de trabalho.",
  "Could not create a temporary export file.": "N\xE3o foi poss\xEDvel criar um arquivo tempor\xE1rio de exporta\xE7\xE3o.",
  "Token usage exported to {{format}}: {{path}}": "Uso de tokens exportado para {{format}}: {{path}}",
  "Failed to export token usage stats: {{error}}": "Falha ao exportar estat\xEDsticas de uso de tokens: {{error}}",
  "Unclosed quote in arguments.": "Aspas n\xE3o fechadas nos argumentos.",
  "Note: generation timing (TTFT/TPS) belongs to generation metrics.": "Observa\xE7\xE3o: o tempo de gera\xE7\xE3o (TTFT/TPS) pertence \xE0s m\xE9tricas de gera\xE7\xE3o.",
  "exit the cli": "sair da cli",
  "Manage workspace directories": "Gerenciar diret\xF3rios do workspace",
  "Add directories to the workspace. Use comma to separate multiple paths": "Adicionar diret\xF3rios ao workspace. Use v\xEDrgula para separar v\xE1rios caminhos",
  "Show all directories in the workspace": "Mostrar todos os diret\xF3rios no workspace",
  "set external editor preference": "definir prefer\xEAncia de editor externo",
  "Select Editor": "Selecionar Editor",
  "Editor Preference": "Prefer\xEAncia de Editor",
  "These editors are currently supported. Please note that some editors cannot be used in sandbox mode.": "Estes editores s\xE3o suportados atualmente. Note que alguns editores n\xE3o podem ser usados no modo sandbox.",
  "Your preferred editor is:": "Seu editor preferido \xE9:",
  "Manage extensions": "Gerenciar extens\xF5es",
  "Manage installed extensions": "Gerenciar extens\xF5es instaladas",
  "Disable an extension": "Desativar uma extens\xE3o",
  "Enable an extension": "Ativar uma extens\xE3o",
  "Install an extension from a git repo or local path": "Instalar uma extens\xE3o de um reposit\xF3rio git ou caminho local",
  "Uninstall an extension": "Desinstalar uma extens\xE3o",
  "No extensions installed.": "Nenhuma extens\xE3o instalada.",
  'Extension "{{name}}" not found.': 'Extens\xE3o "{{name}}" n\xE3o encontrada.',
  "No extensions to update.": "Nenhuma extens\xE3o para atualizar.",
  "Usage: /extensions install <source>": "Uso: /extensions install <fonte>",
  'Installing extension from "{{source}}"...': 'Instalando extens\xE3o de "{{source}}"...',
  'Extension "{{name}}" installed successfully.': 'Extens\xE3o "{{name}}" instalada com sucesso.',
  'Failed to install extension from "{{source}}": {{error}}': 'Falha ao instalar extens\xE3o de "{{source}}": {{error}}',
  "Do you want to continue? [Y/n]: ": "Voc\xEA deseja continuar? [Y/n]: ",
  "Do you want to continue?": "Voc\xEA deseja continuar?",
  'Installing extension "{{name}}".': 'Instalando extens\xE3o "{{name}}".',
  "**Extensions may introduce unexpected behavior. Ensure you have investigated the extension source and trust the author.**": "**As extens\xF5es podem introduzir comportamentos inesperados. Certifique-se de ter investigado a fonte da extens\xE3o e confie no autor.**",
  "This extension will run the following MCP servers:": "Esta extens\xE3o executar\xE1 os seguintes MCP servers:",
  local: "local",
  remote: "remoto",
  "This extension will add the following commands: {{commands}}.": "Esta extens\xE3o adicionar\xE1 os seguintes comandos: {{commands}}.",
  "This extension will append info to your QWEN.md context using {{fileName}}": "Esta extens\xE3o anexar\xE1 informa\xE7\xF5es ao seu contexto QWEN.md usando {{fileName}}",
  "This extension will install the following skills:": "Esta extens\xE3o instalar\xE1 as seguintes habilidades:",
  "This extension will install the following subagents:": "Esta extens\xE3o instalar\xE1 os seguintes subagentes:",
  'Installation cancelled for "{{name}}".': 'Instala\xE7\xE3o cancelada para "{{name}}".',
  "You are installing an extension from {{originSource}}. Some features may not work perfectly with Qwen Code.": "Voc\xEA est\xE1 instalando uma extens\xE3o de {{originSource}}. Alguns recursos podem n\xE3o funcionar perfeitamente com o Qwen Code.",
  "--ref and --auto-update are not applicable for marketplace extensions.": "--ref e --auto-update n\xE3o s\xE3o aplic\xE1veis para extens\xF5es de marketplace.",
  'Extension "{{name}}" installed successfully and enabled.': 'Extens\xE3o "{{name}}" instalada com sucesso e ativada.',
  "The github URL, local path, or marketplace source (marketplace-url:plugin-name) of the extension to install.": "A URL do github, caminho local ou fonte do marketplace (marketplace-url:plugin-name) da extens\xE3o para instalar.",
  "The git ref to install from.": "A refer\xEAncia git para instalar.",
  "Enable auto-update for this extension.": "Ativar atualiza\xE7\xE3o autom\xE1tica para esta extens\xE3o.",
  "Enable pre-release versions for this extension.": "Ativar vers\xF5es de pr\xE9-lan\xE7amento para esta extens\xE3o.",
  "Acknowledge the security risks of installing an extension and skip the confirmation prompt.": "Reconhecer os riscos de seguran\xE7a de instalar uma extens\xE3o e pular o prompt de confirma\xE7\xE3o.",
  "The source argument must be provided.": "O argumento fonte deve ser fornecido.",
  'Extension "{{name}}" successfully uninstalled.': 'Extens\xE3o "{{name}}" desinstalada com sucesso.',
  "Uninstalls an extension.": "Desinstala uma extens\xE3o.",
  "The name or source path of the extension to uninstall.": "O nome ou caminho da fonte da extens\xE3o para desinstalar.",
  "Please include the name of the extension to uninstall as a positional argument.": "Inclua o nome da extens\xE3o para desinstalar como um argumento posicional.",
  "Enables an extension.": "Ativa uma extens\xE3o.",
  "The name of the extension to enable.": "O nome da extens\xE3o para ativar.",
  "The scope to enable the extenison in. If not set, will be enabled in all scopes.": "O escopo para ativar a extens\xE3o. Se n\xE3o definido, ser\xE1 ativada em todos os escopos.",
  'Extension "{{name}}" successfully enabled for scope "{{scope}}".': 'Extens\xE3o "{{name}}" ativada com sucesso para o escopo "{{scope}}".',
  'Extension "{{name}}" successfully enabled in all scopes.': 'Extens\xE3o "{{name}}" ativada com sucesso em todos os escopos.',
  "Invalid scope: {{scope}}. Please use one of {{scopes}}.": "Escopo inv\xE1lido: {{scope}}. Use um de {{scopes}}.",
  "Disables an extension.": "Desativa uma extens\xE3o.",
  "The name of the extension to disable.": "O nome da extens\xE3o para desativar.",
  "The scope to disable the extenison in.": "O escopo para desativar a extens\xE3o.",
  'Extension "{{name}}" successfully disabled for scope "{{scope}}".': 'Extens\xE3o "{{name}}" desativada com sucesso para o escopo "{{scope}}".',
  'Extension "{{name}}" successfully updated: {{oldVersion}} \u2192 {{newVersion}}.': 'Extens\xE3o "{{name}}" atualizada com sucesso: {{oldVersion}} \u2192 {{newVersion}}.',
  'Unable to install extension "{{name}}" due to missing install metadata': 'N\xE3o foi poss\xEDvel instalar a extens\xE3o "{{name}}" devido \xE0 falta de metadados de instala\xE7\xE3o',
  'Extension "{{name}}" is already up to date.': 'A extens\xE3o "{{name}}" j\xE1 est\xE1 atualizada.',
  "Updates all extensions or a named extension to the latest version.": "Atualiza todas as extens\xF5es ou uma extens\xE3o nomeada para a \xFAltima vers\xE3o.",
  "Update all extensions.": "Atualizar todas as extens\xF5es.",
  "Either an extension name or --all must be provided": "Um nome de extens\xE3o ou --all deve ser fornecido",
  "Lists installed extensions.": "Lista as extens\xF5es instaladas.",
  "Link extension failed to install.": "Falha ao instalar link da extens\xE3o.",
  'Extension "{{name}}" linked successfully and enabled.': 'Extens\xE3o "{{name}}" vinculada com sucesso e ativada.',
  "Links an extension from a local path. Updates made to the local path will always be reflected.": "Vincula uma extens\xE3o de um caminho local. Atualiza\xE7\xF5es feitas no caminho local sempre ser\xE3o refletidas.",
  "The name of the extension to link.": "O nome da extens\xE3o para vincular.",
  "Set a specific setting for an extension.": "Define uma configura\xE7\xE3o espec\xEDfica para uma extens\xE3o.",
  "Name of the extension to configure.": "Nome da extens\xE3o para configurar.",
  "The setting to configure (name or env var).": "A configura\xE7\xE3o para configurar (nome ou var env).",
  "The scope to set the setting in.": "O escopo para definir a configura\xE7\xE3o.",
  "List all settings for an extension.": "Listar todas as configura\xE7\xF5es de uma extens\xE3o.",
  "Name of the extension.": "Nome da extens\xE3o.",
  'Extension "{{name}}" has no settings to configure.': 'A extens\xE3o "{{name}}" n\xE3o tem configura\xE7\xF5es para configurar.',
  'Settings for "{{name}}":': 'Configura\xE7\xF5es para "{{name}}":',
  "(user)": "(usu\xE1rio)",
  "[not set]": "[n\xE3o definido]",
  "[value stored in keychain]": "[valor armazenado no chaveiro]",
  "Value:": "Valor:",
  "Manage extension settings.": "Gerenciar configura\xE7\xF5es de extens\xE3o.",
  "You need to specify a command (set or list).": "Voc\xEA precisa especificar um comando (set ou list).",
  // ============================================================================
  // Plugin Choice / Marketplace
  // ============================================================================
  "No plugins available in this marketplace.": "Nenhum plugin dispon\xEDvel neste marketplace.",
  'Select a plugin to install from marketplace "{{name}}":': 'Selecione um plugin para instalar do marketplace "{{name}}":',
  "Plugin selection cancelled.": "Sele\xE7\xE3o de plugin cancelada.",
  'Select a plugin from "{{name}}"': 'Selecione um plugin de "{{name}}"',
  "Use \u2191\u2193 or j/k to navigate, Enter to select, Escape to cancel": "Use \u2191\u2193 ou j/k para navegar, Enter para selecionar, Escape para cancelar",
  "{{count}} more above": "{{count}} mais acima",
  "{{count}} more below": "{{count}} mais abaixo",
  "manage IDE integration": "gerenciar integra\xE7\xE3o com IDE",
  "check status of IDE integration": "verificar status da integra\xE7\xE3o com IDE",
  "install required IDE companion for {{ideName}}": "instalar companion IDE necess\xE1rio para {{ideName}}",
  "enable IDE integration": "ativar integra\xE7\xE3o com IDE",
  "disable IDE integration": "desativar integra\xE7\xE3o com IDE",
  "IDE integration is not supported in your current environment. To use this feature, run Qwen Code in one of these supported IDEs: VS Code or VS Code forks.": "A integra\xE7\xE3o com IDE n\xE3o \xE9 suportada no seu ambiente atual. Para usar este recurso, execute o Qwen Code em um destes IDEs suportados: VS Code ou forks do VS Code.",
  "Set up GitHub Actions": "Configurar GitHub Actions",
  "Configure terminal keybindings for multiline input (VS Code, Cursor, Windsurf, Trae)": "Configurar atalhos de terminal para entrada multilinhas (VS Code, Cursor, Windsurf, Trae)",
  "Please restart your terminal for the changes to take effect.": "Reinicie seu terminal para que as altera\xE7\xF5es tenham efeito.",
  "Failed to configure terminal: {{error}}": "Falha ao configurar terminal: {{error}}",
  "Could not determine {{terminalName}} config path on Windows: APPDATA environment variable is not set.": "N\xE3o foi poss\xEDvel determinar o caminho de configura\xE7\xE3o de {{terminalName}} no Windows: vari\xE1vel de ambiente APPDATA n\xE3o est\xE1 definida.",
  "{{terminalName}} keybindings.json exists but is not a valid JSON array. Please fix the file manually or delete it to allow automatic configuration.": "{{terminalName}} keybindings.json existe mas n\xE3o \xE9 um array JSON v\xE1lido. Corrija o arquivo manualmente ou exclua-o para permitir a configura\xE7\xE3o autom\xE1tica.",
  "File: {{file}}": "Arquivo: {{file}}",
  "Failed to parse {{terminalName}} keybindings.json. The file contains invalid JSON. Please fix the file manually or delete it to allow automatic configuration.": "Falha ao analisar {{terminalName}} keybindings.json. O arquivo cont\xE9m JSON inv\xE1lido. Corrija o arquivo manualmente ou exclua-o para permitir a configura\xE7\xE3o autom\xE1tica.",
  "Error: {{error}}": "Erro: {{error}}",
  "Shift+Enter binding already exists": "Atalho Shift+Enter j\xE1 existe",
  "Ctrl+Enter binding already exists": "Atalho Ctrl+Enter j\xE1 existe",
  "Existing keybindings detected. Will not modify to avoid conflicts.": "Atalhos existentes detectados. N\xE3o ser\xE3o modificados para evitar conflitos.",
  "Please check and modify manually if needed: {{file}}": "Verifique e modifique manualmente se necess\xE1rio: {{file}}",
  "Added Shift+Enter and Ctrl+Enter keybindings to {{terminalName}}.": "Adicionados atalhos Shift+Enter e Ctrl+Enter para {{terminalName}}.",
  "Modified: {{file}}": "Modificado: {{file}}",
  "{{terminalName}} keybindings already configured.": "Atalhos de {{terminalName}} j\xE1 configurados.",
  "Failed to configure {{terminalName}}.": "Falha ao configurar {{terminalName}}.",
  "Your terminal is already configured for an optimal experience with multiline input (Shift+Enter and Ctrl+Enter).": "Seu terminal j\xE1 est\xE1 configurado para uma experi\xEAncia ideal com entrada multilinhas (Shift+Enter e Ctrl+Enter).",
  // ============================================================================
  // Commands - Hooks
  // ============================================================================
  "Manage Qwen Code hooks": "Gerenciar hooks do Qwen Code",
  "List all configured hooks": "Listar todos os hooks configurados",
  // Hooks - Dialog
  Hooks: "Hooks",
  "Loading hooks...": "Carregando hooks...",
  "Error loading hooks:": "Erro ao carregar hooks:",
  "Press Escape to close": "Pressione Escape para fechar",
  "Press Escape, Ctrl+C, or Ctrl+D to cancel": "Pressione Escape, Ctrl+C ou Ctrl+D para cancelar",
  "Press Space, Enter, or Escape to dismiss": "Pressione Space, Enter ou Escape para dispensar",
  "No hook selected": "Nenhum hook selecionado",
  // Hooks - List Step
  "No hook events found.": "Nenhum evento de hook encontrado.",
  "{{count}} hook configured": "{{count}} hook configurado",
  "{{count}} hooks configured": "{{count}} hooks configurados",
  "This menu is read-only. To add or modify hooks, edit settings.json directly or ask Qwen Code.": "Este menu \xE9 somente leitura. Para adicionar ou modificar hooks, edite settings.json diretamente ou pergunte ao Qwen Code.",
  "Enter to select \xB7 Esc to cancel": "Enter para selecionar \xB7 Esc para cancelar",
  // Hooks - Detail Step
  "Exit codes:": "C\xF3digos de sa\xEDda:",
  "Configured hooks:": "Hooks configurados:",
  "No hooks configured for this event.": "Nenhum hook configurado para este evento.",
  "To add hooks, edit settings.json directly or ask Qwen.": "Para adicionar hooks, edite settings.json diretamente ou pergunte ao Qwen.",
  "Enter to select \xB7 Esc to go back": "Enter para selecionar \xB7 Esc para voltar",
  // Hooks - Config Detail Step
  "Hook details": "Detalhes do Hook",
  "Event:": "Evento:",
  "Extension:": "Extens\xE3o:",
  "Desc:": "Descri\xE7\xE3o:",
  "No hook config selected": "Nenhuma configura\xE7\xE3o de hook selecionada",
  "To modify or remove this hook, edit settings.json directly or ask Qwen to help.": "Para modificar ou remover este hook, edite settings.json diretamente ou pergunte ao Qwen.",
  // Hooks - Disabled Step
  "Hook Configuration - Disabled": "Configura\xE7\xE3o de Hook - Desativado",
  "All hooks are currently disabled. You have {{count}} that are not running.": "Todos os hooks est\xE3o desativados. Voc\xEA tem {{count}} que n\xE3o est\xE3o em execu\xE7\xE3o.",
  "{{count}} configured hook": "{{count}} hook configurado",
  "{{count}} configured hooks": "{{count}} hooks configurados",
  "When hooks are disabled:": "Quando os hooks est\xE3o desativados:",
  "No hook commands will execute": "Nenhum comando de hook ser\xE1 executado",
  "StatusLine will not be displayed": "StatusLine n\xE3o ser\xE1 exibido",
  "Tool operations will proceed without hook validation": "As opera\xE7\xF5es de ferramentas prosseguir\xE3o sem valida\xE7\xE3o de hook",
  'To re-enable hooks, remove "disableAllHooks" from settings.json or ask Qwen Code.': 'Para reativar os hooks, remova "disableAllHooks" do settings.json ou pergunte ao Qwen Code.',
  // Hooks - Source
  Project: "Projeto",
  User: "Usu\xE1rio",
  Skill: "Habilidade",
  System: "Sistema",
  Extension: "Extens\xE3o",
  "Local Settings": "Configura\xE7\xF5es Locais",
  "User Settings": "Configura\xE7\xF5es do Usu\xE1rio",
  "System Settings": "Configura\xE7\xF5es do Sistema",
  Extensions: "Extens\xF5es",
  "Session (temporary)": "Sess\xE3o (tempor\xE1rio)",
  // Hooks - Event Descriptions (short)
  "Before tool execution": "Antes da execu\xE7\xE3o da ferramenta",
  "After tool execution": "Ap\xF3s a execu\xE7\xE3o da ferramenta",
  "After tool execution fails": "Ap\xF3s a falha da execu\xE7\xE3o da ferramenta",
  "When notifications are sent": "Quando notifica\xE7\xF5es s\xE3o enviadas",
  "When the user submits a prompt": "Quando o usu\xE1rio envia um prompt",
  "When a slash command expands into a prompt": "Quando um comando slash se expande em um prompt",
  "When a new session is started": "Quando uma nova sess\xE3o \xE9 iniciada",
  "Right before Qwen Code concludes its response": "Logo antes do Qwen Code concluir sua resposta",
  "When a subagent (Agent tool call) is started": "Quando um subagente (chamada de ferramenta Agent) \xE9 iniciado",
  "Right before a subagent concludes its response": "Logo antes de um subagente concluir sua resposta",
  "Before conversation compaction": "Antes da compacta\xE7\xE3o da conversa",
  "When a session is ending": "Quando uma sess\xE3o est\xE1 terminando",
  "When a permission dialog is displayed": "Quando um di\xE1logo de permiss\xE3o \xE9 exibido",
  "When a new todo item is created": "Quando um novo item todo \xE9 criado",
  "When a todo item is marked as completed": "Quando um item todo \xE9 marcado como conclu\xEDdo",
  // Hooks - Event Descriptions (detailed)
  "Input to command is JSON of tool call arguments.": "A entrada para o comando \xE9 JSON dos argumentos da chamada da ferramenta.",
  'Input to command is JSON with fields "inputs" (tool call arguments) and "response" (tool call response).': 'A entrada para o comando \xE9 JSON com campos "inputs" (argumentos da chamada da ferramenta) e "response" (resposta da chamada da ferramenta).',
  "Input to command is JSON with tool_name, tool_input, tool_use_id, error, error_type, is_interrupt, and is_timeout.": "A entrada para o comando \xE9 JSON com tool_name, tool_input, tool_use_id, error, error_type, is_interrupt e is_timeout.",
  "Input to command is JSON with notification message and type.": "A entrada para o comando \xE9 JSON com mensagem e tipo de notifica\xE7\xE3o.",
  "Input to command is JSON with original user prompt text.": "A entrada para o comando \xE9 JSON com o texto original do prompt do usu\xE1rio.",
  "Input to command is JSON with command_name, command_args, and expanded prompt text.": "A entrada para o comando \xE9 JSON com command_name, command_args e o texto do prompt expandido.",
  "Input to command is JSON with session start source.": "A entrada para o comando \xE9 JSON com a fonte de in\xEDcio da sess\xE3o.",
  "Input to command is JSON with session end reason.": "A entrada para o comando \xE9 JSON com o motivo do fim da sess\xE3o.",
  "Input to command is JSON with agent_id and agent_type.": "A entrada para o comando \xE9 JSON com agent_id e agent_type.",
  "Input to command is JSON with agent_id, agent_type, and agent_transcript_path.": "A entrada para o comando \xE9 JSON com agent_id, agent_type e agent_transcript_path.",
  "Input to command is JSON with compaction details.": "A entrada para o comando \xE9 JSON com detalhes da compacta\xE7\xE3o.",
  "Input to command is JSON with tool_name, tool_input, and tool_use_id. Output JSON with hookSpecificOutput containing decision to allow or deny.": "A entrada para o comando \xE9 JSON com tool_name, tool_input e tool_use_id. Sa\xEDda \xE9 JSON com hookSpecificOutput contendo decis\xE3o de permitir ou negar.",
  "Input to command is JSON with todo_id, todo_content, todo_status, all_todos, and phase. In validation, output JSON with decision (allow/block/deny) and reason. In postWrite, block/deny is ignored.": "A entrada para o comando \xE9 JSON com todo_id, todo_content, todo_status, all_todos e phase. Em validation, sa\xEDda \xE9 JSON com decision (allow/block/deny) e reason. Em postWrite, block/deny \xE9 ignorado.",
  "Input to command is JSON with todo_id, todo_content, previous_status, all_todos, and phase. In validation, output JSON with decision (allow/block/deny) and reason. In postWrite, block/deny is ignored.": "A entrada para o comando \xE9 JSON com todo_id, todo_content, previous_status, all_todos e phase. Em validation, sa\xEDda \xE9 JSON com decision (allow/block/deny) e reason. Em postWrite, block/deny \xE9 ignorado.",
  // Hooks - Exit Code Descriptions
  "stdout/stderr not shown": "stdout/stderr n\xE3o exibido",
  "show stderr to model and continue conversation": "mostrar stderr ao modelo e continuar conversa",
  "show stderr to user only": "mostrar stderr apenas ao usu\xE1rio",
  "stdout shown in transcript mode (ctrl+o)": "stdout exibido no modo transcri\xE7\xE3o (ctrl+o)",
  "show stderr to model immediately": "mostrar stderr ao modelo imediatamente",
  "show stderr to user only but continue with tool call": "mostrar stderr apenas ao usu\xE1rio mas continuar com chamada de ferramenta",
  "block processing, erase original prompt, and show stderr to user only": "bloquear processamento, apagar prompt original e mostrar stderr apenas ao usu\xE1rio",
  "block expanded prompt submission and show stderr to user only": "bloquear envio do prompt expandido e mostrar stderr apenas ao usu\xE1rio",
  "stdout shown to Qwen": "stdout mostrado ao Qwen",
  "show stderr to user only (blocking errors ignored)": "mostrar stderr apenas ao usu\xE1rio (erros de bloqueio ignorados)",
  "command completes successfully": "comando conclu\xEDdo com sucesso",
  "stdout shown to subagent": "stdout mostrado ao subagente",
  "show stderr to subagent and continue having it run": "mostrar stderr ao subagente e continuar executando",
  "stdout appended as custom compact instructions": "stdout anexado como instru\xE7\xF5es de compacta\xE7\xE3o personalizadas",
  "block compaction": "bloquear compacta\xE7\xE3o",
  "show stderr to user only but continue with compaction": "mostrar stderr apenas ao usu\xE1rio mas continuar com compacta\xE7\xE3o",
  "use hook decision if provided": "usar decis\xE3o do hook se fornecida",
  "allow todo creation": "permitir cria\xE7\xE3o de todo",
  "block todo creation and show reason to model": "bloquear cria\xE7\xE3o de todo e mostrar motivo ao modelo",
  "allow todo completion": "permitir conclus\xE3o de todo",
  "block todo completion and show reason to model": "bloquear conclus\xE3o de todo e mostrar motivo ao modelo",
  // Hooks - Messages
  "Config not loaded.": "Configura\xE7\xE3o n\xE3o carregada.",
  "Hooks are not enabled. Enable hooks in settings to use this feature.": "Hooks n\xE3o est\xE3o ativados. Ative hooks nas configura\xE7\xF5es para usar este recurso.",
  // ============================================================================
  // Commands - Session Export
  // ============================================================================
  "Export current session message history to a file": "Exportar o hist\xF3rico de mensagens da sess\xE3o atual para um arquivo",
  "Export session to HTML format": "Exportar a sess\xE3o para o formato HTML",
  "Export session to JSON format": "Exportar a sess\xE3o para o formato JSON",
  "Export session to JSONL format (one message per line)": "Exportar a sess\xE3o para o formato JSONL (uma mensagem por linha)",
  "Export session to markdown format": "Exportar a sess\xE3o para o formato Markdown",
  // ============================================================================
  // Commands - Insights
  // ============================================================================
  "generate personalized programming insights from your chat history": "Gerar insights personalizados de programa\xE7\xE3o a partir do seu hist\xF3rico de chat",
  // ============================================================================
  // Commands - Session History
  // ============================================================================
  "Resume a previous session": "Retomar uma sess\xE3o anterior",
  "Fork the current conversation into a new session": "Ramificar a conversa atual em uma nova sess\xE3o",
  "Spawn a background agent that inherits the full conversation": "Iniciar um agente em segundo plano que herda toda a conversa",
  "Please provide a directive. Usage: /fork <directive>": "Forne\xE7a uma diretiva. Uso: /fork <diretiva>",
  "Cannot fork while a response or tool call is in progress. Wait for it to finish or resolve the pending tool call.": "N\xE3o \xE9 poss\xEDvel criar um fork enquanto uma resposta ou chamada de ferramenta est\xE1 em andamento. Aguarde a conclus\xE3o ou resolva a chamada de ferramenta pendente.",
  "Cannot fork before the first conversation turn.": "N\xE3o \xE9 poss\xEDvel criar um fork antes da primeira rodada da conversa.",
  "The agent tool is unavailable; cannot fork.": "A ferramenta de agente est\xE1 indispon\xEDvel; n\xE3o \xE9 poss\xEDvel criar um fork.",
  "Failed to launch fork: {{error}}": "Falha ao iniciar o fork: {{error}}",
  "User launched a background fork via /fork: {{directive}}": "O usu\xE1rio iniciou um fork em segundo plano via /fork: {{directive}}",
  "Forked into a background agent. It inherits this conversation and runs without blocking \u2014 track it in the background tasks panel; it reports back when done.": "Fork criado em um agente em segundo plano. Ele herda esta conversa e roda sem bloquear \u2014 acompanhe no painel de tarefas em segundo plano; ele informar\xE1 quando terminar.",
  "Cannot branch while a response or tool call is in progress. Wait for it to finish or resolve the pending tool call.": "N\xE3o \xE9 poss\xEDvel ramificar enquanto uma resposta ou chamada de ferramenta est\xE1 em andamento. Aguarde a conclus\xE3o ou resolva a chamada de ferramenta pendente.",
  "No conversation to branch.": "N\xE3o h\xE1 conversa para ramificar.",
  "Restore a tool call. This will reset the conversation and file history to the state it was in when the tool call was suggested": "Restaurar uma chamada de ferramenta. Isso redefinir\xE1 o hist\xF3rico da conversa e dos arquivos para o estado em que a chamada da ferramenta foi sugerida",
  "Could not detect terminal type. Supported terminals: VS Code, Cursor, Windsurf, and Trae.": "N\xE3o foi poss\xEDvel detectar o tipo de terminal. Terminais suportados: VS Code, Cursor, Windsurf e Trae.",
  'Terminal "{{terminal}}" is not supported yet.': 'O terminal "{{terminal}}" ainda n\xE3o \xE9 suportado.',
  // ============================================================================
  // Commands - Language
  // ============================================================================
  "Invalid language. Available: {{options}}": "Idioma inv\xE1lido. Dispon\xEDveis: {{options}}",
  "Language subcommands do not accept additional arguments.": "Subcomandos de idioma n\xE3o aceitam argumentos adicionais.",
  "Current UI language: {{lang}}": "Idioma atual da interface: {{lang}}",
  "Current LLM output language: {{lang}}": "Idioma atual da sa\xEDda do LLM: {{lang}}",
  "Set UI language": "Definir idioma da interface",
  "Set LLM output language": "Definir idioma de sa\xEDda do LLM",
  "Usage: /language ui [{{options}}]": "Uso: /language ui [{{options}}]",
  "Usage: /language output <language>": "Uso: /language output <idioma>",
  "Example: /language output \u4E2D\u6587": "Exemplo: /language output Portugu\xEAs",
  "Example: /language output English": "Exemplo: /language output Ingl\xEAs",
  "Example: /language output \u65E5\u672C\u8A9E": "Exemplo: /language output Japon\xEAs",
  "UI language changed to {{lang}}": "Idioma da interface alterado para {{lang}}",
  "LLM output language set to {{lang}}": "Idioma de sa\xEDda do LLM definido para {{lang}}",
  "Please restart the application for the changes to take effect.": "Reinicie o aplicativo para que as altera\xE7\xF5es tenham efeito.",
  "Failed to generate LLM output language rule file: {{error}}": "Falha ao gerar arquivo de regra de idioma de sa\xEDda do LLM: {{error}}",
  "Invalid command. Available subcommands:": "Comando inv\xE1lido. Subcomandos dispon\xEDveis:",
  "Available subcommands:": "Subcomandos dispon\xEDveis:",
  "To request additional UI language packs, please open an issue on GitHub.": "Para solicitar pacotes de idiomas de interface adicionais, abra um problema no GitHub.",
  "Available options:": "Op\xE7\xF5es dispon\xEDveis:",
  "Set UI language to {{name}}": "Definir idioma da interface para {{name}}",
  // ============================================================================
  // Commands - Approval Mode
  // ============================================================================
  "Tool Approval Mode": "Modo de Aprova\xE7\xE3o de Ferramenta",
  "Analyze only, do not modify files or execute commands": "Apenas analisar, n\xE3o modificar arquivos nem executar comandos",
  "Require approval for file edits or shell commands": "Exigir aprova\xE7\xE3o para edi\xE7\xF5es de arquivos ou comandos shell",
  "Automatically approve file edits": "Aprovar automaticamente edi\xE7\xF5es de arquivos",
  "Use classifier to automatically approve safe tool calls": "Usar o classificador para aprovar automaticamente chamadas seguras de ferramentas",
  "Automatically approve all tools": "Aprovar automaticamente todas as ferramentas",
  "Workspace approval mode exists and takes priority. User-level change will have no effect.": "O modo de aprova\xE7\xE3o do workspace existe e tem prioridade. A altera\xE7\xE3o no n\xEDvel do usu\xE1rio n\xE3o ter\xE1 efeito.",
  "Apply To": "Aplicar A",
  "Workspace Settings": "Configura\xE7\xF5es do Workspace",
  "Open auto-memory folder": "Abrir pasta de mem\xF3ria autom\xE1tica",
  "Auto-memory: {{status}}": "Mem\xF3ria autom\xE1tica: {{status}}",
  "Auto-dream: {{status}} \xB7 {{lastDream}} \xB7 /dream to run": "Consolida\xE7\xE3o autom\xE1tica: {{status}} \xB7 {{lastDream}} \xB7 /dream para executar",
  "Auto-skill: {{status}}": "Habilidade autom\xE1tica: {{status}}",
  never: "nunca",
  on: "ativado",
  off: "desativado",
  "Remove matching entries from managed auto-memory.": "Remover entradas correspondentes da mem\xF3ria autom\xE1tica gerenciada.",
  "Usage: /forget <memory text to remove>": "Uso: /forget <texto de mem\xF3ria a remover>",
  "No managed auto-memory entries matched: {{query}}": "Nenhuma entrada de mem\xF3ria autom\xE1tica gerenciada correspondeu: {{query}}",
  "Consolidate managed auto-memory topic files.": "Consolidar arquivos de t\xF3picos de mem\xF3ria autom\xE1tica gerenciada.",
  "Could not retrieve tool registry.": "N\xE3o foi poss\xEDvel recuperar o registro de ferramentas.",
  "Successfully authenticated and refreshed tools for '{{name}}'.": "Autenticado com sucesso e ferramentas atualizadas para '{{name}}'.",
  "Re-discovering tools from '{{name}}'...": "Redescobrindo ferramentas de '{{name}}'...",
  "Discovered {{count}} tool(s) from '{{name}}'.": "{{count}} ferramenta(s) descoberta(s) de '{{name}}'.",
  "Authentication complete. Returning to server details...": "Autentica\xE7\xE3o conclu\xEDda. Retornando aos detalhes do servidor...",
  "Authentication successful.": "Autentica\xE7\xE3o bem-sucedida.",
  // =========================================================
  // Commands - Summary
  // ============================================================================
  "Generate a project summary and save it to .qwen/PROJECT_SUMMARY.md": "Gerar um resumo do projeto e salv\xE1-lo em .qwen/PROJECT_SUMMARY.md",
  "No chat client available to generate summary.": "Nenhum cliente de chat dispon\xEDvel para gerar o resumo.",
  "Already generating summary, wait for previous request to complete": "J\xE1 gerando resumo, aguarde a conclus\xE3o da solicita\xE7\xE3o anterior",
  "No conversation found to summarize.": "Nenhuma conversa encontrada para resumir.",
  "Failed to generate project context summary: {{error}}": "Falha ao gerar resumo do contexto do projeto: {{error}}",
  "Saved project summary to {{filePathForDisplay}}.": "Resumo do projeto salvo em {{filePathForDisplay}}.",
  "Saving project summary...": "Salvando resumo do projeto...",
  "Generating project summary...": "Gerando resumo do projeto...",
  "Processing summary...": "Processando resumo...",
  "Project summary generated and saved successfully!": "Resumo do projeto gerado e salvo com sucesso!",
  "Saved to: {{filePath}}": "Salvo em: {{filePath}}",
  "Stopped because": "Parado porque",
  "Failed to generate summary - no text content received from LLM response": "Falha ao gerar resumo - nenhum conte\xFAdo de texto recebido da resposta do LLM",
  // ============================================================================
  // Commands - Model
  // ============================================================================
  "Switch the model for this session (--fast for suggestion model, [model-id] to switch immediately).": "Trocar o modelo para esta sess\xE3o (--fast para modelo de sugest\xF5es)",
  "Set a lighter model for prompt suggestions and speculative execution": "Definir modelo mais leve para sugest\xF5es de prompt e execu\xE7\xE3o especulativa",
  "Content generator configuration not available.": "Configura\xE7\xE3o do gerador de conte\xFAdo n\xE3o dispon\xEDvel.",
  "Authentication type not available.": "Tipo de autentica\xE7\xE3o n\xE3o dispon\xEDvel.",
  "No models available for the current authentication type ({{authType}}).": "Nenhum modelo dispon\xEDvel para o tipo de autentica\xE7\xE3o atual ({{authType}}).",
  // Needs translation
  " (not in model registry)": " (not in model registry)",
  // ============================================================================
  // Commands - Clear
  // ============================================================================
  "Starting a new session, resetting chat, and clearing terminal.": "Iniciando uma nova sess\xE3o, resetando o chat e limpando o terminal.",
  "Starting a new session and clearing.": "Iniciando uma nova sess\xE3o e limpando.",
  // ============================================================================
  // Commands - Compress
  // ============================================================================
  "Already compressing, wait for previous request to complete": "J\xE1 comprimindo, aguarde a conclus\xE3o da solicita\xE7\xE3o anterior",
  "Failed to compress chat history.": "Falha ao comprimir hist\xF3rico do chat.",
  "Failed to compress chat history: {{error}}": "Falha ao comprimir hist\xF3rico do chat: {{error}}",
  "Compressing chat history": "Comprimindo hist\xF3rico do chat",
  "Chat history compressed from {{originalTokens}} to {{newTokens}} tokens.": "Hist\xF3rico do chat comprimido de {{originalTokens}} para {{newTokens}} tokens.",
  "Compression was not beneficial for this history size.": "A compress\xE3o n\xE3o foi ben\xE9fica para este tamanho de hist\xF3rico.",
  "Chat history compression did not reduce size. This may indicate issues with the compression prompt.": "A compress\xE3o do hist\xF3rico do chat n\xE3o reduziu o tamanho. Isso pode indicar problemas com o prompt de compress\xE3o.",
  "Could not compress chat history due to a token counting error.": "N\xE3o foi poss\xEDvel comprimir o hist\xF3rico do chat devido a um erro de contagem de tokens.",
  // ============================================================================
  // Commands - Directory
  // ============================================================================
  "Configuration is not available.": "A configura\xE7\xE3o n\xE3o est\xE1 dispon\xEDvel.",
  "Please provide at least one path to add.": "Forne\xE7a pelo menos um caminho para adicionar.",
  "The /directory add command is not supported in restrictive sandbox profiles. Please use --include-directories when starting the session instead.": "O comando /directory add n\xE3o \xE9 suportado em perfis de sandbox restritivos. Use --include-directories ao iniciar a sess\xE3o.",
  "Error adding '{{path}}': {{error}}": "Erro ao adicionar '{{path}}': {{error}}",
  "Successfully added QWEN.md files from the following directories if there are:\n- {{directories}}": "Arquivos QWEN.md adicionados com sucesso dos seguintes diret\xF3rios, se houverem:\n- {{directories}}",
  "Error refreshing memory: {{error}}": "Erro ao atualizar mem\xF3ria: {{error}}",
  "Successfully added directories:\n- {{directories}}": "Diret\xF3rios adicionados com sucesso:\n- {{directories}}",
  "Current workspace directories:\n{{directories}}": "Diret\xF3rios atuais do workspace:\n{{directories}}",
  // ============================================================================
  // Commands - Docs
  // ============================================================================
  "Please open the following URL in your browser to view the documentation:\n{{url}}": "Abra a seguinte URL no seu navegador para ver a documenta\xE7\xE3o:\n{{url}}",
  "Opening documentation in your browser: {{url}}": "Abrindo documenta\xE7\xE3o no seu navegador: {{url}}",
  // ============================================================================
  // Dialogs - Tool Confirmation
  // ============================================================================
  "Do you want to proceed?": "Voc\xEA deseja prosseguir?",
  "Yes, allow once": "Sim, permitir uma vez",
  "Allow always": "Permitir sempre",
  Yes: "Sim",
  No: "N\xE3o",
  "No (esc)": "N\xE3o (esc)",
  // MCP Management - Core translations
  "Manage MCP servers": "Gerenciar MCP servers",
  "Server Detail": "Detalhes do servidor",
  Tools: "Ferramentas",
  "Tool Detail": "Detalhes da ferramenta",
  "Loading...": "Carregando...",
  "Unknown step": "Etapa desconhecida",
  "Esc to back": "Esc para voltar",
  "\u2191\u2193 to navigate \xB7 Enter to select \xB7 Esc to close": "\u2191\u2193 navegar \xB7 Enter selecionar \xB7 Esc fechar",
  "\u2191\u2193 to navigate \xB7 Enter to select \xB7 Esc to back": "\u2191\u2193 navegar \xB7 Enter selecionar \xB7 Esc voltar",
  "\u2191\u2193 to navigate \xB7 Enter to confirm \xB7 Esc to back": "\u2191\u2193 navegar \xB7 Enter confirmar \xB7 Esc voltar",
  "User Settings (global)": "Configura\xE7\xF5es do usu\xE1rio (global)",
  "Workspace Settings (project-specific)": "Configura\xE7\xF5es do workspace (espec\xEDfico do projeto)",
  "Disable server:": "Desativar servidor:",
  "Select where to add the server to the exclude list:": "Selecione onde adicionar o servidor \xE0 lista de exclus\xE3o:",
  "Press Enter to confirm, Esc to cancel": "Enter para confirmar, Esc para cancelar",
  Disable: "Desativar",
  Enable: "Ativar",
  Authenticate: "Autenticar",
  "Re-authenticate": "Reautenticar",
  "Clear Authentication": "Limpar autentica\xE7\xE3o",
  disabled: "desativado",
  enabled: "ativado",
  "Server:": "Servidor:",
  Reconnect: "Reconectar",
  "View tools": "Ver ferramentas",
  "Source:": "Fonte:",
  "Command:": "Comando:",
  "Working Directory:": "Diret\xF3rio de trabalho:",
  "No server selected": "Nenhum servidor selecionado",
  "Error:": "Erro:",
  tool: "ferramenta",
  tools: "ferramentas",
  connected: "conectado",
  connecting: "conectando",
  disconnected: "desconectado",
  error: "erro",
  // MCP Server List
  "User MCPs": "MCPs do usu\xE1rio",
  "Project MCPs": "MCPs do projeto",
  "Extension MCPs": "MCPs de extens\xE3o",
  server: "servidor",
  servers: "servidores",
  "Add MCP servers to your settings to get started.": "Adicione MCP servers \xE0s suas configura\xE7\xF5es para come\xE7ar.",
  "Run qwen --debug to see error logs": "Execute qwen --debug para ver os logs de erro",
  // MCP OAuth Authentication
  "OAuth Authentication": "Autentica\xE7\xE3o OAuth",
  "Authenticating... Please complete the login in your browser.": "Autenticando... Por favor, conclua o login no seu navegador.",
  // MCP Tool List
  "No tools available for this server.": "Nenhuma ferramenta dispon\xEDvel para este servidor.",
  destructive: "destrutivo",
  "read-only": "somente leitura",
  "open-world": "mundo aberto",
  idempotent: "idempotente",
  "Tools for {{serverName}}": "Ferramentas para {{serverName}}",
  "{{current}}/{{total}}": "{{current}}/{{total}}",
  // MCP Tool Detail
  required: "obrigat\xF3rio",
  Parameters: "Par\xE2metros",
  "No tool selected": "Nenhuma ferramenta selecionada",
  Server: "Servidor",
  // Invalid tool related translations
  "{{count}} invalid tools": "{{count}} ferramentas inv\xE1lidas",
  invalid: "inv\xE1lido",
  "invalid: {{reason}}": "inv\xE1lido: {{reason}}",
  "missing name": "nome ausente",
  "missing description": "descri\xE7\xE3o ausente",
  "(unnamed)": "(sem nome)",
  "Warning: This tool cannot be called by the LLM": "Aviso: Esta ferramenta n\xE3o pode ser chamada pelo LLM",
  Reason: "Motivo",
  "Tools must have both name and description to be used by the LLM.": "As ferramentas devem ter tanto nome quanto descri\xE7\xE3o para serem usadas pelo LLM.",
  "Modify in progress:": "Modifica\xE7\xE3o em progresso:",
  "Save and close external editor to continue": "Salve e feche o editor externo para continuar",
  "Apply this change?": "Aplicar esta altera\xE7\xE3o?",
  "Yes, allow always": "Sim, permitir sempre",
  "Modify with external editor": "Modificar com editor externo",
  "No, suggest changes (esc)": "N\xE3o, sugerir altera\xE7\xF5es (esc)",
  "Allow execution of: '{{command}}'?": "Permitir a execu\xE7\xE3o de: '{{command}}'?",
  "Always allow in this project": "Sempre permitir neste projeto",
  "Always allow {{action}} in this project": "Sempre permitir {{action}} neste projeto",
  "Always allow for this user": "Sempre permitir para este usu\xE1rio",
  "Always allow {{action}} for this user": "Sempre permitir {{action}} para este usu\xE1rio",
  "Yes, restore previous mode ({{mode}})": "Sim, restaurar modo anterior ({{mode}})",
  "Yes, and auto-accept edits": "Sim, e aceitar edi\xE7\xF5es automaticamente",
  "Yes, and manually approve edits": "Sim, e aprovar edi\xE7\xF5es manualmente",
  "No, keep planning (esc)": "N\xE3o, continuar planejando (esc)",
  "URLs to fetch:": "URLs para buscar:",
  "MCP Server: {{server}}": "MCP Server: {{server}}",
  "Tool: {{tool}}": "Ferramenta: {{tool}}",
  'Allow execution of MCP tool "{{tool}}" from server "{{server}}"?': 'Permitir a execu\xE7\xE3o de MCP tool "{{tool}}" de MCP server "{{server}}"?',
  // ============================================================================
  // Dialogs - Shell Confirmation
  // ============================================================================
  "Shell Command Execution": "Execu\xE7\xE3o de Comando Shell",
  "A custom command wants to run the following shell commands:": "Um comando personalizado deseja executar os seguintes comandos shell:",
  // ============================================================================
  // Dialogs - Welcome Back
  // ============================================================================
  "Current Plan:": "Plano Atual:",
  "Progress: {{done}}/{{total}} tasks completed": "Progresso: {{done}}/{{total}} tarefas conclu\xEDdas",
  ", {{inProgress}} in progress": ", {{inProgress}} em progresso",
  "Pending Tasks:": "Tarefas Pendentes:",
  "What would you like to do?": "O que voc\xEA gostaria de fazer?",
  "Choose how to proceed with your session:": "Escolha como proceder com sua sess\xE3o:",
  "Start new chat session": "Iniciar nova sess\xE3o de chat",
  "Continue previous conversation": "Continuar conversa anterior",
  "\u{1F44B} Welcome back! (Last updated: {{timeAgo}})": "\u{1F44B} Bem-vindo de volta! (\xDAltima atualiza\xE7\xE3o: {{timeAgo}})",
  "\u{1F3AF} Overall Goal:": "\u{1F3AF} Objetivo Geral:",
  "Connect a Provider": "Conectar um provedor",
  "You must connect a provider to proceed. Press Ctrl+C again to exit.": "Voc\xEA deve conectar um provedor para prosseguir. Pressione Ctrl+C novamente para sair.",
  "Terms of Services and Privacy Notice": "Termos de Servi\xE7o e Aviso de Privacidade",
  "Qwen OAuth": "Qwen OAuth",
  "Discontinued \u2014 switch to Coding Plan or API Key": "Descontinuado \u2014 mude para Coding Plan ou API Key",
  "Qwen OAuth free tier was discontinued on 2026-04-15. Please select Coding Plan or API Key instead.": "O n\xEDvel gratuito do Qwen OAuth foi descontinuado em 2026-04-15. Selecione Coding Plan ou API Key.",
  "Qwen OAuth free tier was discontinued on 2026-04-15. Please select a model from another provider or run /auth to switch.": "O n\xEDvel gratuito do Qwen OAuth foi descontinuado em 2026-04-15. Por favor, selecione um modelo de outro provedor ou execute /auth para trocar.",
  "\n\u26A0 Qwen OAuth free tier was discontinued on 2026-04-15. Please select another option.\n": "\n\u26A0 O n\xEDvel gratuito do Qwen OAuth foi descontinuado em 2026-04-15. Selecione outra op\xE7\xE3o.\n",
  "Paid \xB7 Up to 6,000 requests/5 hrs \xB7 All Alibaba Cloud Coding Plan Models": "Pago \xB7 At\xE9 6.000 solicita\xE7\xF5es/5 hrs \xB7 Todos os modelos Alibaba Cloud Coding Plan",
  "Alibaba Cloud Coding Plan": "Alibaba Cloud Coding Plan",
  "Bring your own API key": "Traga sua pr\xF3pria API Key",
  "Browser-based authentication with third-party providers (e.g. OpenRouter, ModelScope)": "Autentica\xE7\xE3o baseada em navegador com provedores terceiros (por exemplo, OpenRouter, ModelScope)",
  "Authentication is enforced to be {{enforcedType}}, but you are currently using {{currentType}}.": "A autentica\xE7\xE3o \xE9 for\xE7ada para {{enforcedType}}, mas voc\xEA est\xE1 usando {{currentType}} no momento.",
  "Qwen OAuth Authentication": "Autentica\xE7\xE3o Qwen OAuth",
  "Please visit this URL to authorize:": "Visite esta URL para autorizar:",
  "Waiting for authorization": "Aguardando autoriza\xE7\xE3o",
  "Time remaining:": "Tempo restante:",
  "Qwen OAuth Authentication Timeout": "Tempo Limite de Autentica\xE7\xE3o Qwen OAuth",
  "OAuth token expired (over {{seconds}} seconds). Please select authentication method again.": "Token OAuth expirado (mais de {{seconds}} segundos). Selecione o m\xE9todo de autentica\xE7\xE3o novamente.",
  "Press any key to return to authentication type selection.": "Pressione qualquer tecla para retornar \xE0 sele\xE7\xE3o do tipo de autentica\xE7\xE3o.",
  "Waiting for Qwen OAuth authentication...": "Aguardando autentica\xE7\xE3o Qwen OAuth...",
  "Authentication timed out. Please try again.": "A autentica\xE7\xE3o expirou. Tente novamente.",
  "Waiting for auth... (Press ESC or CTRL+C to cancel)": "Aguardando autentica\xE7\xE3o... (Pressione ESC ou CTRL+C para cancelar)",
  "Missing API key for OpenAI-compatible auth. Set settings.security.auth.apiKey, or set the {{envKeyHint}} environment variable.": "API Key ausente para autentica\xE7\xE3o compat\xEDvel com OpenAI. Defina settings.security.auth.apiKey ou a vari\xE1vel de ambiente {{envKeyHint}}.",
  "{{envKeyHint}} environment variable not found. Please set it in your .env file or environment variables.": "Vari\xE1vel de ambiente {{envKeyHint}} n\xE3o encontrada. Defina-a no seu arquivo .env ou vari\xE1veis de ambiente.",
  "{{envKeyHint}} environment variable not found (or set settings.security.auth.apiKey). Please set it in your .env file or environment variables.": "Vari\xE1vel de ambiente {{envKeyHint}} n\xE3o encontrada (ou defina settings.security.auth.apiKey). Defina-a no seu arquivo .env ou vari\xE1veis de ambiente.",
  "Missing API key for OpenAI-compatible auth. Set the {{envKeyHint}} environment variable.": "API Key ausente para autentica\xE7\xE3o compat\xEDvel com OpenAI. Defina a vari\xE1vel de ambiente {{envKeyHint}}.",
  "Anthropic provider missing required baseUrl in modelProviders[].baseUrl.": "Provedor Anthropic sem a baseUrl necess\xE1ria em modelProviders[].baseUrl.",
  "ANTHROPIC_BASE_URL environment variable not found.": "Vari\xE1vel de ambiente ANTHROPIC_BASE_URL n\xE3o encontrada.",
  "Invalid auth method selected.": "M\xE9todo de autentica\xE7\xE3o inv\xE1lido selecionado.",
  "Failed to authenticate. Message: {{message}}": "Falha ao autenticar. Mensagem: {{message}}",
  "Authenticated successfully with {{authType}} credentials.": "Autenticado com sucesso com credenciais {{authType}}.",
  'Invalid QWEN_DEFAULT_AUTH_TYPE value: "{{value}}". Valid values are: {{validValues}}': 'Valor QWEN_DEFAULT_AUTH_TYPE inv\xE1lido: "{{value}}". Valores v\xE1lidos s\xE3o: {{validValues}}',
  // ============================================================================
  // Dialogs - Model
  // ============================================================================
  "Select Model": "Selecionar Modelo",
  "API Key": "API Key",
  "(default)": "(padr\xE3o)",
  "(not set)": "(n\xE3o definido)",
  Modality: "Modalidade",
  "Context Window": "Janela de Contexto",
  text: "texto",
  "text-only": "somente texto",
  image: "imagem",
  pdf: "PDF",
  audio: "\xE1udio",
  video: "v\xEDdeo",
  "not set": "n\xE3o definido",
  none: "nenhum",
  unknown: "desconhecido",
  // ============================================================================
  // Dialogs - Permissions
  // ============================================================================
  "Manage folder trust settings": "Gerenciar configura\xE7\xF5es de confian\xE7a de pasta",
  "Manage permission rules": "Gerenciar permission rules",
  Allow: "Permitir",
  Ask: "Perguntar",
  Deny: "Negar",
  Workspace: "\xC1rea de trabalho",
  "Qwen Code won't ask before using allowed tools.": "O Qwen Code n\xE3o perguntar\xE1 antes de usar ferramentas permitidas.",
  "Qwen Code will ask before using these tools.": "O Qwen Code perguntar\xE1 antes de usar essas ferramentas.",
  "Qwen Code is not allowed to use denied tools.": "O Qwen Code n\xE3o tem permiss\xE3o para usar ferramentas negadas.",
  "Manage trusted directories for this workspace.": "Gerenciar diret\xF3rios confi\xE1veis para esta \xE1rea de trabalho.",
  "Any use of the {{tool}} tool": "Qualquer uso da ferramenta {{tool}}",
  "{{tool}} commands matching '{{pattern}}'": "Comandos {{tool}} correspondentes a '{{pattern}}'",
  "From user settings": "Das configura\xE7\xF5es do usu\xE1rio",
  "From project settings": "Das configura\xE7\xF5es do projeto",
  "From session": "Da sess\xE3o",
  "Project settings": "Configura\xE7\xF5es do projeto",
  "Checked in at .qwen/settings.json": "Registrado em .qwen/settings.json",
  "User settings": "Configura\xE7\xF5es do usu\xE1rio",
  "Saved in at ~/.qwen/settings.json": "Salvo em ~/.qwen/settings.json",
  "Add a new rule\u2026": "Adicionar nova regra\u2026",
  "Add {{type}} permission rule": "Adicionar {{type}} permission rule",
  "Permission rules are a tool name, optionally followed by a specifier in parentheses.": "permission rules s\xE3o um nome de ferramenta, opcionalmente seguido por um especificador entre par\xEAnteses.",
  "e.g.,": "ex.",
  or: "ou",
  "Enter permission rule\u2026": "Insira permission rule\u2026",
  "Enter to submit \xB7 Esc to cancel": "Enter para enviar \xB7 Esc para cancelar",
  "Where should this rule be saved?": "Onde esta regra deve ser salva?",
  "Enter to confirm \xB7 Esc to cancel": "Enter para confirmar \xB7 Esc para cancelar",
  "Delete {{type}} rule?": "Excluir regra {{type}}?",
  "Are you sure you want to delete this permission rule?": "Tem certeza de que deseja excluir esta permission rule?",
  "Permissions:": "Permiss\xF5es:",
  "(\u2190/\u2192 or tab to cycle)": "(\u2190/\u2192 ou Tab para alternar)",
  "Press \u2191\u2193 to navigate \xB7 Enter to select \xB7 Type to search \xB7 Esc to cancel": "\u2191\u2193 para navegar \xB7 Enter para selecionar \xB7 Digite para pesquisar \xB7 Esc para cancelar",
  "Search\u2026": "Pesquisar\u2026",
  // Workspace directory management
  "Add directory\u2026": "Adicionar diret\xF3rio\u2026",
  "Add directory to workspace": "Adicionar diret\xF3rio \xE0 \xE1rea de trabalho",
  "Qwen Code can read files in the workspace, and make edits when auto-accept edits is on.": "O Qwen Code pode ler arquivos na \xE1rea de trabalho e fazer edi\xE7\xF5es quando a aceita\xE7\xE3o autom\xE1tica est\xE1 ativada.",
  "Qwen Code will be able to read files in this directory and make edits when auto-accept edits is on.": "O Qwen Code poder\xE1 ler arquivos neste diret\xF3rio e fazer edi\xE7\xF5es quando a aceita\xE7\xE3o autom\xE1tica est\xE1 ativada.",
  "Enter the path to the directory:": "Insira o caminho do diret\xF3rio:",
  "Enter directory path\u2026": "Insira o caminho do diret\xF3rio\u2026",
  "Tab to complete \xB7 Enter to add \xB7 Esc to cancel": "Tab para completar \xB7 Enter para adicionar \xB7 Esc para cancelar",
  "Remove directory?": "Remover diret\xF3rio?",
  "Are you sure you want to remove this directory from the workspace?": "Tem certeza de que deseja remover este diret\xF3rio da \xE1rea de trabalho?",
  "  (Original working directory)": "  (Diret\xF3rio de trabalho original)",
  "  (from settings)": "  (das configura\xE7\xF5es)",
  "Directory does not exist.": "O diret\xF3rio n\xE3o existe.",
  "Path is not a directory.": "O caminho n\xE3o \xE9 um diret\xF3rio.",
  "This directory is already in the workspace.": "Este diret\xF3rio j\xE1 est\xE1 na \xE1rea de trabalho.",
  "Already covered by existing directory: {{dir}}": "J\xE1 coberto pelo diret\xF3rio existente: {{dir}}",
  // ============================================================================
  // Status Bar
  // ============================================================================
  "Using:": "Usando:",
  "{{count}} open file": "{{count}} arquivo aberto",
  "{{count}} open files": "{{count}} arquivos abertos",
  "(ctrl+g to view)": "(ctrl+g para ver)",
  "{{count}} {{name}} file": "{{count}} arquivo {{name}}",
  "{{count}} {{name}} files": "{{count}} arquivos {{name}}",
  "{{count}} MCP server": "{{count}} MCP server",
  "{{count}} MCP servers": "{{count}} MCP servers",
  "{{count}} Blocked": "{{count}} Bloqueados",
  "(ctrl+t to view)": "(ctrl+t para ver)",
  "(ctrl+t to toggle)": "(ctrl+t para alternar)",
  "Press Ctrl+C again to exit.": "Pressione Ctrl+C novamente para sair.",
  "Press Ctrl+D again to exit.": "Pressione Ctrl+D novamente para sair.",
  "Press Esc again to clear.": "Pressione Esc novamente para limpar.",
  "Press \u2191 to edit queued messages": "Pressione \u2191 para editar mensagens na fila",
  // ============================================================================
  // MCP Status
  // ============================================================================
  "No MCP servers configured.": "Nenhum MCP servers configurado.",
  "\u23F3 MCP servers are starting up ({{count}} initializing)...": "\u23F3 MCP servers est\xE3o iniciando ({{count}} inicializando)...",
  "Note: First startup may take longer. Tool availability will update automatically.": "Nota: A primeira inicializa\xE7\xE3o pode demorar mais. A disponibilidade da ferramenta ser\xE1 atualizada automaticamente.",
  "Configured MCP servers:": "MCP servers configurados:",
  Ready: "Pronto",
  "Starting... (first startup may take longer)": "Iniciando... (a primeira inicializa\xE7\xE3o pode demorar mais)",
  Disconnected: "Desconectado",
  "{{count}} tool": "{{count}} ferramenta",
  "{{count}} tools": "{{count}} ferramentas",
  "{{count}} prompt": "{{count}} prompt",
  "{{count}} prompts": "{{count}} prompts",
  "(from {{extensionName}})": "(de {{extensionName}})",
  OAuth: "OAuth",
  "OAuth expired": "OAuth expirado",
  "OAuth not authenticated": "OAuth n\xE3o autenticado",
  "tools and prompts will appear when ready": "ferramentas e prompts aparecer\xE3o quando estiverem prontos",
  "{{count}} tools cached": "{{count}} ferramentas em cache",
  "Tools:": "Ferramentas:",
  "Parameters:": "Par\xE2metros:",
  "Prompts:": "Prompts:",
  "Resources:": "Recursos:",
  Blocked: "Bloqueado",
  "\u{1F4A1} Tips:": "\u{1F4A1} Dicas:",
  "to show server and tool descriptions": "para mostrar descri\xE7\xF5es de servidores e ferramentas",
  "to show tool parameter schemas": "para mostrar tool parameter schemas",
  "to hide descriptions": "para ocultar descri\xE7\xF5es",
  "to authenticate with OAuth-enabled servers": "para autenticar com servidores habilitados para OAuth",
  Press: "Pressione",
  "to toggle tool descriptions on/off": "para alternar descri\xE7\xF5es de ferramentas ligadas/desligadas",
  "Starting OAuth authentication for MCP server '{{name}}'...": "Iniciando autentica\xE7\xE3o OAuth para MCP server '{{name}}'...",
  // ============================================================================
  // Startup Tips
  // ============================================================================
  "Tips:": "Dicas:",
  "Use /compress when the conversation gets long to summarize history and free up context.": "Use /compress quando a conversa ficar longa para resumir o hist\xF3rico e liberar contexto.",
  "Start a fresh idea with /clear or /new; the previous session stays available in history.": "Comece uma nova ideia com /clear ou /new; a sess\xE3o anterior permanece dispon\xEDvel no hist\xF3rico.",
  "Use /bug to submit issues to the maintainers when something goes off.": "Use /bug para enviar problemas aos mantenedores quando algo der errado.",
  "Switch auth type quickly with /auth.": "Troque o tipo de autentica\xE7\xE3o rapidamente com /auth.",
  "You can run any shell commands from Qwen Code using ! (e.g. !ls).": "Voc\xEA pode executar quaisquer comandos shell do Qwen Code usando ! (ex: !ls).",
  "Type / to open the command popup; Tab autocompletes slash commands and saved prompts.": "Digite / para abrir o popup de comandos; Tab autocompleta comandos de barra e prompts salvos.",
  "You can resume a previous conversation by running qwen --continue or qwen --resume.": "Voc\xEA pode retomar uma conversa anterior executando qwen --continue ou qwen --resume.",
  "You can switch permission mode quickly with Shift+Tab or /approval-mode.": "Voc\xEA pode alternar o modo de permiss\xE3o rapidamente com Shift+Tab ou /approval-mode.",
  "Try /insight to generate personalized insights from your chat history.": "Experimente /insight para gerar insights personalizados do seu hist\xF3rico de conversas.",
  "Press Ctrl+O to toggle compact mode \u2014 hide tool output and thinking for a cleaner view.": "Pressione Ctrl+O para alternar o modo compacto \u2014 ocultar sa\xEDda de ferramentas e racioc\xEDnio.",
  "Add a QWEN.md file to give Qwen Code persistent project context.": "Adicione um arquivo QWEN.md para dar ao Qwen Code um contexto persistente do projeto.",
  "Use /btw to ask a quick side question without disrupting the conversation.": "Use /btw para fazer uma pergunta lateral r\xE1pida sem interromper a conversa.",
  "Context is almost full! Run /compress now or start /new to continue.": "O contexto est\xE1 quase cheio! Execute /compress agora ou inicie /new para continuar.",
  "Context is getting full. Use /compress to free up space.": "O contexto est\xE1 ficando cheio. Use /compress para liberar espa\xE7o.",
  "Long conversation? /compress summarizes history to free context.": "Conversa longa? /compress resume o hist\xF3rico para liberar contexto.",
  // ============================================================================
  // Exit Screen / Stats
  // ============================================================================
  "Agent powering down. Goodbye!": "Agente desligando. Adeus!",
  "To continue this session, run": "Para continuar esta sess\xE3o, execute",
  "Interaction Summary": "Resumo da Intera\xE7\xE3o",
  "Session ID:": "ID da Sess\xE3o:",
  "Tool Calls:": "Chamadas de Ferramenta:",
  "Success Rate:": "Taxa de Sucesso:",
  "User Agreement:": "Acordo do Usu\xE1rio:",
  reviewed: "revisado",
  "Code Changes:": "Altera\xE7\xF5es de C\xF3digo:",
  Performance: "Desempenho",
  "Wall Time:": "Tempo Total:",
  "Agent Active:": "Agente Ativo:",
  "API Time:": "Tempo de API:",
  "Tool Time:": "Tempo de Ferramenta:",
  "Session Stats": "Estat\xEDsticas da Sess\xE3o",
  "Model Usage": "Uso do Modelo",
  Reqs: "Reqs",
  "Input Tokens": "Tokens de Entrada",
  "Output Tokens": "Tokens de Sa\xEDda",
  "Savings Highlight:": "Destaque de Economia:",
  "of input tokens were served from the cache, reducing costs.": "de tokens de entrada foram servidos do cache, reduzindo custos.",
  "Tip: For a full token breakdown, run `/stats model`.": "Dica: Para um detalhamento completo de tokens, execute `/stats model`.",
  "Model Stats For Nerds": "Estat\xEDsticas de Modelo Para Nerds",
  "Tool Stats For Nerds": "Estat\xEDsticas de Ferramenta Para Nerds",
  Metric: "M\xE9trica",
  API: "API",
  Requests: "Solicita\xE7\xF5es",
  Errors: "Erros",
  "Avg Latency": "Lat\xEAncia M\xE9dia",
  Tokens: "Tokens",
  Total: "Total",
  Prompt: "Prompt",
  Cached: "Cacheado",
  Thoughts: "Pensamentos",
  Output: "Sa\xEDda",
  "No API calls have been made in this session.": "Nenhuma chamada de API foi feita nesta sess\xE3o.",
  "Tool Name": "Nome da Ferramenta",
  Calls: "Chamadas",
  "Success Rate": "Taxa de Sucesso",
  "Avg Duration": "Dura\xE7\xE3o M\xE9dia",
  "User Decision Summary": "Resumo de Decis\xE3o do Usu\xE1rio",
  "Total Reviewed Suggestions:": "Total de Sugest\xF5es Revisadas:",
  " \xBB Accepted:": " \xBB Aceitas:",
  " \xBB Rejected:": " \xBB Rejeitadas:",
  " \xBB Modified:": " \xBB Modificadas:",
  " Overall Agreement Rate:": " Taxa Geral de Acordo:",
  "No tool calls have been made in this session.": "Nenhuma chamada de ferramenta foi feita nesta sess\xE3o.",
  "Session start time is unavailable, cannot calculate stats.": "Hora de in\xEDcio da sess\xE3o indispon\xEDvel, n\xE3o \xE9 poss\xEDvel calcular estat\xEDsticas.",
  Activity: "Atividade",
  Efficiency: "Efici\xEAncia",
  Today: "Hoje",
  "Token Trend": "Tend\xEAncia de Tokens",
  "Cache Hit Rate": "Taxa de cache",
  "Tool Success": "Sucesso de ferramentas",
  "Tool Leaderboard": "Ranking de ferramentas",
  Time: "Tempo",
  Success: "Sucesso",
  Cache: "Cache",
  Latency: "Lat\xEAncia",
  "Code Impact": "Impacto no c\xF3digo",
  net: "l\xEDquido",
  streak: "sequ\xEAncia",
  best: "recorde",
  // ============================================================================
  // Command Format Migration
  // ============================================================================
  "Command Format Migration": "Migra\xE7\xE3o de Formato de Comando",
  "Found {{count}} TOML command file:": "Encontrado {{count}} arquivo de comando TOML:",
  "Found {{count}} TOML command files:": "Encontrados {{count}} arquivos de comando TOML:",
  "Current tasks": "Tarefas atuais",
  "Background tasks": "Tarefas em segundo plano",
  "No tasks currently running": "Nenhuma tarefa em execu\xE7\xE3o",
  "No entry to show.": "Nenhuma entrada para mostrar.",
  "needs approval": "precisa de aprova\xE7\xE3o",
  "rejected \u2014 edit config to re-approve": "rejeitado \u2014 edite a configura\xE7\xE3o para reaprovar",
  "Background agent needs approval": "Agente em segundo plano precisa de aprova\xE7\xE3o",
  "Approve or deny the request above": "Aprove ou negue a solicita\xE7\xE3o acima",
  Running: "Em execu\xE7\xE3o",
  Paused: "Pausado",
  Completed: "Conclu\xEDdo",
  Failed: "Falhou",
  Stopped: "Parado",
  Shell: "Shell",
  Monitor: "Monitor",
  Command: "Comando",
  Dream: "Dream",
  "[dream] memory consolidation": "[dream] consolida\xE7\xE3o de mem\xF3ria",
  "[dream] memory consolidation (reviewing {{count}} session)": "[dream] consolida\xE7\xE3o de mem\xF3ria (revisando {{count}} sess\xE3o)",
  "[dream] memory consolidation (reviewing {{count}} sessions)": "[dream] consolida\xE7\xE3o de mem\xF3ria (revisando {{count}} sess\xF5es)",
  "... and {{count}} more": "... e mais {{count}}",
  "The TOML format is deprecated. Would you like to migrate them to Markdown format?": "O formato TOML est\xE1 obsoleto. Voc\xEA gostaria de migr\xE1-los para o formato Markdown?",
  "(Backups will be created and original files will be preserved)": "(Backups ser\xE3o criados e arquivos originais ser\xE3o preservados)",
  // ============================================================================
  // Loading Phrases
  // ============================================================================
  "Waiting for user confirmation...": "Aguardando confirma\xE7\xE3o do usu\xE1rio...",
  WITTY_LOADING_PHRASES: [
    "Estou com sorte",
    "Enviando maravilhas...",
    "Pintando os serifos de volta...",
    "Navegando pelo mofo limoso...",
    "Consultando os esp\xEDritos digitais...",
    "Reticulando splines...",
    "Aquecendo os hamsters da IA...",
    "Perguntando \xE0 concha m\xE1gica...",
    "Gerando r\xE9plica espirituosa...",
    "Polindo os algoritmos...",
    "N\xE3o apresse a perfei\xE7\xE3o (ou meu c\xF3digo)...",
    "Preparando bytes frescos...",
    "Contando el\xE9trons...",
    "Engajando processadores cognitivos...",
    "Verificando erros de sintaxe no universo...",
    "Um momento, otimizando o humor...",
    "Embaralhando piadas...",
    "Desembara\xE7ando redes neurais...",
    "Compilando brilhantismo...",
    "Carregando humor.exe...",
    "Invocando a nuvem da sabedoria...",
    "Preparando uma resposta espirituosa...",
    "S\xF3 um segundo, estou depurando a realidade...",
    "Confundindo as op\xE7\xF5es...",
    "Sintonizando as frequ\xEAncias c\xF3smicas...",
    "Criando uma resposta digna da sua paci\xEAncia...",
    "Compilando os 1s e 0s...",
    "Resolvendo depend\xEAncias... e crises existenciais...",
    "Desfragmentando mem\xF3rias... tanto RAM quanto pessoais...",
    "Reiniciando o m\xF3dulo de humor...",
    "Fazendo cache do essencial (principalmente memes de gatos)...",
    "Otimizando para velocidade absurda",
    "Trocando bits... n\xE3o conte para os bytes...",
    "Coletando lixo... volto j\xE1...",
    "Montando a internet...",
    "Convertendo caf\xE9 em c\xF3digo...",
    "Atualizando a sintaxe da realidade...",
    "Reconectando as sinapses...",
    "Procurando um ponto e v\xEDrgula perdido...",
    "Lubrificando as engrenagens da m\xE1quina...",
    "Pr\xE9-aquecendo os servidores...",
    "Calibrando o capacitor de fluxo...",
    "Engajando o motor de improbabilidade...",
    "Canalizando a For\xE7a...",
    "Alinhando as estrelas para uma resposta ideal...",
    "Assim dizemos todos...",
    "Carregando a pr\xF3xima grande ideia...",
    "S\xF3 um momento, estou na zona...",
    "Preparando para deslumbr\xE1-lo com brilhantismo...",
    "S\xF3 um tique, estou polindo minha intelig\xEAncia...",
    "Segure firme, estou criando uma obra-prima...",
    "S\xF3 um instante, estou depurando o universo...",
    "S\xF3 um momento, estou alinhando os pixels...",
    "S\xF3 um segundo, estou otimizando o humor...",
    "S\xF3 um momento, estou ajustando os algoritmos...",
    "Velocidade de dobra engajada...",
    "Minerando mais cristais de Dilithium...",
    "N\xE3o entre em p\xE2nico...",
    "Seguindo o coelho branco...",
    "A verdade est\xE1 l\xE1 fora... em algum lugar...",
    "Soprando o cartucho...",
    "Carregando... Fa\xE7a um barrel roll!",
    "Aguardando o respawn...",
    "Terminando a Kessel Run em menos de 12 parsecs...",
    "O bolo n\xE3o \xE9 uma mentira, s\xF3 ainda est\xE1 carregando...",
    "Mexendo na tela de cria\xE7\xE3o de personagem...",
    "S\xF3 um momento, estou encontrando o meme certo...",
    "Pressionando 'A' para continuar...",
    "Pastoreando gatos digitais...",
    "Polindo os pixels...",
    "Encontrando um trocadilho adequado para a tela de carregamento...",
    "Distraindo voc\xEA com esta frase espirituosa...",
    "Quase l\xE1... provavelmente...",
    "Nossos hamsters est\xE3o trabalhando o mais r\xE1pido que podem...",
    "Dando um tapinha na cabe\xE7a do Cloudy...",
    "Acariciando o gato...",
    "Dando um Rickroll no meu chefe...",
    "Never gonna give you up, never gonna let you down...",
    "Tocando o baixo...",
    "Provando as amoras...",
    "Estou indo longe, estou indo pela velocidade...",
    "Isso \xE9 vida real? Ou \xE9 apenas fantasia?...",
    "Tenho um bom pressentimento sobre isso...",
    "Cutucando o urso...",
    "Fazendo pesquisa sobre os \xFAltimos memes...",
    "Descobrindo como tornar isso mais espirituoso...",
    "Hmmm... deixe-me pensar...",
    "O que voc\xEA chama de um peixe sem olhos? Um pxe...",
    "Por que o computador foi \xE0 terapia? Porque tinha muitos bytes...",
    "Por que programadores n\xE3o gostam da natureza? Porque tem muitos bugs...",
    "Por que programadores preferem o modo escuro? Porque a luz atrai bugs...",
    "Por que o desenvolvedor faliu? Porque usou todo o seu cache...",
    "O que voc\xEA pode fazer com um l\xE1pis quebrado? Nada, ele n\xE3o tem ponta...",
    "Aplicando manuten\xE7\xE3o percussiva...",
    "Procurando a orienta\xE7\xE3o correta do USB...",
    "Garantindo que a fuma\xE7a m\xE1gica permane\xE7a dentro dos fios...",
    "Tentando sair do Vim...",
    "Girando a roda do hamster...",
    "Isso n\xE3o \xE9 um bug, \xE9 um recurso n\xE3o documentado...",
    "Engajar.",
    "Eu voltarei... com uma resposta.",
    "Meu outro processo \xE9 uma TARDIS...",
    "Comungando com o esp\xEDrito da m\xE1quina...",
    "Deixando os pensamentos marinarem...",
    "Lembrei agora onde coloquei minhas chaves...",
    "Ponderando a orbe...",
    "Eu vi coisas que voc\xEAs n\xE3o acreditariam... como um usu\xE1rio que l\xEA mensagens de carregamento.",
    "Iniciando olhar pensativo...",
    "Qual \xE9 o lanche favorito de um computador? Microchips.",
    "Por que desenvolvedores Java usam \xF3culos? Porque eles n\xE3o C#.",
    "Carregando o laser... pew pew!",
    "Dividindo por zero... s\xF3 brincando!",
    "Procurando por um supervisor adulto... digo, processando.",
    "Fazendo bip boop.",
    "Buffering... porque at\xE9 as IAs precisam de um momento.",
    "Entrela\xE7ando part\xEDculas qu\xE2nticas para uma resposta mais r\xE1pida...",
    "Polindo o cromo... nos algoritmos.",
    "Voc\xEA n\xE3o est\xE1 entretido? (Trabalhando nisso!)",
    "Invocando os gremlins do c\xF3digo... para ajudar, \xE9 claro.",
    "S\xF3 esperando o som da conex\xE3o discada terminar...",
    "Recalibrando o humor\xF4metro.",
    "Minha outra tela de carregamento \xE9 ainda mais engra\xE7ada.",
    "Tenho quase certeza que tem um gato andando no teclado em algum lugar...",
    "Aumentando... Aumentando... Ainda carregando.",
    "N\xE3o \xE9 um bug, \xE9 um recurso... desta tela de carregamento.",
    "Voc\xEA j\xE1 tentou desligar e ligar de novo? (A tela de carregamento, n\xE3o eu.)",
    "Construindo pilares adicionais..."
  ],
  // ============================================================================
  // Extension Settings Input
  // ============================================================================
  "Enter value...": "Digite o valor...",
  "Enter sensitive value...": "Digite o valor sens\xEDvel...",
  "Press Enter to submit, Escape to cancel": "Pressione Enter para enviar, Escape para cancelar",
  // ============================================================================
  // Command Migration Tool
  // ============================================================================
  "Markdown file already exists: {{filename}}": "Arquivo Markdown j\xE1 existe: {{filename}}",
  "TOML Command Format Deprecation Notice": "Aviso de Obsolesc\xEAncia do Formato de Comando TOML",
  "Found {{count}} command file(s) in TOML format:": "Encontrado(s) {{count}} arquivo(s) de comando no formato TOML:",
  "The TOML format for commands is being deprecated in favor of Markdown format.": "O formato TOML para comandos est\xE1 sendo descontinuado em favor do formato Markdown.",
  "Markdown format is more readable and easier to edit.": "O formato Markdown \xE9 mais leg\xEDvel e f\xE1cil de editar.",
  "You can migrate these files automatically using:": "Voc\xEA pode migrar esses arquivos automaticamente usando:",
  "Or manually convert each file:": "Ou converter manualmente cada arquivo:",
  'TOML: prompt = "..." / description = "..."': 'TOML: prompt = "..." / description = "..."',
  "Markdown: YAML frontmatter + content": "Markdown: YAML frontmatter + conte\xFAdo",
  "The migration tool will:": "A ferramenta de migra\xE7\xE3o ir\xE1:",
  "Convert TOML files to Markdown": "Converter arquivos TOML para Markdown",
  "Create backups of original files": "Criar backups dos arquivos originais",
  "Preserve all command functionality": "Preservar toda a funcionalidade do comando",
  "TOML format will continue to work for now, but migration is recommended.": "O formato TOML continuar\xE1 a funcionar por enquanto, mas a migra\xE7\xE3o \xE9 recomendada.",
  // ============================================================================
  // Extensions - Explore Command
  // ============================================================================
  "Open extensions page in your browser": "Abrir p\xE1gina de extens\xF5es no seu navegador",
  "Unknown extensions source: {{source}}.": "Fonte de extens\xF5es desconhecida: {{source}}.",
  "Would open extensions page in your browser: {{url}} (skipped in test environment)": "Abriria a p\xE1gina de extens\xF5es no seu navegador: {{url}} (pulado no ambiente de teste)",
  "View available extensions at {{url}}": "Ver extens\xF5es dispon\xEDveis em {{url}}",
  "Opening extensions page in your browser: {{url}}": "Abrindo p\xE1gina de extens\xF5es no seu navegador: {{url}}",
  "Failed to open browser. Check out the extensions gallery at {{url}}": "Falha ao abrir o navegador. Confira a galeria de extens\xF5es em {{url}}",
  // ============================================================================
  // Custom API Key Configuration
  // ============================================================================
  "You can configure your API key and models in settings.json": "Voc\xEA pode configurar sua API Key e modelos em settings.json",
  "Refer to the documentation for setup instructions": "Consulte a documenta\xE7\xE3o para instru\xE7\xF5es de configura\xE7\xE3o",
  // ============================================================================
  // Coding Plan Authentication
  // ============================================================================
  "API key cannot be empty.": "A API Key n\xE3o pode estar vazia.",
  "You can get your Coding Plan API key here": "Voc\xEA pode obter sua API Key do Coding Plan aqui",
  "Failed to update Coding Plan configuration: {{message}}": "Falha ao atualizar a configura\xE7\xE3o do Coding Plan: {{message}}",
  // ============================================================================
  // Auth Dialog - View Titles and Labels
  // ============================================================================
  "Coding Plan": "Coding Plan",
  Custom: "Personalizado",
  "Select Region for Coding Plan": "Selecionar regi\xE3o do Coding Plan",
  "Choose based on where your account is registered": "Escolha com base em onde sua conta est\xE1 registrada",
  "Enter Coding Plan API Key": "Inserir API Key do Coding Plan",
  // ============================================================================
  // Coding Plan International Updates
  // ============================================================================
  "New model configurations are available for {{region}}. Update now?": "Novas configura\xE7\xF5es de modelo est\xE3o dispon\xEDveis para o {{region}}. Atualizar agora?",
  '{{region}} configuration updated successfully. Model switched to "{{model}}".': 'Configura\xE7\xE3o do {{region}} atualizada com sucesso. Modelo alterado para "{{model}}".',
  // ============================================================================
  // Context Usage Component
  // ============================================================================
  "Context Usage": "Uso do Contexto",
  "% used": "% usado",
  "% context used": "% contexto usado",
  "Context exceeds limit! Use /compress or /clear to reduce.": "Contexto excede o limite! Use /compress ou /clear para reduzir.",
  "No API response yet. Send a message to see actual usage.": "Ainda n\xE3o h\xE1 resposta da API. Envie uma mensagem para ver o uso real.",
  "Estimated pre-conversation overhead": "Sobrecarga estimada pr\xE9-conversa",
  "Context window": "Janela de Contexto",
  tokens: "tokens",
  Used: "Usado",
  Free: "Livre",
  "Autocompact buffer": "Buffer de autocompacta\xE7\xE3o",
  "Usage by category": "Uso por categoria",
  "System prompt": "Prompt do sistema",
  "Built-in tools": "Ferramentas integradas",
  "MCP tools": "MCP tools",
  "Memory files": "Arquivos de mem\xF3ria",
  Skills: "Habilidades",
  Messages: "Mensagens",
  "Run /context detail for per-item breakdown.": "Execute /context detail para detalhamento por item.",
  active: "ativo",
  "body loaded": "conte\xFAdo carregado",
  memory: "mem\xF3ria",
  "{{region}} configuration updated successfully.": "Configura\xE7\xE3o do {{region}} atualizada com sucesso.",
  "Authenticated successfully with {{region}}. API key and model configs saved to settings.json.": "Autenticado com sucesso com {{region}}. API Key e configura\xE7\xF5es de modelo salvas em settings.json.",
  "Tip: Use /model to switch between available Coding Plan models.": "Dica: Use /model para alternar entre os modelos dispon\xEDveis do Coding Plan.",
  "Type something...": "Digite algo...",
  Submit: "Enviar",
  "Submit answers": "Enviar respostas",
  Cancel: "Cancelar",
  "Your answers:": "Suas respostas:",
  "(not answered)": "(n\xE3o respondido)",
  "Ready to submit your answers?": "Pronto para enviar suas respostas?",
  "\u2191/\u2193: Navigate | \u2190/\u2192: Switch tabs | Enter: Select": "\u2191/\u2193: Navegar | \u2190/\u2192: Alternar abas | Enter: Selecionar",
  "\u2191/\u2193: Navigate | Enter: Select | Esc: Cancel": "\u2191/\u2193: Navegar | Enter: Selecionar | Esc: Cancelar",
  "Authenticate using Qwen OAuth": "Autenticar usando Qwen OAuth",
  "Authenticate using Alibaba Cloud Coding Plan": "Autenticar usando Alibaba Cloud Coding Plan",
  "Region for Coding Plan (china/global)": "Regi\xE3o para Coding Plan (china/global)",
  "API key for Coding Plan": "API Key para Coding Plan",
  "Show current authentication status": "Mostrar status atual de autentica\xE7\xE3o",
  "Authentication completed successfully.": "Autentica\xE7\xE3o conclu\xEDda com sucesso.",
  "Starting Qwen OAuth authentication...": "Iniciando autentica\xE7\xE3o Qwen OAuth...",
  "Successfully authenticated with Qwen OAuth.": "Autenticado com sucesso via Qwen OAuth.",
  "Failed to authenticate with Qwen OAuth: {{error}}": "Falha ao autenticar com Qwen OAuth: {{error}}",
  "Processing Alibaba Cloud Coding Plan authentication...": "Processando autentica\xE7\xE3o Alibaba Cloud Coding Plan...",
  "Successfully authenticated with Alibaba Cloud Coding Plan.": "Autenticado com sucesso via Alibaba Cloud Coding Plan.",
  "Failed to authenticate with Coding Plan: {{error}}": "Falha ao autenticar com Coding Plan: {{error}}",
  "\u963F\u91CC\u4E91\u767E\u70BC (aliyun.com)": "\u963F\u91CC\u4E91\u767E\u70BC (aliyun.com)",
  Global: "Global",
  "Alibaba Cloud (alibabacloud.com)": "Alibaba Cloud (alibabacloud.com)",
  "Select region for Coding Plan:": "Selecione a regi\xE3o para Coding Plan:",
  "Enter your Coding Plan API key: ": "Insira sua API Key do Coding Plan: ",
  "Select authentication method:": "Selecione o m\xE9todo de autentica\xE7\xE3o:",
  "\n=== Authentication Status ===\n": "\n=== Status de Autentica\xE7\xE3o ===\n",
  "\u26A0\uFE0F  No authentication method configured.\n": "\u26A0\uFE0F  Nenhum m\xE9todo de autentica\xE7\xE3o configurado.\n",
  "Run one of the following commands to get started:\n": "Execute um dos seguintes comandos para come\xE7ar:\n",
  "  qwen auth qwen-oauth     - Authenticate with Qwen OAuth (discontinued)": "  qwen auth qwen-oauth     - Autenticar com Qwen OAuth (descontinuado)",
  "Or simply run:": "Ou simplesmente execute:",
  "  qwen auth                - Interactive authentication setup\n": "  qwen auth                - Configura\xE7\xE3o interativa de autentica\xE7\xE3o\n",
  "\u2713 Authentication Method: Qwen OAuth": "\u2713 M\xE9todo de autentica\xE7\xE3o: Qwen OAuth",
  "  Type: Free tier (discontinued 2026-04-15)": "  Tipo: N\xEDvel gratuito (descontinuado 2026-04-15)",
  "  Limit: No longer available": "  Limite: N\xE3o mais dispon\xEDvel",
  "Qwen OAuth free tier was discontinued on 2026-04-15. Run /auth to switch to Coding Plan, OpenRouter, Fireworks AI, or another provider.": "O n\xEDvel gratuito do Qwen OAuth foi descontinuado em 2026-04-15. Execute /auth para mudar para Coding Plan, OpenRouter, Fireworks AI ou outro provedor.",
  "\u2713 Authentication Method: Alibaba Cloud Coding Plan": "\u2713 M\xE9todo de autentica\xE7\xE3o: Alibaba Cloud Coding Plan",
  "Global - Alibaba Cloud": "Global - Alibaba Cloud",
  "  Region: {{region}}": "  Regi\xE3o: {{region}}",
  "  Current Model: {{model}}": "  Modelo atual: {{model}}",
  "  Config Version: {{version}}": "  Vers\xE3o da configura\xE7\xE3o: {{version}}",
  "  Status: API key configured\n": "  Status: API Key configurada\n",
  "\u26A0\uFE0F  Authentication Method: Alibaba Cloud Coding Plan (Incomplete)": "\u26A0\uFE0F  M\xE9todo de autentica\xE7\xE3o: Alibaba Cloud Coding Plan (Incompleto)",
  "  Issue: API key not found in environment or settings\n": "  Problema: API Key n\xE3o encontrada no ambiente ou configura\xE7\xF5es\n",
  "  Run `qwen auth coding-plan` to re-configure.\n": "  Execute `qwen auth coding-plan` para reconfigurar.\n",
  "\u2713 Authentication Method: {{type}}": "\u2713 M\xE9todo de autentica\xE7\xE3o: {{type}}",
  "  Status: Configured\n": "  Status: Configurado\n",
  "Failed to check authentication status: {{error}}": "Falha ao verificar status de autentica\xE7\xE3o: {{error}}",
  "Select an option:": "Selecione uma op\xE7\xE3o:",
  "Raw mode not available. Please run in an interactive terminal.": "Modo raw n\xE3o dispon\xEDvel. Execute em um terminal interativo.",
  "(Use \u2191 \u2193 arrows to navigate, Enter to select, Ctrl+C to exit)\n": "(Use \u2191 \u2193 para navegar, Enter para selecionar, Ctrl+C para sair)\n",
  "Hide tool output and thinking for a cleaner view (toggle with Ctrl+O).": "Ocultar sa\xEDda da ferramenta e racioc\xEDnio para uma visualiza\xE7\xE3o mais limpa (alternar com Ctrl+O).",
  "Press Ctrl+O to show full tool output": "Pressione Ctrl+O para exibir a sa\xEDda completa da ferramenta",
  "Switch to plan mode or exit plan mode": "Alternar para o modo de planejamento ou sair do modo de planejamento",
  "Exited plan mode. Previous approval mode restored.": "Modo de planejamento encerrado. Modo de aprova\xE7\xE3o anterior restaurado.",
  "Enabled plan mode. The agent will analyze and plan without executing tools.": "Modo de planejamento ativado. O agente analisar\xE1 e planejar\xE1 sem executar ferramentas.",
  'Already in plan mode. Use "/plan exit" to exit plan mode.': 'J\xE1 est\xE1 no modo de planejamento. Use "/plan exit" para sair do modo de planejamento.',
  'Not in plan mode. Use "/plan" to enter plan mode first.': 'N\xE3o est\xE1 no modo de planejamento. Use "/plan" para entrar primeiro no modo de planejamento.',
  "Set up Qwen Code's status line UI": "Configurar a interface da barra de status do Qwen Code",
  // === Core: added from PR #3328 ===
  "Open the memory manager.": "Abrir o gerenciador de mem\xF3ria.",
  "Save a durable memory to the memory system.": "Salvar uma mem\xF3ria dur\xE1vel no sistema de mem\xF3ria.",
  prompts: "Prompts (sugest\xF5es)",
  "Open MCP management dialog": "Abrir di\xE1logo de gerenciamento MCP",
  "Manage extension settings": "Gerenciar configura\xE7\xF5es da extens\xE3o",
  "Manage Extensions": "Gerenciar extens\xF5es",
  "Extension Details": "Detalhes da extens\xE3o",
  "View Extension": "Ver extens\xE3o",
  "Update Extension": "Atualizar extens\xE3o",
  "Disable Extension": "Desativar extens\xE3o",
  "Enable Extension": "Ativar extens\xE3o",
  "Uninstall Extension": "Desinstalar extens\xE3o",
  "Select Scope": "Selecionar escopo",
  "User Scope": "Escopo do usu\xE1rio",
  "Workspace Scope": "Escopo do workspace",
  "No extensions found.": "Nenhuma extens\xE3o encontrada.",
  'Are you sure you want to uninstall extension "{{name}}"?': 'Tem certeza de que deseja desinstalar a extens\xE3o "{{name}}"?',
  "This action cannot be undone.": "Esta a\xE7\xE3o n\xE3o pode ser desfeita.",
  'Extension "{{name}}" updated successfully.': 'Extens\xE3o "{{name}}" atualizada com sucesso.',
  "Name:": "Nome:",
  "MCP Servers:": "MCP Servers:",
  "Settings:": "Configura\xE7\xF5es:",
  "View Details": "Ver detalhes",
  "Update failed:": "Falha na atualiza\xE7\xE3o:",
  "Updating {{name}}...": "Atualizando {{name}}...",
  "Update complete!": "Atualiza\xE7\xE3o conclu\xEDda!",
  "User (global)": "Usu\xE1rio (global)",
  "Workspace (project-specific)": "Workspace (espec\xEDfico do projeto)",
  'Disable "{{name}}" - Select Scope': 'Desativar "{{name}}" - selecionar escopo',
  'Enable "{{name}}" - Select Scope': 'Ativar "{{name}}" - selecionar escopo',
  "No extension selected": "Nenhuma extens\xE3o selecionada",
  "{{count}} extensions installed": "{{count}} extens\xF5es instaladas",
  "up to date": "atualizada",
  "update available": "atualiza\xE7\xE3o dispon\xEDvel",
  "checking...": "verificando...",
  "not updatable": "n\xE3o atualiz\xE1vel",
  "Ask a quick side question without affecting the main conversation": "Fazer uma pergunta r\xE1pida paralela sem afetar a conversa principal",
  "Manage Arena sessions": "Gerenciar sess\xF5es da Arena",
  "Start an Arena session with multiple models competing on the same task": "Iniciar uma sess\xE3o da Arena com v\xE1rios modelos competindo na mesma tarefa",
  "Stop the current Arena session": "Parar a sess\xE3o atual da Arena",
  "Show the current Arena session status": "Mostrar o status da sess\xE3o atual da Arena",
  "Select a model result and merge its diff into the current workspace": "Selecionar o resultado de um modelo e mesclar seu diff ao workspace atual",
  "No running Arena session found.": "Nenhuma sess\xE3o Arena em execu\xE7\xE3o encontrada.",
  "No Arena session found. Start one with /arena start.": "Nenhuma sess\xE3o Arena encontrada. Inicie uma com /arena start.",
  "Arena session is still running. Wait for it to complete or use /arena stop first.": "A sess\xE3o Arena ainda est\xE1 em execu\xE7\xE3o. Aguarde a conclus\xE3o ou use /arena stop primeiro.",
  "No successful agent results to select from. All agents failed or were cancelled.": "Nenhum resultado de agente bem-sucedido para selecionar. Todos os agentes falharam ou foram cancelados.",
  "Use /arena stop to end the session.": "Use /arena stop para encerrar a sess\xE3o.",
  'No idle agent found matching "{{name}}".': 'Nenhum agente ocioso encontrado correspondendo a "{{name}}".',
  "Failed to apply changes from {{label}}: {{error}}": "Falha ao aplicar altera\xE7\xF5es de {{label}}: {{error}}",
  "Applied changes from {{label}} to workspace. Arena session complete.": "Altera\xE7\xF5es de {{label}} aplicadas ao workspace. Sess\xE3o Arena conclu\xEDda.",
  "Discard all Arena results and clean up worktrees?": "Descartar todos os resultados da Arena e limpar as \xE1rvores de trabalho?",
  "Arena results discarded. All worktrees cleaned up.": "Resultados da Arena descartados. Todas as \xE1rvores de trabalho foram limpas.",
  "Arena is not supported in non-interactive mode. Use interactive mode to start an Arena session.": "Arena n\xE3o \xE9 suportado no modo n\xE3o interativo. Use o modo interativo para iniciar uma sess\xE3o Arena.",
  "Arena is not supported in non-interactive mode. Use interactive mode to stop an Arena session.": "Arena n\xE3o \xE9 suportado no modo n\xE3o interativo. Use o modo interativo para parar uma sess\xE3o Arena.",
  "Arena is not supported in non-interactive mode.": "Arena n\xE3o \xE9 suportado no modo n\xE3o interativo.",
  "An Arena session exists. Use /arena stop or /arena select to end it before starting a new one.": "J\xE1 existe uma sess\xE3o Arena. Use /arena stop ou /arena select para encerr\xE1-la antes de iniciar uma nova.",
  "Usage: /arena start --models model1,model2 <task>": "Uso: /arena start --models model1,model2 <tarefa>",
  "Models to compete (required, at least 2)": "Modelos para competir (obrigat\xF3rio, pelo menos 2)",
  "Format: authType:modelId or just modelId": "Formato: authType:modelId ou apenas modelId",
  "Arena requires at least 2 models. Use --models model1,model2 to specify.": "Arena requer pelo menos 2 modelos. Use --models model1,model2 para especificar.",
  'Arena started with {{count}} agents on task: "{{task}}"\nModels:\n{{modelList}}': 'Arena iniciada com {{count}} agentes na tarefa: "{{task}}"\nModelos:\n{{modelList}}',
  "Arena panes are running in tmux. Attach with: `{{command}}`": "Os pain\xE9is Arena est\xE3o em execu\xE7\xE3o no tmux. Anexar com: `{{command}}`",
  "[{{label}}] failed: {{error}}": "[{{label}}] falhou: {{error}}",
  "Loading suggestions...": "Carregando sugest\xF5es...",
  'Show context window usage breakdown. Use "/context detail" for per-item breakdown.': 'Mostrar o detalhamento do uso da janela de contexto. Use "/context detail" para ver o detalhamento por item.',
  "Show per-item context usage breakdown.": "Mostrar o detalhamento do uso de contexto por item.",
  // === Missing key backfill ===
  "\u2191 to manage attachments": "\u2191 para gerenciar anexos",
  "\u2190 \u2192 select, Delete to remove, \u2193 to exit": "\u2190 \u2192 selecionar, Delete para remover, \u2193 para sair",
  "Attachments: ": "Anexos: ",
  "(tab to cycle)": "(Tab para alternar)",
  "Updating...": "Atualizando...",
  Unknown: "Desconhecido",
  Error: "Erro",
  "Version:": "Vers\xE3o:",
  "Use '/extensions install' to install your first extension.": "Use '/extensions install' para instalar sua primeira extens\xE3o.",
  "The name of the extension to update.": "O nome da extens\xE3o a ser atualizada.",
  "Path:": "Caminho:",
  "Type:": "Tipo:",
  "Release tag:": "Tag de lan\xE7amento:",
  "Enabled (User):": "Ativado (usu\xE1rio):",
  "Enabled (Workspace):": "Ativado (workspace):",
  "Context files:": "Arquivos de contexto:",
  "Skills:": "Habilidades:",
  "Agents:": "Agentes:",
  "MCP servers:": "MCP servers:",
  "Press c to copy the authorization URL to your clipboard.": "Pressione c para copiar a URL de autoriza\xE7\xE3o para a \xE1rea de transfer\xEAncia.",
  "Copy request sent to your terminal. If paste is empty, copy the URL above manually.": "Solicita\xE7\xE3o de c\xF3pia enviada ao terminal. Se a colagem estiver vazia, copie manualmente a URL acima.",
  "Cannot write to terminal \u2014 copy the URL above manually.": "N\xE3o foi poss\xEDvel escrever no terminal \u2014 copie manualmente a URL acima.",
  "You can switch permission mode quickly with Tab or /approval-mode.": "Voc\xEA pode alternar rapidamente o modo de permiss\xE3o com Tab ou /approval-mode.",
  "Retrying in {{seconds}} seconds\u2026 (attempt {{attempt}}/{{maxRetries}})": "Tentando novamente em {{seconds}} segundos\u2026 (tentativa {{attempt}}/{{maxRetries}})",
  "Press Ctrl+Y to retry": "Pressione Ctrl+Y para tentar novamente",
  "No failed request to retry.": "Nenhuma solicita\xE7\xE3o com falha para repetir.",
  "to retry last request": "para repetir a \xFAltima solicita\xE7\xE3o",
  'Invalid API key. Coding Plan API keys start with "sk-sp-". Please check.': 'API Key inv\xE1lida. As API Keys do Coding Plan come\xE7am com "sk-sp-". Verifique.',
  "Lock release warning": "Aviso de libera\xE7\xE3o de bloqueio",
  "Metadata write warning": "Aviso de grava\xE7\xE3o de metadados",
  "Subsequent dreams may be skipped as locked until the next session's staleness sweep cleans the file.": "Dreams posteriores podem ser ignorados como bloqueados at\xE9 que a pr\xF3xima varredura de sess\xF5es obsoletas limpe o arquivo.",
  "The scheduler gate did not see this dream's timestamp; the next dream cycle may re-fire sooner than usual.": "O gate do agendador n\xE3o viu o timestamp deste dream; o pr\xF3ximo ciclo de dream pode disparar novamente antes do normal.",
  // === History collapse/expand commands ===
  "Set history to collapse by default when resuming a session": "Set history to collapse by default when resuming a session",
  "Set history to expand by default when resuming a session": "Set history to expand by default when resuming a session",
  "Expand the currently collapsed history transcript": "Expand the currently collapsed history transcript",
  "Control history display preferences and visibility": "Control history display preferences and visibility",
  "History will be collapsed by default for future resumed sessions.": "History will be collapsed by default for future resumed sessions.",
  "History will be expanded by default for future resumed sessions.": "History will be expanded by default for future resumed sessions.",
  "History is already expanded in this session.": "History is already expanded in this session.",
  "Usage: /history collapse-on-resume|expand-on-resume|expand-now": "Usage: /history collapse-on-resume|expand-on-resume|expand-now",
  "History collapsed: {{n}} messages hidden. Use /history expand-now to show.": "Hist\xF3rico recolhido: {{n}} mensagens ocultas. Use /history expand-now para mostrar.",
  // === Same-as-English optimization ===
  "(workspace)": "(espa\xE7o de trabalho)",
  "Ref:": "Refer\xEAncia:",
  Runtime: "Tempo de execu\xE7\xE3o",
  Status: "Estado",
  "Status:": "Estado:",
  Use: "Uso",
  "\u4E2D\u56FD (China)": "China",
  "\u4E2D\u56FD (China) - \u963F\u91CC\u4E91\u767E\u70BC": "China - \u963F\u91CC\u4E91\u767E\u70BC",
  // Stats Dashboard — Category 2
  "Activity Heatmap": "Mapa de Atividade",
  Less: "Menos",
  More: "Mais",
  Sessions: "Sess\xF5es",
  Duration: "Dura\xE7\xE3o",
  Projects: "Projetos",
  "Loading stats...": "Carregando estat\xEDsticas...",
  "(no data)": "(sem dados)",
  d: "d",
  h: "h",
  m: "m",
  Input: "Entrada",
  Models: "Modelos",
  "All time": "Todo o per\xEDodo",
  "Last 7 days": "\xDAltimos 7 dias",
  "Last 30 days": "\xDAltimos 30 dias",
  "Show usage statistics dashboard.": "Exibir painel de estat\xEDsticas de uso.",
  // Stats Dashboard — keyboard hints (not translated)
  "tab \xB7 esc": "tab \xB7 esc",
  "tab \xB7 r dates \xB7 \u2190\u2192 month \xB7 esc": "tab \xB7 r dates \xB7 \u2190\u2192 month \xB7 esc",
  "tab \xB7 r dates \xB7 esc": "tab \xB7 r dates \xB7 esc",
  // Stats Dashboard — missing labels
  "API Requests": "Requisi\xE7\xF5es API",
  "Tool Calls": "Chamadas de Ferramenta",
  "Success rate": "Taxa de sucesso",
  "Code Changes": "Altera\xE7\xF5es de C\xF3digo",
  Tool: "Ferramenta",
  reqs: "reqs",
  in: "ent.",
  out: "sa\xEDda",
  "In/Out": "Ent/Sa\xEDda"
};
export {
  pt_default as default
};
/**
 * @license
 * Copyright 2025 Qwen
 * SPDX-License-Identifier: Apache-2.0
 */
