// @qwen-code/qwen-code/chunks/src-KUV5J5Q2.js
// Force strict mode and setup for ESM
"use strict";
import {
  AGENT_WORKTREE_HEX_LENGTH,
  AGENT_WORKTREE_PREFIX,
  AGENT_WORKTREE_SLUG_PATTERN,
  ALIBABA_PROVIDERS,
  ALL_PROVIDERS,
  APPROVAL_MODES,
  APPROVAL_MODE_INFO,
  ARENA_MAX_AGENTS,
  AUTO_MEMORY_SCHEMA_VERSION,
  AUTO_MEMORY_TYPES,
  AUTO_MODE_DENIAL_LIMITS,
  AUTO_SKILL_THRESHOLD,
  AbortError,
  AgentCore,
  AgentEventEmitter,
  AgentEventType,
  AgentHeadless,
  AgentInteractive,
  AgentStatus,
  AgentTerminateMode,
  ApprovalMode,
  ArenaAgentClient,
  ArenaEventEmitter,
  ArenaEventType,
  ArenaManager,
  ArenaSessionStatus,
  AsyncMessageQueue,
  AuthProviderType,
  BACKGROUND_AGENT_CONCURRENCY_ENV,
  BASELINE_COMMIT_MESSAGE,
  BTW_MAX_INPUT_LENGTH,
  BUNDLE_CHUNK_DIR,
  BackgroundAgentResumeService,
  BackgroundShellRegistry,
  BackgroundTaskRegistry,
  BudgetExhaustedError,
  BuiltinAgentRegistry,
  CODE_ACTION_KIND_LABELS,
  CODING_PLAN_CHINA_BASE_URL,
  CODING_PLAN_ENV_KEY,
  CODING_PLAN_GLOBAL_BASE_URL,
  COMMAND_SUBSTITUTION_WARNING,
  CUSTOM_API_KEY_ENV_PREFIX,
  ChatRecordingService,
  CloseDiffRequestSchema,
  ColorMode,
  CompressionStatus,
  ConditionalRulesRegistry,
  Config,
  ContextState,
  CoreToolScheduler,
  DEFAULT_AUTO_DREAM_MIN_HOURS,
  DEFAULT_AUTO_DREAM_MIN_SESSIONS,
  DEFAULT_BACKGROUND_AGENT_CONTINUATION_MESSAGE,
  DEFAULT_BUILTIN_SUBAGENT_TYPE,
  DEFAULT_LSP_COMMAND_CHECK_TIMEOUT_MS,
  DEFAULT_LSP_DOCUMENT_OPEN_DELAY_MS,
  DEFAULT_LSP_DOCUMENT_RETRY_DELAY_MS,
  DEFAULT_LSP_MAX_RESTARTS,
  DEFAULT_LSP_REQUEST_TIMEOUT_MS,
  DEFAULT_LSP_SOCKET_MAX_RETRY_DELAY_MS,
  DEFAULT_LSP_SOCKET_RETRY_DELAY_MS,
  DEFAULT_LSP_STARTUP_TIMEOUT_MS,
  DEFAULT_LSP_WARMUP_DELAY_MS,
  DEFAULT_LSP_WORKSPACE_SYMBOL_WARMUP_DELAY_MS,
  DEFAULT_MAX_CONCURRENT_BACKGROUND_AGENTS,
  DEFAULT_MAX_INLINE_MEDIA_BYTES,
  DEFAULT_STOP_HOOK_BLOCK_CAP,
  DEFAULT_TOOL_OUTPUT_BATCH_BUDGET,
  DEFAULT_TOOL_RESULTS_TOTAL_CHARS_THRESHOLD,
  DEFAULT_TRUNCATE_TOOL_OUTPUT_LINES,
  DEFAULT_TRUNCATE_TOOL_OUTPUT_THRESHOLD,
  DIAGNOSTIC_SEVERITY_LABELS,
  DISPLAY_MODE,
  DREAM_TASK_TYPE,
  DiscoveredTool,
  EXCLUDED_TOOLS_FOR_SUBAGENTS,
  EXTENSIONS_CONFIG_FILENAME,
  EXTENSIONS_DIRECTORY_NAME,
  EXTENSION_SETTINGS_FILENAME,
  EXTERNAL_MESSAGE_PREFIX,
  EXTRACT_TASK_TYPE,
  ExtensionManager,
  ExtensionPreferencesStore,
  ExtensionSettingScope,
  ExtensionUpdateState,
  FILE_HISTORY_DIR,
  FORK_SUBAGENT_TYPE,
  FileDiscoveryService,
  FileEncoding,
  FileHistoryService,
  FileReadCache,
  FileSchema,
  FileSearchFactory,
  GOAL_HOOK_TIMEOUT_MS,
  GOAL_HOOK_TIMEOUT_SECONDS,
  GeminiChat,
  GeminiClient,
  GeminiEventType,
  GitIgnoreParser,
  GitWorktreeService,
  HookRegistry,
  HookSystem,
  IDEConnectionStatus,
  IDE_DEFINITIONS,
  IDE_MAX_OPEN_FILES,
  IDE_MAX_SELECTED_TEXT_LENGTH,
  IDE_REQUEST_TIMEOUT_MS,
  IDLE_SPECULATION,
  INITIAL_FOLLOWUP_STATE,
  INSTALL_METADATA_FILENAME,
  ITermBackend,
  IdeClient,
  IdeContextNotificationSchema,
  IdeContextSchema,
  IdeContextStore,
  IdeDiffAcceptedNotificationSchema,
  IdeDiffClosedNotificationSchema,
  IdeDiffRejectedNotificationSchema,
  Ignore,
  InProcessBackend,
  InvalidMcpConfigError,
  InvalidStreamError,
  JsonFormatter,
  LITE_READ_BUF_SIZE,
  Logger,
  LspConfigLoader,
  LspConnectionFactory,
  LspResponseNormalizer,
  LspServerManager,
  MAX_CONCURRENT_BACKGROUND_AGENTS,
  MAX_CONCURRENT_MONITORS,
  MAX_DIFF_SIZE_BYTES,
  MAX_FILES,
  MAX_FILES_FOR_DETAILS,
  MAX_GOAL_ITERATIONS,
  MAX_LINES_PER_FILE,
  MAX_NOTIFICATION_OUTPUT_TAIL_BYTES,
  MAX_RETAINED_AGENT_FIELD_CHARS,
  MAX_RETAINED_ANSI_OUTPUT_LINES,
  MAX_RETAINED_FILE_CONTENT_CHARS,
  MAX_RETAINED_FILE_DIFF_CHARS,
  MAX_RETAINED_SNAPSHOTS,
  MAX_RETAINED_TERMINAL_AGENTS,
  MAX_RETAINED_TERMINAL_MONITORS,
  MAX_RETAINED_TERMINAL_SHELLS,
  MAX_RETAINED_TERMINAL_WORKFLOWS,
  MAX_RETAINED_TOOL_RESULT_DISPLAY_CHARS,
  MAX_SNAPSHOTS,
  MAX_TRANSCRIPT_MESSAGES,
  MCPDiscoveryState,
  MCPOAuthProvider,
  MCPOAuthTokenStorage,
  MCPServerConfig,
  MCP_BUDGET_REARM_FRACTION,
  MCP_BUDGET_WARN_FRACTION,
  MCP_DEFAULT_TIMEOUT_MSEC,
  MOCK_TOOL_GET_CONFIRMATION_DETAILS,
  MOCK_TOOL_GET_DEFAULT_PERMISSION,
  McpBudgetWouldExceedError,
  McpClient,
  McpClientManager,
  McpServerSpawnFailedError,
  McpTransportPool,
  MemoryManager,
  MessageBus,
  MessageBusType,
  MessageSenderType,
  MockModifiableTool,
  MockModifiableToolInvocation,
  MockTool,
  ModelRegistry,
  ModelsConfig,
  MonitorRegistry,
  NativeLspClient,
  NativeLspService,
  OAUTH_AUTH_URL_EVENT,
  OAUTH_DISPLAY_MESSAGE_EVENT,
  OAuthUtils,
  OPENROUTER_BASE_URL,
  OPENROUTER_ENV_KEY,
  OpenDiffRequestSchema,
  OverlayFs,
  POOLED_TRANSPORTS_DEFAULT,
  PROVIDER_METADATA_NS,
  PermissionManager,
  ProviderInstallError,
  QWEN_CODE_COMPANION_EXTENSION_NAME,
  QWEN_CODE_SIMPLE_ENV_VAR,
  REDACTED_URL_CREDENTIAL,
  REQUESTY_BASE_URL,
  REQUESTY_ENV_KEY,
  RUNTIME_SNAPSHOT_PREFIX,
  RUNTIME_STATUS_SCHEMA_VERSION,
  RipGrepTool,
  SAFE_TOOL_ALLOWLIST,
  SESSION_TITLE_MAX_LENGTH,
  SHELL_SELF_KILL_REJECTION,
  SHELL_TOOL_NAMES,
  SKILL_REVIEW_TASK_TYPE,
  STOP_HOOK_BLOCK_CAP_ENV,
  SUGGESTION_PROMPT,
  SYMBOL_KIND_LABELS,
  SYSTEM_REMINDER_CLOSE,
  SYSTEM_REMINDER_OPEN,
  SdkControlClientTransport,
  SendMessageType,
  SessionService,
  SettingScope,
  ShellExecutionService,
  SkillActivationRegistry,
  SkillError,
  SkillManager,
  SleepInhibitor,
  SourceRegistryStore,
  StandardFileSystemService,
  StreamEventType,
  SubagentError,
  SubagentErrorCode,
  SubagentManager,
  SubagentValidator,
  TERMINAL_CSI_REGEX,
  TERMINAL_OSC_REGEX,
  TERMINAL_SHIFT_DCS_REGEX,
  THIRD_PARTY_PROVIDERS,
  TOKEN_PLAN_BASE_URL,
  TOKEN_PLAN_ENV_KEY,
  TOOL_NAME_ALIASES,
  TOOL_USE_SUMMARY_SYSTEM_PROMPT,
  TeamEventEmitter,
  TeamEventType,
  TeamManager,
  TmuxBackend,
  ToolRegistry,
  TrustGateError,
  Turn,
  WORKFLOW_NAME_PATTERN,
  WORKTREES_DIR,
  WORKTREE_BRANCH_PREFIX,
  WORKTREE_SESSION_FILE,
  WorkflowRunRegistry,
  WorkspaceMcpBudget,
  WorkspaceMemoryFileTooLargeError,
  WorkspaceMemoryWriteTimeoutError,
  __resetActiveGoalStoreForTests,
  _resetCachedIdeServerHost,
  _resetRipGrepCachesForTest,
  _resetRipgrepUtilsCachesForTest,
  abortGoalForStopHookCap,
  abortSpeculation,
  acceptSpeculation,
  acquireSleepInhibitor,
  activeGoalEquals,
  aggregateUsage,
  alibabaStandardProvider,
  allowEditorTypeInSandbox,
  appendToLastTextPart,
  applyAutoModeDecision,
  applyProviderInstallPlan,
  approxBase64Bytes,
  buildAddedMcpToolsReminder,
  buildAddedSkillsReminder,
  buildApiHistoryFromConversation,
  buildAvailableSkillsReminder,
  buildBackgroundEntryLabel,
  buildBtwCacheSafeParams,
  buildBtwPrompt,
  buildDeferredToolsReminder,
  buildFallbackApproachSummary,
  buildHumanReadableRuleLabel,
  buildInstallPlan,
  buildMcpServerInstructionsReminder,
  buildModelIdContext,
  buildPermissionCheckContext,
  buildPermissionRules,
  buildProviderTemplate,
  buildShellExecWarnings,
  buildStartupContextReminder,
  buildTeammatePromptAddendum,
  canUseRipgrep,
  checkArgumentSafety,
  checkCommandPermissions,
  checkForExtensionUpdate,
  checkHasEditorType,
  checkNpmUpdate,
  clampInlineMediaPart,
  cleanSummary,
  cleanupTmpDir,
  clearActiveGoal,
  clearCacheSafeParams,
  clearGitDirCache,
  clearGoalTerminalObserver,
  clearRuntimeStatus,
  clearSessionCommits,
  clearWelcomeBackState,
  clearWorktreeSession,
  cliSource,
  cloneFromGit,
  codingPlanProvider,
  collectMemoryDiagnostics,
  collectText,
  collectToolCallIdsFromHistory,
  commandExists,
  compactStringForHistory,
  compactStringForRecording,
  compactToolResultDisplayForHistory,
  compactToolResultDisplayForRecording,
  computeModelListVersion,
  computeThresholds,
  computeUniqueBranchTitle,
  computedSource,
  connectAndDiscover,
  connectToMcpServer,
  connectionIdOf,
  convertClaudeAgentConfig,
  convertClaudePluginPackage,
  convertClaudePluginStandalone,
  convertClaudeToQwenConfig,
  convertColorToHex,
  convertToFunctionResponse,
  convertTomlToMarkdown,
  copyExtension,
  crawlCache_exports,
  createDefaultAutoMemoryExtractCursor,
  createDefaultAutoMemoryIndex,
  createDefaultAutoMemoryMetadata,
  createDenialState,
  createDuplicateProviderToolCallResponse,
  createFollowupController,
  createForkedChat,
  createGoalStopHookCallback,
  createInstructionsLoadedCallback,
  createStreamableHttpCompatibilityFetch,
  createTmpDir,
  createToolUseSummaryMessage,
  createTransport,
  customProvider,
  decodeTagName,
  dedupeToolCallsById,
  deepseekProvider,
  defaultSource,
  delay,
  deserializeSnapshots,
  detectBackend,
  detectCommandSubstitution,
  detectIdeFromEnv,
  detectLineEnding,
  detectSelfKillCommand,
  discoverMcpTools,
  discoverPlugins,
  discoverPrompts,
  discoverResources,
  discoverTools,
  doesToolInvocationMatch,
  downloadFromArchiveUrl,
  downloadFromGitHubRelease,
  downloadFromNpmRegistry,
  editorCommands,
  encodeInsightErrorMessage,
  encodeInsightProgressMessage,
  encodeInsightReadyMessage,
  encodeTagName,
  encodeTextFileContent,
  ensureAutoMemoryScaffold,
  ensureCrlfLineEndings,
  ensureMacBinarySigned,
  ensureRipgrepHealthy,
  ensureToolResultPairing,
  ensureUserAutoMemoryScaffold,
  envLayer,
  escapeShellArg,
  evaluateAutoMode,
  evaluatePermissionFlow,
  evaluatePermissionRules,
  evaluateToolCall,
  execCommand,
  executeToolCall,
  extractArchiveFile,
  extractFile,
  extractJsonStringField,
  extractLastJsonStringField,
  extractLastJsonStringFields,
  extractLastUserPrompt,
  extractShellOperations,
  extractToolFilePaths,
  fetchGitDiff,
  fetchGitDiffHunks,
  filter,
  findExistingProviderModels,
  findGitRoot,
  findProviderByCredentials,
  findProviderById,
  findReleaseAsset,
  findRepeatedDuplicateProviderToolCall,
  fireNotificationHook,
  firePermissionRequestHook,
  firePostToolUseFailureHook,
  firePostToolUseHook,
  firePreToolUseHook,
  flatMapTextParts,
  formatClassifierBlockMessage,
  formatDateForContext,
  formatDenialStateLog,
  formatLineStats,
  formatMemoryUsage,
  formatStopHookBlockingCapWarning,
  generateAgentWorktreeSlug,
  generateCustomEnvKey,
  generatePromptSuggestion,
  generateSessionRecap,
  generateToolUseId,
  generateToolUseSummary,
  getActiveGoal,
  getAllProviderBaseUrls,
  getArenaSystemReminder,
  getAutoModePermissionDeniedReason,
  getBuiltinRipgrep,
  getCacheSafeParams,
  getCommandRoot,
  getCommandRoots,
  getCompressionPrompt,
  getCoreSystemPrompt,
  getCustomSystemPrompt,
  getDefaultBaseUrlForProtocol,
  getDefaultModelIds,
  getDiffCommand,
  getDirectoryContextString,
  getEditorExecutable,
  getEffectivePermissionForConfirmation,
  getEnvContents,
  getEnvironmentContext,
  getExtensionDescription,
  getExtensionDisplayName,
  getExtensionId,
  getExternalEditorCommand,
  getFilterReason,
  getFolderStructure,
  getFunctionCalls,
  getFunctionCallsAsJson,
  getFunctionCallsFromParts,
  getFunctionCallsFromPartsAsJson,
  getGitBranch,
  getGitRepoName,
  getIdeInstaller,
  getIdeServerHost,
  getInitialChatHistory,
  getInsightPrompt,
  getLastGoalTerminal,
  getMCPDiscoveryState,
  getMCPServerPrompts,
  getMaxInlineMediaBytes,
  getMcpOAuthDialogInstruction,
  getMonitorOutputPath,
  getPlanModeSystemReminder,
  getProjectSummaryInfo,
  getProjectSummaryPrompt,
  getProviderToolCallId,
  getRateLimitErrorDetails,
  getRateLimitRetryDelayMs,
  getRecentGitStatus,
  getResponseText,
  getResponseTextFromParts,
  getResumePromptTokenCount,
  getResumeTokenCounts,
  getRuleDisplayName,
  getSavedWorkflowDirs,
  getScopedEnvContents,
  getShellAbortReasonKind,
  getShellConfiguration,
  getSpecifierKind,
  getStartupContextLength,
  getStructuredResponse,
  getStructuredResponseFromParts,
  getSubagentsRootDir,
  getThoughtText,
  getTimeRangeBounds,
  getWelcomeBackState,
  globalMemoryManager,
  hasImageParts,
  hasNetworkTransport,
  hasNonFinalTopLevelBackgroundOperator,
  hasShellSubstitution,
  hasUnsafeMonitorBackgroundOperator,
  hashMcpServerConfig,
  hashValue,
  hookEventSupportsMatcher,
  hydrateString,
  ideContextStore,
  idealabProvider,
  injectPermissionRulesIfMissing,
  invokeMcpPrompt,
  isApiError,
  isApproveOutcome,
  isAutoEditApproved,
  isBareMode,
  isBinary,
  isClaudePluginConfig,
  isCommandAllowed,
  isCommandAvailable,
  isCommandNeedsPermission,
  isDenialFallbackReason,
  isDestructiveCommand,
  isEditorAvailable,
  isEnabled,
  isGatedMcpScope,
  isGenericQuotaExceededError,
  isGitRepository,
  isImageCapable,
  isImagePart,
  isInSafeToolAllowlist,
  isInvalidArgumentError,
  isModifiableDeclarativeTool,
  isPlanModeBlocked,
  isProQuotaExceededError,
  isQwenQuotaExceededError,
  isRateLimitError,
  isSchemaDepthError,
  isScopedNpmPackage,
  isSdkMcpServerConfig,
  isSessionRuntimeActive,
  isSettledStatus,
  isStructuredError,
  isSuccessStatus,
  isSupportedArchivePath,
  isSupportedArchiveUrl,
  isSystemReminderContent,
  isTerminalEditor,
  isTerminalStatus,
  isTomlFormat,
  isToolEnabled,
  isTransientCapacityError,
  isUnattendedMode,
  isUsableImagePart,
  isValidEditorType,
  isValidGitSha,
  isValidNonThoughtTextPart,
  isValidRefName,
  isWindows,
  judgeGoal,
  layer,
  listMcpPrompts,
  listMcpResources,
  listSavedWorkflows,
  listWorkflowSnapshots,
  loadIgnoreRules,
  loadMarketplaceConfigFromSource,
  loadServerHierarchicalMemory,
  loadUsageHistory,
  makeFakeConfig,
  markDuplicateProviderToolCallResponseSent,
  matchesCommandPattern,
  matchesDomainPattern,
  matchesMcpPattern,
  matchesPathPattern,
  matchesRule,
  maybePromptForSettings,
  mcpServerRequiresOAuth,
  mcpTransportOf,
  mergeClaudeConfigs,
  metricsToUsageRecord,
  minimaxProvider,
  modelProvidersSource,
  modelRegistryKey,
  modelscopeProvider,
  modifyWithEditor,
  needsConfirmation,
  needsUtf8Bom,
  normalizeClaudeMcpServer,
  normalizeContent,
  normalizeModelToolCallIds,
  normalizeMonitorCommand,
  normalizeParts,
  normalizeProxyUrl,
  normalizeSkillPriority,
  normalizeStopHookBlockingCap,
  notifyGoalTerminal,
  openDiff,
  openRouterProvider,
  oversizedMediaPlaceholder,
  parse,
  parseAndFormatApiError,
  parseConnectionId,
  parseDeletedFromNameStatus,
  parseExtensionSourceType,
  parseGitDiff,
  parseGitHubRepoForReleases,
  parseGitNumstat,
  parseInsightMessage,
  parseInstallSource,
  parseNpmPackageSource,
  parsePositiveIntegerEnv,
  parseRule,
  parseRules,
  parseShortstat,
  parseThought,
  partListUnionToString,
  partToString,
  passesAcceptEditsFastPath,
  performVariableReplacement,
  persistPermissionOutcome,
  persistSessionUsage,
  populateMcpServerCommand,
  prependToFirstTextPart,
  promptForSetting,
  promptIdContext,
  providerMatchesCredentials,
  readAutoMemoryIndex,
  readGitHead,
  readLastJsonStringFieldSync,
  readLastJsonStringFieldsSync,
  readManyFiles,
  readPathFromWorkspace,
  readRuntimeStatus,
  readUserAutoMemoryIndex,
  readWorktreeSession,
  readWorktreeSessionMarker,
  recordAllow,
  recordBlock,
  recordFallbackApprove,
  recordGoalIteration,
  recordStartupEvent,
  recordUnavailable,
  recursivelyHydrateStrings,
  redactStructuredOutputArgsForRecording,
  redactUrlCredentials,
  registerGoalHook,
  registerSessionCommit,
  repairOrphanedToolUseTurns,
  replaceImagesWithText,
  replayUiTelemetryFromConversation,
  requestyProvider,
  resetDenialState,
  resetUtf8BomCache,
  resolveBaseUrl,
  resolveBranchName,
  resolveBundleDir,
  resolveCommandPath,
  resolveExtensionConfigLocale,
  resolveField,
  resolveGitDir,
  resolveLocalizableString,
  resolveMaxConcurrentBackgroundAgents,
  resolveMetadataKey,
  resolveModelConfig,
  resolveModelId,
  resolveNpmRegistry,
  resolveOptionalField,
  resolveOwnsModel,
  resolvePathFromEnv,
  resolvePathPattern,
  resolveProviderProtocol,
  resolveRipgrep,
  resolveSavedWorkflowScript,
  resolveStopHookBlockingCap,
  resolveToolName,
  restoreWorktreeContext,
  retryWithBackoff,
  rewritePathArgs,
  runForkedAgent,
  runRipgrep,
  runVisionBridge,
  runWithForkedChatModel,
  safeAgentId,
  safeLiteralReplace,
  sanitizeTitle,
  sanitizeToolCallResultForRecording,
  saveCacheSafeParams,
  saveWelcomeBackRestartChoice,
  saveWorkflowScript,
  selectVisionBridgeModel,
  serializeSnapshot,
  serializeTerminalToObject,
  serializeTerminalToText,
  sessionIdContext,
  setActiveGoal,
  setGoalTerminalObserver,
  setLastGoalTerminal,
  setMCPDiscoveryState,
  setStartupEventSink,
  settingsSource,
  shouldDefaultToNodePty,
  shouldFallback,
  shouldFilterSuggestion,
  shouldFirePermissionDeniedForAutoMode,
  shouldForceAutoModeReviewForAllow,
  shouldRunAutoModeForCall,
  shouldRunVisionBridge,
  shouldShowStep,
  sleepInhibitor,
  splitCommands,
  splitCompoundCommand,
  splitConditionalSkills,
  splitImageParts,
  startSpeculation,
  stringify,
  stripAnsiAndControl,
  stripRuntimeSnapshotPrefix,
  stripShellWrapper,
  stripStartupContext,
  stripTerminalControlSequences,
  stripTrailingBackgroundAmp,
  subagentGenerator,
  substituteHookVariables,
  summarizeUnifiedDiff,
  templateString,
  toSnapshot,
  tokenPlanProvider,
  toolMatchesRuleToolName,
  truncateJson,
  tryGenerateSessionTitle,
  unescapeJsonString,
  unregisterGoalHook,
  updateSetting,
  validateName,
  validateVariables,
  validateWorkflowName,
  watchRepoBranch,
  worktreeBranchForSlug,
  writeRuntimeStatus,
  writeWorkflowSnapshot,
  writeWorkspaceContextFile,
  writeWorktreeSession,
  writeWorktreeSessionMarker,
  zaiProvider
} from "./chunk-YAINVQ3Q.js";
import "./chunk-K5PGHDBN.js";
import "./chunk-SKAZWEV5.js";
import {
  clearAllInboxes,
  clearInbox,
  consumeUnread,
  disposeInboxLocks,
  forwardApproval,
  getInboxPath,
  getLeader,
  readInbox,
  registerLeader,
  sendStructuredMessage,
  unregisterLeader,
  wrapConfirmWithBadge,
  writeMessage
} from "./chunk-XJYUPHRC.js";
import {
  RECIPROCAL_CALLER,
  TaskOwnershipError,
  assertValidTaskId,
  blockTask,
  claimTask,
  createTask,
  deleteTask,
  getAgentStatuses,
  getTask,
  getTaskPath,
  listTasks,
  notifyTasksUpdated,
  onTasksUpdated,
  releaseOwnedTask,
  resetTaskList,
  unassignTeammateTasks,
  updateTask
} from "./chunk-QLDPRILK.js";
import {
  assignTeammateColor,
  classifyShutdownResponse,
  clearTeammateColors,
  createTeamFile,
  deleteTeamDirs,
  findMemberById,
  findMemberByName,
  formatAgentId,
  generateUniqueTeammateName,
  getInboxesDir,
  getTasksDir,
  getTeamDir,
  getTeamFilePath,
  getTeamsRootDir,
  listTeamNames,
  readTeamFile,
  sanitizeName,
  setMemberActive,
  tryReclaimStaleTeam,
  writeTeamFile
} from "./chunk-BVSVIN45.js";
import {
  runSideQuery
} from "./chunk-SBBN4BWR.js";
import {
  CronScheduler,
  WAKEUP_MAX_SECONDS,
  WAKEUP_MIN_SECONDS,
  buildMissedCronNotification,
  clampWakeupSeconds
} from "./chunk-S7AKE7HL.js";
import "./chunk-2MPVVENX.js";
import "./chunk-ZTQ26VBE.js";
import "./chunk-O2VBQ3Y2.js";
import {
  InputFormat,
  OutputFormat
} from "./chunk-MLZQVCF3.js";
import {
  MAX_MCP_RESOURCE_BLOB_CHARS,
  MAX_MCP_RESOURCE_TEXT_CHARS,
  emptyMcpResourceText,
  formatMcpResourceContents,
  summarizeMcpResource
} from "./chunk-LQ7TMOCE.js";
import {
  INBOXES_DIR,
  LEADER_NAME,
  MAX_TEAMMATES,
  TASKS_DIR,
  TEAMMATE_COLORS,
  TEAMS_DIR,
  TEAM_CONFIG_FILENAME,
  getAgentName,
  getTeamName,
  getTeammateColor,
  getTeammateContext,
  isInProcessTeammate,
  isTeamLead,
  isTeammate,
  resolveActiveTeamName,
  runWithTeammateIdentity,
  teammateIdentityStore
} from "./chunk-LD2XBG6Z.js";
import {
  applySkillAllowedTools,
  buildSkillLlmContent
} from "./chunk-HYCC3XM3.js";
import {
  AgentType,
  DefaultHookOutput,
  HOOKS_CONFIG_FIELDS,
  HookEventName,
  HookPhase,
  HookType,
  HooksConfigSource,
  MAX_USER_PROMPT_EXPANSION_ADDITIONAL_CONTEXT_LENGTH,
  NotificationType,
  PermissionMode,
  PermissionRequestHookOutput,
  PostCompactTrigger,
  PostToolBatchHookOutput,
  PostToolUseFailureHookOutput,
  PostToolUseHookOutput,
  PreCompactTrigger,
  PreToolUseHookOutput,
  SessionEndReason,
  SessionStartSource,
  StopHookOutput,
  UserPromptExpansionHookOutput,
  createHookOutput,
  detectTodoChanges,
  getHookKey,
  sanitizeUserPromptExpansionAdditionalContext
} from "./chunk-NL4C3H6C.js";
import {
  DEFAULT_FILE_FILTERING_OPTIONS,
  DEFAULT_MEMORY_FILE_FILTERING_OPTIONS,
  DEFAULT_QWEN_CUSTOM_IGNORE_FILE_NAMES,
  QwenIgnoreParser,
  WorkspaceContext,
  isPathWithinRoot
} from "./chunk-JOLIREYR.js";
import {
  KeychainTokenStorage
} from "./chunk-DDMOU62Q.js";
import {
  AGENT_CONTEXT_FILENAME,
  ApiCancelEvent,
  ApiErrorEvent,
  ApiRequestEvent,
  ApiRequestPhase,
  ApiResponseEvent,
  AuthEvent,
  AuthType,
  BINARY_EXTENSIONS,
  BINARY_FILE_PATTERNS,
  COMMON_DIRECTORY_EXCLUDES,
  COMMON_IGNORE_PATTERNS,
  ConversationFinishedEvent,
  DAEMON_TRACEPARENT_META_KEY,
  DAEMON_TRACESTATE_META_KEY,
  DEFAULT_CONTEXT_FILENAME,
  DEFAULT_ENCODING,
  DEFAULT_FILE_EXCLUDES,
  DEFAULT_OTLP_ENDPOINT,
  DEFAULT_SENSITIVE_SPAN_ATTRIBUTE_MAX_LENGTH,
  DEFAULT_TELEMETRY_TARGET,
  DiscoveredMCPTool,
  EVENT_API_ERROR,
  EVENT_API_RESPONSE,
  EVENT_TOOL_CALL,
  EndSessionEvent,
  ExtensionDisableEvent,
  ExtensionEnableEvent,
  ExtensionInstallEvent,
  ExtensionUninstallEvent,
  FileExclusions,
  FileOperation,
  FlashFallbackEvent,
  IdeConnectionEvent,
  IdeConnectionType,
  KittySequenceOverflowEvent,
  LOCAL_CONTEXT_FILENAME,
  LoopType,
  MAIN_SOURCE,
  MCPServerStatus,
  MEDIA_FILE_PATTERNS,
  MEMORY_SECTION_HEADER,
  MODEL_GENERATION_CONFIG_FIELDS,
  MemoryDreamEvent,
  MemoryExtractEvent,
  MemoryMetricType,
  MemoryRecallEvent,
  ModelSlashCommandEvent,
  NextSpeakerCheckEvent,
  OpenAILogger,
  PYTHON_EXCLUDES,
  PerformanceMetricType,
  PromptSuggestionEvent,
  QWEN_OAUTH_MODELS,
  QwenLogger,
  ReadFileTool,
  RipgrepFallbackEvent,
  SENSITIVE_SPAN_ATTRIBUTE_MAX_LENGTH_LIMIT,
  SYSTEM_FILE_EXCLUDES,
  SemanticAttributes,
  SkillLaunchEvent,
  SlashCommandStatus,
  SpeculationEvent,
  StartSessionEvent,
  TelemetryTarget,
  ToolCallEvent,
  ToolExecutionPhase,
  ToolOutputTruncatedEvent,
  UiTelemetryService,
  UserFeedbackEvent,
  UserFeedbackRating,
  UserPromptEvent,
  UserRetryEvent,
  WorkflowKeywordEvent,
  WorkflowRunEvent,
  _recoverObjectsFromLine,
  _resetEnsuredDirsCacheForTest,
  addDaemonRequestAttribute,
  addMCPStatusChangeListener,
  addModelOutputAttributes,
  addSystemPromptAttributes,
  addToolInputAttributes,
  addToolResultAttributes,
  addToolSchemaAttributes,
  addUserPromptAttributes,
  apiResponseEventToTokenUsageRecord,
  areSensitiveSpanAttributesEnabled,
  captureDaemonTelemetryContext,
  countFileLines,
  countLines,
  createContentGenerator,
  createContentGeneratorConfig,
  createDaemonBridgeTelemetry,
  decodeBufferWithEncodingInfo,
  detectBOM,
  detectEncodingFromBuffer,
  detectFileEncoding,
  detectFileType,
  emitDaemonLog,
  endHookSpan,
  endInteractionSpan,
  endLLMRequestSpan,
  endSubagentSpan,
  endToolBlockedOnUserSpan,
  endToolExecutionSpan,
  endToolSpan,
  exists,
  exportTokenUsageSummary,
  extractDaemonTraceContext,
  extractExtensionsFromPatterns,
  fileExists,
  forceFlushMetrics,
  formatTokenUsageSummaryAsCsv,
  formatTokenUsageSummaryAsJson,
  generateValidName,
  getActiveInteractionSpan,
  getAllGeminiMdFilenames,
  getAllMCPServerStatuses,
  getCachedEncodingForBuffer,
  getCurrentGeminiMdFilename,
  getMCPServerStatus,
  getSpecificMimeType,
  getSystemEncoding,
  getTokenUsageFilePath,
  hashDaemonWorkspace,
  initializeDaemonMetrics,
  initializeTelemetry,
  injectDaemonTraceContext,
  isBinaryFile,
  isPerformanceMonitoringActive,
  isTelemetrySdkInitialized,
  isValidSensitiveSpanAttributeMaxLength,
  isWithinRoot,
  logApiCancel,
  logApiError,
  logApiRequest,
  logApiResponse,
  logArenaAgentCompleted,
  logArenaSessionEnded,
  logArenaSessionStarted,
  logAuth,
  logChatCompression,
  logConversationFinishedEvent,
  logExtensionDisable,
  logExtensionEnable,
  logExtensionInstallEvent,
  logExtensionUninstall,
  logExtensionUpdateEvent,
  logFlashFallback,
  logIdeConnection,
  logKittySequenceOverflow,
  logMemoryDream,
  logMemoryExtract,
  logMemoryRecall,
  logModelSlashCommand,
  logNextSpeakerCheck,
  logPromptSuggestion,
  logRipgrepFallback,
  logSkillLaunch,
  logSlashCommand,
  logSpeculation,
  logStartSession,
  logToolCall,
  logToolOutputTruncated,
  logUserFeedback,
  logUserPrompt,
  logUserRetry,
  logWorkflowKeyword,
  logWorkflowRun,
  makeArenaAgentCompletedEvent,
  makeArenaSessionEndedEvent,
  makeArenaSessionStartedEvent,
  makeChatCompressionEvent,
  makeSlashCommandEvent,
  openaiLogger,
  parseBooleanEnvFlag,
  parseLineTolerant,
  parseTelemetryTargetValue,
  processSingleFileContent,
  queryTokenUsage,
  read,
  readFileWithEncoding,
  readFileWithEncodingInfo,
  readFileWithLineAndLimit,
  readLines,
  recordApiErrorMetrics,
  recordApiRequestBreakdown,
  recordApiResponseMetrics,
  recordApiRetry,
  recordArenaAgentCompletedMetrics,
  recordArenaSessionEndedMetrics,
  recordArenaSessionStartedMetrics,
  recordBaselineComparison,
  recordContentRetry,
  recordContentRetryFailure,
  recordCpuUsage,
  recordDaemonBridgeError,
  recordDaemonCancel,
  recordDaemonChannelLifecycle,
  recordDaemonError,
  recordDaemonHttpRequest,
  recordDaemonHttpResponse,
  recordDaemonPromptDuration,
  recordDaemonPromptQueueWait,
  recordDaemonSessionLifecycle,
  recordFileOperationMetric,
  recordInvalidChunk,
  recordMemoryDreamMetrics,
  recordMemoryExtractMetrics,
  recordMemoryRecallMetrics,
  recordMemoryUsage,
  recordPerformanceRegression,
  recordPerformanceScore,
  recordSkillInvocation,
  recordStartupPerformance,
  recordTokenEfficiency,
  recordTokenUsageFromApiResponse,
  recordTokenUsageFromApiResponseBestEffort,
  recordTokenUsageMetrics,
  recordToolCallMetrics,
  recordToolExecutionBreakdown,
  recordToolQueueDepth,
  refreshSessionContext,
  registerDaemonGaugeCallbacks,
  removeMCPServerStatus,
  removeMCPStatusChangeListener,
  resetEncodingCache,
  resolveContentGeneratorConfigWithSources,
  resolveOpenAILogDir,
  resolveTelemetrySettings,
  runInSubagentSpanContext,
  runInToolSpanContext,
  runWithDaemonTelemetryContext,
  sanitizeHookName,
  setGeminiMdFilename,
  shutdownTelemetry,
  startHookSpan,
  startInteractionSpan,
  startLLMRequestSpan,
  startSubagentSpan,
  startToolBlockedOnUserSpan,
  startToolExecutionSpan,
  startToolSpan,
  truncateContent,
  truncateSpanError,
  uiTelemetryService,
  updateMCPServerStatus,
  validateModelConfig,
  windowsCodePageToEncoding,
  withDaemonBridgeSpan,
  withDaemonRequestSpan,
  withDaemonSpan,
  withInteractionSpan,
  write,
  writeLine,
  writeLineSync
} from "./chunk-QNKBEJO2.js";
import "./chunk-3PJXIDKI.js";
import "./chunk-UWCTAVOD.js";
import {
  DEFAULT_QWEN_EMBEDDING_MODEL,
  DEFAULT_QWEN_FLASH_MODEL,
  DEFAULT_QWEN_MODEL,
  MAINLINE_CODER_MODEL
} from "./chunk-OFEVLU4C.js";
import "./chunk-OMX7CUOE.js";
import {
  AUTO_MEMORY_CONSOLIDATION_LOCK_FILENAME,
  AUTO_MEMORY_DIRNAME,
  AUTO_MEMORY_EXTRACT_CURSOR_FILENAME,
  AUTO_MEMORY_INDEX_FILENAME,
  AUTO_MEMORY_METADATA_FILENAME,
  TEAM_AUTO_MEMORY_DIRNAME,
  USER_AUTO_MEMORY_DIRNAME,
  clearAutoMemoryRootCache,
  getAutoMemoryConsolidationLockPath,
  getAutoMemoryExtractCursorPath,
  getAutoMemoryFilePath,
  getAutoMemoryIndexPath,
  getAutoMemoryMetadataPath,
  getAutoMemoryProjectStateDir,
  getAutoMemoryRoot,
  getAutoMemoryTopicFilename,
  getAutoMemoryTopicPath,
  getMemoryBaseDir,
  getTeamAutoMemoryIndexPath,
  getTeamAutoMemoryRoot,
  getUserAutoMemoryIndexPath,
  getUserAutoMemoryRoot,
  getUserAutoMemoryTopicPath,
  isAnyAutoMemPath,
  isAutoMemPath,
  isTeamAutoMemPath,
  isUserAutoMemPath
} from "./chunk-RD6YJJJ6.js";
import "./chunk-SKQZZVAV.js";
import "./chunk-BUUQ5HPX.js";
import {
  AgentStatistics,
  BaseDeclarativeTool,
  BaseToolInvocation,
  CONCURRENCY_SAFE_KINDS,
  DeclarativeTool,
  Kind,
  MUTATOR_KINDS,
  SchemaValidator,
  ToolConfirmationOutcome,
  ToolDisplayNames,
  ToolDisplayNamesMigration,
  ToolErrorType,
  ToolNames,
  ToolNamesMigration,
  hasCycleInSchema,
  isTool
} from "./chunk-BX5YO7UW.js";
import {
  CredentialsClearRequiredError,
  QWEN_CREDENTIAL_FILE_MODE,
  QwenOAuth2Client,
  QwenOAuth2Event,
  QwenOAuthPollError,
  cacheQwenCredentials,
  clearCachedCredentialFile,
  clearQwenCredentials,
  generateCodeChallenge,
  generateCodeVerifier,
  generatePKCEPair,
  getQwenOAuthClient,
  isDeviceAuthorizationSuccess,
  isDeviceTokenPending,
  isDeviceTokenSuccess,
  isErrorResponse,
  qwenOAuth2Events
} from "./chunk-UO5GVXQE.js";
import "./chunk-KMGMKW5Q.js";
import {
  isBrowserCommandBlocked,
  openBrowserSecurely,
  shouldAttemptBrowserLaunch,
  shouldLaunchBrowser
} from "./chunk-22IFUCVR.js";
import {
  atomicWriteFile,
  atomicWriteFileSync,
  atomicWriteJSON,
  renameWithRetry,
  renameWithRetrySync
} from "./chunk-3227DZO4.js";
import {
  DEFAULT_OUTPUT_TOKEN_LIMIT,
  DEFAULT_TOKEN_LIMIT,
  ESCALATED_MAX_TOKENS,
  SUPPORTED_IMAGE_MIME_TYPES,
  TextTokenizer,
  detectRuntime,
  getOrCreateSharedDispatcher,
  getSupportedImageFormatsString,
  getUnsupportedImageFormatWarning,
  hasExplicitOutputLimit,
  isSupportedImageMimeType,
  knownTokenLimit,
  normalize,
  parsePositiveIntegerEnvValue,
  redactProxyCredentials,
  tokenLimit
} from "./chunk-U3KSMM7G.js";
import {
  RuntimeDiagnosticsCollector,
  isRuntimeDiagnosticsEnabled,
  runtimeDiagnostics,
  summarizeAnthropicWireRequest,
  summarizeOpenAIWireRequest
} from "./chunk-BKXBARJZ.js";
import "./chunk-EJYJYJZ2.js";
import "./chunk-55ZMG67I.js";
import "./chunk-H6BD2ELD.js";
import "./chunk-5IFG2VC4.js";
import "./chunk-64WXLC72.js";
import {
  BadRequestError,
  FatalAuthenticationError,
  FatalBudgetExceededError,
  FatalCancellationError,
  FatalConfigError,
  FatalError,
  FatalInputError,
  FatalSandboxError,
  FatalToolExecutionError,
  FatalTurnLimitedError,
  ForbiddenError,
  GOOGLE_ACCOUNTS_FILENAME,
  PATH_ARG_KEYS,
  QWEN_DIR,
  SHELL_SPECIAL_CHARS,
  Storage,
  UnauthorizedError,
  _resetValidatePathCacheForTest,
  createDebugLogger,
  escapePath,
  expandHomeDir,
  formatTraceparent,
  getErrorMessage,
  getErrorStatus,
  getErrorType,
  getProjectHash,
  getTraceContext,
  isAbortError,
  isDebugLoggingDegraded,
  isNodeError,
  isSubpath,
  isSubpaths,
  makeRelative,
  resetDebugLoggingState,
  resolveAndValidatePath,
  resolvePath,
  runWithDebugLogSession,
  sanitizeCwd,
  setDebugLogSession,
  shortenPath,
  tildeifyPath,
  toFriendlyError,
  unescapePath,
  updateSymlink,
  validatePath
} from "./chunk-UX4SGCGN.js";
import "./chunk-ZERZSAZL.js";
import "./chunk-QN5NZ3UQ.js";
import "./chunk-BR4QREVK.js";
import {
  SpanStatusCode,
  ValueType
} from "./chunk-Z2Z3GUXZ.js";
import "./chunk-A4BMJM77.js";
import "./chunk-J2S4EL5Y.js";
export {
  AGENT_CONTEXT_FILENAME,
  AGENT_WORKTREE_HEX_LENGTH,
  AGENT_WORKTREE_PREFIX,
  AGENT_WORKTREE_SLUG_PATTERN,
  ALIBABA_PROVIDERS,
  ALL_PROVIDERS,
  APPROVAL_MODES,
  APPROVAL_MODE_INFO,
  ARENA_MAX_AGENTS,
  AUTO_MEMORY_CONSOLIDATION_LOCK_FILENAME,
  AUTO_MEMORY_DIRNAME,
  AUTO_MEMORY_EXTRACT_CURSOR_FILENAME,
  AUTO_MEMORY_INDEX_FILENAME,
  AUTO_MEMORY_METADATA_FILENAME,
  AUTO_MEMORY_SCHEMA_VERSION,
  AUTO_MEMORY_TYPES,
  AUTO_MODE_DENIAL_LIMITS,
  AUTO_SKILL_THRESHOLD,
  AbortError,
  AgentCore,
  AgentEventEmitter,
  AgentEventType,
  AgentHeadless,
  AgentInteractive,
  AgentStatistics,
  AgentStatus,
  AgentTerminateMode,
  AgentType,
  ApiCancelEvent,
  ApiErrorEvent,
  ApiRequestEvent,
  ApiRequestPhase,
  ApiResponseEvent,
  ApprovalMode,
  ArenaAgentClient,
  ArenaEventEmitter,
  ArenaEventType,
  ArenaManager,
  ArenaSessionStatus,
  AsyncMessageQueue,
  AuthEvent,
  AuthProviderType,
  AuthType,
  BACKGROUND_AGENT_CONCURRENCY_ENV,
  BASELINE_COMMIT_MESSAGE,
  BINARY_EXTENSIONS,
  BINARY_FILE_PATTERNS,
  BTW_MAX_INPUT_LENGTH,
  BUNDLE_CHUNK_DIR,
  BackgroundAgentResumeService,
  BackgroundShellRegistry,
  BackgroundTaskRegistry,
  BadRequestError,
  BaseDeclarativeTool,
  BaseToolInvocation,
  BudgetExhaustedError,
  BuiltinAgentRegistry,
  CODE_ACTION_KIND_LABELS,
  CODING_PLAN_CHINA_BASE_URL,
  CODING_PLAN_ENV_KEY,
  CODING_PLAN_GLOBAL_BASE_URL,
  COMMAND_SUBSTITUTION_WARNING,
  COMMON_DIRECTORY_EXCLUDES,
  COMMON_IGNORE_PATTERNS,
  CONCURRENCY_SAFE_KINDS,
  CUSTOM_API_KEY_ENV_PREFIX,
  ChatRecordingService,
  CloseDiffRequestSchema,
  ColorMode,
  CompressionStatus,
  ConditionalRulesRegistry,
  Config,
  ContextState,
  ConversationFinishedEvent,
  CoreToolScheduler,
  CredentialsClearRequiredError,
  CronScheduler,
  DAEMON_TRACEPARENT_META_KEY,
  DAEMON_TRACESTATE_META_KEY,
  DEFAULT_AUTO_DREAM_MIN_HOURS,
  DEFAULT_AUTO_DREAM_MIN_SESSIONS,
  DEFAULT_BACKGROUND_AGENT_CONTINUATION_MESSAGE,
  DEFAULT_BUILTIN_SUBAGENT_TYPE,
  DEFAULT_CONTEXT_FILENAME,
  DEFAULT_ENCODING,
  DEFAULT_FILE_EXCLUDES,
  DEFAULT_FILE_FILTERING_OPTIONS,
  DEFAULT_LSP_COMMAND_CHECK_TIMEOUT_MS,
  DEFAULT_LSP_DOCUMENT_OPEN_DELAY_MS,
  DEFAULT_LSP_DOCUMENT_RETRY_DELAY_MS,
  DEFAULT_LSP_MAX_RESTARTS,
  DEFAULT_LSP_REQUEST_TIMEOUT_MS,
  DEFAULT_LSP_SOCKET_MAX_RETRY_DELAY_MS,
  DEFAULT_LSP_SOCKET_RETRY_DELAY_MS,
  DEFAULT_LSP_STARTUP_TIMEOUT_MS,
  DEFAULT_LSP_WARMUP_DELAY_MS,
  DEFAULT_LSP_WORKSPACE_SYMBOL_WARMUP_DELAY_MS,
  DEFAULT_MAX_CONCURRENT_BACKGROUND_AGENTS,
  DEFAULT_MAX_INLINE_MEDIA_BYTES,
  DEFAULT_MEMORY_FILE_FILTERING_OPTIONS,
  DEFAULT_OTLP_ENDPOINT,
  DEFAULT_OUTPUT_TOKEN_LIMIT,
  DEFAULT_QWEN_CUSTOM_IGNORE_FILE_NAMES,
  DEFAULT_QWEN_EMBEDDING_MODEL,
  DEFAULT_QWEN_FLASH_MODEL,
  DEFAULT_QWEN_MODEL,
  DEFAULT_SENSITIVE_SPAN_ATTRIBUTE_MAX_LENGTH,
  DEFAULT_STOP_HOOK_BLOCK_CAP,
  DEFAULT_TELEMETRY_TARGET,
  DEFAULT_TOKEN_LIMIT,
  DEFAULT_TOOL_OUTPUT_BATCH_BUDGET,
  DEFAULT_TOOL_RESULTS_TOTAL_CHARS_THRESHOLD,
  DEFAULT_TRUNCATE_TOOL_OUTPUT_LINES,
  DEFAULT_TRUNCATE_TOOL_OUTPUT_THRESHOLD,
  DIAGNOSTIC_SEVERITY_LABELS,
  DISPLAY_MODE,
  DREAM_TASK_TYPE,
  DeclarativeTool,
  DefaultHookOutput,
  DiscoveredMCPTool,
  DiscoveredTool,
  ESCALATED_MAX_TOKENS,
  EVENT_API_ERROR,
  EVENT_API_RESPONSE,
  EVENT_TOOL_CALL,
  EXCLUDED_TOOLS_FOR_SUBAGENTS,
  EXTENSIONS_CONFIG_FILENAME,
  EXTENSIONS_DIRECTORY_NAME,
  EXTENSION_SETTINGS_FILENAME,
  EXTERNAL_MESSAGE_PREFIX,
  EXTRACT_TASK_TYPE,
  EndSessionEvent,
  ExtensionDisableEvent,
  ExtensionEnableEvent,
  ExtensionInstallEvent,
  ExtensionManager,
  ExtensionPreferencesStore,
  ExtensionSettingScope,
  ExtensionUninstallEvent,
  ExtensionUpdateState,
  FILE_HISTORY_DIR,
  FORK_SUBAGENT_TYPE,
  FatalAuthenticationError,
  FatalBudgetExceededError,
  FatalCancellationError,
  FatalConfigError,
  FatalError,
  FatalInputError,
  FatalSandboxError,
  FatalToolExecutionError,
  FatalTurnLimitedError,
  FileDiscoveryService,
  FileEncoding,
  FileExclusions,
  FileHistoryService,
  FileOperation,
  FileReadCache,
  FileSchema,
  FileSearchFactory,
  FlashFallbackEvent,
  ForbiddenError,
  GOAL_HOOK_TIMEOUT_MS,
  GOAL_HOOK_TIMEOUT_SECONDS,
  GOOGLE_ACCOUNTS_FILENAME,
  GeminiChat,
  GeminiClient,
  GeminiEventType,
  GitIgnoreParser,
  GitWorktreeService,
  HOOKS_CONFIG_FIELDS,
  HookEventName,
  HookPhase,
  HookRegistry,
  HookSystem,
  HookType,
  HooksConfigSource,
  IDEConnectionStatus,
  IDE_DEFINITIONS,
  IDE_MAX_OPEN_FILES,
  IDE_MAX_SELECTED_TEXT_LENGTH,
  IDE_REQUEST_TIMEOUT_MS,
  IDLE_SPECULATION,
  INBOXES_DIR,
  INITIAL_FOLLOWUP_STATE,
  INSTALL_METADATA_FILENAME,
  ITermBackend,
  IdeClient,
  IdeConnectionEvent,
  IdeConnectionType,
  IdeContextNotificationSchema,
  IdeContextSchema,
  IdeContextStore,
  IdeDiffAcceptedNotificationSchema,
  IdeDiffClosedNotificationSchema,
  IdeDiffRejectedNotificationSchema,
  Ignore,
  InProcessBackend,
  InputFormat,
  InvalidMcpConfigError,
  InvalidStreamError,
  JsonFormatter,
  KeychainTokenStorage,
  Kind,
  KittySequenceOverflowEvent,
  LEADER_NAME,
  LITE_READ_BUF_SIZE,
  LOCAL_CONTEXT_FILENAME,
  Logger,
  LoopType,
  LspConfigLoader,
  LspConnectionFactory,
  LspResponseNormalizer,
  LspServerManager,
  MAINLINE_CODER_MODEL,
  MAIN_SOURCE,
  MAX_CONCURRENT_BACKGROUND_AGENTS,
  MAX_CONCURRENT_MONITORS,
  MAX_DIFF_SIZE_BYTES,
  MAX_FILES,
  MAX_FILES_FOR_DETAILS,
  MAX_GOAL_ITERATIONS,
  MAX_LINES_PER_FILE,
  MAX_MCP_RESOURCE_BLOB_CHARS,
  MAX_MCP_RESOURCE_TEXT_CHARS,
  MAX_NOTIFICATION_OUTPUT_TAIL_BYTES,
  MAX_RETAINED_AGENT_FIELD_CHARS,
  MAX_RETAINED_ANSI_OUTPUT_LINES,
  MAX_RETAINED_FILE_CONTENT_CHARS,
  MAX_RETAINED_FILE_DIFF_CHARS,
  MAX_RETAINED_SNAPSHOTS,
  MAX_RETAINED_TERMINAL_AGENTS,
  MAX_RETAINED_TERMINAL_MONITORS,
  MAX_RETAINED_TERMINAL_SHELLS,
  MAX_RETAINED_TERMINAL_WORKFLOWS,
  MAX_RETAINED_TOOL_RESULT_DISPLAY_CHARS,
  MAX_SNAPSHOTS,
  MAX_TEAMMATES,
  MAX_TRANSCRIPT_MESSAGES,
  MAX_USER_PROMPT_EXPANSION_ADDITIONAL_CONTEXT_LENGTH,
  MCPDiscoveryState,
  MCPOAuthProvider,
  MCPOAuthTokenStorage,
  MCPServerConfig,
  MCPServerStatus,
  MCP_BUDGET_REARM_FRACTION,
  MCP_BUDGET_WARN_FRACTION,
  MCP_DEFAULT_TIMEOUT_MSEC,
  MEDIA_FILE_PATTERNS,
  MEMORY_SECTION_HEADER,
  MOCK_TOOL_GET_CONFIRMATION_DETAILS,
  MOCK_TOOL_GET_DEFAULT_PERMISSION,
  MODEL_GENERATION_CONFIG_FIELDS,
  MUTATOR_KINDS,
  McpBudgetWouldExceedError,
  McpClient,
  McpClientManager,
  McpServerSpawnFailedError,
  McpTransportPool,
  MemoryDreamEvent,
  MemoryExtractEvent,
  MemoryManager,
  MemoryMetricType,
  MemoryRecallEvent,
  MessageBus,
  MessageBusType,
  MessageSenderType,
  MockModifiableTool,
  MockModifiableToolInvocation,
  MockTool,
  ModelRegistry,
  ModelSlashCommandEvent,
  ModelsConfig,
  MonitorRegistry,
  NativeLspClient,
  NativeLspService,
  NextSpeakerCheckEvent,
  NotificationType,
  OAUTH_AUTH_URL_EVENT,
  OAUTH_DISPLAY_MESSAGE_EVENT,
  OAuthUtils,
  OPENROUTER_BASE_URL,
  OPENROUTER_ENV_KEY,
  OpenAILogger,
  OpenDiffRequestSchema,
  OutputFormat,
  OverlayFs,
  PATH_ARG_KEYS,
  POOLED_TRANSPORTS_DEFAULT,
  PROVIDER_METADATA_NS,
  PYTHON_EXCLUDES,
  PerformanceMetricType,
  PermissionManager,
  PermissionMode,
  PermissionRequestHookOutput,
  PostCompactTrigger,
  PostToolBatchHookOutput,
  PostToolUseFailureHookOutput,
  PostToolUseHookOutput,
  PreCompactTrigger,
  PreToolUseHookOutput,
  PromptSuggestionEvent,
  ProviderInstallError,
  QWEN_CODE_COMPANION_EXTENSION_NAME,
  QWEN_CODE_SIMPLE_ENV_VAR,
  QWEN_CREDENTIAL_FILE_MODE,
  QWEN_DIR,
  QWEN_OAUTH_MODELS,
  QwenIgnoreParser,
  QwenLogger,
  QwenOAuth2Client,
  QwenOAuth2Event,
  QwenOAuthPollError,
  RECIPROCAL_CALLER,
  REDACTED_URL_CREDENTIAL,
  REQUESTY_BASE_URL,
  REQUESTY_ENV_KEY,
  RUNTIME_SNAPSHOT_PREFIX,
  RUNTIME_STATUS_SCHEMA_VERSION,
  ReadFileTool,
  RipGrepTool,
  RipgrepFallbackEvent,
  RuntimeDiagnosticsCollector,
  SAFE_TOOL_ALLOWLIST,
  SENSITIVE_SPAN_ATTRIBUTE_MAX_LENGTH_LIMIT,
  SESSION_TITLE_MAX_LENGTH,
  SHELL_SELF_KILL_REJECTION,
  SHELL_SPECIAL_CHARS,
  SHELL_TOOL_NAMES,
  SKILL_REVIEW_TASK_TYPE,
  STOP_HOOK_BLOCK_CAP_ENV,
  SUGGESTION_PROMPT,
  SUPPORTED_IMAGE_MIME_TYPES,
  SYMBOL_KIND_LABELS,
  SYSTEM_FILE_EXCLUDES,
  SYSTEM_REMINDER_CLOSE,
  SYSTEM_REMINDER_OPEN,
  SchemaValidator,
  SdkControlClientTransport,
  SemanticAttributes,
  SendMessageType,
  SessionEndReason,
  SessionService,
  SessionStartSource,
  SettingScope,
  ShellExecutionService,
  SkillActivationRegistry,
  SkillError,
  SkillLaunchEvent,
  SkillManager,
  SlashCommandStatus,
  SleepInhibitor,
  SourceRegistryStore,
  SpanStatusCode,
  SpeculationEvent,
  StandardFileSystemService,
  StartSessionEvent,
  StopHookOutput,
  Storage,
  StreamEventType,
  SubagentError,
  SubagentErrorCode,
  SubagentManager,
  SubagentValidator,
  TASKS_DIR,
  TEAMMATE_COLORS,
  TEAMS_DIR,
  TEAM_AUTO_MEMORY_DIRNAME,
  TEAM_CONFIG_FILENAME,
  TERMINAL_CSI_REGEX,
  TERMINAL_OSC_REGEX,
  TERMINAL_SHIFT_DCS_REGEX,
  THIRD_PARTY_PROVIDERS,
  TOKEN_PLAN_BASE_URL,
  TOKEN_PLAN_ENV_KEY,
  TOOL_NAME_ALIASES,
  TOOL_USE_SUMMARY_SYSTEM_PROMPT,
  TaskOwnershipError,
  TeamEventEmitter,
  TeamEventType,
  TeamManager,
  TelemetryTarget,
  TextTokenizer,
  TmuxBackend,
  ToolCallEvent,
  ToolConfirmationOutcome,
  ToolDisplayNames,
  ToolDisplayNamesMigration,
  ToolErrorType,
  ToolExecutionPhase,
  ToolNames,
  ToolNamesMigration,
  ToolOutputTruncatedEvent,
  ToolRegistry,
  TrustGateError,
  Turn,
  USER_AUTO_MEMORY_DIRNAME,
  UiTelemetryService,
  UnauthorizedError,
  UserFeedbackEvent,
  UserFeedbackRating,
  UserPromptEvent,
  UserPromptExpansionHookOutput,
  UserRetryEvent,
  ValueType,
  WAKEUP_MAX_SECONDS,
  WAKEUP_MIN_SECONDS,
  WORKFLOW_NAME_PATTERN,
  WORKTREES_DIR,
  WORKTREE_BRANCH_PREFIX,
  WORKTREE_SESSION_FILE,
  WorkflowKeywordEvent,
  WorkflowRunEvent,
  WorkflowRunRegistry,
  WorkspaceContext,
  WorkspaceMcpBudget,
  WorkspaceMemoryFileTooLargeError,
  WorkspaceMemoryWriteTimeoutError,
  __resetActiveGoalStoreForTests,
  _recoverObjectsFromLine,
  _resetCachedIdeServerHost,
  _resetEnsuredDirsCacheForTest,
  _resetRipGrepCachesForTest,
  _resetRipgrepUtilsCachesForTest,
  _resetValidatePathCacheForTest,
  abortGoalForStopHookCap,
  abortSpeculation,
  acceptSpeculation,
  acquireSleepInhibitor,
  activeGoalEquals,
  addDaemonRequestAttribute,
  addMCPStatusChangeListener,
  addModelOutputAttributes,
  addSystemPromptAttributes,
  addToolInputAttributes,
  addToolResultAttributes,
  addToolSchemaAttributes,
  addUserPromptAttributes,
  aggregateUsage,
  alibabaStandardProvider,
  allowEditorTypeInSandbox,
  apiResponseEventToTokenUsageRecord,
  appendToLastTextPart,
  applyAutoModeDecision,
  applyProviderInstallPlan,
  applySkillAllowedTools,
  approxBase64Bytes,
  areSensitiveSpanAttributesEnabled,
  assertValidTaskId,
  assignTeammateColor,
  atomicWriteFile,
  atomicWriteFileSync,
  atomicWriteJSON,
  blockTask,
  buildAddedMcpToolsReminder,
  buildAddedSkillsReminder,
  buildApiHistoryFromConversation,
  buildAvailableSkillsReminder,
  buildBackgroundEntryLabel,
  buildBtwCacheSafeParams,
  buildBtwPrompt,
  buildDeferredToolsReminder,
  buildFallbackApproachSummary,
  buildHumanReadableRuleLabel,
  buildInstallPlan,
  buildMcpServerInstructionsReminder,
  buildMissedCronNotification,
  buildModelIdContext,
  buildPermissionCheckContext,
  buildPermissionRules,
  buildProviderTemplate,
  buildShellExecWarnings,
  buildSkillLlmContent,
  buildStartupContextReminder,
  buildTeammatePromptAddendum,
  cacheQwenCredentials,
  canUseRipgrep,
  captureDaemonTelemetryContext,
  checkArgumentSafety,
  checkCommandPermissions,
  checkForExtensionUpdate,
  checkHasEditorType,
  checkNpmUpdate,
  claimTask,
  clampInlineMediaPart,
  clampWakeupSeconds,
  classifyShutdownResponse,
  cleanSummary,
  cleanupTmpDir,
  clearActiveGoal,
  clearAllInboxes,
  clearAutoMemoryRootCache,
  clearCacheSafeParams,
  clearCachedCredentialFile,
  clearGitDirCache,
  clearGoalTerminalObserver,
  clearInbox,
  clearQwenCredentials,
  clearRuntimeStatus,
  clearSessionCommits,
  clearTeammateColors,
  clearWelcomeBackState,
  clearWorktreeSession,
  cliSource,
  cloneFromGit,
  codingPlanProvider,
  collectMemoryDiagnostics,
  collectText,
  collectToolCallIdsFromHistory,
  commandExists,
  compactStringForHistory,
  compactStringForRecording,
  compactToolResultDisplayForHistory,
  compactToolResultDisplayForRecording,
  computeModelListVersion,
  computeThresholds,
  computeUniqueBranchTitle,
  computedSource,
  connectAndDiscover,
  connectToMcpServer,
  connectionIdOf,
  consumeUnread,
  convertClaudeAgentConfig,
  convertClaudePluginPackage,
  convertClaudePluginStandalone,
  convertClaudeToQwenConfig,
  convertColorToHex,
  convertToFunctionResponse,
  convertTomlToMarkdown,
  copyExtension,
  countFileLines,
  countLines,
  crawlCache_exports as crawlCache,
  createContentGenerator,
  createContentGeneratorConfig,
  createDaemonBridgeTelemetry,
  createDebugLogger,
  createDefaultAutoMemoryExtractCursor,
  createDefaultAutoMemoryIndex,
  createDefaultAutoMemoryMetadata,
  createDenialState,
  createDuplicateProviderToolCallResponse,
  createFollowupController,
  createForkedChat,
  createGoalStopHookCallback,
  createHookOutput,
  createInstructionsLoadedCallback,
  createStreamableHttpCompatibilityFetch,
  createTask,
  createTeamFile,
  createTmpDir,
  createToolUseSummaryMessage,
  createTransport,
  customProvider,
  decodeBufferWithEncodingInfo,
  decodeTagName,
  dedupeToolCallsById,
  deepseekProvider,
  defaultSource,
  delay,
  deleteTask,
  deleteTeamDirs,
  deserializeSnapshots,
  detectBOM,
  detectBackend,
  detectCommandSubstitution,
  detectEncodingFromBuffer,
  detectFileEncoding,
  detectFileType,
  detectIdeFromEnv,
  detectLineEnding,
  detectRuntime,
  detectSelfKillCommand,
  detectTodoChanges,
  discoverMcpTools,
  discoverPlugins,
  discoverPrompts,
  discoverResources,
  discoverTools,
  disposeInboxLocks,
  doesToolInvocationMatch,
  downloadFromArchiveUrl,
  downloadFromGitHubRelease,
  downloadFromNpmRegistry,
  editorCommands,
  emitDaemonLog,
  emptyMcpResourceText,
  encodeInsightErrorMessage,
  encodeInsightProgressMessage,
  encodeInsightReadyMessage,
  encodeTagName,
  encodeTextFileContent,
  endHookSpan,
  endInteractionSpan,
  endLLMRequestSpan,
  endSubagentSpan,
  endToolBlockedOnUserSpan,
  endToolExecutionSpan,
  endToolSpan,
  ensureAutoMemoryScaffold,
  ensureCrlfLineEndings,
  ensureMacBinarySigned,
  ensureRipgrepHealthy,
  ensureToolResultPairing,
  ensureUserAutoMemoryScaffold,
  envLayer,
  escapePath,
  escapeShellArg,
  evaluateAutoMode,
  evaluatePermissionFlow,
  evaluatePermissionRules,
  evaluateToolCall,
  execCommand,
  executeToolCall,
  exists,
  expandHomeDir,
  exportTokenUsageSummary,
  extractArchiveFile,
  extractDaemonTraceContext,
  extractExtensionsFromPatterns,
  extractFile,
  extractJsonStringField,
  extractLastJsonStringField,
  extractLastJsonStringFields,
  extractLastUserPrompt,
  extractShellOperations,
  extractToolFilePaths,
  fetchGitDiff,
  fetchGitDiffHunks,
  fileExists,
  filter,
  findExistingProviderModels,
  findGitRoot,
  findMemberById,
  findMemberByName,
  findProviderByCredentials,
  findProviderById,
  findReleaseAsset,
  findRepeatedDuplicateProviderToolCall,
  fireNotificationHook,
  firePermissionRequestHook,
  firePostToolUseFailureHook,
  firePostToolUseHook,
  firePreToolUseHook,
  flatMapTextParts,
  forceFlushMetrics,
  formatAgentId,
  formatClassifierBlockMessage,
  formatDateForContext,
  formatDenialStateLog,
  formatLineStats,
  formatMcpResourceContents,
  formatMemoryUsage,
  formatStopHookBlockingCapWarning,
  formatTokenUsageSummaryAsCsv,
  formatTokenUsageSummaryAsJson,
  formatTraceparent,
  forwardApproval,
  generateAgentWorktreeSlug,
  generateCodeChallenge,
  generateCodeVerifier,
  generateCustomEnvKey,
  generatePKCEPair,
  generatePromptSuggestion,
  generateSessionRecap,
  generateToolUseId,
  generateToolUseSummary,
  generateUniqueTeammateName,
  generateValidName,
  getActiveGoal,
  getActiveInteractionSpan,
  getAgentName,
  getAgentStatuses,
  getAllGeminiMdFilenames,
  getAllMCPServerStatuses,
  getAllProviderBaseUrls,
  getArenaSystemReminder,
  getAutoMemoryConsolidationLockPath,
  getAutoMemoryExtractCursorPath,
  getAutoMemoryFilePath,
  getAutoMemoryIndexPath,
  getAutoMemoryMetadataPath,
  getAutoMemoryProjectStateDir,
  getAutoMemoryRoot,
  getAutoMemoryTopicFilename,
  getAutoMemoryTopicPath,
  getAutoModePermissionDeniedReason,
  getBuiltinRipgrep,
  getCacheSafeParams,
  getCachedEncodingForBuffer,
  getCommandRoot,
  getCommandRoots,
  getCompressionPrompt,
  getCoreSystemPrompt,
  getCurrentGeminiMdFilename,
  getCustomSystemPrompt,
  getDefaultBaseUrlForProtocol,
  getDefaultModelIds,
  getDiffCommand,
  getDirectoryContextString,
  getEditorExecutable,
  getEffectivePermissionForConfirmation,
  getEnvContents,
  getEnvironmentContext,
  getErrorMessage,
  getErrorStatus,
  getErrorType,
  getExtensionDescription,
  getExtensionDisplayName,
  getExtensionId,
  getExternalEditorCommand,
  getFilterReason,
  getFolderStructure,
  getFunctionCalls,
  getFunctionCallsAsJson,
  getFunctionCallsFromParts,
  getFunctionCallsFromPartsAsJson,
  getGitBranch,
  getGitRepoName,
  getHookKey,
  getIdeInstaller,
  getIdeServerHost,
  getInboxPath,
  getInboxesDir,
  getInitialChatHistory,
  getInsightPrompt,
  getLastGoalTerminal,
  getLeader,
  getMCPDiscoveryState,
  getMCPServerPrompts,
  getMCPServerStatus,
  getMaxInlineMediaBytes,
  getMcpOAuthDialogInstruction,
  getMemoryBaseDir,
  getMonitorOutputPath,
  getOrCreateSharedDispatcher,
  getPlanModeSystemReminder,
  getProjectHash,
  getProjectSummaryInfo,
  getProjectSummaryPrompt,
  getProviderToolCallId,
  getQwenOAuthClient,
  getRateLimitErrorDetails,
  getRateLimitRetryDelayMs,
  getRecentGitStatus,
  getResponseText,
  getResponseTextFromParts,
  getResumePromptTokenCount,
  getResumeTokenCounts,
  getRuleDisplayName,
  getSavedWorkflowDirs,
  getScopedEnvContents,
  getShellAbortReasonKind,
  getShellConfiguration,
  getSpecificMimeType,
  getSpecifierKind,
  getStartupContextLength,
  getStructuredResponse,
  getStructuredResponseFromParts,
  getSubagentsRootDir,
  getSupportedImageFormatsString,
  getSystemEncoding,
  getTask,
  getTaskPath,
  getTasksDir,
  getTeamAutoMemoryIndexPath,
  getTeamAutoMemoryRoot,
  getTeamDir,
  getTeamFilePath,
  getTeamName,
  getTeammateColor,
  getTeammateContext,
  getTeamsRootDir,
  getThoughtText,
  getTimeRangeBounds,
  getTokenUsageFilePath,
  getTraceContext,
  getUnsupportedImageFormatWarning,
  getUserAutoMemoryIndexPath,
  getUserAutoMemoryRoot,
  getUserAutoMemoryTopicPath,
  getWelcomeBackState,
  globalMemoryManager,
  hasCycleInSchema,
  hasExplicitOutputLimit,
  hasImageParts,
  hasNetworkTransport,
  hasNonFinalTopLevelBackgroundOperator,
  hasShellSubstitution,
  hasUnsafeMonitorBackgroundOperator,
  hashDaemonWorkspace,
  hashMcpServerConfig,
  hashValue,
  hookEventSupportsMatcher,
  hydrateString,
  ideContextStore,
  idealabProvider,
  initializeDaemonMetrics,
  initializeTelemetry,
  injectDaemonTraceContext,
  injectPermissionRulesIfMissing,
  invokeMcpPrompt,
  isAbortError,
  isAnyAutoMemPath,
  isApiError,
  isApproveOutcome,
  isAutoEditApproved,
  isAutoMemPath,
  isBareMode,
  isBinary,
  isBinaryFile,
  isBrowserCommandBlocked,
  isClaudePluginConfig,
  isCommandAllowed,
  isCommandAvailable,
  isCommandNeedsPermission,
  isDebugLoggingDegraded,
  isDenialFallbackReason,
  isDestructiveCommand,
  isDeviceAuthorizationSuccess,
  isDeviceTokenPending,
  isDeviceTokenSuccess,
  isEditorAvailable,
  isEnabled,
  isErrorResponse,
  isGatedMcpScope,
  isGenericQuotaExceededError,
  isGitRepository,
  isImageCapable,
  isImagePart,
  isInProcessTeammate,
  isInSafeToolAllowlist,
  isInvalidArgumentError,
  isModifiableDeclarativeTool,
  isNodeError,
  isPathWithinRoot,
  isPerformanceMonitoringActive,
  isPlanModeBlocked,
  isProQuotaExceededError,
  isQwenQuotaExceededError,
  isRateLimitError,
  isRuntimeDiagnosticsEnabled,
  isSchemaDepthError,
  isScopedNpmPackage,
  isSdkMcpServerConfig,
  isSessionRuntimeActive,
  isSettledStatus,
  isStructuredError,
  isSubpath,
  isSubpaths,
  isSuccessStatus,
  isSupportedArchivePath,
  isSupportedArchiveUrl,
  isSupportedImageMimeType,
  isSystemReminderContent,
  isTeamAutoMemPath,
  isTeamLead,
  isTeammate,
  isTelemetrySdkInitialized,
  isTerminalEditor,
  isTerminalStatus,
  isTomlFormat,
  isTool,
  isToolEnabled,
  isTransientCapacityError,
  isUnattendedMode,
  isUsableImagePart,
  isUserAutoMemPath,
  isValidEditorType,
  isValidGitSha,
  isValidNonThoughtTextPart,
  isValidRefName,
  isValidSensitiveSpanAttributeMaxLength,
  isWindows,
  isWithinRoot,
  judgeGoal,
  knownTokenLimit,
  layer,
  listMcpPrompts,
  listMcpResources,
  listSavedWorkflows,
  listTasks,
  listTeamNames,
  listWorkflowSnapshots,
  loadIgnoreRules,
  loadMarketplaceConfigFromSource,
  loadServerHierarchicalMemory,
  loadUsageHistory,
  logApiCancel,
  logApiError,
  logApiRequest,
  logApiResponse,
  logArenaAgentCompleted,
  logArenaSessionEnded,
  logArenaSessionStarted,
  logAuth,
  logChatCompression,
  logConversationFinishedEvent,
  logExtensionDisable,
  logExtensionEnable,
  logExtensionInstallEvent,
  logExtensionUninstall,
  logExtensionUpdateEvent,
  logFlashFallback,
  logIdeConnection,
  logKittySequenceOverflow,
  logMemoryDream,
  logMemoryExtract,
  logMemoryRecall,
  logModelSlashCommand,
  logNextSpeakerCheck,
  logPromptSuggestion,
  logRipgrepFallback,
  logSkillLaunch,
  logSlashCommand,
  logSpeculation,
  logStartSession,
  logToolCall,
  logToolOutputTruncated,
  logUserFeedback,
  logUserPrompt,
  logUserRetry,
  logWorkflowKeyword,
  logWorkflowRun,
  makeArenaAgentCompletedEvent,
  makeArenaSessionEndedEvent,
  makeArenaSessionStartedEvent,
  makeChatCompressionEvent,
  makeFakeConfig,
  makeRelative,
  makeSlashCommandEvent,
  markDuplicateProviderToolCallResponseSent,
  matchesCommandPattern,
  matchesDomainPattern,
  matchesMcpPattern,
  matchesPathPattern,
  matchesRule,
  maybePromptForSettings,
  mcpServerRequiresOAuth,
  mcpTransportOf,
  mergeClaudeConfigs,
  metricsToUsageRecord,
  minimaxProvider,
  modelProvidersSource,
  modelRegistryKey,
  modelscopeProvider,
  modifyWithEditor,
  needsConfirmation,
  needsUtf8Bom,
  normalize,
  normalizeClaudeMcpServer,
  normalizeContent,
  normalizeModelToolCallIds,
  normalizeMonitorCommand,
  normalizeParts,
  normalizeProxyUrl,
  normalizeSkillPriority,
  normalizeStopHookBlockingCap,
  notifyGoalTerminal,
  notifyTasksUpdated,
  onTasksUpdated,
  openBrowserSecurely,
  openDiff,
  openRouterProvider,
  openaiLogger,
  oversizedMediaPlaceholder,
  parse,
  parseAndFormatApiError,
  parseBooleanEnvFlag,
  parseConnectionId,
  parseDeletedFromNameStatus,
  parseExtensionSourceType,
  parseGitDiff,
  parseGitHubRepoForReleases,
  parseGitNumstat,
  parseInsightMessage,
  parseInstallSource,
  parseLineTolerant,
  parseNpmPackageSource,
  parsePositiveIntegerEnv,
  parsePositiveIntegerEnvValue,
  parseRule,
  parseRules,
  parseShortstat,
  parseTelemetryTargetValue,
  parseThought,
  partListUnionToString,
  partToString,
  passesAcceptEditsFastPath,
  performVariableReplacement,
  persistPermissionOutcome,
  persistSessionUsage,
  populateMcpServerCommand,
  prependToFirstTextPart,
  processSingleFileContent,
  promptForSetting,
  promptIdContext,
  providerMatchesCredentials,
  queryTokenUsage,
  qwenOAuth2Events,
  read,
  readAutoMemoryIndex,
  readFileWithEncoding,
  readFileWithEncodingInfo,
  readFileWithLineAndLimit,
  readGitHead,
  readInbox,
  readLastJsonStringFieldSync,
  readLastJsonStringFieldsSync,
  readLines,
  readManyFiles,
  readPathFromWorkspace,
  readRuntimeStatus,
  readTeamFile,
  readUserAutoMemoryIndex,
  readWorktreeSession,
  readWorktreeSessionMarker,
  recordAllow,
  recordApiErrorMetrics,
  recordApiRequestBreakdown,
  recordApiResponseMetrics,
  recordApiRetry,
  recordArenaAgentCompletedMetrics,
  recordArenaSessionEndedMetrics,
  recordArenaSessionStartedMetrics,
  recordBaselineComparison,
  recordBlock,
  recordContentRetry,
  recordContentRetryFailure,
  recordCpuUsage,
  recordDaemonBridgeError,
  recordDaemonCancel,
  recordDaemonChannelLifecycle,
  recordDaemonError,
  recordDaemonHttpRequest,
  recordDaemonHttpResponse,
  recordDaemonPromptDuration,
  recordDaemonPromptQueueWait,
  recordDaemonSessionLifecycle,
  recordFallbackApprove,
  recordFileOperationMetric,
  recordGoalIteration,
  recordInvalidChunk,
  recordMemoryDreamMetrics,
  recordMemoryExtractMetrics,
  recordMemoryRecallMetrics,
  recordMemoryUsage,
  recordPerformanceRegression,
  recordPerformanceScore,
  recordSkillInvocation,
  recordStartupEvent,
  recordStartupPerformance,
  recordTokenEfficiency,
  recordTokenUsageFromApiResponse,
  recordTokenUsageFromApiResponseBestEffort,
  recordTokenUsageMetrics,
  recordToolCallMetrics,
  recordToolExecutionBreakdown,
  recordToolQueueDepth,
  recordUnavailable,
  recursivelyHydrateStrings,
  redactProxyCredentials,
  redactStructuredOutputArgsForRecording,
  redactUrlCredentials,
  refreshSessionContext,
  registerDaemonGaugeCallbacks,
  registerGoalHook,
  registerLeader,
  registerSessionCommit,
  releaseOwnedTask,
  removeMCPServerStatus,
  removeMCPStatusChangeListener,
  renameWithRetry,
  renameWithRetrySync,
  repairOrphanedToolUseTurns,
  replaceImagesWithText,
  replayUiTelemetryFromConversation,
  requestyProvider,
  resetDebugLoggingState,
  resetDenialState,
  resetEncodingCache,
  resetTaskList,
  resetUtf8BomCache,
  resolveActiveTeamName,
  resolveAndValidatePath,
  resolveBaseUrl,
  resolveBranchName,
  resolveBundleDir,
  resolveCommandPath,
  resolveContentGeneratorConfigWithSources,
  resolveExtensionConfigLocale,
  resolveField,
  resolveGitDir,
  resolveLocalizableString,
  resolveMaxConcurrentBackgroundAgents,
  resolveMetadataKey,
  resolveModelConfig,
  resolveModelId,
  resolveNpmRegistry,
  resolveOpenAILogDir,
  resolveOptionalField,
  resolveOwnsModel,
  resolvePath,
  resolvePathFromEnv,
  resolvePathPattern,
  resolveProviderProtocol,
  resolveRipgrep,
  resolveSavedWorkflowScript,
  resolveStopHookBlockingCap,
  resolveTelemetrySettings,
  resolveToolName,
  restoreWorktreeContext,
  retryWithBackoff,
  rewritePathArgs,
  runForkedAgent,
  runInSubagentSpanContext,
  runInToolSpanContext,
  runRipgrep,
  runSideQuery,
  runVisionBridge,
  runWithDaemonTelemetryContext,
  runWithDebugLogSession,
  runWithForkedChatModel,
  runWithTeammateIdentity,
  runtimeDiagnostics,
  safeAgentId,
  safeLiteralReplace,
  sanitizeCwd,
  sanitizeHookName,
  sanitizeName,
  sanitizeTitle,
  sanitizeToolCallResultForRecording,
  sanitizeUserPromptExpansionAdditionalContext,
  saveCacheSafeParams,
  saveWelcomeBackRestartChoice,
  saveWorkflowScript,
  selectVisionBridgeModel,
  sendStructuredMessage,
  serializeSnapshot,
  serializeTerminalToObject,
  serializeTerminalToText,
  sessionIdContext,
  setActiveGoal,
  setDebugLogSession,
  setGeminiMdFilename,
  setGoalTerminalObserver,
  setLastGoalTerminal,
  setMCPDiscoveryState,
  setMemberActive,
  setStartupEventSink,
  settingsSource,
  shortenPath,
  shouldAttemptBrowserLaunch,
  shouldDefaultToNodePty,
  shouldFallback,
  shouldFilterSuggestion,
  shouldFirePermissionDeniedForAutoMode,
  shouldForceAutoModeReviewForAllow,
  shouldLaunchBrowser,
  shouldRunAutoModeForCall,
  shouldRunVisionBridge,
  shouldShowStep,
  shutdownTelemetry,
  sleepInhibitor,
  splitCommands,
  splitCompoundCommand,
  splitConditionalSkills,
  splitImageParts,
  startHookSpan,
  startInteractionSpan,
  startLLMRequestSpan,
  startSpeculation,
  startSubagentSpan,
  startToolBlockedOnUserSpan,
  startToolExecutionSpan,
  startToolSpan,
  stringify,
  stripAnsiAndControl,
  stripRuntimeSnapshotPrefix,
  stripShellWrapper,
  stripStartupContext,
  stripTerminalControlSequences,
  stripTrailingBackgroundAmp,
  subagentGenerator,
  substituteHookVariables,
  summarizeAnthropicWireRequest,
  summarizeMcpResource,
  summarizeOpenAIWireRequest,
  summarizeUnifiedDiff,
  teammateIdentityStore,
  templateString,
  tildeifyPath,
  toFriendlyError,
  toSnapshot,
  tokenLimit,
  tokenPlanProvider,
  toolMatchesRuleToolName,
  truncateContent,
  truncateJson,
  truncateSpanError,
  tryGenerateSessionTitle,
  tryReclaimStaleTeam,
  uiTelemetryService,
  unassignTeammateTasks,
  unescapeJsonString,
  unescapePath,
  unregisterGoalHook,
  unregisterLeader,
  updateMCPServerStatus,
  updateSetting,
  updateSymlink,
  updateTask,
  validateModelConfig,
  validateName,
  validatePath,
  validateVariables,
  validateWorkflowName,
  watchRepoBranch,
  windowsCodePageToEncoding,
  withDaemonBridgeSpan,
  withDaemonRequestSpan,
  withDaemonSpan,
  withInteractionSpan,
  worktreeBranchForSlug,
  wrapConfirmWithBadge,
  write,
  writeLine,
  writeLineSync,
  writeMessage,
  writeRuntimeStatus,
  writeTeamFile,
  writeWorkflowSnapshot,
  writeWorkspaceContextFile,
  writeWorktreeSession,
  writeWorktreeSessionMarker,
  zaiProvider
};
