// @qwen-code/qwen-code/chunks/zh-TW-TR3XYGE5.js
// Force strict mode and setup for ESM
"use strict";
import {
  init_esbuild_shims
} from "./chunk-A4BMJM77.js";
import "./chunk-J2S4EL5Y.js";

// packages/cli/src/i18n/locales/zh-TW.js
init_esbuild_shims();
var zh_TW_default = {
  "Cannot disable an extension-provided MCP server here.": "\u7121\u6CD5\u5728\u6B64\u8655\u505C\u7528\u64F4\u5C55\u63D0\u4F9B\u7684 MCP \u4F3A\u670D\u5668\u3002",
  'Cleared authentication for "{{name}}".': '\u5DF2\u6E05\u7A7A "{{name}}" \u7684\u8A8D\u8B49\u8CC7\u8A0A\u3002',
  'MCP "{{name}}" disabled for all projects.': 'MCP "{{name}}" \u5DF2\u5728\u6240\u6709\u5C08\u6848\u4E2D\u505C\u7528\u3002',
  'Enable extension "{{name}}" to manage this MCP server.': '\u555F\u7528\u64F4\u5C55 "{{name}}" \u5F8C\u624D\u80FD\u7BA1\u7406\u6B64 MCP \u4F3A\u670D\u5668\u3002',
  "Extension-provided MCP servers cannot be favorited.": "\u64F4\u5C55\u63D0\u4F9B\u7684 MCP \u4F3A\u670D\u5668\u7121\u6CD5\u55AE\u7368\u6536\u85CF\u3002",
  "User level": "\u4F7F\u7528\u8005\u5C64\u7D1A",
  "Project level": "\u5C08\u6848\u5C64\u7D1A",
  // ==========================================================================
  // Extensions manager dialog (Installed / Discover / Sources tabs)
  // ==========================================================================
  " \xB7 {{marketplace}} (Tab to clear)": " \xB7 {{marketplace}}\uFF08Tab \u6E05\u9664\uFF09",
  '"{{name}}" {{state}}.': '"{{name}}" {{state}}\u3002',
  "(Tab / \u2190\u2192 to switch)": "\uFF08Tab / \u2190\u2192 \u5207\u63DB\uFF09",
  "+ Add new marketplace": "+ \u65B0\u589E\u5E02\u5834\u4F86\u6E90",
  "+ Install a new extension": "+ \u5B89\u88DD\u4E00\u500B\u65B0\u64F4\u5C55",
  Actions: "\u64CD\u4F5C",
  "Add Marketplace": "\u65B0\u589E\u5E02\u5834\u4F86\u6E90",
  "Add a marketplace in the Sources tab to discover extensions.": "\u5728\u300C\u4F86\u6E90\u300D\u5206\u9801\u4E2D\u65B0\u589E\u5E02\u5834\u4F86\u6E90\u4EE5\u767C\u73FE\u64F4\u5C55\u3002",
  "Add new": "\u65B0\u589E",
  "Add to Favorites": "\u52A0\u5165\u6536\u85CF",
  'Added "{{name}}" to favorites.': '\u5DF2\u5C07 "{{name}}" \u52A0\u5165\u6536\u85CF\u3002',
  'Added marketplace "{{name}}".': '\u5DF2\u65B0\u589E\u5E02\u5834\u4F86\u6E90 "{{name}}"\u3002',
  "Adding...": "\u65B0\u589E\u4E2D...",
  "Back to extension list": "\u8FD4\u56DE\u64F4\u5C55\u6E05\u55AE",
  "Browse extensions ({{count}})": "\u700F\u89BD\u64F4\u5C55\uFF08{{count}}\uFF09",
  "By: {{a}}": "\u4F5C\u8005\uFF1A{{a}}",
  "Change scope": "\u8B8A\u66F4\u4F5C\u7528\u57DF",
  'Change scope for "{{name}}":': '\u8B8A\u66F4 "{{name}}" \u7684\u4F5C\u7528\u57DF\uFF1A',
  "Changing scope...": "\u6B63\u5728\u8B8A\u66F4\u4F5C\u7528\u57DF...",
  'Uninstalling "{{name}}"...': '\u6B63\u5728\u5378\u8F09 "{{name}}"...',
  'Update available for "{{name}}".': '"{{name}}" \u6709\u53EF\u7528\u66F4\u65B0\u3002',
  '"{{name}}" is already up to date.': '"{{name}}" \u5DF2\u662F\u6700\u65B0\u3002',
  'Checking "{{name}}" for updates...': '\u6B63\u5728\u6AA2\u67E5 "{{name}}" \u7684\u66F4\u65B0...',
  '"{{name}}" does not support update checks.': '"{{name}}" \u4E0D\u652F\u63F4\u6AA2\u67E5\u66F4\u65B0\u3002',
  '"{{name}}" cannot be update-checked (Claude marketplace plugins update by reinstalling).': '"{{name}}" \u7121\u6CD5\u6AA2\u67E5\u66F4\u65B0\uFF08Claude \u5E02\u5834\u6E90\u63D2\u4EF6\u9700\u5378\u8F09\u5F8C\u91CD\u88DD\u4F86\u66F4\u65B0\uFF09\u3002',
  'Failed to check "{{name}}" for updates.': '\u6AA2\u67E5 "{{name}}" \u7684\u66F4\u65B0\u5931\u6557\u3002',
  "Claude plugin marketplace": "Claude \u5916\u639B\u5E02\u5834",
  Commands: "\u547D\u4EE4",
  "Components:": "\u5143\u4EF6\uFF1A",
  "Could not load this marketplace.": "\u7121\u6CD5\u8F09\u5165\u6B64\u5E02\u5834\u4F86\u6E90\u3002",
  "Current: {{scope}}": "\u76EE\u524D\uFF1A{{scope}}",
  Disabled: "\u5DF2\u7981\u7528",
  Discover: "\u767C\u73FE",
  'Disabling "{{name}}"...': '\u6B63\u5728\u7981\u7528 "{{name}}"...',
  'Disabling MCP "{{name}}"...': '\u6B63\u5728\u7981\u7528 MCP "{{name}}"...',
  "Discover extensions": "\u767C\u73FE\u64F4\u5C55",
  "Discovering extensions...": "\u6B63\u5728\u767C\u73FE\u64F4\u5C55...",
  'Enabling "{{name}}"...': '\u6B63\u5728\u555F\u7528 "{{name}}"...',
  'Enabling MCP "{{name}}"...': '\u6B63\u5728\u555F\u7528 MCP "{{name}}"...',
  "Enter extension source:": "\u8F38\u5165\u64F4\u5C55\u4F86\u6E90\uFF1A",
  "Enter marketplace source (Claude format):": "\u8F38\u5165\u5E02\u5834\u4F86\u6E90\u4F4D\u5740\uFF08Claude \u683C\u5F0F\uFF09\uFF1A",
  "Examples:": "\u7BC4\u4F8B\uFF1A",
  "Extension details": "\u64F4\u5C55\u8A73\u60C5",
  "Extension v{{version}}": "\u64F4\u5C55 v{{version}}",
  "Extensions are not available in this environment.": "\u76EE\u524D\u74B0\u5883\u4E2D\u64F4\u5C55\u4E0D\u53EF\u7528\u3002",
  "Failed to open {{url}}": "\u958B\u555F {{url}} \u5931\u6557",
  Favorites: "\u6536\u85CF",
  "Global (User Scope)": "\u5168\u57DF\uFF08\u4F7F\u7528\u8005\u4F5C\u7528\u57DF\uFF09",
  "Install Extension": "\u5B89\u88DD\u64F4\u5C55",
  "Install for the current workspace (project scope)": "\u70BA\u76EE\u524D\u5DE5\u4F5C\u5340\u5B89\u88DD\uFF08\u5C08\u6848\u4F5C\u7528\u57DF\uFF09",
  "Install for you (user scope)": "\u50C5\u70BA\u4F60\u5B89\u88DD\uFF08\u4F7F\u7528\u8005\u4F5C\u7528\u57DF\uFF09",
  "Install {{count}} extension(s) to which scope?": "\u5C07 {{count}} \u500B\u64F4\u5C55\u5B89\u88DD\u5230\u54EA\u500B\u4F5C\u7528\u57DF\uFF1F",
  Installed: "\u5DF2\u5B89\u88DD",
  'Installed extension "{{name}}".': '\u5DF2\u5B89\u88DD\u64F4\u5C55 "{{name}}"\u3002',
  "Installed extensions ({{count}}):": "\u5DF2\u5B89\u88DD\u7684\u64F4\u5C55\uFF08{{count}}\uFF09\uFF1A",
  "Installed {{count}} extension(s).": "\u5DF2\u5B89\u88DD {{count}} \u500B\u64F4\u5C55\u3002",
  "{{name}}: installed, but the scope rollback failed \u2014 it may be disabled at all scopes; re-enable it from the Installed tab.": "{{name}}\uFF1A\u5DF2\u5B89\u88DD,\u4F46\u4F5C\u7528\u57DF\u56DE\u6EFE\u5931\u6557 \u2014\u2014 \u8A72\u64F4\u5C55\u53EF\u80FD\u5728\u6240\u6709\u4F5C\u7528\u57DF\u5747\u88AB\u505C\u7528;\u8ACB\u5728\u300C\u5DF2\u5B89\u88DD\u300D\u9801\u91CD\u65B0\u555F\u7528\u3002",
  'Could not change scope, and the rollback also failed \u2014 "{{name}}" may be disabled at all scopes. Re-enable it from the Installed tab. ({{error}})': "\u7121\u6CD5\u8B8A\u66F4\u4F5C\u7528\u57DF,\u4E14\u56DE\u6EFE\u4E5F\u5931\u6557 \u2014\u2014\u300C{{name}}\u300D\u53EF\u80FD\u5728\u6240\u6709\u4F5C\u7528\u57DF\u5747\u88AB\u505C\u7528\u3002\u8ACB\u5728\u300C\u5DF2\u5B89\u88DD\u300D\u9801\u91CD\u65B0\u555F\u7528\u3002({{error}})",
  "Installed {{ok}}, failed {{fail}}: {{detail}}": "\u6210\u529F {{ok}} \u500B\uFF0C\u5931\u6557 {{fail}} \u500B\uFF1A{{detail}}",
  "Installing...": "\u5B89\u88DD\u4E2D...",
  "Last updated: {{date}}": "\u6700\u8FD1\u66F4\u65B0\uFF1A{{date}}",
  MCP: "MCP",
  'MCP "{{name}}" {{state}}.': 'MCP "{{name}}" {{state}}\u3002',
  "MCP servers": "MCP \u4F3A\u670D\u5668",
  "Mark for Update": "\u6A19\u8A18\u70BA\u5F85\u66F4\u65B0",
  Marketplaces: "\u5E02\u5834\u4F86\u6E90",
  "No extensions discovered.": "\u672A\u767C\u73FE\u4EFB\u4F55\u64F4\u5C55\u3002",
  "No extensions match your search.": "\u6C92\u6709\u8207\u641C\u5C0B\u76F8\u7B26\u7684\u64F4\u5C55\u3002",
  "No extensions or marketplaces added yet.": "\u5C1A\u672A\u65B0\u589E\u4EFB\u4F55\u64F4\u5C55\u6216\u5E02\u5834\u4F86\u6E90\u3002",
  "No homepage available.": "\u6C92\u6709\u53EF\u7528\u7684\u4E3B\u9801\u3002",
  "No installable extensions selected.": "\u672A\u9078\u53D6\u53EF\u5B89\u88DD\u7684\u64F4\u5C55\u3002",
  "No plugins or MCP servers installed.": "\u5C1A\u672A\u5B89\u88DD\u4EFB\u4F55\u5916\u639B\u6216 MCP \u4F3A\u670D\u5668\u3002",
  None: "\u7121",
  "Note: Uninstall permanently removes this extension.": "\u6CE8\u610F\uFF1A\u5378\u8F09\u5C07\u6C38\u4E45\u79FB\u9664\u6B64\u64F4\u5C55\u3002",
  "Open homepage": "\u958B\u555F\u4E3B\u9801",
  "Project (Workspace)": "\u5C08\u6848\uFF08\u5DE5\u4F5C\u5340\uFF09",
  "Refreshed {{count}} extension(s).": "\u5DF2\u5237\u65B0 {{count}} \u500B\u64F4\u5145\u3002",
  "Remove from Favorites": "\u5F9E\u6536\u85CF\u4E2D\u79FB\u9664",
  "Remove marketplace": "\u79FB\u9664\u5E02\u5834\u4F86\u6E90",
  'Remove marketplace "{{name}}"?': '\u79FB\u9664\u5E02\u5834\u4F86\u6E90 "{{name}}"\uFF1F',
  'Removed "{{name}}" from favorites.': '\u5DF2\u5C07 "{{name}}" \u5F9E\u6536\u85CF\u4E2D\u79FB\u9664\u3002',
  'Removed marketplace "{{name}}".': '\u5DF2\u79FB\u9664\u5E02\u5834\u4F86\u6E90 "{{name}}"\u3002',
  "Scope:": "\u4F5C\u7528\u57DF\uFF1A",
  'Set "{{name}}" scope to {{scope}}.': '\u5DF2\u5C07 "{{name}}" \u7684\u4F5C\u7528\u57DF\u8A2D\u70BA {{scope}}\u3002',
  Sources: "\u4F86\u6E90",
  "Type to search \xB7 Space to toggle \xB7 Enter to view \xB7 Ctrl+R refresh \xB7 Esc to go back": "\u8F38\u5165\u4EE5\u641C\u5C0B \xB7 Space \u5207\u63DB \xB7 Enter \u67E5\u770B \xB7 Ctrl+R \u5237\u65B0 \xB7 Esc \u8FD4\u56DE",
  Uninstall: "\u5378\u8F09",
  'Uninstalled "{{name}}".': '\u5DF2\u5378\u8F09 "{{name}}"\u3002',
  "Update Now": "\u7ACB\u5373\u66F4\u65B0",
  "Update marketplace": "\u66F4\u65B0\u5E02\u5834\u4F86\u6E90",
  "Update marketplace (last updated {{date}})": "\u66F4\u65B0\u5E02\u5834\u4F86\u6E90\uFF08\u6700\u8FD1\u66F4\u65B0 {{date}}\uFF09",
  'Could not update marketplace "{{name}}".': '\u7121\u6CD5\u66F4\u65B0\u5E02\u5834\u4F86\u6E90 "{{name}}"\u3002',
  'Updated "{{name}}".': '\u5DF2\u66F4\u65B0 "{{name}}"\u3002',
  'Updated marketplace "{{name}}".': '\u5DF2\u66F4\u65B0\u5E02\u5834\u4F86\u6E90 "{{name}}"\u3002',
  "Use the Discover tab to find and install plugins.": "\u4F7F\u7528\u300C\u767C\u73FE\u300D\u5206\u9801\u5C0B\u627E\u4E26\u5B89\u88DD\u64F4\u5C55\u3002",
  "Version: {{v}}": "\u7248\u672C\uFF1A{{v}}",
  "Will install:": "\u5C07\u5B89\u88DD\uFF1A",
  "Would open: {{url}}": "\u5C07\u958B\u555F\uFF1A{{url}}",
  "Y/Enter to confirm \xB7 N/Esc to cancel": "Y/Enter \u78BA\u8A8D \xB7 N/Esc \u53D6\u6D88",
  "Press R to retry \xB7 Esc to go back": "\u6309 R \u91CD\u8A66 \xB7 Esc \u8FD4\u56DE",
  "Enter to select \xB7 R refresh \xB7 Esc to go back": "Enter \u9078\u64C7 \xB7 R \u5237\u65B0 \xB7 Esc \u8FD4\u56DE",
  "from {{marketplace}}": "\u4F86\u81EA {{marketplace}}",
  installed: "\u5DF2\u5B89\u88DD",
  "{{count}} Agents": "{{count}} \u500B\u667A\u80FD\u9AD4",
  "{{count}} Commands": "{{count}} \u500B\u547D\u4EE4",
  "{{count}} MCP": "{{count}} \u500B MCP",
  "{{count}} Skills": "{{count}} \u500B\u6280\u80FD",
  "{{count}} available extensions": "{{count}} \u500B\u53EF\u7528\u64F4\u5C55",
  "\u2191 more above": "\u2191 \u4E0A\u65B9\u66F4\u591A",
  "\u2191\u2193 navigate \xB7 Enter open \xB7 d remove marketplace \xB7 Esc close": "\u2191\u2193 \u5C0E\u89BD \xB7 Enter \u958B\u555F \xB7 d \u79FB\u9664\u5E02\u5834\u4F86\u6E90 \xB7 Esc \u95DC\u9589",
  "\u2191\u2193 navigate \xB7 Enter select \xB7 Esc close": "\u2191\u2193 \u5C0E\u89BD \xB7 Enter \u9078\u64C7 \xB7 Esc \u95DC\u9589",
  "\u2191\u2193 navigate \xB7 Enter select \xB7 d remove marketplace \xB7 Esc close": "\u2191\u2193 \u5C0E\u89BD \xB7 Enter \u9078\u64C7 \xB7 d \u79FB\u9664\u5E02\u5834\u4F86\u6E90 \xB7 Esc \u95DC\u9589",
  "\u2191\u2193 navigate \xB7 Space enable/disable \xB7 f favorite \xB7 Enter details \xB7 Esc close": "\u2191\u2193 \u5C0E\u89BD \xB7 Space \u555F\u7528/\u7981\u7528 \xB7 f \u6536\u85CF \xB7 Enter \u67E5\u770B\u8A73\u60C5 \xB7 Esc \u95DC\u9589",
  "\u2193 more below": "\u2193 \u4E0B\u65B9\u66F4\u591A",
  "\u26A0 Make sure you trust an extension before installing, updating, or using it. We cannot verify what MCP servers, files, or other software an extension includes, or that it works as intended. See the extension homepage for more information.": "\u26A0 \u5728\u5B89\u88DD\u3001\u66F4\u65B0\u6216\u4F7F\u7528\u64F4\u5C55\u524D\uFF0C\u8ACB\u78BA\u4FDD\u4F60\u4FE1\u4EFB\u5B83\u3002\u6211\u5011\u7121\u6CD5\u9A57\u8B49\u64F4\u5C55\u5305\u542B\u54EA\u4E9B MCP \u4F3A\u670D\u5668\u3001\u6A94\u6848\u6216\u5176\u4ED6\u8EDF\u9AD4\uFF0C\u4E5F\u7121\u6CD5\u4FDD\u8B49\u5176\u6309\u9810\u671F\u904B\u4F5C\u3002\u66F4\u591A\u8CC7\u8A0A\u8ACB\u67E5\u770B\u64F4\u5C55\u4E3B\u9801\u3002",
  // Tool display names (chat-stream badge labels)
  // ----------------------------------------------------------------------------
  // Keyed by `toolDisplayName.<English display name>` (from core
  // `ToolDisplayNames`); a missing key falls back to the English display name
  // via `localizeToolDisplayName`. A product name (e.g. `Notebook`) is kept
  // verbatim inside an otherwise-translated label.
  // ============================================================================
  "toolDisplayName.Edit": "\u7DE8\u8F2F",
  "toolDisplayName.WriteFile": "\u5BEB\u5165\u6A94\u6848",
  "toolDisplayName.ReadFile": "\u8B80\u53D6\u6A94\u6848",
  "toolDisplayName.Grep": "Grep",
  "toolDisplayName.Glob": "Glob",
  "toolDisplayName.Shell": "\u904B\u884C\u547D\u4EE4",
  "toolDisplayName.Shell Command": "Shell \u547D\u4EE4",
  "toolDisplayName.TodoList": "\u4EFB\u52D9\u6E05\u55AE",
  "toolDisplayName.SaveMemory": "\u5132\u5B58\u8A18\u61B6",
  "toolDisplayName.Agent": "Agent",
  "toolDisplayName.Artifact": "\u88FD\u54C1",
  "toolDisplayName.Skill": "\u6280\u80FD",
  "toolDisplayName.EnterPlanMode": "\u9032\u5165\u8A08\u756B\u6A21\u5F0F",
  "toolDisplayName.ExitPlanMode": "\u9000\u51FA\u8A08\u756B\u6A21\u5F0F",
  "toolDisplayName.WebFetch": "\u7DB2\u8DEF\u64F7\u53D6",
  "toolDisplayName.WebSearch": "\u7DB2\u8DEF\u641C\u5C0B",
  "toolDisplayName.ListFiles": "\u5217\u51FA\u6A94\u6848",
  "toolDisplayName.Lsp": "LSP",
  "toolDisplayName.AskUserQuestion": "\u8A62\u554F\u4F7F\u7528\u8005",
  "toolDisplayName.CronCreate": "\u5EFA\u7ACB\u5B9A\u6642\u4EFB\u52D9",
  "toolDisplayName.CronList": "\u5B9A\u6642\u4EFB\u52D9\u6E05\u55AE",
  "toolDisplayName.CronDelete": "\u522A\u9664\u5B9A\u6642\u4EFB\u52D9",
  "toolDisplayName.LoopWakeup": "\u5FAA\u74B0\u559A\u9192",
  "toolDisplayName.TaskCreate": "\u5EFA\u7ACB\u4EFB\u52D9",
  "toolDisplayName.TaskUpdate": "\u66F4\u65B0\u4EFB\u52D9",
  "toolDisplayName.TaskList": "\u4EFB\u52D9\u5217\u8868",
  "toolDisplayName.TaskStop": "\u505C\u6B62\u4EFB\u52D9",
  "toolDisplayName.TeamCreate": "\u5EFA\u7ACB\u5718\u968A",
  "toolDisplayName.TeamDelete": "\u522A\u9664\u5718\u968A",
  "toolDisplayName.SendMessage": "\u50B3\u9001\u8A0A\u606F",
  "toolDisplayName.StructuredOutput": "\u7D50\u69CB\u5316\u8F38\u51FA",
  "toolDisplayName.Monitor": "\u76E3\u63A7",
  "toolDisplayName.NotebookEdit": "\u7DE8\u8F2F Notebook",
  "toolDisplayName.ToolSearch": "\u5DE5\u5177\u641C\u5C0B",
  "toolDisplayName.EnterWorktree": "\u9032\u5165 Worktree",
  "toolDisplayName.ExitWorktree": "\u9000\u51FA Worktree",
  "toolDisplayName.Workflow": "\u5DE5\u4F5C\u6D41\u7A0B",
  "toolDisplayName.ReadMcpResource": "\u8B80\u53D6 MCP \u8CC7\u6E90",
  "\u2191 to manage attachments": "\u2191 \u7BA1\u7406\u9644\u4EF6",
  "\u2190 \u2192 select, Delete to remove, \u2193 to exit": "\u2190 \u2192 \u9078\u64C7\uFF0CDelete \u522A\u9664\uFF0C\u2193 \u9000\u51FA",
  "Attachments: ": "\u9644\u4EF6\uFF1A",
  "Basics:": "\u57FA\u790E\u529F\u80FD\uFF1A",
  "Add context": "\u6DFB\u52A0\u4E0A\u4E0B\u6587",
  "Use {{symbol}} to specify files for context (e.g., {{example}}) to target specific files or folders.": "\u4F7F\u7528 {{symbol}} \u6307\u5B9A\u6A94\u6848\u4F5C\u70BA\u4E0A\u4E0B\u6587\uFF08\u4F8B\u5982\uFF0C{{example}}\uFF09\uFF0C\u7528\u65BC\u5B9A\u4F4D\u7279\u5B9A\u6A94\u6848\u6216\u6A94\u6848\u593E",
  "@": "@",
  "@src/myFile.ts": "@src/myFile.ts",
  "Shell mode": "Shell \u6A21\u5F0F",
  "YOLO mode": "YOLO \u6A21\u5F0F",
  "Auto mode": "Auto \u6A21\u5F0F",
  "plan mode": "\u898F\u5283\u6A21\u5F0F",
  "auto-accept edits": "\u81EA\u52D5\u63A5\u53D7\u7DE8\u8F2F",
  "Accepting edits": "\u63A5\u53D7\u7DE8\u8F2F",
  "(shift + tab to cycle)": "(Shift + Tab \u5207\u63DB)",
  "(tab to cycle)": "(\u6309 Tab \u5207\u63DB)",
  "Execute shell commands via {{symbol}} (e.g., {{example1}}) or use natural language (e.g., {{example2}}).": "\u901A\u904E {{symbol}} \u57F7\u884C shell \u547D\u4EE4\uFF08\u4F8B\u5982\uFF0C{{example1}}\uFF09\u6216\u4F7F\u7528\u81EA\u7136\u8A9E\u8A00\uFF08\u4F8B\u5982\uFF0C{{example2}}\uFF09",
  "!": "!",
  "!npm run start": "!npm run start",
  "Commands:": "\u547D\u4EE4\uFF1A",
  "shell command": "shell \u547D\u4EE4",
  "Model Context Protocol command (from external servers)": "Model Context Protocol \u547D\u4EE4\uFF08\u4F86\u81EA\u5916\u90E8\u4F3A\u670D\u5668\uFF09",
  "Keyboard Shortcuts:": "\u9375\u76E4\u5FEB\u6377\u9375\uFF1A",
  "Toggle this help display": "\u5207\u63DB\u6B64\u5E6B\u52A9\u986F\u793A",
  "Toggle shell mode": "\u5207\u63DB\u547D\u4EE4\u884C\u6A21\u5F0F",
  "Open command menu": "\u6253\u958B\u547D\u4EE4\u9078\u55AE",
  "Add file context": "\u6DFB\u52A0\u6A94\u6848\u4E0A\u4E0B\u6587",
  "Accept suggestion / Autocomplete": "\u63A5\u53D7\u5EFA\u8B70 / \u81EA\u52D5\u88DC\u5168",
  "Reverse search history": "\u53CD\u5411\u641C\u7D22\u6B77\u53F2",
  "Press ? again to close": "\u518D\u6B21\u6309 ? \u95DC\u9589",
  "for shell mode": "\u547D\u4EE4\u884C\u6A21\u5F0F",
  "for commands": "\u547D\u4EE4\u9078\u55AE",
  "for file paths": "\u6A94\u6848\u8DEF\u5F91",
  "to clear input": "\u6E05\u7A7A\u8F38\u5165",
  "to cycle approvals": "\u5207\u63DB\u5BE9\u6279\u6A21\u5F0F",
  "to quit": "\u9000\u51FA",
  "for newline": "\u63DB\u884C",
  "to clear screen": "\u6E05\u5C4F",
  "to search history": "\u641C\u7D22\u6B77\u53F2",
  "to paste images": "\u7C98\u8CBC\u5716\u7247",
  "for external editor": "\u5916\u90E8\u7DE8\u8F2F\u5668",
  "to toggle compact mode": "\u5207\u63DB\u7DCA\u6E4A\u6A21\u5F0F",
  "Jump through words in the input": "\u5728\u8F38\u5165\u4E2D\u6309\u55AE\u8A5E\u8DF3\u8F49",
  "Close dialogs, cancel requests, or quit application": "\u95DC\u9589\u5C0D\u8A71\u6846\u3001\u53D6\u6D88\u8ACB\u6C42\u6216\u9000\u51FA\u61C9\u7528\u7A0B\u5E8F",
  "New line": "\u63DB\u884C",
  "New line (Alt+Enter works for certain linux distros)": "\u63DB\u884C\uFF08\u67D0\u4E9B Linux \u767C\u884C\u7248\u652F\u6301 Alt+Enter\uFF09",
  "Clear the screen": "\u6E05\u5C4F",
  "Open input in external editor": "\u5728\u5916\u90E8\u7DE8\u8F2F\u5668\u4E2D\u6253\u958B\u8F38\u5165",
  "Send message": "\u767C\u9001\u6D88\u606F",
  "Initializing...": "\u6B63\u5728\u521D\u59CB\u5316...",
  "Connecting to MCP servers... ({{connected}}/{{total}})": "\u6B63\u5728\u9023\u63A5\u5230 MCP servers... ({{connected}}/{{total}})",
  "Type your message or @path/to/file": "\u8F38\u5165\u60A8\u7684\u6D88\u606F\u6216 @ \u6A94\u6848\u8DEF\u5F91",
  "? for shortcuts": "\u6309 ? \u67E5\u770B\u5FEB\u6377\u9375",
  "Press 'i' for INSERT mode and 'Esc' for NORMAL mode.": "\u6309 'i' \u9032\u5165\u63D2\u5165\u6A21\u5F0F\uFF0C\u6309 'Esc' \u9032\u5165\u666E\u901A\u6A21\u5F0F",
  "Cancel operation / Clear input (double press)": "\u53D6\u6D88\u64CD\u4F5C / \u6E05\u7A7A\u8F38\u5165\uFF08\u96D9\u64CA\uFF09",
  "Cycle approval modes": "\u5FAA\u74B0\u5207\u63DB\u5BE9\u6279\u6A21\u5F0F",
  "Cycle through your prompt history": "\u5FAA\u74B0\u700F\u89BD\u63D0\u793A\u6B77\u53F2",
  "For a full list of shortcuts, see {{docPath}}": "\u5B8C\u6574\u5FEB\u6377\u9375\u5217\u8868\uFF0C\u8ACB\u53C3\u95B1 {{docPath}}",
  "docs/keyboard-shortcuts.md": "docs/keyboard-shortcuts.md",
  "for help on Qwen Code": "\u7372\u53D6 Qwen Code \u5E6B\u52A9",
  "show version info": "\u986F\u793A\u7248\u672C\u4FE1\u606F",
  "show paths for current session files and logs": "\u986F\u793A\u76EE\u524D\u6703\u8A71\u6A94\u6848\u548C\u65E5\u8A8C\u8DEF\u5F91",
  "submit a bug report": "\u63D0\u4EA4\u932F\u8AA4\u5831\u544A",
  Status: "\u72C0\u614B",
  "Qwen Code": "Qwen Code",
  Runtime: "\u904B\u884C\u74B0\u5883",
  OS: "\u64CD\u4F5C\u7CFB\u7D71",
  Auth: "\u8A8D\u8B49",
  Model: "\u6A21\u578B",
  "Fast Model": "\u5FEB\u901F\u6A21\u578B",
  Sandbox: "\u6C99\u7BB1",
  "Session ID": "\u6703\u8A71 ID",
  "Base URL": "Base URL",
  Proxy: "\u4EE3\u7406",
  "Memory Usage": "\u5167\u5B58\u4F7F\u7528",
  "IDE Client": "IDE \u5BA2\u6236\u7AEF",
  "Analyzes the project and creates a tailored QWEN.md file.": "\u5206\u6790\u9805\u76EE\u4E26\u5275\u5EFA\u5B9A\u88FD\u7684 QWEN.md \u6A94\u6848",
  "List available Qwen Code tools. Usage: /tools [desc]": "\u5217\u51FA\u53EF\u7528\u7684 Qwen Code \u5DE5\u5177\u3002\u7528\u6CD5\uFF1A/tools [desc]",
  "Open the skills panel (browse, search, toggle, pick).": "\u958B\u555F\u6280\u80FD\u9762\u677F\uFF08\u700F\u89BD\u3001\u641C\u5C0B\u3001\u555F\u505C\u3001\u9078\u64C7\uFF09\u3002",
  "Move this session to a new working directory": "\u5C07\u6B64\u6703\u8A71\u79FB\u52D5\u5230\u65B0\u7684\u5DE5\u4F5C\u76EE\u9304",
  "Manage Skills": "\u7BA1\u7406\u6280\u80FD",
  "Skills configuration saved.": "\u6280\u80FD\u8A2D\u5B9A\u5DF2\u5132\u5B58\u3002",
  "Skills configuration saved, but refresh failed: {{error}}. Restart to ensure the new state is applied.": "\u6280\u80FD\u8A2D\u5B9A\u5DF2\u5132\u5B58\uFF0C\u4F46\u91CD\u65B0\u6574\u7406\u5931\u6557\uFF1A{{error}}\u3002\u8ACB\u91CD\u65B0\u555F\u52D5\u4EE5\u78BA\u4FDD\u65B0\u72C0\u614B\u751F\u6548\u3002",
  "Workspace is untrusted; workspace settings are ignored by the merged config. Run /trust first to persist skills changes here, or edit ~/.qwen/settings.json directly to manage skills at user scope.": "\u76EE\u524D\u5DE5\u4F5C\u5340\u672A\u53D7\u4FE1\u4EFB\uFF0C\u5DE5\u4F5C\u5340\u8A2D\u5B9A\u6703\u88AB\u5408\u4F75\u8A2D\u5B9A\u5FFD\u7565\u3002\u8ACB\u5148\u57F7\u884C /trust\uFF0C\u6216\u76F4\u63A5\u7DE8\u8F2F ~/.qwen/settings.json \u5728\u4F7F\u7528\u8005\u7BC4\u570D\u7BA1\u7406\u6280\u80FD\u3002",
  "SkillManager not available.": "SkillManager \u4E0D\u53EF\u7528\u3002",
  "Loading skills\u2026": "\u6B63\u5728\u8F09\u5165\u6280\u80FD\u2026",
  "Failed to load skills: {{error}}": "\u8F09\u5165\u6280\u80FD\u5931\u6557\uFF1A{{error}}",
  "Failed to save skills configuration: {{error}}": "\u5132\u5B58\u6280\u80FD\u8A2D\u5B9A\u5931\u6557\uFF1A{{error}}",
  "All available skills are disabled. Edit ~/.qwen/settings.json or .qwen/settings.json (skills.disabled) to re-enable.": "\u6240\u6709\u53EF\u7528\u6280\u80FD\u7686\u5DF2\u505C\u7528\u3002\u8ACB\u7DE8\u8F2F ~/.qwen/settings.json \u6216 .qwen/settings.json\uFF08skills.disabled\uFF09\u4EE5\u91CD\u65B0\u555F\u7528\u3002",
  "Press esc to close.": "\u6309 Esc \u95DC\u9589\u3002",
  "{{count}} skills \xB7 ": "{{count}} \u500B\u6280\u80FD \xB7 ",
  "{{matched}} / {{total}} skills \xB7 ": "{{matched}} / {{total}} \u500B\u6280\u80FD \xB7 ",
  "Space toggle \xB7 Enter pick (fill input) \xB7 Esc save & exit \xB7 workspace scope": "\u7A7A\u767D\u9375 \u555F\u505C \xB7 \u56DE\u8ECA \u9078\u53D6(\u586B\u5165\u8F38\u5165\u6846) \xB7 Esc \u5132\u5B58\u4E26\u96E2\u958B \xB7 \u5DE5\u4F5C\u5340\u7BC4\u570D",
  "Search:": "\u641C\u5C0B\uFF1A",
  "type to filter\u2026": "\u8F38\u5165\u4EE5\u7BE9\u9078\u2026",
  "No skills are currently available.": "\u76EE\u524D\u6C92\u6709\u53EF\u7528\u7684\u6280\u80FD\u3002",
  "All available skills are locked at a higher scope (see below).": "\u6240\u6709\u53EF\u7528\u6280\u80FD\u90FD\u88AB\u66F4\u9AD8\u7BC4\u570D\u9396\u5B9A\uFF08\u8A73\u898B\u4E0B\u65B9\uFF09\u3002",
  "No skills match the search.": "\u6C92\u6709\u7B26\u5408\u641C\u5C0B\u689D\u4EF6\u7684\u6280\u80FD\u3002",
  "Locked by higher-scope settings (cannot toggle here):": "\u88AB\u66F4\u9AD8\u7BC4\u570D\u8A2D\u5B9A\u9396\u5B9A\uFF08\u6B64\u8655\u7121\u6CD5\u5207\u63DB\uFF09\uFF1A",
  "higher scope": "\u66F4\u9AD8\u7BC4\u570D",
  "  {{name}} {{description}}  [locked: {{scope}}]": "  {{name}} {{description}}  [\u5DF2\u9396\u5B9A\uFF1A{{scope}}]",
  "\u2191/\u2193 navigate \xB7 backspace edits search": "\u2191/\u2193 \u5C0E\u89BD \xB7 \u5012\u9000 \u7DE8\u8F2F\u641C\u5C0B",
  Bundled: "\u5167\u5EFA",
  "Available Qwen Code CLI tools:": "\u53EF\u7528\u7684 Qwen Code CLI \u5DE5\u5177\uFF1A",
  "No tools available": "\u6C92\u6709\u53EF\u7528\u5DE5\u5177",
  "View or change the approval mode for tool usage": "\u67E5\u770B\u6216\u66F4\u6539\u5DE5\u5177\u4F7F\u7528\u7684\u5BE9\u6279\u6A21\u5F0F",
  'Invalid approval mode "{{arg}}". Valid modes: {{modes}}': '\u7121\u6548\u7684\u5BE9\u6279\u6A21\u5F0F "{{arg}}"\u3002\u6709\u6548\u6A21\u5F0F\uFF1A{{modes}}',
  'Approval mode set to "{{mode}}"': '\u5BE9\u6279\u6A21\u5F0F\u5DF2\u8A2D\u7F6E\u70BA "{{mode}}"',
  "View or change the language setting": "\u67E5\u770B\u6216\u66F4\u6539\u8A9E\u8A00\u8A2D\u7F6E",
  "List background tasks (text dump \u2014 interactive dialog opens via the footer pill)": "\u5217\u51FA\u80CC\u666F\u4EFB\u52D9\uFF08\u6587\u5B57\u5217\u8868\uFF1B\u4E92\u52D5\u5F0F\u5C0D\u8A71\u6846\u53EF\u900F\u904E\u9801\u8173\u4E2D\u7684\u300C\u80CC\u666F\u4EFB\u52D9\u300D\u5165\u53E3\u958B\u555F\uFF09",
  "Delete a previous session": "\u522A\u9664\u5148\u524D\u7684\u6703\u8A71",
  "Run installation and environment diagnostics": "\u57F7\u884C\u5B89\u88DD\u8207\u74B0\u5883\u8A3A\u65B7",
  "Browse dynamic model catalogs and choose which models stay enabled locally": "\u700F\u89BD\u52D5\u614B\u6A21\u578B\u76EE\u9304\uFF0C\u4E26\u9078\u64C7\u8981\u5728\u672C\u6A5F\u4FDD\u6301\u555F\u7528\u7684\u6A21\u578B",
  "Generate a one-line session recap now": "\u7ACB\u5373\u751F\u6210\u4E00\u689D\u55AE\u884C\u6703\u8A71\u56DE\u9867",
  "Rename the current conversation. --auto lets the fast model pick a title.": "\u91CD\u65B0\u547D\u540D\u76EE\u524D\u7684\u5C0D\u8A71\u3002--auto \u6703\u8B93\u5FEB\u901F\u6A21\u578B\u81EA\u52D5\u7522\u751F\u6A19\u984C\u3002",
  "Rewind conversation to a previous turn": "\u5C07\u5C0D\u8A71\u56DE\u9000\u5230\u5148\u524D\u7684\u67D0\u4E00\u8F2A",
  "Rewind Conversation": "\u56DE\u9000\u5C0D\u8A71",
  "No user turns to rewind to.": "\u6C92\u6709\u53EF\u56DE\u9000\u7684\u4F7F\u7528\u8005\u5C0D\u8A71\u8F2A\u6B21\u3002",
  "Rewind to: ": "\u56DE\u9000\u5230\uFF1A",
  "Restore code and conversation": "\u6062\u5FA9\u7A0B\u5F0F\u78BC\u548C\u5C0D\u8A71",
  "Restore conversation only": "\u50C5\u6062\u5FA9\u5C0D\u8A71",
  "Restore code only": "\u50C5\u6062\u5FA9\u7A0B\u5F0F\u78BC",
  "Never mind": "\u7B97\u4E86",
  "Computing file changes...": "\u6B63\u5728\u8A08\u7B97\u6A94\u6848\u8B8A\u66F4...",
  "Restoring...": "\u6B63\u5728\u6062\u5FA9...",
  "Restored {{count}} file(s).": "\u5DF2\u6062\u5FA9 {{count}} \u500B\u6A94\u6848\u3002",
  "Failed to restore files: {{error}}": "\u6062\u5FA9\u6A94\u6848\u5931\u6557\uFF1A{{error}}",
  "Rewind failed: {{error}}": "\u56DE\u9000\u5931\u6557\uFF1A{{error}}",
  "Cannot rewind conversation: no active model client.": "\u7121\u6CD5\u56DE\u9000\u5C0D\u8A71\uFF1A\u6A21\u578B\u5BA2\u6236\u7AEF\u672A\u555F\u7528\u3002",
  "Code restored, but conversation could not be rewound (no active client).": "\u7A0B\u5F0F\u78BC\u5DF2\u6062\u5FA9\uFF0C\u4F46\u5C0D\u8A71\u7121\u6CD5\u56DE\u9000\uFF08\u6A21\u578B\u5BA2\u6236\u7AEF\u672A\u555F\u7528\uFF09\u3002",
  "Conversation rewound. Edit your prompt and press Enter to continue.": "\u5C0D\u8A71\u5DF2\u56DE\u9000\u3002\u4FEE\u6539\u63D0\u793A\u5F8C\u6309 Enter \u7E7C\u7E8C\u3002",
  "Rewinding does not affect files edited manually or via shell commands.": "\u56DE\u9000\u4E0D\u6703\u5F71\u97FF\u624B\u52D5\u7DE8\u8F2F\u6216\u900F\u904E shell \u547D\u4EE4\u4FEE\u6539\u7684\u6A94\u6848\u3002",
  "Cannot rewind to a turn that was compressed. Try a more recent turn.": "\u7121\u6CD5\u56DE\u9000\u5230\u5DF2\u88AB\u58D3\u7E2E\u7684\u8F2A\u6B21\uFF0C\u8ACB\u5617\u8A66\u66F4\u8FD1\u4E00\u4E9B\u7684\u8F2A\u6B21\u3002",
  "File restore is unavailable for this turn (no captured file changes, or this turn predates the current session).": "\u8A72\u8F2A\u6B21\u7121\u6CD5\u9084\u539F\u6A94\u6848\uFF08\u6C92\u6709\u64F7\u53D6\u5230\u6A94\u6848\u8B8A\u66F4\uFF0C\u6216\u8A72\u8F2A\u6B21\u5C6C\u65BC\u672C\u6B21\u6703\u8A71\u4E4B\u524D\uFF09\u3002",
  "(+{{insertions}} -{{deletions}} in {{count}} file)": "(+{{insertions}} -{{deletions}}\uFF0C{{count}} \u500B\u6A94\u6848)",
  "(+{{insertions}} -{{deletions}} in {{count}} files)": "(+{{insertions}} -{{deletions}}\uFF0C{{count}} \u500B\u6A94\u6848)",
  "Failed to restore {{count}} file(s): {{files}}": "\u6062\u5FA9 {{count}} \u500B\u6A94\u6848\u5931\u6557\uFF1A{{files}}",
  "Cannot restore files: this turn was created before file checkpointing was enabled.": "\u7121\u6CD5\u6062\u5FA9\u6A94\u6848\uFF1A\u8A72\u8F2A\u5C0D\u8A71\u5EFA\u7ACB\u6642\u5C1A\u672A\u555F\u7528\u6A94\u6848\u6AA2\u67E5\u9EDE\u529F\u80FD\u3002",
  "No files needed to be restored.": "\u6C92\u6709\u6A94\u6848\u9700\u8981\u6062\u5FA9\u3002",
  "\u2191\u2193 to navigate \xB7 Enter to select \xB7 Esc to go back": "\u2191\u2193 \u5C0E\u89BD \xB7 Enter \u9078\u53D6 \xB7 Esc \u8FD4\u56DE",
  "\u2191\u2193 to navigate \xB7 Enter to select \xB7 Esc to cancel": "\u2191\u2193 \u5C0E\u89BD \xB7 Enter \u9078\u53D6 \xB7 Esc \u53D6\u6D88",
  "Enter/Y to confirm \xB7 Esc/N to go back": "Enter/Y \u78BA\u8A8D \xB7 Esc/N \u8FD4\u56DE",
  "change the theme": "\u66F4\u6539\u4E3B\u984C",
  "Select Theme": "\u9078\u64C7\u4E3B\u984C",
  Preview: "\u9810\u89BD",
  "(Use Enter to select, Tab to configure scope)": "\uFF08\u4F7F\u7528 Enter \u9078\u64C7\uFF0CTab \u914D\u7F6E\u4F5C\u7528\u57DF\uFF09",
  "(Use Enter to apply scope, Tab to go back)": "\uFF08\u4F7F\u7528 Enter \u61C9\u7528\u4F5C\u7528\u57DF\uFF0CTab \u8FD4\u56DE\uFF09",
  "Theme configuration unavailable due to NO_COLOR env variable.": "\u7531\u65BC NO_COLOR \u74B0\u5883\u8B8A\u91CF\uFF0C\u4E3B\u984C\u914D\u7F6E\u4E0D\u53EF\u7528\u3002",
  'Theme "{{themeName}}" not found.': '\u672A\u627E\u5230\u4E3B\u984C "{{themeName}}"\u3002',
  'Theme "{{themeName}}" not found in selected scope.': '\u5728\u6240\u9078\u4F5C\u7528\u57DF\u4E2D\u672A\u627E\u5230\u4E3B\u984C "{{themeName}}"\u3002',
  "Clear conversation history and free up context": "\u6E05\u9664\u5C0D\u8A71\u6B77\u53F2\u4E26\u91CB\u653E\u4E0A\u4E0B\u6587",
  "Compresses the context by replacing it with a summary.": "\u901A\u904E\u6458\u8981\u66FF\u63DB\u4F86\u58D3\u7E2E\u4E0A\u4E0B\u6587",
  "Fast context compression without AI. Strips old tool outputs and thinking parts.": "\u7121\u9700 AI \u7684\u5FEB\u901F\u4E0A\u4E0B\u6587\u58D3\u7E2E\u3002\u6E05\u7406\u820A\u5DE5\u5177\u8F38\u51FA\u4E26\u525D\u96E2\u601D\u8003\u904E\u7A0B\u3002",
  "open full Qwen Code documentation in your browser": "\u5728\u700F\u89BD\u5668\u4E2D\u6253\u958B\u5B8C\u6574\u7684 Qwen Code \u6587\u6A94",
  "Configuration not available.": "\u914D\u7F6E\u4E0D\u53EF\u7528",
  "Connect an LLM provider": "\u9023\u63A5 LLM \u63D0\u4F9B\u5546",
  "Copy to clipboard: reply, code (by lang), LaTeX, or Mermaid. N = Nth-latest message, index = block number": "\u8907\u88FD\u5230\u526A\u8CBC\u7C3F\uFF1AAI \u56DE\u61C9\u3001\u7A0B\u5F0F\u78BC\u5340\u584A\uFF08\u53EF\u4F9D\u8A9E\u8A00\u7BE9\u9078\uFF09\u3001LaTeX \u6216 Mermaid\u3002N \u70BA\u5012\u6578\u7B2C N \u5247\u8A0A\u606F\uFF0Cindex \u70BA\u7A0B\u5F0F\u78BC\u5340\u584A\u5E8F\u865F",
  "Show working-tree change stats versus HEAD": "\u986F\u793A\u5DE5\u4F5C\u5340\u76F8\u5C0D HEAD \u7684\u8B8A\u66F4\u7D71\u8A08",
  "Could not determine current working directory.": "\u7121\u6CD5\u78BA\u5B9A\u7576\u524D\u5DE5\u4F5C\u76EE\u9304\u3002",
  "Failed to compute git diff stats": "\u8A08\u7B97 git diff \u7D71\u8A08\u5931\u6557",
  "No diff available. Either this is not a git repository, HEAD is missing, or a merge/rebase/cherry-pick/revert is in progress.": "\u7121\u53EF\u7528 diff\u3002\u53EF\u80FD\u4E0D\u662F Git \u5009\u5EAB\u3001HEAD \u7F3A\u5931\uFF0C\u6216\u6B63\u5728\u57F7\u884C merge/rebase/cherry-pick/revert\u3002",
  "Clean working tree \u2014 no changes against HEAD.": "\u5DE5\u4F5C\u5340\u4E7E\u6DE8 \u2014\u2014 \u8207 HEAD \u7121\u5DEE\u7570\u3002",
  "{{count}} file changed, +{{added}} / -{{removed}}": "{{count}} \u500B\u6A94\u6848\u8B8A\u66F4\uFF0C+{{added}} / -{{removed}}",
  "{{count}} files changed, +{{added}} / -{{removed}}": "{{count}} \u500B\u6A94\u6848\u8B8A\u66F4\uFF0C+{{added}} / -{{removed}}",
  "{{count}} file changed": "{{count}} \u500B\u6A94\u6848\u8B8A\u66F4",
  "{{count}} files changed": "{{count}} \u500B\u6A94\u6848\u8B8A\u66F4",
  "\u2026and {{hidden}} more (showing first {{shown}})": "\u2026\u9084\u6709 {{hidden}} \u500B\uFF08\u50C5\u986F\u793A\u524D {{shown}} \u500B\uFF09",
  "(binary)": "\uFF08\u4E8C\u9032\u4F4D\uFF09",
  "(binary, new)": "\uFF08\u4E8C\u9032\u4F4D\uFF0C\u65B0\u589E\uFF09",
  "(new)": "\uFF08\u65B0\u589E\uFF09",
  "(new, partial)": "\uFF08\u65B0\u589E\uFF0C\u90E8\u5206\u7D71\u8A08\uFF09",
  "(deleted)": "\uFF08\u5DF2\u522A\u9664\uFF09",
  "(binary, deleted)": "\uFF08\u4E8C\u9032\u4F4D\uFF0C\u5DF2\u522A\u9664\uFF09",
  "Manage subagents for specialized task delegation.": "\u7BA1\u7406\u7528\u65BC\u5C08\u9580\u4EFB\u52D9\u59D4\u6D3E\u7684\u5B50\u667A\u80FD\u9AD4",
  "Manage existing subagents (view, edit, delete).": "\u7BA1\u7406\u73FE\u6709\u5B50\u667A\u80FD\u9AD4\uFF08\u67E5\u770B\u3001\u7DE8\u8F2F\u3001\u522A\u9664\uFF09",
  "Create a new subagent with guided setup.": "\u901A\u904E\u5F15\u5C0E\u5F0F\u8A2D\u7F6E\u5275\u5EFA\u65B0\u7684\u5B50\u667A\u80FD\u9AD4",
  Agents: "\u667A\u80FD\u9AD4",
  "Choose Action": "\u9078\u64C7\u64CD\u4F5C",
  "Edit {{name}}": "\u7DE8\u8F2F {{name}}",
  "Edit Tools: {{name}}": "\u7DE8\u8F2F\u5DE5\u5177: {{name}}",
  "Edit Color: {{name}}": "\u7DE8\u8F2F\u984F\u8272: {{name}}",
  "Delete {{name}}": "\u522A\u9664 {{name}}",
  "Unknown Step": "\u672A\u77E5\u6B65\u9A5F",
  "Esc to close": "\u6309 Esc \u95DC\u9589",
  "Enter to select, \u2191\u2193 to navigate, Esc to close": "Enter \u9078\u64C7\uFF0C\u2191\u2193 \u5C0E\u822A\uFF0CEsc \u95DC\u9589",
  "Esc to go back": "\u6309 Esc \u8FD4\u56DE",
  "Enter to confirm, Esc to cancel": "Enter \u78BA\u8A8D\uFF0CEsc \u53D6\u6D88",
  "Enter to select, \u2191\u2193 to navigate, Esc to go back": "Enter \u9078\u64C7\uFF0C\u2191\u2193 \u5C0E\u822A\uFF0CEsc \u8FD4\u56DE",
  "Enter to submit, Esc to go back": "Enter \u63D0\u4EA4\uFF0CEsc \u8FD4\u56DE",
  "Invalid step: {{step}}": "\u7121\u6548\u6B65\u9A5F: {{step}}",
  "No subagents found.": "\u672A\u627E\u5230\u5B50\u667A\u80FD\u9AD4\u3002",
  "Use '/agents create' to create your first subagent.": "\u4F7F\u7528 '/agents create' \u5275\u5EFA\u60A8\u7684\u7B2C\u4E00\u500B\u5B50\u667A\u80FD\u9AD4\u3002",
  "(built-in)": "\uFF08\u5167\u7F6E\uFF09",
  "(overridden by project level agent)": "\uFF08\u5DF2\u88AB\u9805\u76EE\u7D1A\u667A\u80FD\u9AD4\u8986\u84CB\uFF09",
  "Project Level ({{path}})": "\u9805\u76EE\u7D1A ({{path}})",
  "User Level ({{path}})": "\u7528\u6236\u7D1A ({{path}})",
  "Built-in Agents": "\u5167\u7F6E\u667A\u80FD\u9AD4",
  "Extension Agents": "\u64F4\u5C55\u667A\u80FD\u9AD4",
  "Using: {{count}} agents": "\u4F7F\u7528\u4E2D: {{count}} \u500B\u667A\u80FD\u9AD4",
  "View Agent": "\u67E5\u770B\u667A\u80FD\u9AD4",
  "Edit Agent": "\u7DE8\u8F2F\u667A\u80FD\u9AD4",
  "Delete Agent": "\u522A\u9664\u667A\u80FD\u9AD4",
  Back: "\u8FD4\u56DE",
  "No agent selected": "\u672A\u9078\u64C7\u667A\u80FD\u9AD4",
  "File Path: ": "\u6A94\u6848\u8DEF\u5F91: ",
  "Tools: ": "\u5DE5\u5177: ",
  "Color: ": "\u984F\u8272: ",
  "Description:": "\u63CF\u8FF0:",
  "System Prompt:": "\u7CFB\u7D71\u63D0\u793A:",
  "Open in editor": "\u5728\u7DE8\u8F2F\u5668\u4E2D\u6253\u958B",
  "Edit tools": "\u7DE8\u8F2F\u5DE5\u5177",
  "Edit color": "\u7DE8\u8F2F\u984F\u8272",
  "\u274C Error:": "\u274C \u932F\u8AA4:",
  'Are you sure you want to delete agent "{{name}}"?': '\u60A8\u78BA\u5B9A\u8981\u522A\u9664\u667A\u80FD\u9AD4 "{{name}}" \u55CE\uFF1F',
  "Project Level (.qwen/agents/)": "\u9805\u76EE\u7D1A (.qwen/agents/)",
  "User Level (~/.qwen/agents/)": "\u7528\u6236\u7D1A (~/.qwen/agents/)",
  "\u2705 Subagent Created Successfully!": "\u2705 \u5B50\u667A\u80FD\u9AD4\u5275\u5EFA\u6210\u529F\uFF01",
  'Subagent "{{name}}" has been saved to {{level}} level.': '\u5B50\u667A\u80FD\u9AD4 "{{name}}" \u5DF2\u4FDD\u5B58\u5230 {{level}} \u7D1A\u5225\u3002',
  "Name: ": "\u540D\u7A31: ",
  "Location: ": "\u4F4D\u7F6E: ",
  "\u274C Error saving subagent:": "\u274C \u4FDD\u5B58\u5B50\u667A\u80FD\u9AD4\u6642\u51FA\u932F:",
  "Warnings:": "\u8B66\u544A:",
  'Name "{{name}}" already exists at {{level}} level - will overwrite existing subagent': '\u540D\u7A31 "{{name}}" \u5728 {{level}} \u7D1A\u5225\u5DF2\u5B58\u5728 - \u5C07\u8986\u84CB\u73FE\u6709\u5B50\u667A\u80FD\u9AD4',
  'Name "{{name}}" exists at user level - project level will take precedence': '\u540D\u7A31 "{{name}}" \u5728\u7528\u6236\u7D1A\u5225\u5B58\u5728 - \u9805\u76EE\u7D1A\u5225\u5C07\u512A\u5148',
  'Name "{{name}}" exists at project level - existing subagent will take precedence': '\u540D\u7A31 "{{name}}" \u5728\u9805\u76EE\u7D1A\u5225\u5B58\u5728 - \u73FE\u6709\u5B50\u667A\u80FD\u9AD4\u5C07\u512A\u5148',
  "Description is over {{length}} characters": "\u63CF\u8FF0\u8D85\u904E {{length}} \u500B\u5B57\u7B26",
  "System prompt is over {{length}} characters": "\u7CFB\u7D71\u63D0\u793A\u8D85\u904E {{length}} \u500B\u5B57\u7B26",
  "Step {{n}}: Choose Location": "\u6B65\u9A5F {{n}}: \u9078\u64C7\u4F4D\u7F6E",
  "Step {{n}}: Choose Generation Method": "\u6B65\u9A5F {{n}}: \u9078\u64C7\u751F\u6210\u65B9\u5F0F",
  "Generate with Qwen Code (Recommended)": "\u4F7F\u7528 Qwen Code \u751F\u6210\uFF08\u63A8\u85A6\uFF09",
  "Manual Creation": "\u624B\u52D5\u5275\u5EFA",
  "Describe what this subagent should do and when it should be used. (Be comprehensive for best results)": "\u63CF\u8FF0\u6B64\u5B50\u667A\u80FD\u9AD4\u61C9\u8A72\u505A\u4EC0\u9EBC\u4EE5\u53CA\u4F55\u6642\u4F7F\u7528\u5B83\u3002\uFF08\u70BA\u4E86\u7372\u5F97\u6700\u4F73\u6548\u679C\uFF0C\u8ACB\u5168\u9762\u63CF\u8FF0\uFF09",
  "e.g., Expert code reviewer that reviews code based on best practices...": "\u4F8B\u5982\uFF1A\u5C08\u696D\u7684\u4EE3\u78BC\u5BE9\u67E5\u54E1\uFF0C\u6839\u64DA\u6700\u4F73\u5BE6\u8E10\u5BE9\u67E5\u4EE3\u78BC...",
  "Generating subagent configuration...": "\u6B63\u5728\u751F\u6210\u5B50\u667A\u80FD\u9AD4\u914D\u7F6E...",
  "Failed to generate subagent: {{error}}": "\u751F\u6210\u5B50\u667A\u80FD\u9AD4\u5931\u6557: {{error}}",
  "Step {{n}}: Describe Your Subagent": "\u6B65\u9A5F {{n}}: \u63CF\u8FF0\u60A8\u7684\u5B50\u667A\u80FD\u9AD4",
  "Step {{n}}: Enter Subagent Name": "\u6B65\u9A5F {{n}}: \u8F38\u5165\u5B50\u667A\u80FD\u9AD4\u540D\u7A31",
  "Step {{n}}: Enter System Prompt": "\u6B65\u9A5F {{n}}: \u8F38\u5165\u7CFB\u7D71\u63D0\u793A",
  "Step {{n}}: Enter Description": "\u6B65\u9A5F {{n}}: \u8F38\u5165\u63CF\u8FF0",
  "Step {{n}}: Select Tools": "\u6B65\u9A5F {{n}}: \u9078\u64C7\u5DE5\u5177",
  "All Tools (Default)": "\u6240\u6709\u5DE5\u5177\uFF08\u9ED8\u8A8D\uFF09",
  "All Tools": "\u6240\u6709\u5DE5\u5177",
  "Read-only Tools": "\u53EA\u8B80\u5DE5\u5177",
  "Read & Edit Tools": "\u8B80\u53D6\u548C\u7DE8\u8F2F\u5DE5\u5177",
  "Read & Edit & Execution Tools": "\u8B80\u53D6\u3001\u7DE8\u8F2F\u548C\u57F7\u884C\u5DE5\u5177",
  "All tools selected, including MCP tools": "\u5DF2\u9078\u64C7\u6240\u6709\u5DE5\u5177\uFF0C\u5305\u62EC MCP tools",
  "Selected tools:": "\u5DF2\u9078\u64C7\u7684\u5DE5\u5177:",
  "Read-only tools:": "\u53EA\u8B80\u5DE5\u5177:",
  "Edit tools:": "\u7DE8\u8F2F\u5DE5\u5177:",
  "Execution tools:": "\u57F7\u884C\u5DE5\u5177:",
  "Step {{n}}: Choose Background Color": "\u6B65\u9A5F {{n}}: \u9078\u64C7\u80CC\u666F\u984F\u8272",
  "Step {{n}}: Confirm and Save": "\u6B65\u9A5F {{n}}: \u78BA\u8A8D\u4E26\u4FDD\u5B58",
  "Esc to cancel": "\u6309 Esc \u53D6\u6D88",
  "Press Enter to save, e to save and edit, Esc to go back": "\u6309 Enter \u4FDD\u5B58\uFF0Ce \u4FDD\u5B58\u4E26\u7DE8\u8F2F\uFF0CEsc \u8FD4\u56DE",
  "Press Enter to continue, {{navigation}}Esc to {{action}}": "\u6309 Enter \u7E7C\u7E8C\uFF0C{{navigation}}Esc {{action}}",
  cancel: "\u53D6\u6D88",
  "go back": "\u8FD4\u56DE",
  "\u2191\u2193 to navigate, ": "\u2191\u2193 \u5C0E\u822A\uFF0C",
  "Enter a clear, unique name for this subagent.": "\u70BA\u6B64\u5B50\u667A\u80FD\u9AD4\u8F38\u5165\u4E00\u500B\u6E05\u6670\u3001\u552F\u4E00\u7684\u540D\u7A31\u3002",
  "e.g., Code Reviewer": "\u4F8B\u5982\uFF1A\u4EE3\u78BC\u5BE9\u67E5\u54E1",
  "Name cannot be empty.": "\u540D\u7A31\u4E0D\u80FD\u70BA\u7A7A\u3002",
  "Write the system prompt that defines this subagent's behavior. Be comprehensive for best results.": "\u7DE8\u5BEB\u5B9A\u7FA9\u6B64\u5B50\u667A\u80FD\u9AD4\u884C\u70BA\u7684\u7CFB\u7D71\u63D0\u793A\u3002\u70BA\u4E86\u7372\u5F97\u6700\u4F73\u6548\u679C\uFF0C\u8ACB\u5168\u9762\u63CF\u8FF0\u3002",
  "e.g., You are an expert code reviewer...": "\u4F8B\u5982\uFF1A\u60A8\u662F\u4E00\u4F4D\u5C08\u696D\u7684\u4EE3\u78BC\u5BE9\u67E5\u54E1...",
  "System prompt cannot be empty.": "\u7CFB\u7D71\u63D0\u793A\u4E0D\u80FD\u70BA\u7A7A\u3002",
  "Describe when and how this subagent should be used.": "\u63CF\u8FF0\u4F55\u6642\u4EE5\u53CA\u5982\u4F55\u4F7F\u7528\u6B64\u5B50\u667A\u80FD\u9AD4\u3002",
  "e.g., Reviews code for best practices and potential bugs.": "\u4F8B\u5982\uFF1A\u5BE9\u67E5\u4EE3\u78BC\u4EE5\u67E5\u627E\u6700\u4F73\u5BE6\u8E10\u548C\u6F5B\u5728\u932F\u8AA4\u3002",
  "Description cannot be empty.": "\u63CF\u8FF0\u4E0D\u80FD\u70BA\u7A7A\u3002",
  "Failed to launch editor: {{error}}": "\u555F\u52D5\u7DE8\u8F2F\u5668\u5931\u6557: {{error}}",
  "Failed to save and edit subagent: {{error}}": "\u4FDD\u5B58\u4E26\u7DE8\u8F2F\u5B50\u667A\u80FD\u9AD4\u5931\u6557: {{error}}",
  "Manage Extensions": "\u7BA1\u7406\u64F4\u5C55",
  "Extension Details": "\u64F4\u5C55\u8A73\u60C5",
  "View Extension": "\u67E5\u770B\u64F4\u5C55",
  "Update Extension": "\u66F4\u65B0\u64F4\u5C55",
  "Disable Extension": "\u7981\u7528\u64F4\u5C55",
  "Enable Extension": "\u555F\u7528\u64F4\u5C55",
  "Uninstall Extension": "\u5378\u8F09\u64F4\u5C55",
  "Select Scope": "\u9078\u64C7\u4F5C\u7528\u57DF",
  "User Scope": "\u7528\u6236\u4F5C\u7528\u57DF",
  "Workspace Scope": "\u5DE5\u4F5C\u5340\u4F5C\u7528\u57DF",
  "No extensions found.": "\u672A\u627E\u5230\u64F4\u5C55\u3002",
  "Updating...": "\u66F4\u65B0\u4E2D...",
  Unknown: "\u672A\u77E5",
  Error: "\u932F\u8AA4",
  "Stopped because": "\u505C\u6B62\u539F\u56E0",
  "Version:": "\u7248\u672C\uFF1A",
  "Status:": "\u72C0\u614B\uFF1A",
  'Are you sure you want to uninstall extension "{{name}}"?': '\u78BA\u5B9A\u8981\u5378\u8F09\u64F4\u5C55 "{{name}}" \u55CE\uFF1F',
  "This action cannot be undone.": "\u6B64\u64CD\u4F5C\u7121\u6CD5\u64A4\u92B7\u3002",
  'Extension "{{name}}" updated successfully.': '\u64F4\u5C55 "{{name}}" \u66F4\u65B0\u6210\u529F\u3002',
  "Name:": "\u540D\u7A31\uFF1A",
  "MCP Servers:": "MCP Servers\uFF1A",
  "Settings:": "\u8A2D\u7F6E\uFF1A",
  active: "\u5DF2\u555F\u7528",
  disabled: "\u5DF2\u7981\u7528",
  enabled: "\u5DF2\u555F\u7528",
  "View Details": "\u67E5\u770B\u8A73\u60C5",
  "Update failed:": "\u66F4\u65B0\u5931\u6557\uFF1A",
  "Updating {{name}}...": "\u6B63\u5728\u66F4\u65B0 {{name}}...",
  "Update complete!": "\u66F4\u65B0\u5B8C\u6210\uFF01",
  "User (global)": "\u7528\u6236\uFF08\u5168\u5C40\uFF09",
  "Workspace (project-specific)": "\u5DE5\u4F5C\u5340\uFF08\u9805\u76EE\u7279\u5B9A\uFF09",
  'Disable "{{name}}" - Select Scope': '\u7981\u7528 "{{name}}" - \u9078\u64C7\u4F5C\u7528\u57DF',
  'Enable "{{name}}" - Select Scope': '\u555F\u7528 "{{name}}" - \u9078\u64C7\u4F5C\u7528\u57DF',
  "No extension selected": "\u672A\u9078\u64C7\u64F4\u5C55",
  "{{count}} extensions installed": "\u5DF2\u5B89\u88DD {{count}} \u500B\u64F4\u5C55",
  "Use '/extensions install' to install your first extension.": "\u4F7F\u7528 '/extensions install' \u5B89\u88DD\u60A8\u7684\u7B2C\u4E00\u500B\u64F4\u5C55\u3002",
  "up to date": "\u5DF2\u662F\u6700\u65B0",
  "update available": "\u6709\u53EF\u7528\u66F4\u65B0",
  "checking...": "\u6AA2\u67E5\u4E2D...",
  "not updatable": "\u4E0D\u53EF\u66F4\u65B0",
  error: "\u932F\u8AA4",
  "View and edit Qwen Code settings": "\u67E5\u770B\u548C\u7DE8\u8F2F Qwen Code \u8A2D\u7F6E",
  Settings: "\u8A2D\u7F6E",
  "To see changes, Qwen Code must be restarted. Press r to exit and apply changes now.": "\u8981\u67E5\u770B\u66F4\u6539\uFF0C\u5FC5\u9808\u91CD\u555F Qwen Code\u3002\u6309 r \u9000\u51FA\u4E26\u7ACB\u5373\u61C9\u7528\u66F4\u6539\u3002",
  "Vim Mode": "Vim \u6A21\u5F0F",
  "Attribution: commit": "\u7F72\u540D\uFF1A\u63D0\u4EA4",
  "Terminal Bell Notification": "\u7D42\u7AEF\u97FF\u9234\u901A\u77E5",
  "Enable Usage Statistics": "\u555F\u7528\u4F7F\u7528\u7D71\u8A08",
  Theme: "\u4E3B\u984C",
  "Preferred Editor": "\u9996\u9078\u7DE8\u8F2F\u5668",
  "Auto-connect to IDE": "\u81EA\u52D5\u9023\u63A5\u5230 IDE",
  "Debug Keystroke Logging": "\u8ABF\u8A66\u6309\u9375\u8A18\u9304",
  "Language: UI": "\u8A9E\u8A00\uFF1A\u754C\u9762",
  "Language: Model": "\u8A9E\u8A00\uFF1A\u6A21\u578B",
  "Output Format": "\u8F38\u51FA\u683C\u5F0F",
  "Hide Window Title": "\u96B1\u85CF\u7A97\u53E3\u6A19\u984C",
  "Show Status in Title": "\u5728\u6A19\u984C\u4E2D\u986F\u793A\u72C0\u614B",
  "Hide Tips": "\u96B1\u85CF\u63D0\u793A",
  "Show Line Numbers in Code": "\u5728\u4EE3\u78BC\u4E2D\u986F\u793A\u884C\u865F",
  "Show Citations": "\u986F\u793A\u5F15\u7528",
  "Custom Witty Phrases": "\u81EA\u5B9A\u7FA9\u8A7C\u8AE7\u77ED\u8A9E",
  "Show Welcome Back Dialog": "\u986F\u793A\u6B61\u8FCE\u56DE\u4F86\u5C0D\u8A71\u6846",
  "Enable User Feedback": "\u555F\u7528\u7528\u6236\u53CD\u994B",
  "How is Qwen doing this session? (optional)": "Qwen \u9019\u6B21\u8868\u73FE\u5982\u4F55\uFF1F\uFF08\u53EF\u9078\uFF09",
  Bad: "\u4E0D\u6EFF\u610F",
  Fine: "\u9084\u884C",
  Good: "\u6EFF\u610F",
  Dismiss: "\u5FFD\u7565",
  "Screen Reader Mode": "\u5C4F\u5E55\u95B1\u8B80\u5668\u6A21\u5F0F",
  "Max Session Turns": "\u6700\u5927\u6703\u8A71\u8F2A\u6B21",
  "Skip Next Speaker Check": "\u8DF3\u904E\u4E0B\u4E00\u500B\u8AAA\u8A71\u8005\u6AA2\u67E5",
  "Skip Loop Detection": "\u8DF3\u904E\u5FAA\u74B0\u6AA2\u6E2C",
  "Skip Startup Context": "\u8DF3\u904E\u555F\u52D5\u4E0A\u4E0B\u6587",
  "Enable OpenAI Logging": "\u555F\u7528 OpenAI \u65E5\u8A8C",
  "OpenAI Logging Directory": "OpenAI \u65E5\u8A8C\u76EE\u9304",
  Timeout: "\u8D85\u6642",
  "Max Retries": "\u6700\u5927\u91CD\u8A66\u6B21\u6578",
  "Load Memory From Include Directories": "\u5F9E\u5305\u542B\u76EE\u9304\u52A0\u8F09\u5167\u5B58",
  "Respect .gitignore": "\u9075\u5B88 .gitignore",
  "Respect .qwenignore": "\u9075\u5B88 .qwenignore",
  "Enable Recursive File Search": "\u555F\u7528\u905E\u6B78\u6A94\u6848\u641C\u7D22",
  "Interactive Shell (PTY)": "\u4EA4\u4E92\u5F0F Shell (PTY)",
  "Show Color": "\u986F\u793A\u984F\u8272",
  "Auto Accept": "\u81EA\u52D5\u63A5\u53D7",
  "Use Ripgrep": "\u4F7F\u7528 Ripgrep",
  "Use Builtin Ripgrep": "\u4F7F\u7528\u5167\u7F6E Ripgrep",
  "Tool Output Truncation Threshold": "\u5DE5\u5177\u8F38\u51FA\u622A\u65B7\u95BE\u503C",
  "Tool Output Truncation Lines": "\u5DE5\u5177\u8F38\u51FA\u622A\u65B7\u884C\u6578",
  "Folder Trust": "\u6A94\u6848\u593E\u4FE1\u4EFB",
  "Tool Schema Compliance": "Tool Schema \u517C\u5BB9\u6027",
  "Auto (detect from system)": "\u81EA\u52D5\uFF08\u5F9E\u7CFB\u7D71\u6AA2\u6E2C\uFF09",
  "Auto (detect terminal theme)": "\u81EA\u52D5\uFF08\u6AA2\u6E2C\u7D42\u7AEF\u4E3B\u984C\uFF09",
  Auto: "\u81EA\u52D5",
  Text: "\u6587\u672C",
  JSON: "JSON",
  Plan: "\u898F\u5283",
  "Ask permissions": "\u8ACB\u6C42\u6388\u6B0A",
  "Auto Edit": "\u81EA\u52D5\u7DE8\u8F2F",
  YOLO: "YOLO",
  "toggle vim mode on/off": "\u5207\u63DB vim \u6A21\u5F0F\u958B\u95DC",
  "Show model-specific usage statistics.": "\u986F\u793A\u6A21\u578B\u76F8\u95DC\u7684\u4F7F\u7528\u7D71\u8A08\u8CC7\u8A0A",
  "Show tool-specific usage statistics.": "\u986F\u793A\u5DE5\u5177\u76F8\u95DC\u7684\u4F7F\u7528\u7D71\u8A08\u8CC7\u8A0A",
  "Show skill-specific usage statistics.": "\u986F\u793A\u6280\u80FD\u76F8\u95DC\u7684\u4F7F\u7528\u7D71\u8A08\u8CC7\u8A0A",
  "Show daily token usage statistics.": "\u986F\u793A\u6BCF\u65E5 token \u4F7F\u7528\u7D71\u8A08\u8CC7\u8A0A",
  "Show monthly token usage statistics.": "\u986F\u793A\u6BCF\u6708 token \u4F7F\u7528\u7D71\u8A08\u8CC7\u8A0A",
  "Export token usage statistics to CSV or JSON.": "\u5C07 token \u4F7F\u7528\u7D71\u8A08\u8CC7\u8A0A\u532F\u51FA\u70BA CSV \u6216 JSON",
  "No usage data.": "\u6C92\u6709\u4F7F\u7528\u8CC7\u6599\u3002",
  "{{label}}: {{tokens}} tokens ({{requests}} requests)": "{{label}}\uFF1A{{tokens}} \u500B token\uFF08{{requests}} \u500B\u8ACB\u6C42\uFF09",
  "Daily token usage for {{value}}": "{{value}} \u7684\u6BCF\u65E5 token \u4F7F\u7528\u60C5\u6CC1",
  "Monthly token usage for {{value}}": "{{value}} \u7684\u6BCF\u6708 token \u4F7F\u7528\u60C5\u6CC1",
  "Total: {{tokens}} tokens": "\u7E3D\u8A08\uFF1A{{tokens}} \u500B token",
  "Requests: {{requests}}": "\u8ACB\u6C42\u6578\uFF1A{{requests}}",
  "Breakdown:": "\u660E\u7D30\uFF1A",
  "Input: {{tokens}}": "\u8F38\u5165\uFF1A{{tokens}}",
  "Output: {{tokens}}": "\u8F38\u51FA\uFF1A{{tokens}}",
  "Cached (included in Input): {{tokens}}": "\u5FEB\u53D6\uFF08\u5DF2\u5305\u542B\u5728\u8F38\u5165\u4E2D\uFF09\uFF1A{{tokens}}",
  "Thoughts: {{tokens}}": "\u601D\u8003\uFF1A{{tokens}}",
  "By model:": "\u6309\u6A21\u578B\uFF1A",
  "By auth type:": "\u6309\u8A8D\u8B49\u985E\u578B\uFF1A",
  "By model/auth type:": "\u6309\u6A21\u578B/\u8A8D\u8B49\u985E\u578B\uFF1A",
  "By source:": "\u6309\u4F86\u6E90\uFF1A",
  "Failed to load token usage stats: {{error}}": "\u8F09\u5165 token \u4F7F\u7528\u7D71\u8A08\u8CC7\u8A0A\u5931\u6557\uFF1A{{error}}",
  "Expected --format csv or --format json.": "\u61C9\u70BA --format csv \u6216 --format json\u3002",
  "Expected a file path after --output.": "--output \u5F8C\u61C9\u63D0\u4F9B\u6A94\u6848\u8DEF\u5F91\u3002",
  "Unexpected argument: {{argument}}": "\u672A\u9810\u671F\u7684\u53C3\u6578\uFF1A{{argument}}",
  "Usage: /stats export <daily|monthly> [YYYY-MM-DD|YYYY-MM] [--format csv|json] [--output path]": "\u7528\u6CD5\uFF1A/stats export <daily|monthly> [YYYY-MM-DD|YYYY-MM] [--format csv|json] [--output path]",
  "Token usage export path must be within the project working directory.": "Token \u4F7F\u7528\u532F\u51FA\u8DEF\u5F91\u5FC5\u9808\u4F4D\u65BC\u5C08\u6848\u5DE5\u4F5C\u76EE\u9304\u5167\u3002",
  "Export target does not exist: {{path}}": "\u532F\u51FA\u76EE\u6A19\u4E0D\u5B58\u5728\uFF1A{{path}}",
  "Cannot resolve export path within the working directory.": "\u7121\u6CD5\u5728\u5DE5\u4F5C\u76EE\u9304\u5167\u89E3\u6790\u532F\u51FA\u8DEF\u5F91\u3002",
  "Could not create a temporary export file.": "\u7121\u6CD5\u5EFA\u7ACB\u81E8\u6642\u532F\u51FA\u6A94\u6848\u3002",
  "Token usage exported to {{format}}: {{path}}": "Token \u4F7F\u7528\u60C5\u6CC1\u5DF2\u532F\u51FA\u70BA {{format}}\uFF1A{{path}}",
  "Failed to export token usage stats: {{error}}": "\u532F\u51FA token \u4F7F\u7528\u7D71\u8A08\u8CC7\u8A0A\u5931\u6557\uFF1A{{error}}",
  "Unclosed quote in arguments.": "\u53C3\u6578\u4E2D\u6709\u672A\u9589\u5408\u7684\u5F15\u865F\u3002",
  "Note: generation timing (TTFT/TPS) belongs to generation metrics.": "\u6CE8\u610F\uFF1A\u751F\u6210\u8017\u6642\uFF08TTFT/TPS\uFF09\u5C6C\u65BC\u751F\u6210\u6307\u6A19\u3002",
  "exit the cli": "\u9000\u51FA\u547D\u4EE4\u884C\u754C\u9762",
  "Manage workspace directories": "\u7BA1\u7406\u5DE5\u4F5C\u5340\u76EE\u9304",
  "Add directories to the workspace. Use comma to separate multiple paths": "\u5C07\u76EE\u9304\u6DFB\u52A0\u5230\u5DE5\u4F5C\u5340\u3002\u4F7F\u7528\u9017\u865F\u5206\u9694\u591A\u500B\u8DEF\u5F91",
  "Show all directories in the workspace": "\u986F\u793A\u5DE5\u4F5C\u5340\u4E2D\u7684\u6240\u6709\u76EE\u9304",
  "set external editor preference": "\u8A2D\u7F6E\u5916\u90E8\u7DE8\u8F2F\u5668\u9996\u9078\u9805",
  "Select Editor": "\u9078\u64C7\u7DE8\u8F2F\u5668",
  "Editor Preference": "\u7DE8\u8F2F\u5668\u9996\u9078\u9805",
  "These editors are currently supported. Please note that some editors cannot be used in sandbox mode.": "\u7576\u524D\u652F\u6301\u4EE5\u4E0B\u7DE8\u8F2F\u5668\u3002\u8ACB\u6CE8\u610F\uFF0C\u67D0\u4E9B\u7DE8\u8F2F\u5668\u7121\u6CD5\u5728\u6C99\u7BB1\u6A21\u5F0F\u4E0B\u4F7F\u7528\u3002",
  "Your preferred editor is:": "\u60A8\u7684\u9996\u9078\u7DE8\u8F2F\u5668\u662F\uFF1A",
  "Manage extensions": "\u7BA1\u7406\u64F4\u5C55",
  "Manage installed extensions": "\u7BA1\u7406\u5DF2\u5B89\u88DD\u7684\u64F4\u5C55",
  "Disable an extension": "\u7981\u7528\u64F4\u5C55",
  "Enable an extension": "\u555F\u7528\u64F4\u5C55",
  "Install an extension from a git repo or local path": "\u5F9E Git \u5009\u5EAB\u6216\u672C\u5730\u8DEF\u5F91\u5B89\u88DD\u64F4\u5C55",
  "Uninstall an extension": "\u5378\u8F09\u64F4\u5C55",
  "No extensions installed.": "\u672A\u5B89\u88DD\u64F4\u5C55\u3002",
  'Extension "{{name}}" not found.': '\u672A\u627E\u5230\u64F4\u5C55 "{{name}}"\u3002',
  'The scope to install the extension in: "user" (global, default) or "project" (current workspace only).': '\u5B89\u88DD\u64F4\u5C55\u7684\u4F5C\u7528\u57DF\uFF1A"user"\uFF08\u5168\u57DF\uFF0C\u9810\u8A2D\uFF09\u6216 "project"\uFF08\u50C5\u7576\u524D\u5DE5\u4F5C\u5340\uFF09\u3002',
  'Extension "{{name}}" installed successfully and enabled for the current workspace.': '\u64F4\u5C55 "{{name}}" \u5B89\u88DD\u6210\u529F\uFF0C\u4E26\u5DF2\u5728\u7576\u524D\u5DE5\u4F5C\u5340\u555F\u7528\u3002',
  'Marketplace "{{name}}" not found.': '\u672A\u627E\u5230\u5E02\u5834\u6E90 "{{name}}"\u3002',
  "No marketplace sources added yet.": "\u5C1A\u672A\u6DFB\u52A0\u4EFB\u4F55\u5E02\u5834\u6E90\u3002",
  "No marketplaces added yet.": "\u5C1A\u672A\u6DFB\u52A0\u4EFB\u4F55\u5E02\u5834\u6E90\u3002",
  "Adds a marketplace source (Claude format).": "\u6DFB\u52A0\u4E00\u500B\u5E02\u5834\u6E90\uFF08Claude \u683C\u5F0F\uFF09\u3002",
  "The marketplace source to add: owner/repo (GitHub), a git or https URL, or a local path.": "\u8981\u6DFB\u52A0\u7684\u5E02\u5834\u6E90\uFF1Aowner/repo\uFF08GitHub\uFF09\u3001git \u6216 https URL\uFF0C\u6216\u672C\u5730\u8DEF\u5F91\u3002",
  "Removes a marketplace source.": "\u79FB\u9664\u4E00\u500B\u5E02\u5834\u6E90\u3002",
  "The name of the marketplace to remove.": "\u8981\u79FB\u9664\u7684\u5E02\u5834\u6E90\u540D\u7A31\u3002",
  "Lists configured marketplace sources.": "\u5217\u51FA\u5DF2\u914D\u7F6E\u7684\u5E02\u5834\u6E90\u3002",
  "Re-fetches a marketplace source and its plugin listing.": "\u91CD\u65B0\u62C9\u53D6\u5E02\u5834\u6E90\u53CA\u5176\u63D2\u4EF6\u5217\u8868\u3002",
  "The name of the marketplace to update.": "\u8981\u66F4\u65B0\u7684\u5E02\u5834\u6E90\u540D\u7A31\u3002",
  "Manage marketplace sources for discovering extensions.": "\u7BA1\u7406\u7528\u65BC\u767C\u73FE\u64F4\u5C55\u7684\u5E02\u5834\u6E90\u3002",
  "You need at least one command before continuing.": "\u9700\u8981\u81F3\u5C11\u63D0\u4F9B\u4E00\u500B\u5B50\u547D\u4EE4\u3002",
  "No extensions to update.": "\u6C92\u6709\u53EF\u66F4\u65B0\u7684\u64F4\u5C55\u3002",
  "Usage: /extensions install <source>": "\u7528\u6CD5\uFF1A/extensions install <\u4F86\u6E90>",
  'Installing extension from "{{source}}"...': '\u6B63\u5728\u5F9E "{{source}}" \u5B89\u88DD\u64F4\u5C55...',
  'Extension "{{name}}" installed successfully.': '\u64F4\u5C55 "{{name}}" \u5B89\u88DD\u6210\u529F\u3002',
  'Failed to install extension from "{{source}}": {{error}}': '\u5F9E "{{source}}" \u5B89\u88DD\u64F4\u5C55\u5931\u6557\uFF1A{{error}}',
  "Do you want to continue? [Y/n]: ": "\u662F\u5426\u7E7C\u7E8C\uFF1F[Y/n]\uFF1A",
  "Do you want to continue?": "\u662F\u5426\u7E7C\u7E8C\uFF1F",
  'Installing extension "{{name}}".': '\u6B63\u5728\u5B89\u88DD\u64F4\u5C55 "{{name}}"\u3002',
  "**Extensions may introduce unexpected behavior. Ensure you have investigated the extension source and trust the author.**": "**\u64F4\u5C55\u53EF\u80FD\u6703\u5F15\u5165\u610F\u5916\u884C\u70BA\u3002\u8ACB\u78BA\u4FDD\u60A8\u5DF2\u8ABF\u67E5\u904E\u64F4\u5C55\u6E90\u4E26\u4FE1\u4EFB\u4F5C\u8005\u3002**",
  "This extension will run the following MCP servers:": "\u6B64\u64F4\u5C55\u5C07\u904B\u884C\u4EE5\u4E0B MCP servers\uFF1A",
  local: "\u672C\u5730",
  remote: "\u9060\u7A0B",
  "This extension will add the following commands: {{commands}}.": "\u6B64\u64F4\u5C55\u5C07\u6DFB\u52A0\u4EE5\u4E0B\u547D\u4EE4\uFF1A{{commands}}\u3002",
  "This extension will append info to your QWEN.md context using {{fileName}}": "\u6B64\u64F4\u5C55\u5C07\u4F7F\u7528 {{fileName}} \u5411\u60A8\u7684 QWEN.md \u4E0A\u4E0B\u6587\u8FFD\u52A0\u4FE1\u606F",
  "This extension will install the following skills:": "\u6B64\u64F4\u5C55\u5C07\u5B89\u88DD\u4EE5\u4E0B\u6280\u80FD\uFF1A",
  "This extension will install the following subagents:": "\u6B64\u64F4\u5C55\u5C07\u5B89\u88DD\u4EE5\u4E0B\u5B50\u667A\u80FD\u9AD4\uFF1A",
  'Installation cancelled for "{{name}}".': '\u5DF2\u53D6\u6D88\u5B89\u88DD "{{name}}"\u3002',
  "You are installing an extension from {{originSource}}. Some features may not work perfectly with Qwen Code.": "\u60A8\u6B63\u5728\u5B89\u88DD\u4F86\u81EA {{originSource}} \u7684\u64F4\u5C55\u3002\u67D0\u4E9B\u529F\u80FD\u53EF\u80FD\u7121\u6CD5\u5B8C\u7F8E\u517C\u5BB9 Qwen Code\u3002",
  "--ref and --auto-update are not applicable for marketplace extensions.": "--ref \u548C --auto-update \u4E0D\u9069\u7528\u65BC\u5E02\u5834\u64F4\u5C55\u3002",
  'Extension "{{name}}" installed successfully and enabled.': '\u64F4\u5C55 "{{name}}" \u5B89\u88DD\u6210\u529F\u4E26\u5DF2\u555F\u7528\u3002',
  "The github URL, local path, or marketplace source (marketplace-url:plugin-name) of the extension to install.": "\u8981\u5B89\u88DD\u7684\u64F4\u5C55\u7684 GitHub URL\u3001\u672C\u5730\u8DEF\u5F91\u6216\u5E02\u5834\u6E90\uFF08marketplace-url:plugin-name\uFF09\u3002",
  "The git ref to install from.": "\u8981\u5B89\u88DD\u7684 Git \u5F15\u7528\u3002",
  "--registry is only applicable for npm extensions.": "--registry \u50C5\u9069\u7528\u65BC npm \u64F4\u5C55\u3002",
  "Custom npm registry URL (only for npm extensions).": "\u81EA\u8A02 npm registry URL\uFF08\u50C5\u9069\u7528\u65BC npm \u64F4\u5C55\uFF09\u3002",
  "--ref is not applicable for npm extensions. Use @version suffix instead (e.g. @scope/package@1.2.0).": "--ref \u4E0D\u9069\u7528\u65BC npm \u64F4\u5C55\u3002\u8ACB\u6539\u7528 @version \u5F8C\u7DB4\uFF08\u4F8B\u5982 @scope/package@1.2.0\uFF09\u3002",
  "Installs an extension from a git repository URL, local path, scoped npm package (@scope/name), or claude marketplace (marketplace-url:plugin-name).": "\u5F9E Git \u5009\u5EAB URL\u3001\u672C\u5730\u8DEF\u5F91\u3001\u5E36\u4F5C\u7528\u57DF\u7684 npm \u5957\u4EF6\uFF08@scope/name\uFF09\u6216 Claude \u5E02\u5834\u6E90\uFF08marketplace-url:plugin-name\uFF09\u5B89\u88DD\u64F4\u5C55\u3002",
  Description: "\u63CF\u8FF0",
  "Delete Session": "\u522A\u9664\u6703\u8A71",
  "Enable auto-update for this extension.": "\u70BA\u6B64\u64F4\u5C55\u555F\u7528\u81EA\u52D5\u66F4\u65B0\u3002",
  "Enable pre-release versions for this extension.": "\u70BA\u6B64\u64F4\u5C55\u555F\u7528\u9810\u767C\u4F48\u7248\u672C\u3002",
  "Acknowledge the security risks of installing an extension and skip the confirmation prompt.": "\u78BA\u8A8D\u5B89\u88DD\u64F4\u5C55\u7684\u5B89\u5168\u98A8\u96AA\u4E26\u8DF3\u904E\u78BA\u8A8D\u63D0\u793A\u3002",
  "The source argument must be provided.": "\u5FC5\u9808\u63D0\u4F9B\u4F86\u6E90\u53C3\u6578\u3002",
  'Extension "{{name}}" successfully uninstalled.': '\u64F4\u5C55 "{{name}}" \u5378\u8F09\u6210\u529F\u3002',
  "Uninstalls an extension.": "\u5378\u8F09\u64F4\u5C55\u3002",
  "The name or source path of the extension to uninstall.": "\u8981\u5378\u8F09\u7684\u64F4\u5C55\u7684\u540D\u7A31\u6216\u6E90\u8DEF\u5F91\u3002",
  "Please include the name of the extension to uninstall as a positional argument.": "\u8ACB\u5C07\u8981\u5378\u8F09\u7684\u64F4\u5C55\u540D\u7A31\u4F5C\u70BA\u4F4D\u7F6E\u53C3\u6578\u3002",
  "Enables an extension.": "\u555F\u7528\u64F4\u5C55\u3002",
  "The name of the extension to enable.": "\u8981\u555F\u7528\u7684\u64F4\u5C55\u540D\u7A31\u3002",
  "The scope to enable the extenison in. If not set, will be enabled in all scopes.": "\u555F\u7528\u64F4\u5C55\u7684\u4F5C\u7528\u57DF\u3002\u5982\u679C\u672A\u8A2D\u7F6E\uFF0C\u5C07\u5728\u6240\u6709\u4F5C\u7528\u57DF\u4E2D\u555F\u7528\u3002",
  'Extension "{{name}}" successfully enabled for scope "{{scope}}".': '\u64F4\u5C55 "{{name}}" \u5DF2\u5728\u4F5C\u7528\u57DF "{{scope}}" \u4E2D\u555F\u7528\u3002',
  'Extension "{{name}}" successfully enabled in all scopes.': '\u64F4\u5C55 "{{name}}" \u5DF2\u5728\u6240\u6709\u4F5C\u7528\u57DF\u4E2D\u555F\u7528\u3002',
  "Invalid scope: {{scope}}. Please use one of {{scopes}}.": "\u7121\u6548\u7684\u4F5C\u7528\u57DF\uFF1A{{scope}}\u3002\u8ACB\u4F7F\u7528 {{scopes}} \u4E4B\u4E00\u3002",
  "Disables an extension.": "\u7981\u7528\u64F4\u5C55\u3002",
  "The name of the extension to disable.": "\u8981\u7981\u7528\u7684\u64F4\u5C55\u540D\u7A31\u3002",
  "The scope to disable the extenison in.": "\u7981\u7528\u64F4\u5C55\u7684\u4F5C\u7528\u57DF\u3002",
  'Extension "{{name}}" successfully disabled for scope "{{scope}}".': '\u64F4\u5C55 "{{name}}" \u5DF2\u5728\u4F5C\u7528\u57DF "{{scope}}" \u4E2D\u7981\u7528\u3002',
  'Extension "{{name}}" successfully updated: {{oldVersion}} \u2192 {{newVersion}}.': '\u64F4\u5C55 "{{name}}" \u66F4\u65B0\u6210\u529F\uFF1A{{oldVersion}} \u2192 {{newVersion}}\u3002',
  'Unable to install extension "{{name}}" due to missing install metadata': '\u7531\u65BC\u7F3A\u5C11\u5B89\u88DD\u5143\u6578\u64DA\uFF0C\u7121\u6CD5\u5B89\u88DD\u64F4\u5C55 "{{name}}"',
  'Extension "{{name}}" is already up to date.': '\u64F4\u5C55 "{{name}}" \u5DF2\u662F\u6700\u65B0\u7248\u672C\u3002',
  "Updates all extensions or a named extension to the latest version.": "\u5C07\u6240\u6709\u64F4\u5C55\u6216\u6307\u5B9A\u64F4\u5C55\u66F4\u65B0\u5230\u6700\u65B0\u7248\u672C\u3002",
  "Update all extensions.": "\u66F4\u65B0\u6240\u6709\u64F4\u5C55\u3002",
  "The name of the extension to update.": "\u8981\u66F4\u65B0\u7684\u64F4\u5C55\u540D\u7A31\u3002",
  "Either an extension name or --all must be provided": "\u5FC5\u9808\u63D0\u4F9B\u64F4\u5C55\u540D\u7A31\u6216 --all",
  "List installed extensions": "\u5217\u51FA\u5DF2\u5B89\u88DD\u7684\u64F4\u5C55",
  "Lists installed extensions.": "\u5217\u51FA\u5DF2\u5B89\u88DD\u7684\u64F4\u5C55\u3002",
  "Path:": "\u8DEF\u5F91\uFF1A",
  "Source:": "\u4F86\u6E90\uFF1A",
  "Type:": "\u985E\u578B\uFF1A",
  "Ref:": "\u5F15\u7528\uFF1A",
  "Release tag:": "\u767C\u4F48\u6A19\u7C64\uFF1A",
  "Enabled (User):": "\u5DF2\u555F\u7528\uFF08\u7528\u6236\uFF09\uFF1A",
  "Enabled (Workspace):": "\u5DF2\u555F\u7528\uFF08\u5DE5\u4F5C\u5340\uFF09\uFF1A",
  "Context files:": "\u4E0A\u4E0B\u6587\u6A94\u6848\uFF1A",
  "Skills:": "\u6280\u80FD\uFF1A",
  "Agents:": "\u667A\u80FD\u9AD4\uFF1A",
  "MCP servers:": "MCP servers\uFF1A",
  "Link extension failed to install.": "\u9023\u7D50\u64F4\u5C55\u5B89\u88DD\u5931\u6557\u3002",
  'Extension "{{name}}" linked successfully and enabled.': '\u64F4\u5C55 "{{name}}" \u9023\u7D50\u6210\u529F\u4E26\u5DF2\u555F\u7528\u3002',
  "Links an extension from a local path. Updates made to the local path will always be reflected.": "\u5F9E\u672C\u5730\u8DEF\u5F91\u9023\u7D50\u64F4\u5C55\u3002\u5C0D\u672C\u5730\u8DEF\u5F91\u7684\u66F4\u65B0\u5C07\u59CB\u7D42\u53CD\u6620\u3002",
  "The name of the extension to link.": "\u8981\u9023\u7D50\u7684\u64F4\u5C55\u540D\u7A31\u3002",
  "Set a specific setting for an extension.": "\u70BA\u64F4\u5C55\u8A2D\u7F6E\u7279\u5B9A\u914D\u7F6E\u3002",
  "Name of the extension to configure.": "\u8981\u914D\u7F6E\u7684\u64F4\u5C55\u540D\u7A31\u3002",
  "The setting to configure (name or env var).": "\u8981\u914D\u7F6E\u7684\u8A2D\u7F6E\uFF08\u540D\u7A31\u6216\u74B0\u5883\u8B8A\u91CF\uFF09\u3002",
  "The scope to set the setting in.": "\u8A2D\u7F6E\u914D\u7F6E\u7684\u4F5C\u7528\u57DF\u3002",
  "List all settings for an extension.": "\u5217\u51FA\u64F4\u5C55\u7684\u6240\u6709\u8A2D\u7F6E\u3002",
  "Name of the extension.": "\u64F4\u5C55\u540D\u7A31\u3002",
  'Extension "{{name}}" has no settings to configure.': '\u64F4\u5C55 "{{name}}" \u6C92\u6709\u53EF\u914D\u7F6E\u7684\u8A2D\u7F6E\u3002',
  'Settings for "{{name}}":': '"{{name}}" \u7684\u8A2D\u7F6E\uFF1A',
  "(workspace)": "\uFF08\u5DE5\u4F5C\u5340\uFF09",
  "(user)": "\uFF08\u7528\u6236\uFF09",
  "[not set]": "\uFF3B\u672A\u8A2D\u7F6E\uFF3D",
  "[value stored in keychain]": "\uFF3B\u503C\u5B58\u5132\u5728\u9470\u5319\u4E32\u4E2D\uFF3D",
  "Value:": "\u503C\uFF1A",
  "Manage extension settings.": "\u7BA1\u7406\u64F4\u5C55\u8A2D\u7F6E\u3002",
  "You need to specify a command (set or list).": "\u60A8\u9700\u8981\u6307\u5B9A\u547D\u4EE4\uFF08set \u6216 list\uFF09\u3002",
  "No plugins available in this marketplace.": "\u6B64\u5E02\u5834\u4E2D\u6C92\u6709\u53EF\u7528\u7684\u63D2\u4EF6\u3002",
  'Select a plugin to install from marketplace "{{name}}":': '\u5F9E\u5E02\u5834 "{{name}}" \u4E2D\u9078\u64C7\u8981\u5B89\u88DD\u7684\u63D2\u4EF6\uFF1A',
  "Plugin selection cancelled.": "\u63D2\u4EF6\u9078\u64C7\u5DF2\u53D6\u6D88\u3002",
  'Select a plugin from "{{name}}"': '\u5F9E "{{name}}" \u4E2D\u9078\u64C7\u63D2\u4EF6',
  "Use \u2191\u2193 or j/k to navigate, Enter to select, Escape to cancel": "\u4F7F\u7528 \u2191\u2193 \u6216 j/k \u5C0E\u822A\uFF0CEnter \u9078\u64C7\uFF0CEscape \u53D6\u6D88",
  "{{count}} more above": "\u4E0A\u65B9\u9084\u6709 {{count}} \u9805",
  "{{count}} more below": "\u4E0B\u65B9\u9084\u6709 {{count}} \u9805",
  "manage IDE integration": "\u7BA1\u7406 IDE \u96C6\u6210",
  "check status of IDE integration": "\u6AA2\u67E5 IDE \u96C6\u6210\u72C0\u614B",
  "install required IDE companion for {{ideName}}": "\u5B89\u88DD {{ideName}} \u6240\u9700\u7684 IDE \u914D\u5957\u5DE5\u5177",
  "enable IDE integration": "\u555F\u7528 IDE \u96C6\u6210",
  "disable IDE integration": "\u7981\u7528 IDE \u96C6\u6210",
  "IDE integration is not supported in your current environment. To use this feature, run Qwen Code in one of these supported IDEs: VS Code or VS Code forks.": "\u60A8\u7576\u524D\u74B0\u5883\u4E0D\u652F\u6301 IDE \u96C6\u6210\u3002\u8981\u4F7F\u7528\u6B64\u529F\u80FD\uFF0C\u8ACB\u5728\u4EE5\u4E0B\u652F\u6301\u7684 IDE \u4E4B\u4E00\u4E2D\u904B\u884C Qwen Code\uFF1AVS Code \u6216 VS Code \u5206\u652F\u7248\u672C\u3002",
  "Set up GitHub Actions": "\u8A2D\u7F6E GitHub Actions",
  "Configure terminal keybindings for multiline input (VS Code, Cursor, Windsurf, Trae)": "\u914D\u7F6E\u7D42\u7AEF\u6309\u9375\u7D81\u5B9A\u4EE5\u652F\u6301\u591A\u884C\u8F38\u5165\uFF08VS Code\u3001Cursor\u3001Windsurf\u3001Trae\uFF09",
  "Please restart your terminal for the changes to take effect.": "\u8ACB\u91CD\u555F\u7D42\u7AEF\u4EE5\u4F7F\u66F4\u6539\u751F\u6548\u3002",
  "Failed to configure terminal: {{error}}": "\u914D\u7F6E\u7D42\u7AEF\u5931\u6557\uFF1A{{error}}",
  "Could not determine {{terminalName}} config path on Windows: APPDATA environment variable is not set.": "\u7121\u6CD5\u78BA\u5B9A {{terminalName}} \u5728 Windows \u4E0A\u7684\u914D\u7F6E\u8DEF\u5F91\uFF1A\u672A\u8A2D\u7F6E APPDATA \u74B0\u5883\u8B8A\u91CF\u3002",
  "{{terminalName}} keybindings.json exists but is not a valid JSON array. Please fix the file manually or delete it to allow automatic configuration.": "{{terminalName}} keybindings.json \u5B58\u5728\u4F46\u4E0D\u662F\u6709\u6548\u7684 JSON \u6578\u7D44\u3002\u8ACB\u624B\u52D5\u4FEE\u5FA9\u6A94\u6848\u6216\u522A\u9664\u5B83\u4EE5\u5141\u8A31\u81EA\u52D5\u914D\u7F6E\u3002",
  "File: {{file}}": "\u6A94\u6848\uFF1A{{file}}",
  "Failed to parse {{terminalName}} keybindings.json. The file contains invalid JSON. Please fix the file manually or delete it to allow automatic configuration.": "\u89E3\u6790 {{terminalName}} keybindings.json \u5931\u6557\u3002\u6A94\u6848\u5305\u542B\u7121\u6548\u7684 JSON\u3002\u8ACB\u624B\u52D5\u4FEE\u5FA9\u6A94\u6848\u6216\u522A\u9664\u5B83\u4EE5\u5141\u8A31\u81EA\u52D5\u914D\u7F6E\u3002",
  "Error: {{error}}": "\u932F\u8AA4\uFF1A{{error}}",
  "Shift+Enter binding already exists": "Shift+Enter \u7D81\u5B9A\u5DF2\u5B58\u5728",
  "Ctrl+Enter binding already exists": "Ctrl+Enter \u7D81\u5B9A\u5DF2\u5B58\u5728",
  "Existing keybindings detected. Will not modify to avoid conflicts.": "\u6AA2\u6E2C\u5230\u73FE\u6709\u6309\u9375\u7D81\u5B9A\u3002\u70BA\u907F\u514D\u885D\u7A81\uFF0C\u4E0D\u6703\u4FEE\u6539\u3002",
  "Please check and modify manually if needed: {{file}}": "\u5982\u6709\u9700\u8981\uFF0C\u8ACB\u624B\u52D5\u6AA2\u67E5\u4E26\u4FEE\u6539\uFF1A{{file}}",
  "Added Shift+Enter and Ctrl+Enter keybindings to {{terminalName}}.": "\u5DF2\u70BA {{terminalName}} \u6DFB\u52A0 Shift+Enter \u548C Ctrl+Enter \u6309\u9375\u7D81\u5B9A\u3002",
  "Modified: {{file}}": "\u5DF2\u4FEE\u6539\uFF1A{{file}}",
  "{{terminalName}} keybindings already configured.": "{{terminalName}} \u6309\u9375\u7D81\u5B9A\u5DF2\u914D\u7F6E\u3002",
  "Failed to configure {{terminalName}}.": "\u914D\u7F6E {{terminalName}} \u5931\u6557\u3002",
  "Your terminal is already configured for an optimal experience with multiline input (Shift+Enter and Ctrl+Enter).": "\u60A8\u7684\u7D42\u7AEF\u5DF2\u914D\u7F6E\u70BA\u652F\u6301\u591A\u884C\u8F38\u5165\uFF08Shift+Enter \u548C Ctrl+Enter\uFF09\u7684\u6700\u4F73\u9AD4\u9A57\u3002",
  "Manage Qwen Code hooks": "\u7BA1\u7406 Qwen Code Hook",
  "List all configured hooks": "\u5217\u51FA\u6240\u6709\u5DF2\u914D\u7F6E\u7684 Hook",
  Hooks: "Hook",
  "Loading hooks...": "\u6B63\u5728\u52A0\u8F09 Hook...",
  "Error loading hooks:": "\u52A0\u8F09 Hook \u51FA\u932F\uFF1A",
  "Press Escape to close": "\u6309 Escape \u95DC\u9589",
  "Press Escape, Ctrl+C, or Ctrl+D to cancel": "\u6309 Escape\u3001Ctrl+C \u6216 Ctrl+D \u53D6\u6D88",
  "Press Space, Enter, or Escape to dismiss": "\u6309 Space\u3001Enter \u6216 Escape \u95DC\u9589",
  "No hook selected": "\u672A\u9078\u64C7 Hook",
  "No hook events found.": "\u672A\u627E\u5230 Hook \u4E8B\u4EF6\u3002",
  "{{count}} hook configured": "{{count}} \u500B Hook \u5DF2\u914D\u7F6E",
  "{{count}} hooks configured": "{{count}} \u500B Hook \u5DF2\u914D\u7F6E",
  "This menu is read-only. To add or modify hooks, edit settings.json directly or ask Qwen Code.": "\u6B64\u9078\u55AE\u70BA\u53EA\u8B80\u3002\u8981\u6DFB\u52A0\u6216\u4FEE\u6539 Hook\uFF0C\u8ACB\u76F4\u63A5\u7DE8\u8F2F settings.json \u6216\u8A62\u554F Qwen Code\u3002",
  "Enter to select \xB7 Esc to cancel": "Enter \u9078\u64C7 \xB7 Esc \u53D6\u6D88",
  "Exit codes:": "\u9000\u51FA\u78BC\uFF1A",
  "Configured hooks:": "\u5DF2\u914D\u7F6E\u7684 Hook\uFF1A",
  "No hooks configured for this event.": "\u6B64\u4E8B\u4EF6\u672A\u914D\u7F6E Hook\u3002",
  "To add hooks, edit settings.json directly or ask Qwen.": "\u8981\u6DFB\u52A0 Hook\uFF0C\u8ACB\u76F4\u63A5\u7DE8\u8F2F settings.json \u6216\u8A62\u554F Qwen\u3002",
  "Enter to select \xB7 Esc to go back": "Enter \u9078\u64C7 \xB7 Esc \u8FD4\u56DE",
  "Hook details": "Hook \u8A73\u60C5",
  "Event:": "\u4E8B\u4EF6\uFF1A",
  "Extension:": "\u64F4\u5C55\uFF1A",
  "Desc:": "\u63CF\u8FF0\uFF1A",
  "No hook config selected": "\u672A\u9078\u64C7 Hook \u914D\u7F6E",
  "To modify or remove this hook, edit settings.json directly or ask Qwen to help.": "\u8981\u4FEE\u6539\u6216\u522A\u9664\u6B64 Hook\uFF0C\u8ACB\u76F4\u63A5\u7DE8\u8F2F settings.json \u6216\u8A62\u554F Qwen\u3002",
  "Hook Configuration - Disabled": "Hook \u914D\u7F6E - \u5DF2\u7981\u7528",
  "All hooks are currently disabled. You have {{count}} that are not running.": "\u6240\u6709 Hook \u7576\u524D\u5DF2\u7981\u7528\u3002\u60A8\u6709 {{count}} \u672A\u904B\u884C\u3002",
  "{{count}} configured hook": "{{count}} \u500B\u5DF2\u914D\u7F6E\u7684 Hook",
  "{{count}} configured hooks": "{{count}} \u500B\u5DF2\u914D\u7F6E\u7684 Hook",
  "When hooks are disabled:": "\u7576 Hook \u88AB\u7981\u7528\u6642\uFF1A",
  "No hook commands will execute": "\u4E0D\u6703\u57F7\u884C\u4EFB\u4F55 Hook \u547D\u4EE4",
  "StatusLine will not be displayed": "\u4E0D\u6703\u986F\u793A\u72C0\u614B\u6B04",
  "Tool operations will proceed without hook validation": "\u5DE5\u5177\u64CD\u4F5C\u5C07\u5728\u6C92\u6709 Hook \u9A57\u8B49\u7684\u60C5\u6CC1\u4E0B\u7E7C\u7E8C",
  'To re-enable hooks, remove "disableAllHooks" from settings.json or ask Qwen Code.': '\u8981\u91CD\u65B0\u555F\u7528 Hook\uFF0C\u8ACB\u5F9E settings.json \u4E2D\u522A\u9664 "disableAllHooks" \u6216\u8A62\u554F Qwen Code\u3002',
  Project: "\u9805\u76EE",
  User: "\u7528\u6236",
  Skill: "\u6280\u80FD",
  System: "\u7CFB\u7D71",
  Extension: "\u64F4\u5C55",
  "Local Settings": "\u672C\u5730\u8A2D\u7F6E",
  "User Settings": "\u7528\u6236\u8A2D\u7F6E",
  "System Settings": "\u7CFB\u7D71\u8A2D\u7F6E",
  Extensions: "\u64F4\u5C55",
  "Session (temporary)": "\u6703\u8A71\uFF08\u81E8\u6642\uFF09",
  "Before tool execution": "\u5DE5\u5177\u57F7\u884C\u524D",
  "After tool execution": "\u5DE5\u5177\u57F7\u884C\u5F8C",
  "After tool execution fails": "\u5DE5\u5177\u57F7\u884C\u5931\u6557\u5F8C",
  "When notifications are sent": "\u767C\u9001\u901A\u77E5\u6642",
  "When the user submits a prompt": "\u7528\u6236\u63D0\u4EA4\u63D0\u793A\u6642",
  "When a slash command expands into a prompt": "\u659C\u7DDA\u547D\u4EE4\u5C55\u958B\u70BA\u63D0\u793A\u6642",
  "When a new session is started": "\u65B0\u6703\u8A71\u958B\u59CB\u6642",
  "Right before Qwen Code concludes its response": "Qwen Code \u7D50\u675F\u97FF\u61C9\u4E4B\u524D",
  "When a subagent (Agent tool call) is started": "\u5B50\u667A\u80FD\u9AD4\uFF08Agent \u5DE5\u5177\u8ABF\u7528\uFF09\u555F\u52D5\u6642",
  "Right before a subagent concludes its response": "\u5B50\u667A\u80FD\u9AD4\u7D50\u675F\u97FF\u61C9\u4E4B\u524D",
  "Before conversation compaction": "\u5C0D\u8A71\u58D3\u7E2E\u524D",
  "When a session is ending": "\u6703\u8A71\u7D50\u675F\u6642",
  "When a permission dialog is displayed": "\u986F\u793A\u6B0A\u9650\u5C0D\u8A71\u6846\u6642",
  "Input to command is JSON of tool call arguments.": "\u547D\u4EE4\u8F38\u5165\u70BA\u5DE5\u5177\u8ABF\u7528\u53C3\u6578\u7684 JSON\u3002",
  'Input to command is JSON with fields "inputs" (tool call arguments) and "response" (tool call response).': '\u547D\u4EE4\u8F38\u5165\u70BA\u5305\u542B "inputs"\uFF08\u5DE5\u5177\u8ABF\u7528\u53C3\u6578\uFF09\u548C "response"\uFF08\u5DE5\u5177\u8ABF\u7528\u97FF\u61C9\uFF09\u5B57\u6BB5\u7684 JSON\u3002',
  "Input to command is JSON with tool_name, tool_input, tool_use_id, error, error_type, is_interrupt, and is_timeout.": "\u547D\u4EE4\u8F38\u5165\u70BA\u5305\u542B tool_name\u3001tool_input\u3001tool_use_id\u3001error\u3001error_type\u3001is_interrupt \u548C is_timeout \u7684 JSON\u3002",
  "Input to command is JSON with notification message and type.": "\u547D\u4EE4\u8F38\u5165\u70BA\u5305\u542B\u901A\u77E5\u6D88\u606F\u548C\u985E\u578B\u7684 JSON\u3002",
  "Input to command is JSON with original user prompt text.": "\u547D\u4EE4\u8F38\u5165\u70BA\u5305\u542B\u539F\u59CB\u7528\u6236\u63D0\u793A\u6587\u672C\u7684 JSON\u3002",
  "Input to command is JSON with command_name, command_args, and expanded prompt text.": "\u547D\u4EE4\u8F38\u5165\u70BA\u5305\u542B command_name\u3001command_args \u548C\u5C55\u958B\u5F8C\u63D0\u793A\u6587\u672C\u7684 JSON\u3002",
  "Input to command is JSON with session start source.": "\u547D\u4EE4\u8F38\u5165\u70BA\u5305\u542B\u6703\u8A71\u555F\u52D5\u4F86\u6E90\u7684 JSON\u3002",
  "Input to command is JSON with session end reason.": "\u547D\u4EE4\u8F38\u5165\u70BA\u5305\u542B\u6703\u8A71\u7D50\u675F\u539F\u56E0\u7684 JSON\u3002",
  "Input to command is JSON with agent_id and agent_type.": "\u547D\u4EE4\u8F38\u5165\u70BA\u5305\u542B agent_id \u548C agent_type \u7684 JSON\u3002",
  "Input to command is JSON with agent_id, agent_type, and agent_transcript_path.": "\u547D\u4EE4\u8F38\u5165\u70BA\u5305\u542B agent_id\u3001agent_type \u548C agent_transcript_path \u7684 JSON\u3002",
  "Input to command is JSON with compaction details.": "\u547D\u4EE4\u8F38\u5165\u70BA\u5305\u542B\u58D3\u7E2E\u8A73\u60C5\u7684 JSON\u3002",
  "Input to command is JSON with tool_name, tool_input, and tool_use_id. Output JSON with hookSpecificOutput containing decision to allow or deny.": "\u547D\u4EE4\u8F38\u5165\u70BA\u5305\u542B tool_name\u3001tool_input \u548C tool_use_id \u7684 JSON\u3002\u8F38\u51FA\u5305\u542B hookSpecificOutput \u7684 JSON\uFF0C\u5176\u4E2D\u5305\u542B\u5141\u8A31\u6216\u62D2\u7D55\u7684\u6C7A\u5B9A\u3002",
  "stdout/stderr not shown": "stdout/stderr \u4E0D\u986F\u793A",
  "show stderr to model and continue conversation": "\u5411\u6A21\u578B\u986F\u793A stderr \u4E26\u7E7C\u7E8C\u5C0D\u8A71",
  "show stderr to user only": "\u50C5\u5411\u7528\u6236\u986F\u793A stderr",
  "stdout shown in transcript mode (ctrl+o)": "stdout \u4EE5\u8F49\u9304\u6A21\u5F0F\u986F\u793A (ctrl+o)",
  "show stderr to model immediately": "\u7ACB\u5373\u5411\u6A21\u578B\u986F\u793A stderr",
  "show stderr to user only but continue with tool call": "\u50C5\u5411\u7528\u6236\u986F\u793A stderr \u4F46\u7E7C\u7E8C\u5DE5\u5177\u8ABF\u7528",
  "block processing, erase original prompt, and show stderr to user only": "\u963B\u6B62\u8655\u7406\uFF0C\u64E6\u9664\u539F\u59CB\u63D0\u793A\uFF0C\u50C5\u5411\u7528\u6236\u986F\u793A stderr",
  "block expanded prompt submission and show stderr to user only": "\u963B\u6B62\u63D0\u4EA4\u5C55\u958B\u5F8C\u7684\u63D0\u793A\uFF0C\u4E26\u50C5\u5411\u7528\u6236\u986F\u793A stderr",
  "stdout shown to Qwen": "\u5411 Qwen \u986F\u793A stdout",
  "show stderr to user only (blocking errors ignored)": "\u50C5\u5411\u7528\u6236\u986F\u793A stderr\uFF08\u5FFD\u7565\u963B\u585E\u932F\u8AA4\uFF09",
  "command completes successfully": "\u547D\u4EE4\u6210\u529F\u5B8C\u6210",
  "stdout shown to subagent": "\u5411\u5B50\u667A\u80FD\u9AD4\u986F\u793A stdout",
  "show stderr to subagent and continue having it run": "\u5411\u5B50\u667A\u80FD\u9AD4\u986F\u793A stderr \u4E26\u7E7C\u7E8C\u904B\u884C",
  "stdout appended as custom compact instructions": "stdout \u4F5C\u70BA\u81EA\u5B9A\u7FA9\u58D3\u7E2E\u6307\u4EE4\u8FFD\u52A0",
  "block compaction": "\u963B\u6B62\u58D3\u7E2E",
  "show stderr to user only but continue with compaction": "\u50C5\u5411\u7528\u6236\u986F\u793A stderr \u4F46\u7E7C\u7E8C\u58D3\u7E2E",
  "use hook decision if provided": "\u5982\u679C\u63D0\u4F9B\u5247\u4F7F\u7528 Hook \u6C7A\u5B9A",
  "Config not loaded.": "\u914D\u7F6E\u672A\u52A0\u8F09\u3002",
  "Hooks are not enabled. Enable hooks in settings to use this feature.": "Hook \u672A\u555F\u7528\u3002\u8ACB\u5728\u8A2D\u7F6E\u4E2D\u555F\u7528 Hook \u4EE5\u4F7F\u7528\u6B64\u529F\u80FD\u3002",
  "Export current session message history to a file": "\u5C07\u7576\u524D\u6703\u8A71\u7684\u6D88\u606F\u8A18\u9304\u5C0E\u51FA\u5230\u6A94\u6848",
  "Export session to HTML format": "\u5C07\u6703\u8A71\u5C0E\u51FA\u70BA HTML \u6A94\u6848",
  "Export session to JSON format": "\u5C07\u6703\u8A71\u5C0E\u51FA\u70BA JSON \u6A94\u6848",
  "Export session to JSONL format (one message per line)": "\u5C07\u6703\u8A71\u5C0E\u51FA\u70BA JSONL \u6A94\u6848\uFF08\u6BCF\u884C\u4E00\u689D\u6D88\u606F\uFF09",
  "Export session to markdown format": "\u5C07\u6703\u8A71\u5C0E\u51FA\u70BA Markdown \u6A94\u6848",
  "generate personalized programming insights from your chat history": "\u6839\u64DA\u4F60\u7684\u804A\u5929\u8A18\u9304\u751F\u6210\u500B\u6027\u5316\u7DE8\u7A0B\u6D1E\u5BDF",
  "Resume a previous session": "\u6062\u5FA9\u5148\u524D\u6703\u8A71",
  "Fork the current conversation into a new session": "\u5C07\u76EE\u524D\u5C0D\u8A71\u5206\u652F\u5230\u65B0\u6703\u8A71",
  "Spawn a background agent that inherits the full conversation": "\u555F\u52D5\u7E7C\u627F\u5B8C\u6574\u5C0D\u8A71\u7684\u80CC\u666F\u667A\u80FD\u9AD4",
  "Please provide a directive. Usage: /fork <directive>": "\u8ACB\u63D0\u4F9B\u6307\u4EE4\u3002\u7528\u6CD5\uFF1A/fork <\u6307\u4EE4>",
  "Cannot fork while a response or tool call is in progress. Wait for it to finish or resolve the pending tool call.": "\u56DE\u61C9\u6216\u5DE5\u5177\u547C\u53EB\u6B63\u5728\u9032\u884C\u6642\u7121\u6CD5\u5206\u652F\u3002\u8ACB\u7B49\u5F85\u5176\u5B8C\u6210\u6216\u8655\u7406\u5F85\u78BA\u8A8D\u7684\u5DE5\u5177\u547C\u53EB\u3002",
  "Cannot fork before the first conversation turn.": "\u9996\u6B21\u5C0D\u8A71\u8F2A\u6B21\u524D\u7121\u6CD5\u5206\u652F\u3002",
  "The agent tool is unavailable; cannot fork.": "Agent \u5DE5\u5177\u4E0D\u53EF\u7528\uFF1B\u7121\u6CD5\u5206\u652F\u3002",
  "Failed to launch fork: {{error}}": "\u555F\u52D5\u5206\u652F\u5931\u6557\uFF1A{{error}}",
  "the background agent could not be started.": "\u80CC\u666F\u667A\u80FD\u9AD4\u7121\u6CD5\u555F\u52D5\u3002",
  "User launched a background fork via /fork: {{directive}}": "\u4F7F\u7528\u8005\u900F\u904E /fork \u555F\u52D5\u4E86\u80CC\u666F\u5206\u652F\uFF1A{{directive}}",
  "Forked into a background agent. It inherits this conversation and runs without blocking \u2014 track it in the background tasks panel; it reports back when done.": "\u5DF2\u5206\u652F\u5230\u80CC\u666F\u667A\u80FD\u9AD4\u3002\u5B83\u6703\u7E7C\u627F\u6B64\u5C0D\u8A71\u4E26\u4EE5\u975E\u963B\u585E\u65B9\u5F0F\u57F7\u884C\uFF0C\u53EF\u5728\u80CC\u666F\u4EFB\u52D9\u9762\u677F\u4E2D\u8FFD\u8E64\uFF1B\u5B8C\u6210\u5F8C\u6703\u56DE\u5831\u7D50\u679C\u3002",
  "Cannot branch while a response or tool call is in progress. Wait for it to finish or resolve the pending tool call.": "\u56DE\u61C9\u6216\u5DE5\u5177\u547C\u53EB\u6B63\u5728\u9032\u884C\u6642\u7121\u6CD5\u5206\u652F\u3002\u8ACB\u7B49\u5F85\u5176\u5B8C\u6210\u6216\u8655\u7406\u5F85\u78BA\u8A8D\u7684\u5DE5\u5177\u547C\u53EB\u3002",
  "No conversation to branch.": "\u6C92\u6709\u53EF\u5206\u652F\u7684\u5C0D\u8A71\u3002",
  "Restore a tool call. This will reset the conversation and file history to the state it was in when the tool call was suggested": "\u6062\u5FA9\u67D0\u6B21\u5DE5\u5177\u8ABF\u7528\u3002\u9019\u5C07\u628A\u5C0D\u8A71\u8207\u6A94\u6848\u6B77\u53F2\u91CD\u7F6E\u5230\u63D0\u51FA\u8A72\u5DE5\u5177\u8ABF\u7528\u5EFA\u8B70\u6642\u7684\u72C0\u614B",
  "Could not detect terminal type. Supported terminals: VS Code, Cursor, Windsurf, and Trae.": "\u7121\u6CD5\u6AA2\u6E2C\u7D42\u7AEF\u985E\u578B\u3002\u652F\u6301\u7684\u7D42\u7AEF\uFF1AVS Code\u3001Cursor\u3001Windsurf \u548C Trae\u3002",
  'Terminal "{{terminal}}" is not supported yet.': '\u7D42\u7AEF "{{terminal}}" \u5C1A\u672A\u652F\u6301\u3002',
  "Invalid language. Available: {{options}}": "\u7121\u6548\u7684\u8A9E\u8A00\u3002\u53EF\u7528\u9078\u9805\uFF1A{{options}}",
  "Language subcommands do not accept additional arguments.": "\u8A9E\u8A00\u5B50\u547D\u4EE4\u4E0D\u63A5\u53D7\u984D\u5916\u53C3\u6578",
  "Current UI language: {{lang}}": "\u7576\u524D UI \u8A9E\u8A00\uFF1A{{lang}}",
  "Current LLM output language: {{lang}}": "\u7576\u524D LLM \u8F38\u51FA\u8A9E\u8A00\uFF1A{{lang}}",
  "Set UI language": "\u8A2D\u7F6E UI \u8A9E\u8A00",
  "Set LLM output language": "\u8A2D\u7F6E LLM \u8F38\u51FA\u8A9E\u8A00",
  "Usage: /language ui [{{options}}]": "\u7528\u6CD5\uFF1A/language ui [{{options}}]",
  "Usage: /language output <language>": "\u7528\u6CD5\uFF1A/language output <\u8A9E\u8A00>",
  "Example: /language output \u4E2D\u6587": "\u793A\u4F8B\uFF1A/language output \u4E2D\u6587",
  "Example: /language output English": "\u793A\u4F8B\uFF1A/language output English",
  "Example: /language output \u65E5\u672C\u8A9E": "\u793A\u4F8B\uFF1A/language output \u65E5\u672C\u8A9E",
  "UI language changed to {{lang}}": "UI \u8A9E\u8A00\u5DF2\u66F4\u6539\u70BA {{lang}}",
  "LLM output language set to {{lang}}": "LLM \u8F38\u51FA\u8A9E\u8A00\u5DF2\u8A2D\u7F6E\u70BA {{lang}}",
  "Please restart the application for the changes to take effect.": "\u8ACB\u91CD\u555F\u61C9\u7528\u7A0B\u5E8F\u4EE5\u4F7F\u66F4\u6539\u751F\u6548\u3002",
  "Failed to generate LLM output language rule file: {{error}}": "\u751F\u6210 LLM \u8F38\u51FA\u8A9E\u8A00\u898F\u5247\u6A94\u6848\u5931\u6557\uFF1A{{error}}",
  "Invalid command. Available subcommands:": "\u7121\u6548\u7684\u547D\u4EE4\u3002\u53EF\u7528\u7684\u5B50\u547D\u4EE4\uFF1A",
  "Available subcommands:": "\u53EF\u7528\u7684\u5B50\u547D\u4EE4\uFF1A",
  "To request additional UI language packs, please open an issue on GitHub.": "\u5982\u9700\u8ACB\u6C42\u5176\u4ED6 UI \u8A9E\u8A00\u5305\uFF0C\u8ACB\u5728 GitHub \u4E0A\u63D0\u4EA4 issue",
  "Available options:": "\u53EF\u7528\u9078\u9805\uFF1A",
  "Set UI language to {{name}}": "\u5C07 UI \u8A9E\u8A00\u8A2D\u7F6E\u70BA {{name}}",
  "Tool Approval Mode": "\u5DE5\u5177\u5BE9\u6279\u6A21\u5F0F",
  "Analyze only, do not modify files or execute commands": "\u50C5\u5206\u6790\uFF0C\u4E0D\u4FEE\u6539\u6A94\u6848\u6216\u57F7\u884C\u547D\u4EE4",
  "Require approval for file edits or shell commands": "\u9700\u8981\u6279\u51C6\u6A94\u6848\u7DE8\u8F2F\u6216 shell \u547D\u4EE4",
  "Automatically approve file edits": "\u81EA\u52D5\u6279\u51C6\u6A94\u6848\u7DE8\u8F2F",
  "Use classifier to automatically approve safe tool calls": "\u4F7F\u7528\u5206\u985E\u5668\u81EA\u52D5\u6279\u51C6\u5B89\u5168\u7684\u5DE5\u5177\u8ABF\u7528",
  "Automatically approve all tools": "\u81EA\u52D5\u6279\u51C6\u6240\u6709\u5DE5\u5177",
  "Workspace approval mode exists and takes priority. User-level change will have no effect.": "\u5DE5\u4F5C\u5340\u5BE9\u6279\u6A21\u5F0F\u5DF2\u5B58\u5728\u4E26\u5177\u6709\u512A\u5148\u7D1A\u3002\u7528\u6236\u7D1A\u5225\u7684\u66F4\u6539\u5C07\u7121\u6548\u3002",
  "Apply To": "\u61C9\u7528\u65BC",
  "Workspace Settings": "\u5DE5\u4F5C\u5340\u8A2D\u7F6E",
  "Open auto-memory folder": "\u6253\u958B\u81EA\u52D5\u8A18\u61B6\u6A94\u6848\u593E",
  "Auto-memory: {{status}}": "\u81EA\u52D5\u8A18\u61B6\uFF1A{{status}}",
  "Auto-dream: {{status}} \xB7 {{lastDream}} \xB7 /dream to run": "\u81EA\u52D5\u6574\u7406\uFF1A{{status}} \xB7 {{lastDream}} \xB7 /dream \u7ACB\u5373\u904B\u884C",
  "Auto-skill: {{status}}": "\u81EA\u52D5\u6280\u80FD\uFF1A{{status}}",
  never: "\u5F9E\u672A",
  on: "\u958B",
  off: "\u95DC",
  "Remove matching entries from managed auto-memory.": "\u5F9E\u8A17\u7BA1\u81EA\u52D5\u8A18\u61B6\u4E2D\u522A\u9664\u5339\u914D\u7684\u689D\u76EE\u3002",
  "Usage: /forget <memory text to remove>": "\u7528\u6CD5\uFF1A/forget <\u8981\u522A\u9664\u7684\u8A18\u61B6\u6587\u672C>",
  "No managed auto-memory entries matched: {{query}}": "\u6C92\u6709\u5339\u914D\u7684\u8A17\u7BA1\u81EA\u52D5\u8A18\u61B6\u689D\u76EE\uFF1A{{query}}",
  "Consolidate managed auto-memory topic files.": "\u6574\u7406\u8A17\u7BA1\u81EA\u52D5\u8A18\u61B6\u4E3B\u984C\u6A94\u6848",
  "Import MCP servers from Claude configs": "\u5F9E Claude \u8A2D\u5B9A\u532F\u5165 MCP \u4F3A\u670D\u5668",
  "Open MCP management dialog": "\u6253\u958B MCP \u7BA1\u7406\u5C0D\u8A71\u6846",
  "Could not retrieve tool registry.": "\u7121\u6CD5\u6AA2\u7D22\u5DE5\u5177\u8A3B\u518A\u8868",
  "Successfully authenticated and refreshed tools for '{{name}}'.": "\u6210\u529F\u8A8D\u8B49\u4E26\u5237\u65B0\u4E86 '{{name}}' \u7684\u5DE5\u5177",
  "Re-discovering tools from '{{name}}'...": "\u6B63\u5728\u91CD\u65B0\u767C\u73FE '{{name}}' \u7684\u5DE5\u5177...",
  "Discovered {{count}} tool(s) from '{{name}}'.": "\u5F9E '{{name}}' \u767C\u73FE\u4E86 {{count}} \u500B\u5DE5\u5177\u3002",
  "Authentication complete. Returning to server details...": "\u8A8D\u8B49\u5B8C\u6210\uFF0C\u6B63\u5728\u8FD4\u56DE\u4F3A\u670D\u5668\u8A73\u60C5...",
  "Authentication successful.": "\u8A8D\u8B49\u6210\u529F\u3002",
  "Manage MCP servers": "\u7BA1\u7406 MCP servers",
  "Server Detail": "\u4F3A\u670D\u5668\u8A73\u60C5",
  Tools: "\u5DE5\u5177",
  "Tool Detail": "\u5DE5\u5177\u8A73\u60C5",
  "Loading...": "\u52A0\u8F09\u4E2D...",
  "Unknown step": "\u672A\u77E5\u6B65\u9A5F",
  "Esc to back": "Esc \u8FD4\u56DE",
  "\u2191\u2193 to navigate \xB7 Enter to select \xB7 Esc to close": "\u2191\u2193 \u5C0E\u822A \xB7 Enter \u9078\u64C7 \xB7 Esc \u95DC\u9589",
  "\u2191\u2193 to navigate \xB7 Enter to select \xB7 Esc to back": "\u2191\u2193 \u5C0E\u822A \xB7 Enter \u9078\u64C7 \xB7 Esc \u8FD4\u56DE",
  "\u2191\u2193 to navigate \xB7 Enter to confirm \xB7 Esc to back": "\u2191\u2193 \u5C0E\u822A \xB7 Enter \u78BA\u8A8D \xB7 Esc \u8FD4\u56DE",
  "User Settings (global)": "\u7528\u6236\u8A2D\u7F6E\uFF08\u5168\u5C40\uFF09",
  "Workspace Settings (project-specific)": "\u5DE5\u4F5C\u5340\u8A2D\u7F6E\uFF08\u9805\u76EE\u7D1A\uFF09",
  "Disable server:": "\u7981\u7528\u4F3A\u670D\u5668\uFF1A",
  "Select where to add the server to the exclude list:": "\u9078\u64C7\u5C07\u4F3A\u670D\u5668\u6DFB\u52A0\u5230\u6392\u9664\u5217\u8868\u7684\u4F4D\u7F6E\uFF1A",
  "Press Enter to confirm, Esc to cancel": "\u6309 Enter \u78BA\u8A8D\uFF0CEsc \u53D6\u6D88",
  "View tools": "\u67E5\u770B\u5DE5\u5177",
  "View resources": "\u67E5\u770B\u8CC7\u6E90",
  Reconnect: "\u91CD\u65B0\u9023\u63A5",
  Enable: "\u555F\u7528",
  Disable: "\u7981\u7528",
  Authenticate: "\u8A8D\u8B49",
  "Re-authenticate": "\u91CD\u65B0\u8A8D\u8B49",
  "Clear Authentication": "\u6E05\u7A7A\u8A8D\u8B49",
  "Server:": "\u4F3A\u670D\u5668\uFF1A",
  "Command:": "\u547D\u4EE4\uFF1A",
  "Working Directory:": "\u5DE5\u4F5C\u76EE\u9304\uFF1A",
  "No server selected": "\u672A\u9078\u64C7\u4F3A\u670D\u5668",
  prompts: "\u63D0\u793A\u8A5E",
  "Error:": "\u932F\u8AA4\uFF1A",
  tool: "\u5DE5\u5177",
  tools: "\u500B\u5DE5\u5177",
  resource: "\u8CC7\u6E90",
  resources: "\u500B\u8CC7\u6E90",
  connected: "\u5DF2\u9023\u63A5",
  connecting: "\u9023\u63A5\u4E2D",
  disconnected: "\u5DF2\u65B7\u958B",
  "needs authentication": "\u9700\u8981\u8A8D\u8B49",
  "User MCPs": "\u7528\u6236 MCP",
  "Project MCPs": "\u9805\u76EE MCP",
  "Extension MCPs": "\u64F4\u5C55 MCP",
  server: "\u500B\u4F3A\u670D\u5668",
  servers: "\u500B\u4F3A\u670D\u5668",
  "Add MCP servers to your settings to get started.": "\u8ACB\u5728\u8A2D\u7F6E\u4E2D\u6DFB\u52A0 MCP servers \u4EE5\u958B\u59CB\u4F7F\u7528\u3002",
  "Run qwen --debug to see error logs": "\u904B\u884C qwen --debug \u67E5\u770B\u932F\u8AA4\u65E5\u8A8C",
  "OAuth Authentication": "OAuth \u8A8D\u8B49",
  "Authenticating... Please complete the login in your browser.": "\u8A8D\u8B49\u4E2D... \u8ACB\u5728\u700F\u89BD\u5668\u4E2D\u5B8C\u6210\u767B\u9304\u3002",
  "Press c to copy the authorization URL to your clipboard.": "\u6309 c \u8907\u88FD\u6388\u6B0A URL \u5230\u526A\u8CBC\u677F\u3002",
  "Copy request sent to your terminal. If paste is empty, copy the URL above manually.": "\u5DF2\u5411\u7D42\u7AEF\u767C\u9001\u8907\u88FD\u8ACB\u6C42\uFF1B\u82E5\u7C98\u8CBC\u70BA\u7A7A\uFF0C\u8ACB\u624B\u52D5\u8907\u88FD\u4E0A\u65B9 URL\u3002",
  "Cannot write to terminal \u2014 copy the URL above manually.": "\u7121\u6CD5\u5BEB\u5165\u7D42\u7AEF\uFF0C\u8ACB\u624B\u52D5\u8907\u88FD\u4E0A\u65B9 URL\u3002",
  "No tools available for this server.": "\u6B64\u4F3A\u670D\u5668\u6C92\u6709\u53EF\u7528\u5DE5\u5177\u3002",
  destructive: "\u7834\u58DE\u6027",
  "read-only": "\u53EA\u8B80",
  "open-world": "\u958B\u653E\u4E16\u754C",
  idempotent: "\u51AA\u7B49",
  "Tools for {{serverName}}": "{{serverName}} \u7684\u5DE5\u5177",
  "{{current}}/{{total}}": "{{current}}/{{total}}",
  required: "\u5FC5\u586B",
  Parameters: "\u53C3\u6578",
  "No tool selected": "\u672A\u9078\u64C7\u5DE5\u5177",
  Server: "\u4F3A\u670D\u5668",
  // MCP Resource List/Detail
  "No resources available for this server.": "\u6B64\u4F3A\u670D\u5668\u6C92\u6709\u53EF\u7528\u8CC7\u6E90\u3002",
  "Resources for {{serverName}}": "{{serverName}} \u7684\u8CC7\u6E90",
  "No resource selected": "\u672A\u9078\u64C7\u8CC7\u6E90",
  "Resource Detail": "\u8CC7\u6E90\u8A73\u60C5",
  "URI:": "URI\uFF1A",
  "MIME Type:": "MIME \u985E\u578B\uFF1A",
  "Size:": "\u5927\u5C0F\uFF1A",
  "{{count}} bytes": "{{count}} \u4F4D\u5143\u7D44",
  "Reference in chat": "\u5728\u5C0D\u8A71\u4E2D\u5F15\u7528",
  "MCP resource server": "MCP \u8CC7\u6E90\u4F3A\u670D\u5668",
  "{{count}} invalid tools": "{{count}} \u500B\u7121\u6548\u5DE5\u5177",
  invalid: "\u7121\u6548",
  "invalid: {{reason}}": "\u7121\u6548\uFF1A{{reason}}",
  "missing name": "\u7F3A\u5C11\u540D\u7A31",
  "missing description": "\u7F3A\u5C11\u63CF\u8FF0",
  "(unnamed)": "(\u672A\u547D\u540D)",
  "Warning: This tool cannot be called by the LLM": "\u8B66\u544A\uFF1A\u6B64\u5DE5\u5177\u7121\u6CD5\u88AB LLM \u8ABF\u7528",
  Reason: "\u539F\u56E0",
  "Tools must have both name and description to be used by the LLM.": "\u5DE5\u5177\u5FC5\u9808\u540C\u6642\u5177\u6709\u540D\u7A31\u548C\u63CF\u8FF0\u624D\u80FD\u88AB LLM \u4F7F\u7528\u3002",
  "Generate a project summary and save it to .qwen/PROJECT_SUMMARY.md": "\u751F\u6210\u9805\u76EE\u6458\u8981\u4E26\u4FDD\u5B58\u5230 .qwen/PROJECT_SUMMARY.md",
  "No chat client available to generate summary.": "\u6C92\u6709\u53EF\u7528\u7684\u804A\u5929\u5BA2\u6236\u7AEF\u4F86\u751F\u6210\u6458\u8981",
  "Already generating summary, wait for previous request to complete": "\u6B63\u5728\u751F\u6210\u6458\u8981\uFF0C\u8ACB\u7B49\u5F85\u4E0A\u4E00\u500B\u8ACB\u6C42\u5B8C\u6210",
  "No conversation found to summarize.": "\u672A\u627E\u5230\u8981\u7E3D\u7D50\u7684\u5C0D\u8A71",
  "Failed to generate project context summary: {{error}}": "\u751F\u6210\u9805\u76EE\u4E0A\u4E0B\u6587\u6458\u8981\u5931\u6557\uFF1A{{error}}",
  "Saved project summary to {{filePathForDisplay}}.": "\u9805\u76EE\u6458\u8981\u5DF2\u4FDD\u5B58\u5230 {{filePathForDisplay}}",
  "Saving project summary...": "\u6B63\u5728\u4FDD\u5B58\u9805\u76EE\u6458\u8981...",
  "Generating project summary...": "\u6B63\u5728\u751F\u6210\u9805\u76EE\u6458\u8981...",
  "Processing summary...": "\u6B63\u5728\u8655\u7406\u6458\u8981...",
  "Project summary generated and saved successfully!": "\u9805\u76EE\u6458\u8981\u5DF2\u751F\u6210\u4E26\u6210\u529F\u4FDD\u5B58\uFF01",
  "Saved to: {{filePath}}": "\u5132\u5B58\u81F3\uFF1A{{filePath}}",
  "Failed to generate summary - no text content received from LLM response": "\u751F\u6210\u6458\u8981\u5931\u6557 - \u672A\u5F9E LLM \u97FF\u61C9\u4E2D\u63A5\u6536\u5230\u6587\u672C\u5167\u5BB9",
  "Switch the model for this session (--fast for suggestion model, [model-id] to switch immediately).": "\u5207\u63DB\u6B64\u6703\u8A71\u7684\u6A21\u578B\uFF08--fast \u53EF\u8A2D\u7F6E\u5EFA\u8B70\u6A21\u578B\uFF09",
  "Switch the model for this session (--fast for suggestion model, --voice for voice transcription model, [model-id] to switch immediately).": "\u5207\u63DB\u6B64\u6703\u8A71\u7684\u6A21\u578B\uFF08--fast \u53EF\u8A2D\u7F6E\u5EFA\u8B70\u6A21\u578B\uFF0C--voice \u53EF\u8A2D\u7F6E\u8A9E\u97F3\u8F49\u5BEB\u6A21\u578B\uFF0C[model-id] \u53EF\u7ACB\u5373\u5207\u63DB\uFF09",
  "Switch the model for this session (--fast for suggestion model, --voice for voice transcription model, --vision for the vision bridge model, [model-id] to switch immediately).": "\u5207\u63DB\u6B64\u6703\u8A71\u7684\u6A21\u578B\uFF08--fast \u53EF\u8A2D\u7F6E\u5EFA\u8B70\u6A21\u578B\uFF0C--voice \u53EF\u8A2D\u7F6E\u8A9E\u97F3\u8F49\u5BEB\u6A21\u578B\uFF0C--vision \u53EF\u8A2D\u7F6E\u8996\u89BA\u6A4B\u63A5\u6A21\u578B\uFF0C[model-id] \u53EF\u7ACB\u5373\u5207\u63DB\uFF09",
  "\u26A0 '{{model}}' is not a known image-capable model; the vision bridge may fail on images.": "\u26A0 '{{model}}' \u4E0D\u662F\u5DF2\u77E5\u7684\u5716\u50CF\u80FD\u529B\u6A21\u578B\uFF1B\u8996\u89BA\u6A4B\u63A5\u8655\u7406\u5716\u7247\u6642\u53EF\u80FD\u6703\u5931\u6557\u3002",
  "Set a lighter model for prompt suggestions and speculative execution": "\u8A2D\u7F6E\u7528\u65BC\u8F38\u5165\u5EFA\u8B70\u548C\u63A8\u6E2C\u57F7\u884C\u7684\u8F15\u91CF\u6A21\u578B",
  "Toggle voice dictation input": "\u5207\u63DB\u8A9E\u97F3\u807D\u5BEB\u8F38\u5165",
  "Set the model for voice transcription": "\u8A2D\u5B9A\u8A9E\u97F3\u8F49\u5BEB\u6A21\u578B",
  "Set the image-capable model used to transcribe images for a text-only main model": "\u8A2D\u5B9A\u7528\u65BC\u70BA\u7D14\u6587\u5B57\u4E3B\u6A21\u578B\u8F49\u5BEB\u5716\u50CF\u7684\u5716\u50CF\u80FD\u529B\u6A21\u578B",
  "Select Fast Model": "\u9078\u64C7\u5FEB\u901F\u6A21\u578B",
  "Select Vision Model": "\u9078\u64C7\u8996\u89BA\u6A21\u578B",
  "Select Voice Model": "\u9078\u64C7\u8A9E\u97F3\u6A21\u578B",
  "Vision Model": "\u8996\u89BA\u6A21\u578B",
  "Voice Model": "\u8A9E\u97F3\u6A21\u578B",
  "Selected voice model is unavailable.": "\u6240\u9078\u8A9E\u97F3\u6A21\u578B\u4E0D\u53EF\u7528\u3002",
  "Voice model '{{model}}' is configured more than once. Remove duplicate model ids before selecting it for voice transcription.": "\u8A9E\u97F3\u6A21\u578B '{{model}}' \u88AB\u914D\u7F6E\u4E86\u591A\u6B21\u3002\u8ACB\u5148\u79FB\u9664\u91CD\u8907\u7684\u6A21\u578B ID\uFF0C\u518D\u5C07\u5176\u9078\u70BA\u8A9E\u97F3\u8F49\u5BEB\u6A21\u578B\u3002",
  "Voice dictation: {{status}} (mode: {{mode}}, {{modelText}}).": "\u8A9E\u97F3\u807D\u5BEB\uFF1A{{status}}\uFF08\u6A21\u5F0F\uFF1A{{mode}}\uFF0C{{modelText}}\uFF09\u3002",
  "model: {{voiceModel}}": "\u6A21\u578B\uFF1A{{voiceModel}}",
  "no voice model selected": "\u672A\u9078\u64C7\u8A9E\u97F3\u6A21\u578B",
  "Voice dictation disabled.": "\u8A9E\u97F3\u807D\u5BEB\u5DF2\u505C\u7528\u3002",
  "Usage: /voice [hold|tap|off|status]": "\u7528\u6CD5\uFF1A/voice [hold|tap|off|status]",
  "No voice model selected. Run /model --voice to choose one before enabling voice dictation.": "\u672A\u9078\u64C7\u8A9E\u97F3\u6A21\u578B\u3002\u8ACB\u5148\u57F7\u884C /model --voice \u9078\u64C7\u6A21\u578B\uFF0C\u518D\u555F\u7528\u8A9E\u97F3\u807D\u5BEB\u3002",
  "Voice dictation enabled (tap mode). Tap Space at an empty prompt to start, tap again or pause to stop and submit, using {{voiceModel}}.": "\u8A9E\u97F3\u807D\u5BEB\u5DF2\u555F\u7528\uFF08\u9EDE\u6309\u6A21\u5F0F\uFF09\u3002\u5728\u7A7A\u8F38\u5165\u6846\u4E2D\u9EDE\u6309 Space \u958B\u59CB\uFF0C\u518D\u9EDE\u6309\u4E00\u6B21\u6216\u505C\u9813\u5F8C\u505C\u6B62\u4E26\u63D0\u4EA4\uFF0C\u4F7F\u7528 {{voiceModel}}\u3002",
  "Voice dictation enabled (hold mode). Hold Space at an empty prompt to dictate with {{voiceModel}}.": "\u8A9E\u97F3\u807D\u5BEB\u5DF2\u555F\u7528\uFF08\u6309\u4F4F\u6A21\u5F0F\uFF09\u3002\u5728\u7A7A\u8F38\u5165\u6846\u4E2D\u6309\u4F4F Space\uFF0C\u4F7F\u7528 {{voiceModel}} \u807D\u5BEB\u3002",
  "No models are configured.": "\u672A\u8A2D\u5B9A\u6A21\u578B\u3002",
  "Configured models: {{models}}.": "\u5DF2\u8A2D\u5B9A\u6A21\u578B\uFF1A{{models}}\u3002",
  "Configure a unique model id in settings.modelProviders or run /model --voice to select an available model.": "\u8ACB\u5728 settings.modelProviders \u4E2D\u8A2D\u5B9A\u552F\u4E00\u7684\u6A21\u578B ID\uFF0C\u6216\u57F7\u884C /model --voice \u9078\u64C7\u53EF\u7528\u6A21\u578B\u3002",
  "Voice model '{{modelName}}' is not configured.": "\u8A9E\u97F3\u6A21\u578B '{{modelName}}' \u672A\u8A2D\u5B9A\u3002",
  "Voice model '{{modelName}}' cannot be used for transcription.": "\u8A9E\u97F3\u6A21\u578B '{{modelName}}' \u4E0D\u80FD\u7528\u65BC\u8F49\u5BEB\u3002",
  "Voice model '{{modelName}}' cannot be used for transcription. Configure an OpenAI-compatible model with baseUrl in settings.modelProviders.": "\u8A9E\u97F3\u6A21\u578B '{{modelName}}' \u4E0D\u80FD\u7528\u65BC\u8F49\u5BEB\u3002\u8ACB\u5728 settings.modelProviders \u4E2D\u8A2D\u5B9A\u5E36 baseUrl \u7684 OpenAI \u76F8\u5BB9\u6A21\u578B\u3002",
  "Configure an OpenAI-compatible model with baseUrl in settings.modelProviders.": "\u8ACB\u5728 settings.modelProviders \u4E2D\u8A2D\u5B9A\u5E36 baseUrl \u7684 OpenAI \u76F8\u5BB9\u6A21\u578B\u3002",
  "Microphone access is denied. Enable it for your terminal in System Settings \u2192 Privacy & Security \u2192 Microphone, then restart voice dictation.": "\u9EA5\u514B\u98A8\u5B58\u53D6\u88AB\u62D2\u7D55\u3002\u8ACB\u5728\u7CFB\u7D71\u8A2D\u5B9A \u2192 \u96B1\u79C1\u6B0A\u8207\u5B89\u5168\u6027 \u2192 \u9EA5\u514B\u98A8\u4E2D\u5141\u8A31\u76EE\u524D\u7D42\u7AEF\u6A5F\u5B58\u53D6\uFF0C\u7136\u5F8C\u91CD\u65B0\u555F\u52D5\u8A9E\u97F3\u807D\u5BEB\u3002",
  "Voice dictation is not supported on {{platform}}.": "\u8A9E\u97F3\u807D\u5BEB\u4E0D\u652F\u63F4 {{platform}}\u3002",
  "Voice dictation needs microphone access, which is unavailable in this WSL session. Use WSLg/PulseAudio, or run Qwen Code on a host with a microphone.": "\u8A9E\u97F3\u807D\u5BEB\u9700\u8981\u9EA5\u514B\u98A8\u5B58\u53D6\uFF0C\u4F46\u76EE\u524D WSL \u6703\u8A71\u4E0D\u53EF\u7528\u3002\u8ACB\u4F7F\u7528 WSLg/PulseAudio\uFF0C\u6216\u5728\u5177\u5099\u9EA5\u514B\u98A8\u7684\u4E3B\u6A5F\u4E0A\u57F7\u884C Qwen Code\u3002",
  "Voice dictation needs microphone access. macOS will ask the first time you record \u2014 approve it, then start again. Your first recording may be empty while the dialog is open.": "\u8A9E\u97F3\u807D\u5BEB\u9700\u8981\u9EA5\u514B\u98A8\u5B58\u53D6\u3002macOS \u6703\u5728\u4F60\u9996\u6B21\u9304\u97F3\u6642\u5F48\u51FA\u6388\u6B0A\u8ACB\u6C42\u2014\u2014\u8ACB\u540C\u610F\u5F8C\u91CD\u65B0\u958B\u59CB\u3002\u5F48\u7A97\u958B\u555F\u671F\u9593\u7684\u9996\u6B21\u9304\u97F3\u53EF\u80FD\u70BA\u7A7A\u3002",
  "Voice: recording": "\u8A9E\u97F3\uFF1A\u9304\u97F3\u4E2D",
  "Voice: transcribing": "\u8A9E\u97F3\uFF1A\u8F49\u5BEB\u4E2D",
  "Voice: refining": "\u8A9E\u97F3\uFF1A\u512A\u5316\u4E2D",
  "listening\u2026": "\u8046\u807D\u4E2D\u2026",
  "transcribing\u2026": "\u8F49\u5BEB\u4E2D\u2026",
  "refining\u2026": "\u512A\u5316\u4E2D\u2026",
  "Content generator configuration not available.": "\u5167\u5BB9\u751F\u6210\u5668\u914D\u7F6E\u4E0D\u53EF\u7528",
  "Authentication type not available.": "\u8A8D\u8B49\u985E\u578B\u4E0D\u53EF\u7528",
  "No models available for the current authentication type ({{authType}}).": "\u7576\u524D\u8A8D\u8B49\u985E\u578B ({{authType}}) \u6C92\u6709\u53EF\u7528\u7684\u6A21\u578B",
  // Needs translation
  "Starting a new session, resetting chat, and clearing terminal.": "\u6B63\u5728\u958B\u59CB\u65B0\u6703\u8A71\uFF0C\u91CD\u7F6E\u804A\u5929\u4E26\u6E05\u5C4F\u3002",
  "Starting a new session and clearing.": "\u6B63\u5728\u958B\u59CB\u65B0\u6703\u8A71\u4E26\u6E05\u5C4F\u3002",
  "Already compressing, wait for previous request to complete": "\u6B63\u5728\u58D3\u7E2E\u4E2D\uFF0C\u8ACB\u7B49\u5F85\u4E0A\u4E00\u500B\u8ACB\u6C42\u5B8C\u6210",
  "Failed to compress chat history.": "\u58D3\u7E2E\u804A\u5929\u6B77\u53F2\u5931\u6557",
  "Failed to compress chat history: {{error}}": "\u58D3\u7E2E\u804A\u5929\u6B77\u53F2\u5931\u6557\uFF1A{{error}}",
  "Compressing chat history": "\u6B63\u5728\u58D3\u7E2E\u804A\u5929\u6B77\u53F2",
  "Chat history compressed from {{originalTokens}} to {{newTokens}} tokens.": "\u804A\u5929\u6B77\u53F2\u5DF2\u5F9E {{originalTokens}} \u500B token \u58D3\u7E2E\u5230 {{newTokens}} \u500B token\u3002",
  "Compression was not beneficial for this history size.": "\u5C0D\u65BC\u6B64\u6B77\u53F2\u8A18\u9304\u5927\u5C0F\uFF0C\u58D3\u7E2E\u6C92\u6709\u76CA\u8655\u3002",
  "Chat history compression did not reduce size. This may indicate issues with the compression prompt.": "\u804A\u5929\u6B77\u53F2\u58D3\u7E2E\u672A\u80FD\u6E1B\u5C0F\u5927\u5C0F\u3002\u9019\u53EF\u80FD\u8868\u660E\u58D3\u7E2E\u63D0\u793A\u5B58\u5728\u554F\u984C\u3002",
  "Could not compress chat history due to a token counting error.": "\u7531\u65BC token \u8A08\u6578\u932F\u8AA4\uFF0C\u7121\u6CD5\u58D3\u7E2E\u804A\u5929\u6B77\u53F2\u3002",
  "Configuration is not available.": "\u914D\u7F6E\u4E0D\u53EF\u7528\u3002",
  "Please provide at least one path to add.": "\u8ACB\u63D0\u4F9B\u81F3\u5C11\u4E00\u500B\u8981\u6DFB\u52A0\u7684\u8DEF\u5F91\u3002",
  "The /directory add command is not supported in restrictive sandbox profiles. Please use --include-directories when starting the session instead.": "/directory add \u547D\u4EE4\u5728\u9650\u5236\u6027\u6C99\u7BB1\u914D\u7F6E\u6A94\u6848\u4E2D\u4E0D\u53D7\u652F\u6301\u3002\u8ACB\u6539\u70BA\u5728\u555F\u52D5\u6703\u8A71\u6642\u4F7F\u7528 --include-directories\u3002",
  "Error adding '{{path}}': {{error}}": "\u6DFB\u52A0 '{{path}}' \u6642\u51FA\u932F\uFF1A{{error}}",
  "Successfully added QWEN.md files from the following directories if there are:\n- {{directories}}": "\u5982\u679C\u5B58\u5728\uFF0C\u5DF2\u6210\u529F\u5F9E\u4EE5\u4E0B\u76EE\u9304\u6DFB\u52A0 QWEN.md \u6A94\u6848\uFF1A\n- {{directories}}",
  "Error refreshing memory: {{error}}": "\u5237\u65B0\u5167\u5B58\u6642\u51FA\u932F\uFF1A{{error}}",
  "Successfully added directories:\n- {{directories}}": "\u6210\u529F\u6DFB\u52A0\u76EE\u9304\uFF1A\n- {{directories}}",
  "Current workspace directories:\n{{directories}}": "\u7576\u524D\u5DE5\u4F5C\u5340\u76EE\u9304\uFF1A\n{{directories}}",
  "Please open the following URL in your browser to view the documentation:\n{{url}}": "\u8ACB\u5728\u700F\u89BD\u5668\u4E2D\u6253\u958B\u4EE5\u4E0B URL \u4EE5\u67E5\u770B\u6587\u6A94\uFF1A\n{{url}}",
  "Opening documentation in your browser: {{url}}": "\u6B63\u5728\u700F\u89BD\u5668\u4E2D\u6253\u958B\u6587\u6A94\uFF1A{{url}}",
  "Do you want to proceed?": "\u662F\u5426\u7E7C\u7E8C\uFF1F",
  "Yes, allow once": "\u662F\uFF0C\u5141\u8A31\u4E00\u6B21",
  "Allow always": "\u7E3D\u662F\u5141\u8A31",
  Yes: "\u662F",
  No: "\u5426",
  "No (esc)": "\u5426 (esc)",
  "Modify in progress:": "\u6B63\u5728\u4FEE\u6539\uFF1A",
  "Save and close external editor to continue": "\u4FDD\u5B58\u4E26\u95DC\u9589\u5916\u90E8\u7DE8\u8F2F\u5668\u4EE5\u7E7C\u7E8C",
  "Apply this change?": "\u662F\u5426\u61C9\u7528\u6B64\u66F4\u6539\uFF1F",
  "Yes, allow always": "\u662F\uFF0C\u7E3D\u662F\u5141\u8A31",
  "Modify with external editor": "\u4F7F\u7528\u5916\u90E8\u7DE8\u8F2F\u5668\u4FEE\u6539",
  "No, suggest changes (esc)": "\u5426\uFF0C\u5EFA\u8B70\u66F4\u6539 (esc)",
  "Allow execution of: '{{command}}'?": "\u5141\u8A31\u57F7\u884C\uFF1A'{{command}}'\uFF1F",
  "Always allow in this project": "\u5728\u672C\u9805\u76EE\u4E2D\u7E3D\u662F\u5141\u8A31",
  "Always allow {{action}} in this project": "\u5728\u672C\u9805\u76EE\u4E2D\u7E3D\u662F\u5141\u8A31{{action}}",
  "Always allow for this user": "\u5C0D\u8A72\u7528\u6236\u7E3D\u662F\u5141\u8A31",
  "Always allow {{action}} for this user": "\u5C0D\u8A72\u7528\u6236\u7E3D\u662F\u5141\u8A31{{action}}",
  "Yes, restore previous mode ({{mode}})": "\u662F\uFF0C\u6062\u5FA9\u4E4B\u524D\u7684\u6A21\u5F0F ({{mode}})",
  "Yes, and auto-accept edits": "\u662F\uFF0C\u4E26\u81EA\u52D5\u63A5\u53D7\u7DE8\u8F2F",
  "Yes, and manually approve edits": "\u662F\uFF0C\u4E26\u624B\u52D5\u6279\u51C6\u7DE8\u8F2F",
  "No, keep planning (esc)": "\u5426\uFF0C\u7E7C\u7E8C\u898F\u5283 (esc)",
  "URLs to fetch:": "\u8981\u7372\u53D6\u7684 URL\uFF1A",
  "MCP Server: {{server}}": "MCP Server\uFF1A{{server}}",
  "Tool: {{tool}}": "\u5DE5\u5177\uFF1A{{tool}}",
  'Allow execution of MCP tool "{{tool}}" from server "{{server}}"?': '\u5141\u8A31\u57F7\u884C\u4F86\u81EA MCP server "{{server}}" \u7684 MCP tool "{{tool}}"\uFF1F',
  "Shell Command Execution": "Shell \u547D\u4EE4\u57F7\u884C",
  "A custom command wants to run the following shell commands:": "\u81EA\u5B9A\u7FA9\u547D\u4EE4\u60F3\u8981\u904B\u884C\u4EE5\u4E0B shell \u547D\u4EE4\uFF1A",
  "Current Plan:": "\u7576\u524D\u8A08\u5283\uFF1A",
  "Progress: {{done}}/{{total}} tasks completed": "\u9032\u5EA6\uFF1A\u5DF2\u5B8C\u6210 {{done}}/{{total}} \u500B\u4EFB\u52D9",
  ", {{inProgress}} in progress": "\uFF0C{{inProgress}} \u500B\u9032\u884C\u4E2D",
  "Pending Tasks:": "\u5F85\u8655\u7406\u4EFB\u52D9\uFF1A",
  "What would you like to do?": "\u60A8\u60F3\u8981\u505A\u4EC0\u9EBC\uFF1F",
  "Choose how to proceed with your session:": "\u9078\u64C7\u5982\u4F55\u7E7C\u7E8C\u60A8\u7684\u6703\u8A71\uFF1A",
  "Start new chat session": "\u958B\u59CB\u65B0\u7684\u804A\u5929\u6703\u8A71",
  "Continue previous conversation": "\u7E7C\u7E8C\u4E4B\u524D\u7684\u5C0D\u8A71",
  "\u{1F44B} Welcome back! (Last updated: {{timeAgo}})": "\u{1F44B} \u6B61\u8FCE\u56DE\u4F86\uFF01\uFF08\u6700\u5F8C\u66F4\u65B0\uFF1A{{timeAgo}}\uFF09",
  "\u{1F3AF} Overall Goal:": "\u{1F3AF} \u7E3D\u9AD4\u76EE\u6A19\uFF1A",
  "Connect a Provider": "\u9023\u63A5\u670D\u52D9\u5546",
  "You must connect a provider to proceed. Press Ctrl+C again to exit.": "\u5FC5\u9808\u9023\u63A5\u4E00\u500B\u670D\u52D9\u5546\u624D\u80FD\u7E7C\u7E8C\u3002\u518D\u6B21\u6309 Ctrl+C \u9000\u51FA",
  "Terms of Services and Privacy Notice": "\u670D\u52D9\u689D\u6B3E\u548C\u96B1\u79C1\u8072\u660E",
  "Qwen OAuth": "Qwen OAuth (\u514D\u8CBB)",
  "Discontinued \u2014 switch to Coding Plan or API Key": "\u5DF2\u505C\u7528 \u2014 \u8ACB\u5207\u63DB\u5230 Coding Plan \u6216 API Key",
  "Qwen OAuth free tier was discontinued on 2026-04-15. Please select Coding Plan or API Key instead.": "Qwen OAuth \u514D\u8CBB\u984D\u5EA6\u5DF2\u65BC 2026-04-15 \u505C\u7528\u3002\u8ACB\u9078\u64C7 Coding Plan \u6216 API Key\u3002",
  "Qwen OAuth free tier was discontinued on 2026-04-15. Please select a model from another provider or run /auth to switch.": "Qwen OAuth\u514D\u8CBB\u5C64\u5DF2\u65BC2026-04-15\u505C\u6B62\u670D\u52D9\u3002\u8ACB\u9078\u64C7\u5176\u4ED6\u63D0\u4F9B\u5546\u7684\u6A21\u578B\u6216\u904B\u884C /auth \u5207\u63DB\u3002",
  "\n\u26A0 Qwen OAuth free tier was discontinued on 2026-04-15. Please select another option.\n": "\n\u26A0 Qwen OAuth \u514D\u8CBB\u984D\u5EA6\u5DF2\u65BC 2026-04-15 \u505C\u7528\u3002\u8ACB\u9078\u64C7\u5176\u4ED6\u9078\u9805\u3002\n",
  "Paid \xB7 Up to 6,000 requests/5 hrs \xB7 All Alibaba Cloud Coding Plan Models": "\u4ED8\u8CBB \xB7 \u6BCF 5 \u5C0F\u6642\u6700\u591A 6,000 \u6B21\u8ACB\u6C42 \xB7 \u652F\u6301\u963F\u91CC\u96F2\u767E\u934A Coding Plan \u5168\u90E8\u6A21\u578B",
  "For teams \xB7 Paid \xB7 Up to 6,000 requests/5 hrs \xB7 All Alibaba Cloud Coding Plan Models": "\u9069\u5408\u5718\u968A \xB7 \u4ED8\u8CBB \xB7 \u6BCF 5 \u5C0F\u6642\u6700\u591A 6,000 \u6B21\u8ACB\u6C42 \xB7 \u652F\u63F4\u963F\u91CC\u96F2\u767E\u934A Coding Plan \u5168\u90E8\u6A21\u578B",
  "For individual developers \xB7 Pay per model call \xB7 5-hour/weekly quotas": "\u9069\u5408\u500B\u4EBA\u958B\u767C\u5834\u666F \xB7 \u6309\u6A21\u578B\u8ABF\u7528\u6B21\u6578\u8A08\u8CBB \xB7 \u6BCF 5 \u5C0F\u6642/\u6BCF\u9031\u9650\u984D",
  Subscribe: "\u8A02\u95B1\u8A08\u5283",
  "Paid subscription plans from Alibaba Cloud ModelStudio": "Alibaba Cloud ModelStudio \u4ED8\u8CBB\u8A02\u95B1\u8A08\u5283",
  "Select Subscription Plan": "\u9078\u64C7\u8A02\u95B1\u8A08\u5283",
  "Alibaba Cloud Coding Plan": "\u963F\u91CC\u96F2\u767E\u934A Coding Plan",
  "Alibaba Cloud Token Plan": "\u963F\u91CC\u96F2\u767E\u934A Token Plan",
  "Pay-as-you-go tokens \xB7 Configure ModelStudio standard API key": "\u6309 Token \u4ED8\u8CBB \xB7 \u914D\u7F6E ModelStudio \u6A19\u6E96 API Key",
  "For individuals \xB7 Pay-as-you-go tokens \xB7 Dedicated Token Plan endpoint": "\u9069\u5408\u500B\u4EBA \xB7 \u6309 Token \u4ED8\u8CBB \xB7 \u4F7F\u7528\u7368\u7ACB Token Plan Endpoint",
  "For teams/companies \xB7 Credits deducted by token usage \xB7 Dedicated API key and base URL": "\u9069\u5408\u4E00\u4EBA\u516C\u53F8/\u5718\u968A/\u4F01\u696D \xB7 \u6309 Token \u6D88\u8017\u62B5\u6263 Credits \xB7 \u5C08\u5C6C API Key \u548C Base URL",
  "Token Plan documentation": "Token Plan \u53C3\u8003\u6587\u6A94",
  "Bring your own API key": "\u4F7F\u7528\u81EA\u5DF1\u7684 API Key",
  "Browser-based authentication with third-party providers (e.g. OpenRouter, ModelScope)": "\u57FA\u65BC\u700F\u89BD\u5668\u7684\u7B2C\u4E09\u65B9\u63D0\u4F9B\u5546\u8A8D\u8B49\uFF08\u4F8B\u5982 OpenRouter\u3001ModelScope\uFF09",
  "Authentication is enforced to be {{enforcedType}}, but you are currently using {{currentType}}.": "\u8A8D\u8B49\u65B9\u5F0F\u88AB\u5F37\u5236\u8A2D\u7F6E\u70BA {{enforcedType}}\uFF0C\u4F46\u60A8\u7576\u524D\u4F7F\u7528\u7684\u662F {{currentType}}",
  "Qwen OAuth Authentication": "Qwen OAuth \u8A8D\u8B49",
  "Please visit this URL to authorize:": "\u8ACB\u8A2A\u554F\u6B64 URL \u9032\u884C\u6388\u6B0A\uFF1A",
  "Waiting for authorization": "\u7B49\u5F85\u6388\u6B0A\u4E2D",
  "Time remaining:": "\u5269\u9918\u6642\u9593\uFF1A",
  "Qwen OAuth Authentication Timeout": "Qwen OAuth \u8A8D\u8B49\u8D85\u6642",
  "OAuth token expired (over {{seconds}} seconds). Please select authentication method again.": "OAuth token \u5DF2\u904E\u671F\uFF08\u8D85\u904E {{seconds}} \u79D2\uFF09\u3002\u8ACB\u91CD\u65B0\u9078\u64C7\u8A8D\u8B49\u65B9\u6CD5",
  "Press any key to return to authentication type selection.": "\u6309\u4EFB\u610F\u9375\u8FD4\u56DE\u8A8D\u8B49\u985E\u578B\u9078\u64C7",
  "Waiting for Qwen OAuth authentication...": "\u6B63\u5728\u7B49\u5F85 Qwen OAuth \u8A8D\u8B49...",
  "Authentication timed out. Please try again.": "\u8A8D\u8B49\u8D85\u6642\u3002\u8ACB\u91CD\u8A66\u3002",
  "Waiting for auth... (Press ESC or CTRL+C to cancel)": "\u6B63\u5728\u7B49\u5F85\u8A8D\u8B49...\uFF08\u6309 ESC \u6216 CTRL+C \u53D6\u6D88\uFF09",
  "Missing API key for OpenAI-compatible auth. Set settings.security.auth.apiKey, or set the {{envKeyHint}} environment variable.": "\u7F3A\u5C11 OpenAI \u517C\u5BB9\u8A8D\u8B49\u7684 API Key\u3002\u8ACB\u8A2D\u7F6E settings.security.auth.apiKey \u6216\u8A2D\u7F6E {{envKeyHint}} \u74B0\u5883\u8B8A\u91CF\u3002",
  "{{envKeyHint}} environment variable not found. Please set it in your .env file or environment variables.": "\u672A\u627E\u5230 {{envKeyHint}} \u74B0\u5883\u8B8A\u91CF\u3002\u8ACB\u5728 .env \u6A94\u6848\u6216\u7CFB\u7D71\u74B0\u5883\u8B8A\u91CF\u4E2D\u9032\u884C\u8A2D\u7F6E\u3002",
  "{{envKeyHint}} environment variable not found (or set settings.security.auth.apiKey). Please set it in your .env file or environment variables.": "\u672A\u627E\u5230 {{envKeyHint}} \u74B0\u5883\u8B8A\u91CF\uFF08\u6216\u8A2D\u7F6E settings.security.auth.apiKey\uFF09\u3002\u8ACB\u5728 .env \u6A94\u6848\u6216\u7CFB\u7D71\u74B0\u5883\u8B8A\u91CF\u4E2D\u9032\u884C\u8A2D\u7F6E\u3002",
  "Missing API key for OpenAI-compatible auth. Set the {{envKeyHint}} environment variable.": "\u7F3A\u5C11 OpenAI \u517C\u5BB9\u8A8D\u8B49\u7684 API Key\u3002\u8ACB\u8A2D\u7F6E {{envKeyHint}} \u74B0\u5883\u8B8A\u91CF\u3002",
  "Anthropic provider missing required baseUrl in modelProviders[].baseUrl.": "Anthropic \u63D0\u4F9B\u5546\u7F3A\u5C11\u5FC5\u9700\u7684 baseUrl\uFF0C\u8ACB\u5728 modelProviders[].baseUrl \u4E2D\u914D\u7F6E\u3002",
  "ANTHROPIC_BASE_URL environment variable not found.": "\u672A\u627E\u5230 ANTHROPIC_BASE_URL \u74B0\u5883\u8B8A\u91CF\u3002",
  "Invalid auth method selected.": "\u9078\u64C7\u4E86\u7121\u6548\u7684\u8A8D\u8B49\u65B9\u5F0F\u3002",
  "Failed to authenticate. Message: {{message}}": "\u8A8D\u8B49\u5931\u6557\u3002\u6D88\u606F\uFF1A{{message}}",
  "Authenticated successfully with {{authType}} credentials.": "\u4F7F\u7528 {{authType}} \u6191\u64DA\u6210\u529F\u8A8D\u8B49\u3002",
  'Invalid QWEN_DEFAULT_AUTH_TYPE value: "{{value}}". Valid values are: {{validValues}}': '\u7121\u6548\u7684 QWEN_DEFAULT_AUTH_TYPE \u503C\uFF1A"{{value}}"\u3002\u6709\u6548\u503C\u70BA\uFF1A{{validValues}}',
  "Select Model": "\u9078\u64C7\u6A21\u578B",
  "API Key": "API Key",
  "(default)": "(\u9ED8\u8A8D)",
  "(not set)": "(\u672A\u8A2D\u7F6E)",
  Modality: "\u6A21\u614B",
  "Context Window": "\u4E0A\u4E0B\u6587\u7A97\u53E3",
  text: "\u6587\u672C",
  "text-only": "\u7D14\u6587\u672C",
  image: "\u5716\u50CF",
  pdf: "PDF",
  audio: "\u97F3\u983B",
  video: "\u8996\u983B",
  "not set": "\u672A\u8A2D\u7F6E",
  'Current voice model: {{voiceModel}}\nUse "/model --voice <model-id>" to set voice model.': '\u7576\u524D\u8A9E\u97F3\u6A21\u578B\uFF1A{{voiceModel}}\n\u4F7F\u7528 "/model --voice <model-id>" \u8A2D\u7F6E\u8A9E\u97F3\u6A21\u578B\u3002',
  'Current vision model: {{visionModel}}\nUse "/model --vision <model-id>" to set the vision bridge model.': '\u7576\u524D\u8996\u89BA\u6A21\u578B\uFF1A{{visionModel}}\n\u4F7F\u7528 "/model --vision <model-id>" \u8A2D\u7F6E\u8996\u89BA\u6A4B\u63A5\u6A21\u578B\u3002',
  "Voice model '{{modelName}}' is ambiguous. Configure a unique model id before using /model --voice.": "\u8A9E\u97F3\u6A21\u578B '{{modelName}}' \u4E0D\u552F\u4E00\u3002\u8ACB\u5148\u914D\u7F6E\u552F\u4E00\u7684\u6A21\u578B ID\uFF0C\u518D\u4F7F\u7528 /model --voice\u3002",
  none: "\u7121",
  unknown: "\u672A\u77E5",
  "Manage folder trust settings": "\u7BA1\u7406\u6A94\u6848\u593E\u4FE1\u4EFB\u8A2D\u7F6E",
  "Manage permission rules": "\u7BA1\u7406 permission rules",
  Allow: "\u5141\u8A31",
  Ask: "\u8A62\u554F",
  Deny: "\u62D2\u7D55",
  Workspace: "\u5DE5\u4F5C\u5340",
  "Qwen Code won't ask before using allowed tools.": "Qwen Code \u4F7F\u7528\u5DF2\u5141\u8A31\u7684\u5DE5\u5177\u524D\u4E0D\u6703\u8A62\u554F\u3002",
  "Qwen Code will ask before using these tools.": "Qwen Code \u4F7F\u7528\u9019\u4E9B\u5DE5\u5177\u524D\u6703\u5148\u8A62\u554F\u3002",
  "Qwen Code is not allowed to use denied tools.": "Qwen Code \u4E0D\u5141\u8A31\u4F7F\u7528\u88AB\u62D2\u7D55\u7684\u5DE5\u5177\u3002",
  "Manage trusted directories for this workspace.": "\u7BA1\u7406\u6B64\u5DE5\u4F5C\u5340\u7684\u53D7\u4FE1\u4EFB\u76EE\u9304\u3002",
  "Any use of the {{tool}} tool": "{{tool}} \u5DE5\u5177\u7684\u4EFB\u4F55\u4F7F\u7528",
  "{{tool}} commands matching '{{pattern}}'": "\u5339\u914D '{{pattern}}' \u7684 {{tool}} \u547D\u4EE4",
  "From user settings": "\u4F86\u81EA\u7528\u6236\u8A2D\u7F6E",
  "From project settings": "\u4F86\u81EA\u9805\u76EE\u8A2D\u7F6E",
  "From session": "\u4F86\u81EA\u6703\u8A71",
  "Project settings": "\u9805\u76EE\u8A2D\u7F6E",
  "Checked in at .qwen/settings.json": "\u4FDD\u5B58\u5728 .qwen/settings.json",
  "User settings": "\u7528\u6236\u8A2D\u7F6E",
  "Saved in at ~/.qwen/settings.json": "\u4FDD\u5B58\u5728 ~/.qwen/settings.json",
  "Add a new rule\u2026": "\u6DFB\u52A0\u65B0\u898F\u5247\u2026",
  "Add {{type}} permission rule": "\u6DFB\u52A0 {{type}} permission rule",
  "Permission rules are a tool name, optionally followed by a specifier in parentheses.": "permission rules \u662F\u5DE5\u5177\u540D\u7A31\uFF0C\u53EF\u9078\u5730\u5F8C\u8DDF\u62EC\u865F\u4E2D\u7684\u9650\u5B9A\u7B26\u3002",
  "e.g.,": "\u4F8B\u5982",
  or: "\u6216",
  "Enter permission rule\u2026": "\u8F38\u5165 permission rule\u2026",
  "Enter to submit \xB7 Esc to cancel": "Enter \u63D0\u4EA4 \xB7 Esc \u53D6\u6D88",
  "Where should this rule be saved?": "\u6B64\u898F\u5247\u61C9\u4FDD\u5B58\u5728\u54EA\u88CF\uFF1F",
  "Enter to confirm \xB7 Esc to cancel": "Enter \u78BA\u8A8D \xB7 Esc \u53D6\u6D88",
  "Delete {{type}} rule?": "\u522A\u9664{{type}}\u898F\u5247\uFF1F",
  "Are you sure you want to delete this permission rule?": "\u78BA\u5B9A\u8981\u522A\u9664\u6B64 permission rule \u55CE\uFF1F",
  "Permissions:": "\u6B0A\u9650\uFF1A",
  "(\u2190/\u2192 or tab to cycle)": "\uFF08\u2190/\u2192 \u6216 Tab \u5207\u63DB\uFF09",
  "Press \u2191\u2193 to navigate \xB7 Enter to select \xB7 Type to search \xB7 Esc to cancel": "\u6309 \u2191\u2193 \u5C0E\u822A \xB7 Enter \u9078\u64C7 \xB7 \u8F38\u5165\u641C\u7D22 \xB7 Esc \u53D6\u6D88",
  "Search\u2026": "\u641C\u7D22\u2026",
  "Add directory\u2026": "\u6DFB\u52A0\u76EE\u9304\u2026",
  "Add directory to workspace": "\u6DFB\u52A0\u5DE5\u4F5C\u5340\u76EE\u9304",
  "Qwen Code can read files in the workspace, and make edits when auto-accept edits is on.": "Qwen Code \u53EF\u4EE5\u8B80\u53D6\u5DE5\u4F5C\u5340\u4E2D\u7684\u6A94\u6848\uFF0C\u4E26\u5728\u81EA\u52D5\u63A5\u53D7\u7DE8\u8F2F\u6A21\u5F0F\u958B\u555F\u6642\u9032\u884C\u7DE8\u8F2F\u3002",
  "Qwen Code will be able to read files in this directory and make edits when auto-accept edits is on.": "Qwen Code \u5C07\u80FD\u5920\u8B80\u53D6\u6B64\u76EE\u9304\u4E2D\u7684\u6A94\u6848\uFF0C\u4E26\u5728\u81EA\u52D5\u63A5\u53D7\u7DE8\u8F2F\u6A21\u5F0F\u958B\u555F\u6642\u9032\u884C\u7DE8\u8F2F\u3002",
  "Enter the path to the directory:": "\u8F38\u5165\u76EE\u9304\u8DEF\u5F91\uFF1A",
  "Enter directory path\u2026": "\u8F38\u5165\u76EE\u9304\u8DEF\u5F91\u2026",
  "Tab to complete \xB7 Enter to add \xB7 Esc to cancel": "Tab \u88DC\u5168 \xB7 Enter \u6DFB\u52A0 \xB7 Esc \u53D6\u6D88",
  "Remove directory?": "\u522A\u9664\u76EE\u9304\uFF1F",
  "Are you sure you want to remove this directory from the workspace?": "\u78BA\u5B9A\u8981\u5C07\u6B64\u76EE\u9304\u5F9E\u5DE5\u4F5C\u5340\u4E2D\u79FB\u9664\u55CE\uFF1F",
  "  (Original working directory)": "  \uFF08\u539F\u59CB\u5DE5\u4F5C\u76EE\u9304\uFF09",
  "  (from settings)": "  \uFF08\u4F86\u81EA\u8A2D\u7F6E\uFF09",
  "Directory does not exist.": "\u76EE\u9304\u4E0D\u5B58\u5728\u3002",
  "Path is not a directory.": "\u8DEF\u5F91\u4E0D\u662F\u76EE\u9304\u3002",
  "This directory is already in the workspace.": "\u6B64\u76EE\u9304\u5DF2\u5728\u5DE5\u4F5C\u5340\u4E2D\u3002",
  "Already covered by existing directory: {{dir}}": "\u5DF2\u88AB\u73FE\u6709\u76EE\u9304\u8986\u84CB\uFF1A{{dir}}",
  "Using:": "\u5DF2\u52A0\u8F09: ",
  "{{count}} open file": "{{count}} \u500B\u6253\u958B\u7684\u6A94\u6848",
  "{{count}} open files": "{{count}} \u500B\u6253\u958B\u7684\u6A94\u6848",
  "(ctrl+g to view)": "\uFF08\u6309 ctrl+g \u67E5\u770B\uFF09",
  "{{count}} {{name}} file": "{{count}} \u500B {{name}} \u6A94\u6848",
  "{{count}} {{name}} files": "{{count}} \u500B {{name}} \u6A94\u6848",
  "{{count}} MCP server": "{{count}} \u500B MCP server",
  "{{count}} MCP servers": "{{count}} \u500B MCP servers",
  "{{count}} Blocked": "{{count}} \u500B\u5DF2\u963B\u6B62",
  "(ctrl+t to view)": "\uFF08\u6309 ctrl+t \u67E5\u770B\uFF09",
  "(ctrl+t to toggle)": "\uFF08\u6309 ctrl+t \u5207\u63DB\uFF09",
  "Press Ctrl+C again to exit.": "\u518D\u6B21\u6309 Ctrl+C \u9000\u51FA",
  "Press Ctrl+D again to exit.": "\u518D\u6B21\u6309 Ctrl+D \u9000\u51FA",
  "Press Esc again to clear.": "\u518D\u6B21\u6309 Esc \u6E05\u9664",
  "Press \u2191 to edit queued messages": "\u6309 \u2191 \u7DE8\u8F2F\u6392\u968A\u6D88\u606F",
  "No MCP servers configured.": "\u672A\u914D\u7F6E MCP servers",
  "\u23F3 MCP servers are starting up ({{count}} initializing)...": "\u23F3 MCP servers \u6B63\u5728\u555F\u52D5\uFF08{{count}} \u500B\u6B63\u5728\u521D\u59CB\u5316\uFF09...",
  "Note: First startup may take longer. Tool availability will update automatically.": "\u6CE8\u610F\uFF1A\u9996\u6B21\u555F\u52D5\u53EF\u80FD\u9700\u8981\u66F4\u9577\u6642\u9593\u3002\u5DE5\u5177\u53EF\u7528\u6027\u5C07\u81EA\u52D5\u66F4\u65B0",
  "Configured MCP servers:": "\u5DF2\u914D\u7F6E\u7684 MCP servers\uFF1A",
  Ready: "\u5C31\u7DD2",
  "Starting... (first startup may take longer)": "\u6B63\u5728\u555F\u52D5...\uFF08\u9996\u6B21\u555F\u52D5\u53EF\u80FD\u9700\u8981\u66F4\u9577\u6642\u9593\uFF09",
  Disconnected: "\u5DF2\u65B7\u958B\u9023\u63A5",
  "{{count}} tool": "{{count}} \u500B\u5DE5\u5177",
  "{{count}} tools": "{{count}} \u500B\u5DE5\u5177",
  "{{count}} prompt": "{{count}} \u500B\u63D0\u793A",
  "{{count}} prompts": "{{count}} \u500B\u63D0\u793A",
  "(from {{extensionName}})": "\uFF08\u4F86\u81EA {{extensionName}}\uFF09",
  OAuth: "OAuth",
  "OAuth expired": "OAuth \u5DF2\u904E\u671F",
  "OAuth not authenticated": "OAuth \u672A\u8A8D\u8B49",
  "tools and prompts will appear when ready": "\u5DE5\u5177\u548C\u63D0\u793A\u5C07\u5728\u5C31\u7DD2\u6642\u986F\u793A",
  "{{count}} tools cached": "{{count}} \u500B\u5DE5\u5177\u5DF2\u7DE9\u5B58",
  "Tools:": "\u5DE5\u5177\uFF1A",
  "Parameters:": "\u53C3\u6578\uFF1A",
  "Prompts:": "\u63D0\u793A\uFF1A",
  "Resources:": "\u8CC7\u6E90\uFF1A",
  Blocked: "\u5DF2\u963B\u6B62",
  "\u{1F4A1} Tips:": "\u{1F4A1} \u63D0\u793A\uFF1A",
  Use: "\u4F7F\u7528",
  "to show server and tool descriptions": "\u986F\u793A\u4F3A\u670D\u5668\u548C\u5DE5\u5177\u63CF\u8FF0",
  "to show tool parameter schemas": "\u986F\u793A tool parameter schemas",
  "to hide descriptions": "\u96B1\u85CF\u63CF\u8FF0",
  "to authenticate with OAuth-enabled servers": "\u4F7F\u7528\u652F\u6301 OAuth \u7684\u4F3A\u670D\u5668\u9032\u884C\u8A8D\u8B49",
  Press: "\u6309",
  "to toggle tool descriptions on/off": "\u5207\u63DB\u5DE5\u5177\u63CF\u8FF0\u958B\u95DC",
  "Starting OAuth authentication for MCP server '{{name}}'...": "\u6B63\u5728\u70BA MCP server '{{name}}' \u555F\u52D5 OAuth \u8A8D\u8B49...",
  "Tips:": "\u63D0\u793A\uFF1A",
  "Use /compress when the conversation gets long to summarize history and free up context.": "\u5C0D\u8A71\u8B8A\u9577\u6642\u7528 /compress\uFF0C\u7E3D\u7D50\u6B77\u53F2\u4E26\u91CB\u653E\u4E0A\u4E0B\u6587\u3002",
  "Start a fresh idea with /clear or /new; the previous session stays available in history.": "\u7528 /clear \u6216 /new \u958B\u555F\u65B0\u601D\u8DEF\uFF1B\u4E4B\u524D\u7684\u6703\u8A71\u6703\u4FDD\u7559\u5728\u6B77\u53F2\u8A18\u9304\u4E2D\u3002",
  "Use /bug to submit issues to the maintainers when something goes off.": "\u9047\u5230\u554F\u984C\u6642\uFF0C\u7528 /bug \u5C07\u554F\u984C\u63D0\u4EA4\u7D66\u7DAD\u8B77\u8005\u3002",
  "Switch auth type quickly with /auth.": "\u7528 /auth \u5FEB\u901F\u5207\u63DB\u8A8D\u8B49\u65B9\u5F0F\u3002",
  "You can run any shell commands from Qwen Code using ! (e.g. !ls).": "\u5728 Qwen Code \u4E2D\u4F7F\u7528 ! \u53EF\u904B\u884C\u4EFB\u610F shell \u547D\u4EE4\uFF08\u4F8B\u5982 !ls\uFF09\u3002",
  "Type / to open the command popup; Tab autocompletes slash commands and saved prompts.": "\u8F38\u5165 / \u6253\u958B\u547D\u4EE4\u5F48\u7A97\uFF1B\u6309 Tab \u81EA\u52D5\u88DC\u5168\u659C\u69D3\u547D\u4EE4\u548C\u4FDD\u5B58\u7684\u63D0\u793A\u8A5E\u3002",
  "You can resume a previous conversation by running qwen --continue or qwen --resume.": "\u904B\u884C qwen --continue \u6216 qwen --resume \u53EF\u7E7C\u7E8C\u4E4B\u524D\u7684\u6703\u8A71\u3002",
  "You can switch permission mode quickly with Shift+Tab or /approval-mode.": "\u6309 Shift+Tab \u6216\u8F38\u5165 /approval-mode \u53EF\u5FEB\u901F\u5207\u63DB\u6B0A\u9650\u6A21\u5F0F\u3002",
  "You can switch permission mode quickly with Tab or /approval-mode.": "\u6309 Tab \u6216\u8F38\u5165 /approval-mode \u53EF\u5FEB\u901F\u5207\u63DB\u6B0A\u9650\u6A21\u5F0F\u3002",
  "Try /insight to generate personalized insights from your chat history.": "\u8A66\u8A66 /insight\uFF0C\u5F9E\u804A\u5929\u8A18\u9304\u4E2D\u751F\u6210\u500B\u6027\u5316\u6D1E\u5BDF\u3002",
  "Press Ctrl+O to toggle compact mode \u2014 hide tool output and thinking for a cleaner view.": "\u6309 Ctrl+O \u5207\u63DB\u7DCA\u6E4A\u6A21\u5F0F \u2500\u2500 \u96B1\u85CF\u5DE5\u5177\u8F38\u51FA\u548C\u601D\u8003\u904E\u7A0B\uFF0C\u754C\u9762\u66F4\u7C21\u6F54\u3002",
  "Add a QWEN.md file to give Qwen Code persistent project context.": "\u6DFB\u52A0 QWEN.md \u6A94\u6848\uFF0C\u70BA Qwen Code \u63D0\u4F9B\u6301\u4E45\u7684\u9805\u76EE\u4E0A\u4E0B\u6587\u3002",
  "Use /btw to ask a quick side question without disrupting the conversation.": "\u7528 /btw \u5FEB\u901F\u554F\u4E00\u500B\u5C0F\u554F\u984C\uFF0C\u4E0D\u6703\u6253\u65B7\u7576\u524D\u5C0D\u8A71\u3002",
  "Context is almost full! Run /compress now or start /new to continue.": "\u4E0A\u4E0B\u6587\u5373\u5C07\u7528\u6EFF\uFF01\u8ACB\u7ACB\u5373\u57F7\u884C /compress \u6216\u4F7F\u7528 /new \u958B\u555F\u65B0\u6703\u8A71\u3002",
  "Context is getting full. Use /compress to free up space.": "\u4E0A\u4E0B\u6587\u7A7A\u9593\u4E0D\u8DB3\uFF0C\u7528 /compress \u91CB\u653E\u7A7A\u9593\u3002",
  "Long conversation? /compress summarizes history to free context.": "\u5C0D\u8A71\u592A\u9577\uFF1F\u7528 /compress \u7E3D\u7D50\u6B77\u53F2\uFF0C\u91CB\u653E\u4E0A\u4E0B\u6587\u3002",
  "Agent powering down. Goodbye!": "Qwen Code \u6B63\u5728\u95DC\u9589\uFF0C\u518D\u898B\uFF01",
  "To continue this session, run": "\u8981\u7E7C\u7E8C\u6B64\u6703\u8A71\uFF0C\u8ACB\u904B\u884C",
  "Interaction Summary": "\u4EA4\u4E92\u6458\u8981",
  "Session ID:": "\u6703\u8A71 ID\uFF1A",
  "Tool Calls:": "\u5DE5\u5177\u8ABF\u7528\uFF1A",
  "Success Rate:": "\u6210\u529F\u7387\uFF1A",
  "User Agreement:": "\u7528\u6236\u540C\u610F\u7387\uFF1A",
  reviewed: "\u5DF2\u5BE9\u8988",
  "Code Changes:": "\u4EE3\u78BC\u8B8A\u66F4\uFF1A",
  Performance: "\u6027\u80FD",
  "Wall Time:": "\u7E3D\u8017\u6642\uFF1A",
  "Agent Active:": "\u667A\u80FD\u9AD4\u6D3B\u8E8D\u6642\u9593\uFF1A",
  "API Time:": "API \u6642\u9593\uFF1A",
  "Tool Time:": "\u5DE5\u5177\u6642\u9593\uFF1A",
  "Session Stats": "\u6703\u8A71\u7D71\u8A08",
  "Model Usage": "\u6A21\u578B\u4F7F\u7528\u60C5\u6CC1",
  "Input Tokens": "\u8F38\u5165 token \u6578",
  "Output Tokens": "\u8F38\u51FA token \u6578",
  "Savings Highlight:": "\u7BC0\u7701\u4EAE\u9EDE\uFF1A",
  "of input tokens were served from the cache, reducing costs.": "\u5F9E\u7DE9\u5B58\u8F09\u5165 token \uFF0C\u964D\u4F4E\u4E86\u6210\u672C",
  "Tip: For a full token breakdown, run `/stats model`.": "\u63D0\u793A\uFF1A\u8981\u67E5\u770B\u5B8C\u6574\u7684 token \u660E\u7D30\uFF0C\u8ACB\u904B\u884C `/stats model`",
  "Model Stats For Nerds": "\u6A21\u578B\u7D71\u8A08\uFF08\u6280\u8853\u7D30\u7BC0\uFF09",
  "Tool Stats For Nerds": "\u5DE5\u5177\u7D71\u8A08\uFF08\u6280\u8853\u7D30\u7BC0\uFF09",
  Metric: "\u6307\u6A19",
  API: "API",
  Session: "\u6703\u8A71",
  Activity: "\u6982\u89BD",
  Efficiency: "\u6027\u80FD",
  Success: "\u6210\u529F\u7387",
  Today: "\u4ECA\u5929",
  "Token Trend": "Token \u8DA8\u52E2",
  "Cache Hit Rate": "\u7DE9\u5B58\u547D\u4E2D\u7387",
  "Tool Success": "\u5DE5\u5177\u6210\u529F\u7387",
  "Tool Leaderboard": "\u5DE5\u5177\u6392\u884C",
  Calls: "\u8ABF\u7528\u6B21\u6578",
  Time: "\u8017\u6642",
  Reqs: "\u8ACB\u6C42",
  Cache: "\u7DE9\u5B58",
  Latency: "\u5EF6\u9072",
  "In/Out": "\u8F38\u5165/\u8F38\u51FA",
  "Code Impact": "\u4EE3\u78BC\u8B8A\u66F4",
  "Failed to load stats. Press r to retry.": "\u8F09\u5165\u7D71\u8A08\u5931\u6557\uFF0C\u6309 r \u91CD\u8A66\u3002",
  net: "\u6DE8\u589E",
  streak: "\u9023\u7E8C",
  best: "\u6700\u9577",
  Requests: "\u8ACB\u6C42\u6578",
  Errors: "\u932F\u8AA4\u6578",
  "Avg Latency": "\u5E73\u5747\u5EF6\u9072",
  Tokens: "Token",
  Total: "\u7E3D\u8A08",
  Prompt: "\u63D0\u793A",
  Cached: "\u7DE9\u5B58",
  Thoughts: "\u601D\u8003",
  Output: "\u8F38\u51FA",
  "No API calls have been made in this session.": "\u672C\u6B21\u6703\u8A71\u4E2D\u672A\u9032\u884C\u4EFB\u4F55 API \u8ABF\u7528",
  "Tool Name": "\u5DE5\u5177\u540D\u7A31",
  "Success Rate": "\u6210\u529F\u7387",
  "Avg Duration": "\u5E73\u5747\u8017\u6642",
  "User Decision Summary": "\u7528\u6236\u6C7A\u7B56\u6458\u8981",
  "Total Reviewed Suggestions:": "\u5DF2\u5BE9\u8988\u5EFA\u8B70\u7E3D\u6578\uFF1A",
  " \xBB Accepted:": " \xBB \u5DF2\u63A5\u53D7\uFF1A",
  " \xBB Rejected:": " \xBB \u5DF2\u62D2\u7D55\uFF1A",
  " \xBB Modified:": " \xBB \u5DF2\u4FEE\u6539\uFF1A",
  " Overall Agreement Rate:": " \u7E3D\u9AD4\u540C\u610F\u7387\uFF1A",
  "No tool calls have been made in this session.": "\u672C\u6B21\u6703\u8A71\u4E2D\u672A\u9032\u884C\u4EFB\u4F55\u5DE5\u5177\u8ABF\u7528",
  "Session start time is unavailable, cannot calculate stats.": "\u6703\u8A71\u958B\u59CB\u6642\u9593\u4E0D\u53EF\u7528\uFF0C\u7121\u6CD5\u8A08\u7B97\u7D71\u8A08\u4FE1\u606F",
  "Command Format Migration": "\u547D\u4EE4\u683C\u5F0F\u9077\u79FB",
  "Found {{count}} TOML command file:": "\u767C\u73FE {{count}} \u500B TOML \u547D\u4EE4\u6A94\u6848\uFF1A",
  "Found {{count}} TOML command files:": "\u767C\u73FE {{count}} \u500B TOML \u547D\u4EE4\u6A94\u6848\uFF1A",
  "Current tasks": "\u76EE\u524D\u4EFB\u52D9",
  "... and {{count}} more": "... \u4EE5\u53CA\u5176\u4ED6 {{count}} \u500B",
  "The TOML format is deprecated. Would you like to migrate them to Markdown format?": "TOML \u683C\u5F0F\u5DF2\u68C4\u7528\u3002\u662F\u5426\u5C07\u5B83\u5011\u9077\u79FB\u5230 Markdown \u683C\u5F0F\uFF1F",
  "(Backups will be created and original files will be preserved)": "\uFF08\u5C07\u5275\u5EFA\u5099\u4EFD\uFF0C\u539F\u59CB\u6A94\u6848\u5C07\u4FDD\u7559\uFF09",
  "Waiting for user confirmation...": "\u7B49\u5F85\u7528\u6236\u78BA\u8A8D...",
  WITTY_LOADING_PHRASES: [
    "\u6B63\u5728\u52AA\u529B\u642C\u78DA\uFF0C\u8ACB\u7A0D\u5019...",
    "\u8001\u95C6\u5728\u8EAB\u5F8C\uFF0C\u5FEB\u52A0\u8F09\u554A\uFF01",
    "\u982D\u9AEE\u6389\u5149\u524D\uFF0C\u4E00\u5B9A\u80FD\u52A0\u8F09\u5B8C...",
    "\u4F3A\u670D\u5668\u6B63\u5728\u6DF1\u547C\u5438\uFF0C\u6E96\u5099\u653E\u5927\u62DB...",
    "\u6B63\u5728\u5411\u4F3A\u670D\u5668\u6295\u9935\u5496\u5561...",
    "\u6B63\u5728\u8CE6\u80FD\u5168\u93C8\u8DEF\uFF0C\u5C0B\u627E\u95DC\u9375\u6293\u624B...",
    "\u6B63\u5728\u964D\u672C\u589E\u6548\uFF0C\u512A\u5316\u52A0\u8F09\u8DEF\u5F91...",
    "\u6B63\u5728\u6253\u7834\u90E8\u9580\u58C1\u58D8\uFF0C\u6C89\u6FB1\u65B9\u6CD5\u8AD6...",
    "\u6B63\u5728\u64C1\u62B1\u8B8A\u5316\uFF0C\u8FED\u4EE3\u6838\u5FC3\u50F9\u503C...",
    "\u6B63\u5728\u5C0D\u9F4A\u9846\u7C92\u5EA6\uFF0C\u6253\u78E8\u5E95\u5C64\u908F\u8F2F...",
    "\u5927\u529B\u51FA\u5947\u8DE1\uFF0C\u6B63\u5728\u5F37\u884C\u52A0\u8F09...",
    "\u53EA\u8981\u6211\u4E0D\u5BEB\u4EE3\u78BC\uFF0C\u4EE3\u78BC\u5C31\u6C92\u6709 Bug...",
    "\u6B63\u5728\u628A Bug \u8F49\u5316\u70BA Feature...",
    "\u53EA\u8981\u6211\u4E0D\u5C37\u5C2C\uFF0CBug \u5C31\u8FFD\u4E0D\u4E0A\u6211...",
    "\u6B63\u5728\u8A66\u5716\u7406\u89E3\u53BB\u5E74\u7684\u81EA\u5DF1\u5BEB\u4E86\u4EC0\u9EBC...",
    "\u6B63\u5728\u733F\u529B\u89BA\u9192\u4E2D\uFF0C\u8ACB\u8010\u5FC3\u7B49\u5F85...",
    "\u6B63\u5728\u8A62\u554F\u7522\u54C1\u7D93\u7406\uFF1A\u9019\u9700\u6C42\u662F\u771F\u7684\u55CE\uFF1F",
    "\u6B63\u5728\u7D66\u7522\u54C1\u7D93\u7406\u756B\u9905\uFF0C\u8ACB\u7A0D\u7B49...",
    "\u6BCF\u4E00\u884C\u4EE3\u78BC\uFF0C\u90FD\u5728\u52AA\u529B\u8B93\u4E16\u754C\u8B8A\u5F97\u66F4\u597D\u4E00\u9EDE\u9EDE...",
    "\u6BCF\u4E00\u500B\u5049\u5927\u7684\u60F3\u6CD5\uFF0C\u90FD\u503C\u5F97\u9019\u4EFD\u8010\u5FC3\u7684\u7B49\u5F85...",
    "\u5225\u6025\uFF0C\u7F8E\u597D\u7684\u4E8B\u7269\u7E3D\u662F\u9700\u8981\u4E00\u9EDE\u6642\u9593\u53BB\u919E\u91C0...",
    "\u9858\u4F60\u7684\u4EE3\u78BC\u6C38\u7121 Bug\uFF0C\u9858\u4F60\u7684\u5922\u60F3\u7D42\u5C07\u6210\u771F...",
    "\u54EA\u6015\u53EA\u6709 0.1% \u7684\u9032\u5EA6\uFF0C\u4E5F\u662F\u5728\u5411\u76EE\u6A19\u9760\u8FD1...",
    "\u52A0\u8F09\u7684\u662F\u5B57\u7BC0\uFF0C\u627F\u8F09\u7684\u662F\u5C0D\u6280\u8853\u7684\u71B1\u611B..."
  ],
  "Enter value...": "\u8ACB\u8F38\u5165\u503C...",
  "Enter sensitive value...": "\u8ACB\u8F38\u5165\u654F\u611F\u503C...",
  "Press Enter to submit, Escape to cancel": "\u6309 Enter \u63D0\u4EA4\uFF0CEscape \u53D6\u6D88",
  "Markdown file already exists: {{filename}}": "Markdown \u6A94\u6848\u5DF2\u5B58\u5728\uFF1A{{filename}}",
  "TOML Command Format Deprecation Notice": "TOML \u547D\u4EE4\u683C\u5F0F\u68C4\u7528\u901A\u77E5",
  "Found {{count}} command file(s) in TOML format:": "\u767C\u73FE {{count}} \u500B TOML \u683C\u5F0F\u7684\u547D\u4EE4\u6A94\u6848\uFF1A",
  "The TOML format for commands is being deprecated in favor of Markdown format.": "\u547D\u4EE4\u7684 TOML \u683C\u5F0F\u6B63\u5728\u88AB\u68C4\u7528\uFF0C\u63A8\u85A6\u4F7F\u7528 Markdown \u683C\u5F0F\u3002",
  "Markdown format is more readable and easier to edit.": "Markdown \u683C\u5F0F\u66F4\u6613\u8B80\u3001\u66F4\u6613\u7DE8\u8F2F\u3002",
  "You can migrate these files automatically using:": "\u60A8\u53EF\u4EE5\u4F7F\u7528\u4EE5\u4E0B\u547D\u4EE4\u81EA\u52D5\u9077\u79FB\u9019\u4E9B\u6A94\u6848\uFF1A",
  "Or manually convert each file:": "\u6216\u624B\u52D5\u8F49\u63DB\u6BCF\u500B\u6A94\u6848\uFF1A",
  'TOML: prompt = "..." / description = "..."': 'TOML\uFF1Aprompt = "..." / description = "..."',
  "Markdown: YAML frontmatter + content": "Markdown\uFF1AYAML frontmatter + \u5167\u5BB9",
  "The migration tool will:": "\u9077\u79FB\u5DE5\u5177\u5C07\uFF1A",
  "Convert TOML files to Markdown": "\u5C07 TOML \u6A94\u6848\u8F49\u63DB\u70BA Markdown",
  "Create backups of original files": "\u5275\u5EFA\u539F\u59CB\u6A94\u6848\u7684\u5099\u4EFD",
  "Preserve all command functionality": "\u4FDD\u7559\u6240\u6709\u547D\u4EE4\u529F\u80FD",
  "TOML format will continue to work for now, but migration is recommended.": "TOML \u683C\u5F0F\u76EE\u524D\u4ECD\u53EF\u4F7F\u7528\uFF0C\u4F46\u5EFA\u8B70\u9077\u79FB\u3002",
  "Open extensions page in your browser": "\u5728\u700F\u89BD\u5668\u4E2D\u6253\u958B\u64F4\u5C55\u5E02\u5834\u9801\u9762",
  "Unknown extensions source: {{source}}.": "\u672A\u77E5\u7684\u64F4\u5C55\u4F86\u6E90\uFF1A{{source}}\u3002",
  "Would open extensions page in your browser: {{url}} (skipped in test environment)": "\u5C07\u5728\u700F\u89BD\u5668\u4E2D\u6253\u958B\u64F4\u5C55\u9801\u9762\uFF1A{{url}}\uFF08\u6E2C\u8A66\u74B0\u5883\u4E2D\u5DF2\u8DF3\u904E\uFF09",
  "View available extensions at {{url}}": "\u5728 {{url}} \u67E5\u770B\u53EF\u7528\u64F4\u5C55",
  "Opening extensions page in your browser: {{url}}": "\u6B63\u5728\u700F\u89BD\u5668\u4E2D\u6253\u958B\u64F4\u5C55\u9801\u9762\uFF1A{{url}}",
  "Failed to open browser. Check out the extensions gallery at {{url}}": "\u6253\u958B\u700F\u89BD\u5668\u5931\u6557\u3002\u8ACB\u8A2A\u554F\u64F4\u5C55\u5E02\u5834\uFF1A{{url}}",
  "Retrying in {{seconds}} seconds\u2026 (attempt {{attempt}}/{{maxRetries}})": "\u5C07\u65BC {{seconds}} \u79D2\u5F8C\u91CD\u8A66\u2026\uFF08\u7B2C {{attempt}}/{{maxRetries}} \u6B21\uFF09",
  "Press Ctrl+Y to retry": "\u6309 Ctrl+Y \u91CD\u8A66\u3002",
  "No failed request to retry.": "\u6C92\u6709\u53EF\u91CD\u8A66\u7684\u5931\u6557\u8ACB\u6C42\u3002",
  "to retry last request": "\u91CD\u8A66\u4E0A\u4E00\u6B21\u8ACB\u6C42",
  "API key cannot be empty.": "API Key \u4E0D\u80FD\u70BA\u7A7A\u3002",
  'Invalid API key. Coding Plan API keys start with "sk-sp-". Please check.': '\u7121\u6548\u7684 API Key\uFF0CCoding Plan API Key \u5747\u4EE5 "sk-sp-" \u958B\u982D\uFF0C\u8ACB\u6AA2\u67E5',
  "You can get your Coding Plan API key here": "\u60A8\u53EF\u4EE5\u5728\u9019\u88CF\u7372\u53D6 Coding Plan API Key",
  "You can get your Token Plan API key here": "\u60A8\u53EF\u4EE5\u5728\u9019\u88CF\u7372\u53D6 Token Plan API Key",
  "API key is stored in settings.env. You can migrate it to a .env file for better security.": "API Key \u5DF2\u5B58\u5132\u5728 settings.env \u4E2D\u3002\u60A8\u53EF\u4EE5\u5C07\u5176\u9077\u79FB\u5230 .env \u6A94\u6848\u4EE5\u7372\u5F97\u66F4\u597D\u7684\u5B89\u5168\u6027\u3002",
  "New model configurations are available for Alibaba Cloud Coding Plan. Update now?": "\u963F\u91CC\u96F2\u767E\u934A Coding Plan \u6709\u65B0\u6A21\u578B\u914D\u7F6E\u53EF\u7528\u3002\u662F\u5426\u7ACB\u5373\u66F4\u65B0\uFF1F",
  "Coding Plan configuration updated successfully. New models are now available.": "Coding Plan \u914D\u7F6E\u66F4\u65B0\u6210\u529F\u3002\u65B0\u6A21\u578B\u73FE\u5DF2\u53EF\u7528\u3002",
  "Coding Plan API key not found. Please re-authenticate with Coding Plan.": "\u672A\u627E\u5230 Coding Plan API Key\u3002\u8ACB\u91CD\u65B0\u901A\u904E Coding Plan \u8A8D\u8B49\u3002",
  "Failed to update Coding Plan configuration: {{message}}": "\u66F4\u65B0 Coding Plan \u914D\u7F6E\u5931\u6557\uFF1A{{message}}",
  "You can configure your API key and models in settings.json": "\u60A8\u53EF\u4EE5\u5728 settings.json \u4E2D\u914D\u7F6E API Key \u548C\u6A21\u578B",
  "Refer to the documentation for setup instructions": "\u8ACB\u53C3\u8003\u6587\u6A94\u77AD\u89E3\u914D\u7F6E\u8AAA\u660E",
  "Coding Plan": "Coding Plan",
  Custom: "\u81EA\u5B9A\u7FA9",
  "Select Region for Coding Plan": "\u9078\u64C7 Coding Plan \u5340\u57DF",
  "Choose based on where your account is registered": "\u8ACB\u6839\u64DA\u60A8\u7684\u8CEC\u865F\u8A3B\u518A\u5730\u5340\u9078\u64C7",
  "Enter Coding Plan API Key": "\u8F38\u5165 Coding Plan API Key",
  "Enter Token Plan API Key": "\u8F38\u5165 Token Plan API Key",
  "New model configurations are available for {{region}}. Update now?": "{{region}} \u6709\u65B0\u7684\u6A21\u578B\u914D\u7F6E\u53EF\u7528\u3002\u662F\u5426\u7ACB\u5373\u66F4\u65B0\uFF1F",
  '{{region}} configuration updated successfully. Model switched to "{{model}}".': '{{region}} \u914D\u7F6E\u66F4\u65B0\u6210\u529F\u3002\u6A21\u578B\u5DF2\u5207\u63DB\u81F3 "{{model}}"\u3002',
  "Context Usage": "\u4E0A\u4E0B\u6587\u4F7F\u7528\u60C5\u6CC1",
  "% used": "% \u5DF2\u7528",
  "% context used": "% \u4E0A\u4E0B\u6587\u5DF2\u7528",
  "Context exceeds limit! Use /compress or /clear to reduce.": "\u4E0A\u4E0B\u6587\u8D85\u51FA\u9650\u5236\uFF01\u8ACB\u4F7F\u7528 /compress \u6216 /clear \u4F86\u6E1B\u5C11\u4E0A\u4E0B\u6587\u3002",
  "No API response yet. Send a message to see actual usage.": "\u66AB\u7121 API \u97FF\u61C9\u3002\u767C\u9001\u6D88\u606F\u4EE5\u67E5\u770B\u5BE6\u969B\u4F7F\u7528\u60C5\u6CC1\u3002",
  "Estimated pre-conversation overhead": "\u9810\u4F30\u5C0D\u8A71\u524D\u958B\u92B7",
  "Context window": "\u4E0A\u4E0B\u6587\u7A97\u53E3",
  tokens: "tokens",
  Used: "\u5DF2\u7528",
  Free: "\u7A7A\u9592",
  "Autocompact buffer": "\u81EA\u52D5\u58D3\u7E2E\u7DE9\u885D\u5340",
  "Usage by category": "\u5206\u985E\u7528\u91CF",
  "System prompt": "\u7CFB\u7D71\u63D0\u793A",
  "Built-in tools": "\u5167\u7F6E\u5DE5\u5177",
  "MCP tools": "MCP tools",
  "Memory files": "\u8A18\u61B6\u6A94\u6848",
  Skills: "\u6280\u80FD",
  Messages: "\u6D88\u606F",
  "Run /context detail for per-item breakdown.": "\u904B\u884C /context detail \u67E5\u770B\u8A73\u7D30\u5206\u89E3\u3002",
  'Show context window usage breakdown. Use "/context detail" for per-item breakdown.': '\u986F\u793A\u4E0A\u4E0B\u6587\u7A97\u53E3\u4F7F\u7528\u60C5\u6CC1\u5206\u89E3\u3002\u8F38\u5165 "/context detail" \u67E5\u770B\u8A73\u7D30\u5206\u89E3\u3002',
  "body loaded": "\u5167\u5BB9\u5DF2\u52A0\u8F09",
  memory: "\u8A18\u61B6",
  "{{region}} configuration updated successfully.": "{{region}} \u914D\u7F6E\u66F4\u65B0\u6210\u529F\u3002",
  "Authenticated successfully with {{region}}. API key and model configs saved to settings.json.": "\u6210\u529F\u901A\u904E {{region}} \u8A8D\u8B49\u3002API Key \u548C\u6A21\u578B\u914D\u7F6E\u5DF2\u4FDD\u5B58\u81F3 settings.json\u3002",
  "Tip: Use /model to switch between available Coding Plan models.": "\u63D0\u793A\uFF1A\u4F7F\u7528 /model \u5207\u63DB\u53EF\u7528\u7684 Coding Plan \u6A21\u578B\u3002",
  "Type something...": "\u8F38\u5165\u5167\u5BB9...",
  Submit: "\u63D0\u4EA4",
  "Submit answers": "\u63D0\u4EA4\u7B54\u6848",
  Cancel: "\u53D6\u6D88",
  "Your answers:": "\u60A8\u7684\u7B54\u6848\uFF1A",
  "(not answered)": "(\u672A\u56DE\u7B54)",
  "Ready to submit your answers?": "\u6E96\u5099\u597D\u63D0\u4EA4\u60A8\u7684\u7B54\u6848\u4E86\u55CE\uFF1F",
  "\u2191/\u2193: Navigate | \u2190/\u2192: Switch tabs | Enter: Select": "\u2191/\u2193: \u5C0E\u822A | \u2190/\u2192: \u5207\u63DB\u6A19\u7C64\u9801 | Enter: \u9078\u64C7",
  "\u2191/\u2193: Navigate | Enter: Select | Esc: Cancel": "\u2191/\u2193: \u5C0E\u822A | Enter: \u9078\u64C7 | Esc: \u53D6\u6D88",
  "Authenticate using Qwen OAuth": "\u4F7F\u7528 Qwen OAuth \u9032\u884C\u8A8D\u8B49",
  "Authenticate using Alibaba Cloud Coding Plan": "\u4F7F\u7528\u963F\u91CC\u96F2\u767E\u934A Coding Plan \u9032\u884C\u8A8D\u8B49",
  "Region for Coding Plan (china/global)": "Coding Plan \u5340\u57DF (china/global)",
  "API key for Coding Plan": "Coding Plan \u7684 API Key",
  "Show current authentication status": "\u986F\u793A\u7576\u524D\u8A8D\u8B49\u72C0\u614B",
  "Authentication completed successfully.": "\u8A8D\u8B49\u5B8C\u6210\u3002",
  "Starting Qwen OAuth authentication...": "\u6B63\u5728\u555F\u52D5 Qwen OAuth \u8A8D\u8B49...",
  "Successfully authenticated with Qwen OAuth.": "\u5DF2\u6210\u529F\u901A\u904E Qwen OAuth \u8A8D\u8B49\u3002",
  "Failed to authenticate with Qwen OAuth: {{error}}": "Qwen OAuth \u8A8D\u8B49\u5931\u6557\uFF1A{{error}}",
  "Processing Alibaba Cloud Coding Plan authentication...": "\u6B63\u5728\u8655\u7406\u963F\u91CC\u96F2\u767E\u934A Coding Plan \u8A8D\u8B49...",
  "Successfully authenticated with Alibaba Cloud Coding Plan.": "\u5DF2\u6210\u529F\u901A\u904E\u963F\u91CC\u96F2\u767E\u934A Coding Plan \u8A8D\u8B49\u3002",
  "Failed to authenticate with Coding Plan: {{error}}": "Coding Plan \u8A8D\u8B49\u5931\u6557\uFF1A{{error}}",
  "\u4E2D\u56FD (China)": "\u4E2D\u570B (China)",
  "\u963F\u91CC\u4E91\u767E\u70BC (aliyun.com)": "\u963F\u91CC\u96F2\u767E\u934A (aliyun.com)",
  Global: "\u5168\u7403",
  "Alibaba Cloud (alibabacloud.com)": "Alibaba Cloud (alibabacloud.com)",
  "Select region for Coding Plan:": "\u9078\u64C7 Coding Plan \u5340\u57DF\uFF1A",
  "Enter your Coding Plan API key: ": "\u8ACB\u8F38\u5165\u60A8\u7684 Coding Plan API Key\uFF1A",
  "Select authentication method:": "\u9078\u64C7\u8A8D\u8B49\u65B9\u5F0F\uFF1A",
  "\n=== Authentication Status ===\n": "\n=== \u8A8D\u8B49\u72C0\u614B ===\n",
  "\u26A0\uFE0F  No authentication method configured.\n": "\u26A0\uFE0F  \u672A\u914D\u7F6E\u8A8D\u8B49\u65B9\u5F0F\u3002\n",
  "Run one of the following commands to get started:\n": "\u904B\u884C\u4EE5\u4E0B\u547D\u4EE4\u4E4B\u4E00\u958B\u59CB\u914D\u7F6E\uFF1A\n",
  "  qwen auth qwen-oauth     - Authenticate with Qwen OAuth (discontinued)": "  qwen auth qwen-oauth     - \u4F7F\u7528 Qwen OAuth \u767B\u9304\uFF08\u5DF2\u505C\u7528\uFF09",
  "Or simply run:": "\u6216\u8005\u76F4\u63A5\u904B\u884C\uFF1A",
  "  qwen auth                - Interactive authentication setup\n": "  qwen auth                - \u4EA4\u4E92\u5F0F\u8A8D\u8B49\u914D\u7F6E\n",
  "\u2713 Authentication Method: Qwen OAuth": "\u2713 \u8A8D\u8B49\u65B9\u5F0F\uFF1AQwen OAuth",
  "  Type: Free tier (discontinued 2026-04-15)": "  \u985E\u578B\uFF1A\u514D\u8CBB\u984D\u5EA6\uFF082026-04-15 \u5DF2\u505C\u7528\uFF09",
  "  Limit: No longer available": "  \u9650\u984D\uFF1A\u5DF2\u4E0D\u53EF\u7528",
  "Qwen OAuth free tier was discontinued on 2026-04-15. Run /auth to switch to Coding Plan, OpenRouter, Fireworks AI, or another provider.": "Qwen OAuth \u514D\u8CBB\u984D\u5EA6\u5DF2\u65BC 2026-04-15 \u505C\u7528\u3002\u8ACB\u904B\u884C /auth \u5207\u63DB\u5230 Coding Plan\u3001OpenRouter\u3001Fireworks AI \u6216\u5176\u4ED6\u670D\u52D9\u5546\u3002",
  "\u2713 Authentication Method: Alibaba Cloud Coding Plan": "\u2713 \u8A8D\u8B49\u65B9\u5F0F\uFF1A\u963F\u91CC\u96F2\u767E\u934A Coding Plan",
  "\u4E2D\u56FD (China) - \u963F\u91CC\u4E91\u767E\u70BC": "\u4E2D\u570B (China) - \u963F\u91CC\u96F2\u767E\u934A",
  "Global - Alibaba Cloud": "\u5168\u7403 - Alibaba Cloud",
  "  Region: {{region}}": "  \u5340\u57DF\uFF1A{{region}}",
  "  Current Model: {{model}}": "  \u7576\u524D\u6A21\u578B\uFF1A{{model}}",
  "  Config Version: {{version}}": "  \u914D\u7F6E\u7248\u672C\uFF1A{{version}}",
  "  Status: API key configured\n": "  \u72C0\u614B\uFF1AAPI Key \u5DF2\u914D\u7F6E\n",
  "\u26A0\uFE0F  Authentication Method: Alibaba Cloud Coding Plan (Incomplete)": "\u26A0\uFE0F  \u8A8D\u8B49\u65B9\u5F0F\uFF1A\u963F\u91CC\u96F2\u767E\u934A Coding Plan\uFF08\u4E0D\u5B8C\u6574\uFF09",
  "  Issue: API key not found in environment or settings\n": "  \u554F\u984C\uFF1A\u5728\u74B0\u5883\u8B8A\u91CF\u6216\u8A2D\u7F6E\u4E2D\u672A\u627E\u5230 API Key\n",
  "  Run `qwen auth coding-plan` to re-configure.\n": "  \u904B\u884C `qwen auth coding-plan` \u91CD\u65B0\u914D\u7F6E\u3002\n",
  "\u2713 Authentication Method: {{type}}": "\u2713 \u8A8D\u8B49\u65B9\u5F0F\uFF1A{{type}}",
  "  Status: Configured\n": "  \u72C0\u614B\uFF1A\u5DF2\u914D\u7F6E\n",
  "Failed to check authentication status: {{error}}": "\u6AA2\u67E5\u8A8D\u8B49\u72C0\u614B\u5931\u6557\uFF1A{{error}}",
  "Select an option:": "\u8ACB\u9078\u64C7\uFF1A",
  "Raw mode not available. Please run in an interactive terminal.": "\u539F\u59CB\u6A21\u5F0F\u4E0D\u53EF\u7528\u3002\u8ACB\u5728\u4EA4\u4E92\u5F0F\u7D42\u7AEF\u4E2D\u904B\u884C\u3002",
  "(Use \u2191 \u2193 arrows to navigate, Enter to select, Ctrl+C to exit)\n": "(\u4F7F\u7528 \u2191 \u2193 \u7BAD\u982D\u5C0E\u822A\uFF0CEnter \u9078\u64C7\uFF0CCtrl+C \u9000\u51FA)\n",
  "Hide tool output and thinking for a cleaner view (toggle with Ctrl+O).": "\u7DCA\u6E4A\u6A21\u5F0F\u4E0B\u96B1\u85CF\u5DE5\u5177\u8F38\u51FA\u548C\u601D\u8003\u904E\u7A0B\uFF0C\u754C\u9762\u66F4\u7C21\u6F54\uFF08Ctrl+O \u5207\u63DB\uFF09\u3002",
  "Press Ctrl+O to show full tool output": "\u6309 Ctrl+O \u67E5\u770B\u8A73\u7D30\u5DE5\u5177\u8ABF\u7528\u7D50\u679C",
  "Switch to plan mode or exit plan mode": "\u5207\u63DB\u5230\u8A08\u5283\u6A21\u5F0F\u6216\u9000\u51FA\u8A08\u5283\u6A21\u5F0F",
  "Set a goal \u2014 keep working until the condition is met": "\u8A2D\u5B9A\u76EE\u6A19 \u2014 \u6301\u7E8C\u5DE5\u4F5C\u76F4\u5230\u689D\u4EF6\u6EFF\u8DB3",
  "Exited plan mode. Previous approval mode restored.": "\u5DF2\u9000\u51FA\u8A08\u5283\u6A21\u5F0F\uFF0C\u5DF2\u6062\u5FA9\u4E4B\u524D\u7684\u5BE9\u6279\u6A21\u5F0F\u3002",
  "Enabled plan mode. The agent will analyze and plan without executing tools.": "\u555F\u7528\u8A08\u5283\u6A21\u5F0F\u3002\u667A\u80FD\u9AD4\u5C07\u53EA\u5206\u6790\u548C\u898F\u5283\uFF0C\u800C\u4E0D\u57F7\u884C\u5DE5\u5177\u3002",
  'Already in plan mode. Use "/plan exit" to exit plan mode.': '\u5DF2\u8655\u65BC\u8A08\u5283\u6A21\u5F0F\u3002\u4F7F\u7528 "/plan exit" \u9000\u51FA\u8A08\u5283\u6A21\u5F0F\u3002',
  'Not in plan mode. Use "/plan" to enter plan mode first.': '\u672A\u8655\u65BC\u8A08\u5283\u6A21\u5F0F\u3002\u8ACB\u5148\u4F7F\u7528 "/plan" \u9032\u5165\u8A08\u5283\u6A21\u5F0F\u3002',
  "Set up Qwen Code's status line UI": "\u914D\u7F6E Qwen Code \u7684\u72C0\u614B\u6B04",
  "When a new todo item is created": "\u5EFA\u7ACB\u65B0\u5F85\u8FA6\u4E8B\u9805\u6642",
  "When a todo item is marked as completed": "\u5F85\u8FA6\u4E8B\u9805\u6A19\u8A18\u70BA\u5B8C\u6210\u6642",
  "Input to command is JSON with todo_id, todo_content, todo_status, all_todos, and phase. In validation, output JSON with decision (allow/block/deny) and reason. In postWrite, block/deny is ignored.": "\u547D\u4EE4\u8F38\u5165\u70BA\u5305\u542B todo_id\u3001todo_content\u3001todo_status\u3001all_todos \u548C phase \u7684 JSON\u3002\u5728 validation \u4E2D\uFF0C\u8F38\u51FA\u70BA\u5305\u542B decision\uFF08allow/block/deny\uFF09\u548C reason \u7684 JSON\u3002\u5728 postWrite \u4E2D\uFF0Cblock/deny \u6703\u88AB\u5FFD\u7565\u3002",
  "Input to command is JSON with todo_id, todo_content, previous_status, all_todos, and phase. In validation, output JSON with decision (allow/block/deny) and reason. In postWrite, block/deny is ignored.": "\u547D\u4EE4\u8F38\u5165\u70BA\u5305\u542B todo_id\u3001todo_content\u3001previous_status\u3001all_todos \u548C phase \u7684 JSON\u3002\u5728 validation \u4E2D\uFF0C\u8F38\u51FA\u70BA\u5305\u542B decision\uFF08allow/block/deny\uFF09\u548C reason \u7684 JSON\u3002\u5728 postWrite \u4E2D\uFF0Cblock/deny \u6703\u88AB\u5FFD\u7565\u3002",
  "allow todo creation": "\u5141\u8A31\u5EFA\u7ACB\u5F85\u8FA6\u4E8B\u9805",
  "block todo creation and show reason to model": "\u963B\u6B62\u5EFA\u7ACB\u5F85\u8FA6\u4E8B\u9805\u4E26\u5411\u6A21\u578B\u986F\u793A\u539F\u56E0",
  "allow todo completion": "\u5141\u8A31\u5B8C\u6210\u5F85\u8FA6\u4E8B\u9805",
  "block todo completion and show reason to model": "\u963B\u6B62\u5B8C\u6210\u5F85\u8FA6\u4E8B\u9805\u4E26\u5411\u6A21\u578B\u986F\u793A\u539F\u56E0",
  // === Core: added from PR #3328 ===
  "Open the memory manager.": "\u6253\u958B\u8A18\u61B6\u7BA1\u7406\u5668\u3002",
  "Show current process memory diagnostics": "\u986F\u793A\u76EE\u524D\u7A0B\u5E8F\u7684\u5167\u5B58\u8A3A\u65B7\u3002",
  "Record a CPU profile for Chrome DevTools analysis": "\u9304\u88FD CPU \u6548\u80FD\u5206\u6790\u6A94\u6848\uFF0C\u7528\u65BC Chrome DevTools \u5206\u6790",
  "Roll back a standalone update to the previous version": "\u5C07\u7368\u7ACB\u5B89\u88DD\u56DE\u6EFE\u5230\u4E0A\u4E00\u500B\u7248\u672C",
  "Rollback is not available in ACP mode.": "\u56DE\u6EFE\u5728 ACP \u6A21\u5F0F\u4E0B\u4E0D\u53EF\u7528\u3002",
  "Rollback is only available for standalone installations.": "\u56DE\u6EFE\u50C5\u9069\u7528\u65BC\u7368\u7ACB\u5B89\u88DD\u3002",
  "Rollback successful. Restart your terminal to use the previous version.": "\u56DE\u6EFE\u6210\u529F\u3002\u8ACB\u91CD\u555F\u7D42\u7AEF\u4EE5\u4F7F\u7528\u4E0A\u4E00\u500B\u7248\u672C\u3002",
  "Rollback failed:": "\u56DE\u6EFE\u5931\u6557\uFF1A",
  "Rollback on Windows requires manual intervention. Rename qwen-code.old to qwen-code in your installation directory.": "\u5728 Windows \u4E0A\u56DE\u6EFE\u9700\u8981\u624B\u52D5\u64CD\u4F5C\u3002\u8ACB\u5C07\u5B89\u88DD\u76EE\u9304\u4E2D\u7684 qwen-code.old \u91CD\u65B0\u547D\u540D\u70BA qwen-code\u3002",
  "Save a durable memory to the memory system.": "\u5C07\u6301\u4E45\u8A18\u61B6\u4FDD\u5B58\u5230\u8A18\u61B6\u7CFB\u7D71\u3002",
  "Ask a quick side question without affecting the main conversation": "\u5728\u4E0D\u5F71\u97FF\u4E3B\u5C0D\u8A71\u7684\u60C5\u6CC1\u4E0B\u5FEB\u901F\u63D0\u554F\u65C1\u652F\u554F\u984C",
  "Manage Arena sessions": "\u7BA1\u7406 Arena \u6703\u8A71",
  "Start an Arena session with multiple models competing on the same task": "\u555F\u52D5\u4E00\u500B Arena \u6703\u8A71\uFF0C\u8B93\u591A\u500B\u6A21\u578B\u5728\u540C\u4E00\u4EFB\u52D9\u4E0A\u7AF6\u722D",
  "Stop the current Arena session": "\u505C\u6B62\u7576\u524D\u7684 Arena \u6703\u8A71",
  "Show the current Arena session status": "\u986F\u793A\u7576\u524D\u7684 Arena \u6703\u8A71\u72C0\u614B",
  "Select a model result and merge its diff into the current workspace": "\u9078\u64C7\u4E00\u500B\u6A21\u578B\u7D50\u679C\u4E26\u5C07\u5176\u5DEE\u7570\u5408\u4F75\u5230\u7576\u524D\u5DE5\u4F5C\u5340",
  "No running Arena session found.": "\u672A\u627E\u5230\u6B63\u5728\u57F7\u884C\u7684 Arena \u6703\u8A71\u3002",
  "No Arena session found. Start one with /arena start.": "\u672A\u627E\u5230 Arena \u6703\u8A71\u3002\u8ACB\u4F7F\u7528 /arena start \u555F\u52D5\u4E00\u500B\u3002",
  "Arena session is still running. Wait for it to complete or use /arena stop first.": "Arena \u6703\u8A71\u4ECD\u5728\u57F7\u884C\u4E2D\u3002\u8ACB\u7B49\u5F85\u5176\u5B8C\u6210\uFF0C\u6216\u5148\u4F7F\u7528 /arena stop\u3002",
  "No successful agent results to select from. All agents failed or were cancelled.": "\u6C92\u6709\u53EF\u9078\u64C7\u7684\u6210\u529F\u4EE3\u7406\u7D50\u679C\u3002\u6240\u6709\u4EE3\u7406\u5747\u5931\u6557\u6216\u88AB\u53D6\u6D88\u3002",
  "Use /arena stop to end the session.": "\u4F7F\u7528 /arena stop \u7D50\u675F\u6703\u8A71\u3002",
  'No idle agent found matching "{{name}}".': '\u672A\u627E\u5230\u5339\u914D "{{name}}" \u7684\u7A7A\u9592\u4EE3\u7406\u3002',
  "Failed to apply changes from {{label}}: {{error}}": "\u5F9E {{label}} \u5957\u7528\u8B8A\u66F4\u5931\u6557\uFF1A{{error}}",
  "Applied changes from {{label}} to workspace. Arena session complete.": "\u5DF2\u5C07 {{label}} \u7684\u8B8A\u66F4\u5957\u7528\u5230\u5DE5\u4F5C\u5340\u3002Arena \u6703\u8A71\u5B8C\u6210\u3002",
  "Discard all Arena results and clean up worktrees?": "\u6368\u68C4\u6240\u6709 Arena \u7D50\u679C\u4E26\u6E05\u7406\u5DE5\u4F5C\u6A39\uFF1F",
  "Arena results discarded. All worktrees cleaned up.": "Arena \u7D50\u679C\u5DF2\u6368\u68C4\u3002\u6240\u6709\u5DE5\u4F5C\u6A39\u5DF2\u6E05\u7406\u3002",
  "Arena is not supported in non-interactive mode. Use interactive mode to start an Arena session.": "Arena \u4E0D\u652F\u63F4\u975E\u4E92\u52D5\u6A21\u5F0F\u3002\u8ACB\u4F7F\u7528\u4E92\u52D5\u6A21\u5F0F\u555F\u52D5 Arena \u6703\u8A71\u3002",
  "Arena is not supported in non-interactive mode. Use interactive mode to stop an Arena session.": "Arena \u4E0D\u652F\u63F4\u975E\u4E92\u52D5\u6A21\u5F0F\u3002\u8ACB\u4F7F\u7528\u4E92\u52D5\u6A21\u5F0F\u505C\u6B62 Arena \u6703\u8A71\u3002",
  "Arena is not supported in non-interactive mode.": "Arena \u4E0D\u652F\u63F4\u975E\u4E92\u52D5\u6A21\u5F0F\u3002",
  "An Arena session exists. Use /arena stop or /arena select to end it before starting a new one.": "\u5DF2\u5B58\u5728 Arena \u6703\u8A71\u3002\u8ACB\u4F7F\u7528 /arena stop \u6216 /arena select \u7D50\u675F\u5F8C\u518D\u555F\u52D5\u65B0\u6703\u8A71\u3002",
  "Usage: /arena start --models model1,model2 <task>": "\u7528\u6CD5\uFF1A/arena start --models model1,model2 <task>",
  "Models to compete (required, at least 2)": "\u53C3\u8207\u7AF6\u722D\u7684\u6A21\u578B\uFF08\u5FC5\u9808\uFF0C\u81F3\u5C11 2 \u500B\uFF09",
  "Format: authType:modelId or just modelId": "\u683C\u5F0F\uFF1AauthType:modelId \u6216\u50C5 modelId",
  "Arena requires at least 2 models. Use --models model1,model2 to specify.": "Arena \u81F3\u5C11\u9700\u8981 2 \u500B\u6A21\u578B\u3002\u8ACB\u4F7F\u7528 --models model1,model2 \u6307\u5B9A\u3002",
  'Arena started with {{count}} agents on task: "{{task}}"\nModels:\n{{modelList}}': 'Arena \u5DF2\u555F\u52D5 {{count}} \u500B\u4EE3\u7406\u8655\u7406\u4EFB\u52D9\uFF1A"{{task}}"\n\u6A21\u578B\uFF1A\n{{modelList}}',
  "Arena panes are running in tmux. Attach with: `{{command}}`": "Arena \u9762\u677F\u6B63\u5728 tmux \u4E2D\u57F7\u884C\u3002\u4F7F\u7528\u4EE5\u4E0B\u547D\u4EE4\u9644\u52A0\uFF1A`{{command}}`",
  "[{{label}}] failed: {{error}}": "[{{label}}] \u5931\u6557\uFF1A{{error}}",
  "Loading suggestions...": "\u6B63\u5728\u8F09\u5165\u5EFA\u8B70...",
  "Show per-item context usage breakdown.": "\u986F\u793A\u6BCF\u500B\u9805\u76EE\u7684\u4E0A\u4E0B\u6587\u4F7F\u7528\u60C5\u6CC1\u7D30\u5206\u3002",
  "Manage extension settings": "\u7BA1\u7406\u64F4\u5C55\u8A2D\u7F6E",
  "Background tasks": "\u80CC\u666F\u4EFB\u52D9",
  "No tasks currently running": "\u76EE\u524D\u6C92\u6709\u6B63\u5728\u57F7\u884C\u7684\u4EFB\u52D9",
  "No entry to show.": "\u6C92\u6709\u53EF\u986F\u793A\u7684\u9805\u76EE\u3002",
  "needs approval": "\u5F85\u5BE9\u6279",
  "rejected \u2014 edit config to re-approve": "\u5DF2\u62D2\u7D55 \u2014 \u7DE8\u8F2F\u8A2D\u5B9A\u4EE5\u91CD\u65B0\u5BE9\u6279",
  "Background agent needs approval": "\u80CC\u666F agent \u7B49\u5F85\u5BE9\u6279",
  "Approve or deny the request above": "\u8ACB\u6838\u51C6\u6216\u62D2\u7D55\u4E0A\u65B9\u7684\u8ACB\u6C42",
  Running: "\u57F7\u884C\u4E2D",
  Paused: "\u5DF2\u66AB\u505C",
  Completed: "\u5DF2\u5B8C\u6210",
  Failed: "\u5931\u6557",
  Stopped: "\u5DF2\u505C\u6B62",
  Shell: "Shell",
  Monitor: "\u76E3\u63A7",
  Command: "\u547D\u4EE4",
  Dream: "\u8A18\u61B6\u6574\u7406",
  "[dream] memory consolidation": "[\u8A18\u61B6\u6574\u7406] \u8A18\u61B6\u6574\u7406",
  "[dream] memory consolidation (reviewing {{count}} session)": "[\u8A18\u61B6\u6574\u7406] \u8A18\u61B6\u6574\u7406\uFF08\u6B63\u5728\u6AA2\u8996 {{count}} \u500B\u6703\u8A71\uFF09",
  "[dream] memory consolidation (reviewing {{count}} sessions)": "[\u8A18\u61B6\u6574\u7406] \u8A18\u61B6\u6574\u7406\uFF08\u6B63\u5728\u6AA2\u8996 {{count}} \u500B\u6703\u8A71\uFF09",
  "{{count}} session": "{{count}} \u500B\u6703\u8A71",
  "{{count}} sessions": "{{count}} \u500B\u6703\u8A71",
  "{{count}} topic": "{{count}} \u500B\u4E3B\u984C",
  "{{count}} topics": "{{count}} \u500B\u4E3B\u984C",
  "{{count}} tokens": "{{count}} tokens",
  "{{count}} tool call": "{{count}} \u500B\u5DE5\u5177\u547C\u53EB",
  "{{count}} tool calls": "{{count}} \u500B\u5DE5\u5177\u547C\u53EB",
  "{{count}} event": "{{count}} \u500B\u4E8B\u4EF6",
  "{{count}} events": "{{count}} \u500B\u4E8B\u4EF6",
  "{{count}} dropped": "\u4E1F\u68C4 {{count}} \u884C",
  "pid {{pid}}": "pid {{pid}}",
  "exit {{exitCode}}": "\u9000\u51FA\u78BC {{exitCode}}",
  "Sessions reviewing": "\u6B63\u5728\u6AA2\u8996\u7684\u6703\u8A71",
  Progress: "\u9032\u5EA6",
  "Resume blocked": "\u6062\u5FA9\u53D7\u963B",
  "Working dir": "\u5DE5\u4F5C\u76EE\u9304",
  "Output file": "\u8F38\u51FA\u6A94\u6848",
  "Topics touched ({{count}})": "\u89F8\u53CA\u7684\u4E3B\u984C\uFF08{{count}}\uFF09",
  "{{count}} more": "{{count}} \u500B",
  "Lock release warning": "\u9396\u91CB\u653E\u8B66\u544A",
  "Metadata write warning": "\u5143\u8CC7\u6599\u5BEB\u5165\u8B66\u544A",
  "Subsequent dreams may be skipped as locked until the next session's staleness sweep cleans the file.": "\u5F8C\u7E8C\u8A18\u61B6\u6574\u7406\u53EF\u80FD\u6703\u56E0\u9396\u5B9A\u800C\u7565\u904E\uFF0C\u76F4\u5230\u4E0B\u4E00\u6B21\u6703\u8A71\u904E\u671F\u6E05\u7406\u6E05\u9664\u6B64\u6A94\u6848\u3002",
  "The scheduler gate did not see this dream's timestamp; the next dream cycle may re-fire sooner than usual.": "\u6392\u7A0B\u9580\u63A7\u672A\u770B\u5230\u672C\u6B21\u8A18\u61B6\u6574\u7406\u7684\u6642\u9593\u6233\uFF1B\u4E0B\u4E00\u8F2A\u8A18\u61B6\u6574\u7406\u53EF\u80FD\u6703\u6BD4\u5E73\u6642\u66F4\u65E9\u91CD\u65B0\u89F8\u767C\u3002",
  // Stats Dashboard — Category 2 (missing from zh-TW)
  "Activity Heatmap": "\u6D3B\u52D5\u71B1\u529B\u5716",
  Less: "\u5C11",
  More: "\u591A",
  Sessions: "\u6703\u8A71\u6578",
  Duration: "\u6642\u9577",
  Projects: "\u5C08\u6848\u7D71\u8A08",
  "Loading stats...": "\u8F09\u5165\u7D71\u8A08...",
  "(no data)": "\uFF08\u66AB\u7121\u8CC7\u6599\uFF09",
  d: "\u5929",
  h: "\u6642",
  m: "\u5206",
  Input: "\u8F38\u5165",
  Models: "\u6A21\u578B",
  "All time": "\u6240\u6709\u6642\u9593",
  "Last 7 days": "\u6700\u8FD1 7 \u5929",
  "Last 30 days": "\u6700\u8FD1 30 \u5929",
  "Show usage statistics dashboard.": "\u986F\u793A\u4F7F\u7528\u7D71\u8A08\u9762\u677F\u3002",
  // Stats Dashboard — keyboard hints (not translated)
  "tab \xB7 esc": "tab \xB7 esc",
  "tab \xB7 r dates \xB7 \u2190\u2192 month \xB7 esc": "tab \xB7 r dates \xB7 \u2190\u2192 month \xB7 esc",
  "tab \xB7 r dates \xB7 esc": "tab \xB7 r dates \xB7 esc",
  // Stats Dashboard — missing labels
  "API Requests": "API \u8ACB\u6C42",
  "Tool Calls": "\u5DE5\u5177\u547C\u53EB",
  "Success rate": "\u6210\u529F\u7387",
  "Code Changes": "\u7A0B\u5F0F\u78BC\u8B8A\u66F4",
  Tool: "\u5DE5\u5177",
  reqs: "\u8ACB\u6C42",
  in: "\u8F38\u5165",
  out: "\u8F38\u51FA",
  // statsCommand non-interactive output
  "API requests: {{count}}": "API \u8ACB\u6C42\uFF1A{{count}}",
  "Code changes": "\u7A0B\u5F0F\u78BC\u8B8A\u66F4",
  Cost: "\u8CBB\u7528",
  "Estimated cost: ${{cost}}": "\u9810\u4F30\u8CBB\u7528\uFF1A${{cost}}",
  "Files: +{{added}} / -{{removed}} lines": "\u6A94\u6848\uFF1A+{{added}} / -{{removed}} \u884C",
  "N/A": "N/A",
  Name: "\u540D\u7A31",
  "No model usage data yet.": "\u5C1A\u7121\u6A21\u578B\u4F7F\u7528\u8CC7\u6599\u3002",
  "No tool usage data yet.": "\u5C1A\u7121\u5DE5\u5177\u4F7F\u7528\u8CC7\u6599\u3002",
  "Prompts: {{count}}": "\u63D0\u793A\uFF1A{{count}}",
  "Session duration: {{duration}}": "\u6703\u8A71\u6642\u9577\uFF1A{{duration}}",
  "Tokens \u2014 prompt: {{prompt}}, output: {{output}}": "Token \u2014 \u8F38\u5165\uFF1A{{prompt}}\uFF0C\u8F38\u51FA\uFF1A{{output}}",
  "Tool calls": "\u5DE5\u5177\u547C\u53EB",
  "Tool calls: {{total}} ({{success}} ok, {{fail}} fail)": "\u5DE5\u5177\u547C\u53EB\uFF1A{{total}}\uFF08{{success}} \u6210\u529F\uFF0C{{fail}} \u5931\u6557\uFF09",
  cached: "\u5FEB\u53D6",
  days: "\u5929",
  output: "\u8F38\u51FA",
  prompt: "\u8F38\u5165",
  "\u2191 tabs \xB7 r to cycle dates \xB7 esc to close": "\u2191 tab \u5207\u63DB\u6A19\u7C64 \xB7 r \u5207\u63DB\u6642\u9593\u7BC4\u570D \xB7 esc \u95DC\u9589",
  // === History collapse/expand commands ===
  "Set history to collapse by default when resuming a session": "\u6062\u5FA9\u6703\u8A71\u6642\u9810\u8A2D\u647A\u758A\u6B77\u53F2\u8A18\u9304",
  "Set history to expand by default when resuming a session": "\u6062\u5FA9\u6703\u8A71\u6642\u9810\u8A2D\u5C55\u958B\u6B77\u53F2\u8A18\u9304",
  "Expand the currently collapsed history transcript": "\u5C55\u958B\u7576\u524D\u647A\u758A\u7684\u6B77\u53F2\u8A18\u9304",
  "Control history display preferences and visibility": "\u63A7\u5236\u6B77\u53F2\u8A18\u9304\u986F\u793A\u504F\u597D\u548C\u53EF\u898B\u6027",
  "History will be collapsed by default for future resumed sessions.": "\u672A\u4F86\u6062\u5FA9\u7684\u6703\u8A71\u5C07\u9810\u8A2D\u647A\u758A\u6B77\u53F2\u8A18\u9304\u3002",
  "History will be expanded by default for future resumed sessions.": "\u672A\u4F86\u6062\u5FA9\u7684\u6703\u8A71\u5C07\u9810\u8A2D\u5C55\u958B\u6B77\u53F2\u8A18\u9304\u3002",
  "History is already expanded in this session.": "\u7576\u524D\u6703\u8A71\u7684\u6B77\u53F2\u8A18\u9304\u5DF2\u5C55\u958B\u3002",
  "Usage: /history collapse-on-resume|expand-on-resume|expand-now": "\u7528\u6CD5\uFF1A/history collapse-on-resume|expand-on-resume|expand-now",
  "History collapsed: {{n}} messages hidden. Use /history expand-now to show.": "\u6B77\u53F2\u8A18\u9304\u5DF2\u647A\u758A\uFF1A{{n}} \u689D\u8A0A\u606F\u5DF2\u96B1\u85CF\u3002\u4F7F\u7528 /history expand-now \u5C55\u958B\u3002",
  // === Same-as-English optimization ===
  " (not in model registry)": "\uFF08\u4E0D\u5728\u6A21\u578B\u8A3B\u518A\u8868\u4E2D\uFF09",
  "start server": "\u555F\u52D5\u4F3A\u670D\u5668",
  "No compression needed.": "\u7121\u9700\u58D3\u7E2E\u3002"
};
export {
  zh_TW_default as default
};
/**
 * @license
 * Copyright 2025 Qwen team
 * SPDX-License-Identifier: Apache-2.0
 */
